// @shogo-ai/sdk/dist/react/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/react/index.ts
var react_exports = {};
__export(react_exports, {
  DomainProvider: () => DomainProvider,
  useCollection: () => useCollection,
  useDomain: () => useDomain,
  useDomainReady: () => useDomainReady,
  withDomain: () => withDomain
});
module.exports = __toCommonJS(react_exports);

// src/react/DomainProvider.tsx
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var DomainContext = (0, import_react.createContext)(null);
function DomainProvider({
  createStore,
  storeKey,
  onStoreCreated,
  children
}) {
  const storeRef = (0, import_react.useRef)(null);
  const keyRef = (0, import_react.useRef)(storeKey);
  const [isReady, setIsReady] = (0, import_react.useState)(false);
  if (storeRef.current === null || keyRef.current !== storeKey) {
    keyRef.current = storeKey;
    storeRef.current = createStore();
    if (onStoreCreated) {
      onStoreCreated(storeRef.current);
    }
  }
  (0, import_react.useEffect)(() => {
    setIsReady(true);
  }, []);
  const contextValue = {
    store: storeRef.current,
    isReady
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DomainContext.Provider, { value: contextValue, children });
}
function useDomain() {
  const ctx = (0, import_react.useContext)(DomainContext);
  if (!ctx) {
    throw new Error("useDomain must be used within DomainProvider");
  }
  return ctx.store;
}
function useCollection(collectionName) {
  const store = useDomain();
  const collection = store[collectionName];
  if (!collection) {
    throw new Error(`Collection "${collectionName}" not found in domain store`);
  }
  return collection;
}
function useDomainReady() {
  const ctx = (0, import_react.useContext)(DomainContext);
  if (!ctx) {
    throw new Error("useDomainReady must be used within DomainProvider");
  }
  return { isReady: ctx.isReady };
}
function withDomain(Component) {
  return function WithDomainWrapper(props) {
    const store = useDomain();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Component, { ...props, store });
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DomainProvider,
  useCollection,
  useDomain,
  useDomainReady,
  withDomain
});
