// @shogo-ai/sdk/dist/chunk-NHPUSFVK.js
// src/react/DomainProvider.tsx
import {
  createContext,
  useContext,
  useRef,
  useEffect,
  useState
} from "react";
import { jsx } from "react/jsx-runtime";
var DomainContext = createContext(null);
function DomainProvider({
  createStore,
  storeKey,
  onStoreCreated,
  children
}) {
  const storeRef = useRef(null);
  const keyRef = useRef(storeKey);
  const [isReady, setIsReady] = useState(false);
  if (storeRef.current === null || keyRef.current !== storeKey) {
    keyRef.current = storeKey;
    storeRef.current = createStore();
    if (onStoreCreated) {
      onStoreCreated(storeRef.current);
    }
  }
  useEffect(() => {
    setIsReady(true);
  }, []);
  const contextValue = {
    store: storeRef.current,
    isReady
  };
  return /* @__PURE__ */ jsx(DomainContext.Provider, { value: contextValue, children });
}
function useDomain() {
  const ctx = useContext(DomainContext);
  if (!ctx) {
    throw new Error("useDomain must be used within DomainProvider");
  }
  return ctx.store;
}
function useCollection(collectionName) {
  const store = useDomain();
  const collection = store[collectionName];
  if (!collection) {
    throw new Error(`Collection "${collectionName}" not found in domain store`);
  }
  return collection;
}
function useDomainReady() {
  const ctx = useContext(DomainContext);
  if (!ctx) {
    throw new Error("useDomainReady must be used within DomainProvider");
  }
  return { isReady: ctx.isReady };
}
function withDomain(Component) {
  return function WithDomainWrapper(props) {
    const store = useDomain();
    return /* @__PURE__ */ jsx(Component, { ...props, store });
  };
}

export {
  DomainProvider,
  useDomain,
  useCollection,
  useDomainReady,
  withDomain
};
