// @shogo-ai/sdk/dist/chunk-QGJ2C72D.js
// src/projects/cloud-file-transport.ts
var DEFAULT_CONCURRENCY = 8;
async function getDefaultFs() {
  const isNode = typeof globalThis !== "undefined" && typeof globalThis.process?.versions?.node === "string";
  if (!isNode) {
    throw new Error(
      "CloudFileTransport: no FsAdapter passed and `node:fs/promises` is not available. Provide `fs` in CloudFileTransportOptions when running outside Node/Bun."
    );
  }
  const bundlerOpaqueImport = new Function("s", "return import(s)");
  const mod = await bundlerOpaqueImport("node:fs/promises");
  return {
    readFile: (p) => mod.readFile(p),
    writeFile: (p, d) => mod.writeFile(p, d),
    mkdir: (p, o) => mod.mkdir(p, o).then(() => void 0),
    unlink: (p) => mod.unlink(p),
    stat: async (p) => {
      const s = await mod.stat(p);
      return { size: s.size, mtimeMs: s.mtimeMs, isDirectory: () => s.isDirectory() };
    },
    readdir: async (p, o) => {
      const entries = await mod.readdir(p, o);
      return entries.map((e) => ({
        name: e.name,
        isDirectory: () => e.isDirectory(),
        isFile: () => e.isFile()
      }));
    },
    rename: (s, d) => mod.rename(s, d),
    rm: (p, o) => mod.rm(p, o)
  };
}
function joinUrl(base, ...parts) {
  const trimmed = base.replace(/\/+$/, "");
  const rest = parts.map((p) => p.replace(/^\/+|\/+$/g, "")).filter(Boolean).join("/");
  return `${trimmed}/${rest}`;
}
function joinPath(...parts) {
  return parts.filter(Boolean).map((p, i) => i === 0 ? p.replace(/\/+$/, "") : p.replace(/^\/+|\/+$/g, "")).join("/");
}
function dirOf(p) {
  const i = p.lastIndexOf("/");
  return i === -1 ? "" : p.slice(0, i);
}
function makeMatcher(patterns) {
  if (!patterns?.length) return () => true;
  const regexes = patterns.map((p) => {
    const escaped = p.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*\*/g, "\0").replace(/\*/g, "[^/]*").replace(/\u0000/g, ".*").replace(/\?/g, "[^/]");
    return new RegExp(`^${escaped}$`);
  });
  return (path) => regexes.some((re) => re.test(path));
}
async function pool(items, limit, worker) {
  const total = items.length;
  let next = 0;
  const runners = [];
  const N = Math.min(limit, Math.max(1, total));
  for (let i = 0; i < N; i++) {
    runners.push(
      (async () => {
        while (true) {
          const idx = next++;
          if (idx >= total) return;
          await worker(items[idx], idx, total);
        }
      })()
    );
  }
  await Promise.all(runners);
}
var CloudFileTransport = class {
  constructor(opts) {
    this.fsCached = null;
    if (!opts.apiUrl) throw new Error("CloudFileTransport: apiUrl is required");
    if (!opts.apiKey) throw new Error("CloudFileTransport: apiKey is required");
    if (!opts.projectId) throw new Error("CloudFileTransport: projectId is required");
    if (!opts.localDir) throw new Error("CloudFileTransport: localDir is required");
    this.apiUrl = opts.apiUrl.replace(/\/+$/, "");
    this.apiKey = opts.apiKey;
    this.projectId = opts.projectId;
    this.localDir = opts.localDir;
    this.include = makeMatcher(opts.include);
    this.concurrency = opts.concurrency ?? DEFAULT_CONCURRENCY;
    this.onProgress = opts.onProgress;
    this.fetchImpl = opts.fetchImpl ?? globalThis.fetch.bind(globalThis);
    this.fsOverride = opts.fs;
  }
  // ─── Manifest ────────────────────────────────────────────────────
  /**
   * Fetch the full workspace manifest from the cloud. Returns every
   * file the server is willing to expose (excluded dirs + sensitive
   * patterns already filtered server-side).
   */
  async listManifest() {
    const url = joinUrl(this.apiUrl, "/api/projects", encodeURIComponent(this.projectId), "/workspace/manifest");
    const res = await this.fetchImpl(url, {
      method: "GET",
      headers: this.authHeaders()
    });
    if (!res.ok) {
      throw new Error(`manifest_failed: HTTP ${res.status} ${await safeText(res)}`);
    }
    const body = await res.json();
    return body.files.filter((f) => this.include(f.path));
  }
  // ─── Pull (cloud → local) ────────────────────────────────────────
  /**
   * Pull every manifest entry that's missing or stale locally.
   *
   * Atomicity: the pull writes into `<localDir>.shogo-pull-tmp/` and
   * then renames over `<localDir>` on success, so a Ctrl-C mid-pull
   * never leaves a half-populated workspace dir. The previous content
   * is wiped on commit — callers that want incremental updates should
   * use {@link downloadFiles} directly instead.
   */
  async downloadAll() {
    const stats = { downloaded: 0, uploaded: 0, skipped: 0, deleted: 0, errors: [] };
    const fs = await this.fs();
    const manifest = await this.listManifest();
    const stagingDir = `${this.localDir}.shogo-pull-tmp`;
    try {
      await fs.rm(stagingDir, { recursive: true, force: true });
    } catch {
    }
    await fs.mkdir(stagingDir, { recursive: true });
    await this.downloadFilesInto(manifest, stagingDir, stats);
    if (stats.errors.length === 0) {
      try {
        await fs.rm(this.localDir, { recursive: true, force: true });
      } catch {
      }
      await fs.rename(stagingDir, this.localDir);
    } else {
      throw new Error(
        `Pull aborted with ${stats.errors.length} errors (first: ${stats.errors[0]?.message}). Staging dir left at ${stagingDir}.`
      );
    }
    return stats;
  }
  /**
   * Lower-level pull that downloads `files` into an arbitrary directory.
   * Used by `downloadAll` for staging and by the worker auto-pull for
   * "merge into existing workspace" semantics.
   */
  async downloadFiles(files) {
    const stats = { downloaded: 0, uploaded: 0, skipped: 0, deleted: 0, errors: [] };
    await this.downloadFilesInto(files, this.localDir, stats);
    return stats;
  }
  async downloadFilesInto(files, destDir, stats) {
    if (files.length === 0) return;
    const fs = await this.fs();
    const urls = await this.presignBatch(
      files.map((f) => ({ path: f.path, action: "read" }))
    );
    const urlByPath = new Map(urls.map((u) => [u.path, u.url]));
    await pool(files, this.concurrency, async (entry, idx, total) => {
      const url = urlByPath.get(entry.path);
      if (!url) {
        stats.errors.push({ path: entry.path, message: "No presigned URL returned" });
        return;
      }
      try {
        const resp = await this.fetchImpl(url);
        if (!resp.ok) {
          stats.errors.push({ path: entry.path, message: `HTTP ${resp.status} fetching ${entry.path}` });
          return;
        }
        const buf = new Uint8Array(await resp.arrayBuffer());
        const localPath = joinPath(destDir, entry.path);
        const parent = dirOf(localPath);
        if (parent) await fs.mkdir(parent, { recursive: true });
        await fs.writeFile(localPath, buf);
        stats.downloaded += 1;
        this.onProgress?.({ kind: "download", path: entry.path, bytes: buf.byteLength, index: idx, total });
      } catch (err) {
        stats.errors.push({ path: entry.path, message: err?.message ?? String(err) });
      }
    });
  }
  // ─── Push (local → cloud) ────────────────────────────────────────
  /**
   * Walk `localDir`, build a list of files matching the include filter,
   * and upload each via a presigned PUT. If `opts.deleteRemote` is set,
   * any file present in the remote manifest but missing locally is
   * deleted via `DELETE /api/projects/:id/files/<path>`.
   */
  async uploadAll(opts = {}) {
    const stats = { downloaded: 0, uploaded: 0, skipped: 0, deleted: 0, errors: [] };
    const fs = await this.fs();
    const localFiles = await walkLocal(fs, this.localDir, this.include);
    const urls = await this.presignBatch(
      localFiles.map((f) => ({ path: f.path, action: "write", contentType: f.contentType }))
    );
    const urlByPath = new Map(urls.map((u) => [u.path, u.url]));
    await pool(localFiles, this.concurrency, async (file, idx, total) => {
      const url = urlByPath.get(file.path);
      if (!url) {
        stats.errors.push({ path: file.path, message: "No presigned URL returned" });
        return;
      }
      try {
        const buf = await fs.readFile(joinPath(this.localDir, file.path));
        const resp = await this.fetchImpl(url, {
          method: "PUT",
          body: buf,
          headers: file.contentType ? { "content-type": file.contentType } : void 0
        });
        if (!resp.ok) {
          stats.errors.push({ path: file.path, message: `HTTP ${resp.status} uploading ${file.path}` });
          return;
        }
        stats.uploaded += 1;
        this.onProgress?.({ kind: "upload", path: file.path, bytes: buf.byteLength, index: idx, total });
      } catch (err) {
        stats.errors.push({ path: file.path, message: err?.message ?? String(err) });
      }
    });
    if (opts.deleteRemote) {
      const remote = await this.listManifest();
      const localPaths = new Set(localFiles.map((f) => f.path));
      const toDelete = remote.filter((r) => !localPaths.has(r.path));
      await pool(toDelete, this.concurrency, async (entry, idx, total) => {
        try {
          const url = joinUrl(
            this.apiUrl,
            "/api/projects",
            encodeURIComponent(this.projectId),
            "/files",
            entry.path
          );
          const resp = await this.fetchImpl(url, { method: "DELETE", headers: this.authHeaders() });
          if (!resp.ok && resp.status !== 404) {
            stats.errors.push({ path: entry.path, message: `HTTP ${resp.status} deleting ${entry.path}` });
            return;
          }
          stats.deleted += 1;
          this.onProgress?.({ kind: "delete", path: entry.path, index: idx, total });
        } catch (err) {
          stats.errors.push({ path: entry.path, message: err?.message ?? String(err) });
        }
      });
    }
    return stats;
  }
  /**
   * Upload only the given relative paths (a subset of localDir). Used
   * by the worker watcher to push individual file events without
   * re-walking the entire tree.
   */
  async uploadFiles(relativePaths) {
    const stats = { downloaded: 0, uploaded: 0, skipped: 0, deleted: 0, errors: [] };
    if (relativePaths.length === 0) return stats;
    const fs = await this.fs();
    const fileSpecs = relativePaths.filter((p) => this.include(p)).map((p) => ({ path: p, contentType: guessContentType(p) }));
    if (fileSpecs.length === 0) return stats;
    const urls = await this.presignBatch(
      fileSpecs.map((f) => ({ path: f.path, action: "write", contentType: f.contentType }))
    );
    const urlByPath = new Map(urls.map((u) => [u.path, u.url]));
    await pool(fileSpecs, this.concurrency, async (file, idx, total) => {
      const url = urlByPath.get(file.path);
      if (!url) {
        stats.errors.push({ path: file.path, message: "No presigned URL returned" });
        return;
      }
      try {
        const buf = await fs.readFile(joinPath(this.localDir, file.path));
        const resp = await this.fetchImpl(url, {
          method: "PUT",
          body: buf,
          headers: file.contentType ? { "content-type": file.contentType } : void 0
        });
        if (!resp.ok) {
          stats.errors.push({ path: file.path, message: `HTTP ${resp.status} uploading ${file.path}` });
          return;
        }
        stats.uploaded += 1;
        this.onProgress?.({ kind: "upload", path: file.path, bytes: buf.byteLength, index: idx, total });
      } catch (err) {
        stats.errors.push({ path: file.path, message: err?.message ?? String(err) });
      }
    });
    return stats;
  }
  /** Delete a single remote file. Convenience wrapper around the cloud API. */
  async deleteRemote(path) {
    const url = joinUrl(this.apiUrl, "/api/projects", encodeURIComponent(this.projectId), "/files", path);
    const resp = await this.fetchImpl(url, { method: "DELETE", headers: this.authHeaders() });
    if (!resp.ok && resp.status !== 404) {
      throw new Error(`HTTP ${resp.status} deleting ${path}`);
    }
  }
  // ─── Internals ──────────────────────────────────────────────────
  authHeaders(extra) {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      ...extra ?? {}
    };
  }
  async presignBatch(files) {
    if (files.length === 0) return [];
    const url = joinUrl(this.apiUrl, "/api/projects", encodeURIComponent(this.projectId), "/s3/presign");
    const resp = await this.fetchImpl(url, {
      method: "POST",
      headers: this.authHeaders({ "content-type": "application/json" }),
      body: JSON.stringify({ files })
    });
    if (!resp.ok) {
      throw new Error(`presign_failed: HTTP ${resp.status} ${await safeText(resp)}`);
    }
    const body = await resp.json();
    return body.urls ?? [];
  }
  async fs() {
    if (this.fsOverride) return this.fsOverride;
    if (!this.fsCached) this.fsCached = await getDefaultFs();
    return this.fsCached;
  }
};
async function walkLocal(fs, root, include) {
  const out = [];
  const EXCLUDED = /* @__PURE__ */ new Set(["node_modules", ".git", "dist", "build", ".vite", ".cache"]);
  async function visit(dir, rel) {
    let entries;
    try {
      entries = await fs.readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (EXCLUDED.has(entry.name)) continue;
      const next = rel ? `${rel}/${entry.name}` : entry.name;
      const full = joinPath(dir, entry.name);
      if (entry.isDirectory()) {
        await visit(full, next);
      } else if (entry.isFile()) {
        if (include(next)) {
          out.push({ path: next, contentType: guessContentType(next) });
        }
      }
    }
  }
  await visit(root, "");
  return out;
}
function guessContentType(path) {
  const i = path.lastIndexOf(".");
  if (i === -1) return "application/octet-stream";
  const ext = path.slice(i + 1).toLowerCase();
  switch (ext) {
    case "ts":
    case "tsx":
      return "application/typescript";
    case "js":
    case "jsx":
    case "mjs":
    case "cjs":
      return "application/javascript";
    case "json":
      return "application/json";
    case "md":
      return "text/markdown";
    case "html":
    case "htm":
      return "text/html";
    case "css":
      return "text/css";
    case "svg":
      return "image/svg+xml";
    case "png":
      return "image/png";
    case "jpg":
    case "jpeg":
      return "image/jpeg";
    case "webp":
      return "image/webp";
    case "txt":
      return "text/plain";
    case "yaml":
    case "yml":
      return "application/yaml";
    default:
      return "application/octet-stream";
  }
}
async function safeText(resp) {
  try {
    return (await resp.text()).slice(0, 500);
  } catch {
    return "";
  }
}

export {
  CloudFileTransport
};
