// @shogo-ai/sdk/dist/index.js
import {
  CloudFileTransport
} from "./chunk-QGJ2C72D.js";
import {
  DomainProvider,
  useCollection,
  useDomain,
  useDomainReady,
  withDomain
} from "./chunk-NHPUSFVK.js";

// src/errors.ts
var ShogoError = class _ShogoError extends Error {
  constructor(message, code = "UNKNOWN", status, details) {
    super(message);
    this.name = "ShogoError";
    this.code = code;
    this.status = status;
    this.details = details;
    if ("captureStackTrace" in Error) {
      ;
      Error.captureStackTrace(this, _ShogoError);
    }
  }
  static fromStatus(status, message, details) {
    const codeMap = {
      400: "VALIDATION_ERROR",
      401: "UNAUTHORIZED",
      403: "FORBIDDEN",
      404: "NOT_FOUND",
      409: "CONFLICT",
      429: "RATE_LIMITED",
      500: "SERVER_ERROR"
    };
    const code = codeMap[status] || "UNKNOWN";
    const defaultMessage = message || `Request failed with status ${status}`;
    return new _ShogoError(defaultMessage, code, status, details);
  }
  static networkError(message, details) {
    return new _ShogoError(message, "NETWORK_ERROR", void 0, details);
  }
  static unauthorized(message = "Unauthorized") {
    return new _ShogoError(message, "UNAUTHORIZED", 401);
  }
  static notFound(message = "Not found") {
    return new _ShogoError(message, "NOT_FOUND", 404);
  }
  static validationError(message, details) {
    return new _ShogoError(message, "VALIDATION_ERROR", 400, details);
  }
  toJSON() {
    return {
      name: this.name,
      message: this.message,
      code: this.code,
      status: this.status,
      details: this.details
    };
  }
};
var AuthError = class _AuthError extends ShogoError {
  constructor(message, code = "UNAUTHORIZED", details) {
    super(message, code, code === "UNAUTHORIZED" ? 401 : void 0, details);
    this.name = "AuthError";
  }
  static invalidCredentials() {
    return new _AuthError(
      "Invalid email or password",
      "AUTH_INVALID_CREDENTIALS"
    );
  }
  static userExists(email) {
    return new _AuthError(
      `User with email ${email} already exists`,
      "AUTH_USER_EXISTS"
    );
  }
  static sessionExpired() {
    return new _AuthError(
      "Session has expired. Please sign in again.",
      "AUTH_SESSION_EXPIRED"
    );
  }
  static invalidToken() {
    return new _AuthError(
      "Invalid or expired token",
      "AUTH_INVALID_TOKEN"
    );
  }
  /**
   * Returned when the platform Better Auth allowlist hook denies a
   * sign-in / sign-up because the email is not on this project's
   * allowlist (see `ProjectAuthConfig` and the Studio Settings ->
   * Auth & Database panel).
   */
  static projectNotAllowed(message, details) {
    return new _AuthError(
      message ?? "This project's owner has restricted who can sign in.",
      "AUTH_PROJECT_NOT_ALLOWED",
      details
    );
  }
};
var DatabaseError = class _DatabaseError extends ShogoError {
  constructor(message, code = "DB_QUERY_ERROR", details) {
    super(message, code, void 0, details);
    this.name = "DatabaseError";
  }
  static entityNotFound(model, id) {
    return new _DatabaseError(
      `${model} with id '${id}' not found`,
      "DB_ENTITY_NOT_FOUND"
    );
  }
  static queryError(message, details) {
    return new _DatabaseError(message, "DB_QUERY_ERROR", details);
  }
};

// src/http/client.ts
var HttpClient = class {
  constructor(config) {
    this.mcpSessionId = null;
    this.mcpInitPromise = null;
    /** Cache for in-flight and recently completed GET requests */
    this.requestCache = /* @__PURE__ */ new Map();
    const base = config.baseUrl?.replace(/\/$/, "") || "";
    if (!base && typeof globalThis.window !== "undefined") {
      this.baseUrl = globalThis.window.location.origin;
    } else {
      this.baseUrl = base;
    }
    this.getToken = config.getToken ?? (() => null);
    this.mcpPath = config.mcpPath ?? "/mcp";
    this.authPath = config.authPath ?? "/api/auth";
    this.dedupWindowMs = config.dedupWindowMs ?? 100;
    this.getAuthCookie = config.getAuthCookie ?? null;
    this.credentials = this.getAuthCookie ? "omit" : config.credentials ?? "same-origin";
    this.projectId = config.projectId ?? null;
  }
  /**
   * Update the token getter (called when auth state changes)
   */
  setTokenGetter(getToken) {
    this.getToken = getToken;
  }
  /**
   * Build full URL with optional search params
   */
  buildUrl(path, searchParams) {
    const url = new URL(path, this.baseUrl);
    if (searchParams) {
      Object.entries(searchParams).forEach(([key, value]) => {
        url.searchParams.set(key, value);
      });
    }
    return url.toString();
  }
  /**
   * Get default headers including auth token
   */
  getHeaders(additionalHeaders) {
    const headers = {
      "Content-Type": "application/json",
      ...additionalHeaders
    };
    const token = this.getToken();
    if (token) {
      headers["Authorization"] = `Bearer ${token}`;
    }
    if (this.getAuthCookie) {
      const cookie = this.getAuthCookie();
      if (cookie) {
        headers["Cookie"] = cookie;
      }
    }
    return headers;
  }
  /**
   * Make an HTTP request with optional deduplication for GET requests
   */
  async request(path, options = {}) {
    const { method = "GET", headers, body, searchParams, signal } = options;
    if (method === "GET") {
      const cacheKey = this.buildCacheKey(path, searchParams);
      const cached = this.getCachedRequest(cacheKey);
      if (cached) {
        return cached;
      }
      const promise = this.executeRequest(path, options);
      this.cacheRequest(cacheKey, promise);
      return promise;
    }
    return this.executeRequest(path, options);
  }
  /**
   * Build cache key for request deduplication
   */
  buildCacheKey(path, searchParams) {
    const params = searchParams ? JSON.stringify(searchParams) : "";
    return `GET:${path}:${params}`;
  }
  /**
   * Get a cached request if still valid
   */
  getCachedRequest(cacheKey) {
    const entry = this.requestCache.get(cacheKey);
    if (!entry) return void 0;
    const age = Date.now() - entry.timestamp;
    if (age > this.dedupWindowMs) {
      this.requestCache.delete(cacheKey);
      return void 0;
    }
    return entry.promise;
  }
  /**
   * Cache a request promise
   */
  cacheRequest(cacheKey, promise) {
    this.requestCache.set(cacheKey, {
      promise,
      timestamp: Date.now()
    });
    setTimeout(() => {
      const entry = this.requestCache.get(cacheKey);
      if (entry && Date.now() - entry.timestamp >= this.dedupWindowMs) {
        this.requestCache.delete(cacheKey);
      }
    }, this.dedupWindowMs + 10);
  }
  /**
   * Clear the request cache (e.g., after auth change)
   */
  clearCache() {
    this.requestCache.clear();
  }
  /**
   * Execute an HTTP request (internal, no deduplication)
   */
  async executeRequest(path, options = {}) {
    const { method = "GET", headers, body, searchParams, signal } = options;
    const url = this.buildUrl(path, searchParams);
    const requestHeaders = this.getHeaders(headers);
    try {
      const response = await fetch(url, {
        method,
        headers: requestHeaders,
        body: body ? JSON.stringify(body) : void 0,
        signal,
        credentials: this.credentials
      });
      const data = await this.parseResponse(response);
      if (!response.ok) {
        let errorMessage;
        if (typeof data === "object" && data !== null) {
          const d = data;
          if ("message" in d && d.message) {
            errorMessage = String(d.message);
          } else if ("error" in d && d.error !== null && d.error !== void 0) {
            if (typeof d.error === "string") {
              errorMessage = d.error;
            } else if (typeof d.error === "object") {
              const err = d.error;
              if ("message" in err && err.message) {
                errorMessage = String(err.message);
              }
            }
          }
        }
        throw ShogoError.fromStatus(response.status, errorMessage, data);
      }
      return {
        data,
        status: response.status,
        headers: response.headers
      };
    } catch (error) {
      if (error instanceof ShogoError) {
        throw error;
      }
      if (error instanceof Error && error.name === "AbortError") {
        throw error;
      }
      throw ShogoError.networkError(
        error instanceof Error ? error.message : "Network request failed",
        error
      );
    }
  }
  /**
   * Parse response body
   */
  async parseResponse(response) {
    const contentType = response.headers.get("content-type");
    if (contentType?.includes("application/json")) {
      return response.json();
    }
    return response.text();
  }
  /**
   * GET request
   */
  async get(path, searchParams) {
    return this.request(path, { method: "GET", searchParams });
  }
  /**
   * POST request
   */
  async post(path, body, headers) {
    return this.request(path, { method: "POST", body, headers });
  }
  /**
   * PATCH request
   */
  async patch(path, body, headers) {
    return this.request(path, { method: "PATCH", body, headers });
  }
  /**
   * DELETE request
   */
  async delete(path, searchParams) {
    return this.request(path, { method: "DELETE", searchParams });
  }
  // ==========================================================================
  // Auth-specific methods
  // ==========================================================================
  /**
   * Get auth endpoint URL
   */
  getAuthUrl(endpoint) {
    return `${this.authPath}${endpoint}`;
  }
  /**
   * Make auth request. When the client was constructed with a
   * `projectId`, every auth request carries `X-Shogo-Project-Id` so the
   * Shogo platform's Better Auth before-hook can enforce the project's
   * sign-in allowlist (`ProjectAuthConfig`). Caller-supplied headers
   * win — pass `{ headers: { 'x-shogo-project-id': '' } }` to opt out.
   */
  async authRequest(endpoint, options = {}) {
    if (!this.projectId) {
      return this.request(this.getAuthUrl(endpoint), options);
    }
    const merged = {
      ...options,
      headers: {
        "X-Shogo-Project-Id": this.projectId,
        ...options.headers ?? {}
      }
    };
    return this.request(this.getAuthUrl(endpoint), merged);
  }
  // ==========================================================================
  // MCP-specific methods
  // ==========================================================================
  /**
   * Initialize MCP session if not already initialized
   */
  async ensureMcpInitialized() {
    if (this.mcpSessionId) {
      return;
    }
    if (this.mcpInitPromise) {
      return this.mcpInitPromise;
    }
    this.mcpInitPromise = this.initializeMcpSession();
    try {
      await this.mcpInitPromise;
    } finally {
      this.mcpInitPromise = null;
    }
  }
  /**
   * Initialize MCP session
   */
  async initializeMcpSession() {
    const request = {
      jsonrpc: "2.0",
      id: 1,
      method: "initialize",
      params: {
        protocolVersion: "2024-11-05",
        capabilities: {},
        clientInfo: {
          name: "@shogo-ai/sdk",
          version: "0.1.0"
        }
      }
    };
    const response = await fetch(this.buildUrl(this.mcpPath), {
      method: "POST",
      headers: this.getHeaders(),
      body: JSON.stringify(request),
      credentials: this.credentials
    });
    const sessionId = response.headers.get("mcp-session-id");
    if (sessionId) {
      this.mcpSessionId = sessionId;
    }
    const result = await response.json();
    if (result.error) {
      throw new ShogoError(
        result.error.message,
        "SERVER_ERROR",
        void 0,
        result.error
      );
    }
  }
  /**
   * Call an MCP tool
   */
  async callTool(name, args) {
    await this.ensureMcpInitialized();
    const request = {
      jsonrpc: "2.0",
      id: Date.now(),
      method: "tools/call",
      params: {
        name,
        arguments: args
      }
    };
    const headers = this.getHeaders();
    if (this.mcpSessionId) {
      headers["mcp-session-id"] = this.mcpSessionId;
    }
    const response = await fetch(this.buildUrl(this.mcpPath), {
      method: "POST",
      headers,
      body: JSON.stringify(request),
      credentials: this.credentials
    });
    const result = await response.json();
    if (result.error) {
      throw new ShogoError(
        result.error.message,
        "DB_QUERY_ERROR",
        void 0,
        result.error
      );
    }
    if (result.result?.content?.[0]?.text) {
      try {
        return JSON.parse(result.result.content[0].text);
      } catch {
        return result.result.content[0].text;
      }
    }
    return result.result;
  }
  /**
   * Reset MCP session (e.g., after auth change)
   */
  resetMcpSession() {
    this.mcpSessionId = null;
    this.mcpInitPromise = null;
  }
};

// src/auth/index.ts
var STORAGE_KEY_TOKENS = "auth_tokens";
var STORAGE_KEY_USER = "auth_user";
function maybeProjectNotAllowed(error) {
  if (!(error instanceof ShogoError)) return null;
  if (error.status !== 403) return null;
  const details = error.details;
  if (!details || typeof details !== "object") return null;
  if (details.code !== "project_auth_not_allowed") return null;
  return AuthError.projectNotAllowed(
    typeof details.message === "string" ? details.message : void 0,
    details
  );
}
var ShogoAuth = class {
  constructor(httpClient, storage, config = {}) {
    this.state = {
      user: null,
      session: null,
      isAuthenticated: false,
      isLoading: true
    };
    this.listeners = /* @__PURE__ */ new Set();
    this.initPromise = null;
    this.httpClient = httpClient;
    this.storage = storage;
    this.config = config;
    this.initPromise = this.initialize();
  }
  // ==========================================================================
  // Initialization
  // ==========================================================================
  /**
   * Initialize auth state from storage
   */
  async initialize() {
    try {
      const tokensStr = await this.storage.getItem(STORAGE_KEY_TOKENS);
      const userStr = await this.storage.getItem(STORAGE_KEY_USER);
      if (tokensStr && userStr) {
        const tokens = JSON.parse(tokensStr);
        const user = JSON.parse(userStr);
        if (tokens.expiresAt && Date.now() > tokens.expiresAt) {
          await this.clearAuth();
        } else {
          this.updateState({
            user,
            session: { user, token: tokens.accessToken },
            isAuthenticated: true,
            isLoading: false
          });
          return;
        }
      }
    } catch {
    }
    this.updateState({
      user: null,
      session: null,
      isAuthenticated: false,
      isLoading: false
    });
  }
  /**
   * Wait for auth initialization to complete
   */
  async ready() {
    if (this.initPromise) {
      await this.initPromise;
    }
  }
  // ==========================================================================
  // State Management
  // ==========================================================================
  /**
   * Get current auth state
   */
  getState() {
    return { ...this.state };
  }
  /**
   * Get current user (sync, may be null if loading)
   */
  currentUser() {
    return this.state.user;
  }
  /**
   * Get current session
   */
  async getSession() {
    await this.ready();
    return this.state.session;
  }
  /**
   * Check if user is authenticated
   */
  isAuthenticated() {
    return this.state.isAuthenticated;
  }
  /**
   * Get current access token
   */
  getToken() {
    return this.state.session?.token ?? null;
  }
  /**
   * Subscribe to auth state changes
   */
  onAuthStateChanged(callback) {
    this.listeners.add(callback);
    callback(this.getState());
    return () => {
      this.listeners.delete(callback);
    };
  }
  /**
   * Update auth state and notify listeners
   */
  updateState(newState) {
    this.state = { ...this.state, ...newState };
    this.notifyListeners();
  }
  /**
   * Notify all listeners of state change
   */
  notifyListeners() {
    const state = this.getState();
    this.listeners.forEach((callback) => {
      try {
        callback(state);
      } catch {
      }
    });
  }
  // ==========================================================================
  // Token Management
  // ==========================================================================
  /**
   * Save tokens to storage
   */
  async saveTokens(tokens) {
    const expiresAt = tokens.expiresIn ? Date.now() + tokens.expiresIn * 1e3 : void 0;
    const tokensWithExpiry = {
      ...tokens,
      expiresAt
    };
    await this.storage.setItem(STORAGE_KEY_TOKENS, JSON.stringify(tokensWithExpiry));
  }
  /**
   * Save user to storage
   */
  async saveUser(user) {
    await this.storage.setItem(STORAGE_KEY_USER, JSON.stringify(user));
  }
  /**
   * Clear auth from storage
   */
  async clearAuth() {
    await this.storage.removeItem(STORAGE_KEY_TOKENS);
    await this.storage.removeItem(STORAGE_KEY_USER);
    this.httpClient.resetMcpSession();
  }
  /**
   * Set auth session from tokens (for manual token injection)
   */
  async setSession(tokens, user) {
    await this.saveTokens(tokens);
    const currentUser = user ?? await this.fetchCurrentUser(tokens.accessToken);
    await this.saveUser(currentUser);
    this.updateState({
      user: currentUser,
      session: { user: currentUser, token: tokens.accessToken },
      isAuthenticated: true,
      isLoading: false
    });
    return currentUser;
  }
  // ==========================================================================
  // Auth Flows
  // ==========================================================================
  /**
   * Sign up with email and password
   */
  async signUp(data) {
    this.updateState({ isLoading: true });
    try {
      const response = await this.httpClient.authRequest("/sign-up/email", {
        method: "POST",
        body: {
          email: data.email,
          password: data.password,
          name: data.name,
          ...data.metadata
        }
      });
      const { user, token, expiresIn } = response.data;
      await this.saveTokens({
        accessToken: token,
        expiresIn
      });
      await this.saveUser(user);
      this.updateState({
        user,
        session: { user, token },
        isAuthenticated: true,
        isLoading: false
      });
      return user;
    } catch (error) {
      this.updateState({ isLoading: false });
      const projectDenied = maybeProjectNotAllowed(error);
      if (projectDenied) throw projectDenied;
      if (error instanceof AuthError) {
        throw error;
      }
      throw new AuthError(
        error instanceof Error ? error.message : "Sign up failed",
        "UNKNOWN"
      );
    }
  }
  /**
   * Sign in with email and password
   */
  async signIn(data) {
    this.updateState({ isLoading: true });
    try {
      const response = await this.httpClient.authRequest("/sign-in/email", {
        method: "POST",
        body: {
          email: data.email,
          password: data.password
        }
      });
      const { user, token, expiresIn } = response.data;
      await this.saveTokens({
        accessToken: token,
        expiresIn
      });
      await this.saveUser(user);
      this.updateState({
        user,
        session: { user, token },
        isAuthenticated: true,
        isLoading: false
      });
      return user;
    } catch (error) {
      this.updateState({ isLoading: false });
      const projectDenied = maybeProjectNotAllowed(error);
      if (projectDenied) throw projectDenied;
      if (error instanceof AuthError) {
        throw error;
      }
      throw AuthError.invalidCredentials();
    }
  }
  /**
   * Sign in with OAuth provider
   */
  async signInWithProvider(provider) {
    const redirectUrl = this.config.redirectUrl ?? window?.location?.href;
    const authUrl = new URL(
      this.httpClient.getAuthUrl(`/sign-in/${provider}`),
      window?.location?.origin
    );
    authUrl.searchParams.set("callbackURL", redirectUrl ?? "/");
    if (typeof window !== "undefined") {
      window.location.href = authUrl.toString();
    }
  }
  /**
   * Sign in with Google
   */
  async signInWithGoogle() {
    return this.signInWithProvider("google");
  }
  /**
   * Sign in with GitHub
   */
  async signInWithGitHub() {
    return this.signInWithProvider("github");
  }
  /**
   * Sign in with Apple
   */
  async signInWithApple() {
    return this.signInWithProvider("apple");
  }
  /**
   * Sign in with Microsoft
   */
  async signInWithMicrosoft() {
    return this.signInWithProvider("microsoft");
  }
  /**
   * Sign out
   */
  async signOut() {
    this.updateState({ isLoading: true });
    try {
      await this.httpClient.authRequest("/sign-out", {
        method: "POST"
      }).catch(() => {
      });
    } finally {
      await this.clearAuth();
      this.updateState({
        user: null,
        session: null,
        isAuthenticated: false,
        isLoading: false
      });
    }
  }
  // ==========================================================================
  // User Operations
  // ==========================================================================
  /**
   * Fetch current user from server
   */
  async fetchCurrentUser(token) {
    const headers = {};
    if (token) {
      headers["Authorization"] = `Bearer ${token}`;
    }
    const response = await this.httpClient.authRequest(
      "/session",
      { headers }
    );
    return response.data.user;
  }
  /**
   * Get current user (async, fetches from server)
   */
  async me() {
    await this.ready();
    if (!this.state.isAuthenticated) {
      throw AuthError.invalidToken();
    }
    return this.fetchCurrentUser();
  }
  /**
   * Update current user profile
   */
  async updateProfile(updates) {
    if (!this.state.isAuthenticated) {
      throw AuthError.invalidToken();
    }
    const response = await this.httpClient.authRequest(
      "/update-user",
      {
        method: "POST",
        body: updates
      }
    );
    const user = response.data.user;
    await this.saveUser(user);
    this.updateState({
      user,
      session: this.state.session ? { ...this.state.session, user } : null
    });
    return user;
  }
};

// src/platform/index.ts
var BYOK_PROVIDERS = [
  { id: "anthropic", name: "Anthropic", envKey: "ANTHROPIC_API_KEY", signupUrl: "https://console.anthropic.com/" },
  { id: "openai", name: "OpenAI", envKey: "OPENAI_API_KEY", signupUrl: "https://platform.openai.com/api-keys" }
];
var PlatformApi = class {
  constructor(http) {
    this.http = http;
  }
  // ===========================================================================
  // Platform Config
  // ===========================================================================
  /** Fetch platform configuration and feature flags. */
  async getConfig() {
    const res = await this.http.get("/api/config");
    return res.data;
  }
  // ===========================================================================
  // Cloud API Keys (manage keys for Shogo Local → Cloud auth)
  // ===========================================================================
  /** List active API keys for a workspace. Pass `kind: 'device'` to fetch
   * device-session keys for the Devices UI, or `kind: 'user'` for manually
   * created keys. Omit to list both. */
  async listApiKeys(workspaceId, opts) {
    const query = { workspaceId };
    if (opts?.kind) query.kind = opts.kind;
    const res = await this.http.get(
      "/api/api-keys",
      query
    );
    return res.data?.keys ?? [];
  }
  /** Create a new "user" API key. Returns the full plaintext key (shown only once). */
  async createApiKey(name, workspaceId) {
    const res = await this.http.post(
      "/api/api-keys",
      { name, workspaceId }
    );
    return res.data;
  }
  /** Mint a device-session API key for the current user's device. Cloud API
   * deduplicates by (workspaceId, deviceId) — previous device keys for the
   * same machine are revoked automatically. */
  async createDeviceApiKey(device, opts) {
    const res = await this.http.post(
      "/api/api-keys/device",
      {
        workspaceId: opts?.workspaceId,
        deviceId: device.id,
        deviceName: device.name,
        devicePlatform: device.platform,
        deviceAppVersion: device.appVersion
      }
    );
    return res.data;
  }
  /** Revoke (soft-delete) an API key by ID. For device keys this is the
   * "sign out of device" operation. */
  async revokeApiKey(id) {
    await this.http.delete(`/api/api-keys/${id}`);
  }
  /** Validate an API key without requiring a session. */
  async validateApiKey(key) {
    const res = await this.http.post(
      "/api/api-keys/validate",
      { key }
    );
    return res.data;
  }
  // ===========================================================================
  // Local: Shogo Cloud API Key
  // ===========================================================================
  /** Get the current Shogo Cloud key connection status (local mode). */
  async getShogoKeyStatus() {
    const res = await this.http.get("/api/local/shogo-key");
    return res.data ?? { connected: false };
  }
  /**
   * Connect a Shogo Cloud API key (local mode).
   * Validates the key against the cloud before saving. The cloud endpoint
   * is determined by the API server's `SHOGO_CLOUD_URL` env var and cannot
   * be overridden by callers.
   */
  async connectShogoKey(key) {
    const res = await this.http.request(
      "/api/local/shogo-key",
      { method: "PUT", body: { key } }
    );
    return res.data ?? { ok: false };
  }
  /** Disconnect the Shogo Cloud API key (local mode). */
  async disconnectShogoKey() {
    await this.http.delete("/api/local/shogo-key");
  }
  // ===========================================================================
  // Cloud Login (poll-based device flow)
  // ===========================================================================
  //
  // The interactive sign-in flow is no longer driven through the SDK
  // because both real consumers — Shogo Desktop and the Shogo CLI worker —
  // need to talk to *both* the cloud and a local persistence layer, which
  // doesn't fit cleanly behind a single `this.http` base URL. Use the
  // dedicated implementations instead:
  //
  //   - Desktop: apps/desktop/src/main.ts → runCloudSignIn()
  //   - CLI:    packages/shogo-worker/src/lib/cloud-login.ts → runCloudLogin()
  //
  // Both drive the cloud `/api/cli/login/{start,poll,approve}` endpoints
  // directly (no protocol handler / no localhost listener) and then PUT
  // the minted key into `/api/local/shogo-key` (which validates the key
  // against cloud, writes localConfig, and restarts the instance tunnel).
  //
  // The `connectShogoKey` / `disconnectShogoKey` / `cloudLoginStatus` /
  // `signOutCloud` / `heartbeatCloudLogin` helpers below remain for the
  // CLI / headless / API-key paste paths and for the local-mode session
  // status the Settings UI needs.
  /** List workspaces the signed-in cloud user is a member of.
   * Used by the device-login bridge page (apps/mobile/app/auth/cli-link.tsx,
   * shared between desktop and CLI) to render a workspace picker when
   * the user has more than one membership. Requires an authenticated
   * cloud session (cookie); returns an empty list if unauthenticated. */
  async listMyWorkspaces() {
    const res = await this.http.get(
      "/api/workspaces"
    );
    return res.data?.items ?? [];
  }
  /** Read the current local-mode cloud login status. */
  async cloudLoginStatus() {
    const res = await this.http.get("/api/local/cloud-login/status");
    return res.data ?? { signedIn: false };
  }
  /** Sign out of Shogo Cloud on this device. Wipes the local key and best-
   * effort notifies cloud. */
  async signOutCloud() {
    const res = await this.http.post(
      "/api/local/cloud-login/signout",
      {}
    );
    return res.data ?? { ok: false };
  }
  /** Ping cloud with the stored key to refresh this device's lastSeenAt. */
  async heartbeatCloudLogin(deviceAppVersion) {
    const res = await this.http.post(
      "/api/local/cloud-login/heartbeat",
      deviceAppVersion ? { deviceAppVersion } : {}
    );
    return res.data ?? { ok: false };
  }
  // ===========================================================================
  // Local: LLM Configuration
  // ===========================================================================
  /** Get the local LLM provider configuration. */
  async getLlmConfig() {
    const res = await this.http.get("/api/local/llm-config");
    return res.data?.config ?? {};
  }
  /** Update the local LLM provider configuration. */
  async putLlmConfig(config) {
    await this.http.request("/api/local/llm-config", { method: "PUT", body: config });
  }
  // ===========================================================================
  // Local: Provider API Keys
  // ===========================================================================
  /** Get masked provider API keys keyed by provider id (e.g. `anthropic`,
   * `openai`, `google`, `openrouter`). Missing entries mean no key is set. */
  async getProviderKeyMasks() {
    const res = await this.http.get("/api/local/api-keys");
    return res.data?.keys ?? {};
  }
  /**
   * Save provider API keys. Pass a map keyed by provider id (`anthropic`,
   * `openai`, `google`, `openrouter`) — only provided entries are updated.
   * Pass `''` or `null` to clear a stored key.
   *
   * Legacy camelCase fields (`anthropicApiKey`, `openaiApiKey`,
   * `googleApiKey`) are still accepted for backwards-compatibility.
   */
  async putProviderKeys(keys) {
    await this.http.request("/api/local/api-keys", { method: "PUT", body: keys });
  }
  // ===========================================================================
  // Local: Models Discovery
  // ===========================================================================
  /** List models available from a local LLM provider. */
  async getLocalModels(baseUrl) {
    const res = await this.http.get(
      "/api/local/models",
      { baseUrl }
    );
    return res.data ?? { ok: false, models: [] };
  }
  /** Live catalog from OpenRouter (`/api/v1/models`). Requires an
   * `OPENROUTER_API_KEY` to be configured for user-specific pricing,
   * but works unauthenticated too. */
  async getOpenRouterModels() {
    const res = await this.http.get("/api/local/openrouter/models");
    return res.data ?? { ok: false, models: [] };
  }
  // ===========================================================================
  // Admin: Agent Model Defaults
  // ===========================================================================
  /** Get admin-configured overrides for basic/advanced agent mode models, the
   *  default mode, and the Auto-mode tier models (economy/standard/premium). */
  async getAgentModelDefaults() {
    const res = await this.http.get("/api/admin/settings/agent-models");
    return res.data ?? {
      basic: null,
      advanced: null,
      defaultMode: null,
      autoEconomy: null,
      autoStandard: null,
      autoPremium: null
    };
  }
  /** Set which models the basic/advanced agent modes resolve to, the default
   *  mode, and the per-tier Auto-mode models. Pass null to reset to platform
   *  default. Auto-tier values may be a public alias (e.g. `hoshi-1.0`). */
  async putAgentModelDefaults(overrides) {
    await this.http.request("/api/admin/settings/agent-models", { method: "PUT", body: overrides });
  }
  /** Get the admin-configured model used to generate chat/project titles.
   *  Returns `{ model: null }` when unset (platform default applies). */
  async getTitleGenerationModel() {
    const res = await this.http.get(
      "/api/admin/settings/title-generation-model"
    );
    return res.data ?? { model: null };
  }
  /** Set the model used to generate chat/project titles. Pass null/empty to
   *  reset to the platform default (Haiku). */
  async putTitleGenerationModel(model) {
    await this.http.request("/api/admin/settings/title-generation-model", {
      method: "PUT",
      body: { model }
    });
  }
  // ===========================================================================
  // Admin: Visible Models (catalog allowlist + curated OpenRouter models)
  // ===========================================================================
  /** Read the admin-configured visible-models config (allowlist + OR extras). */
  async getVisibleModelsConfig() {
    const res = await this.http.get(
      "/api/admin/settings/visible-models"
    );
    return res.data ?? { catalogIds: null, openrouterModels: [] };
  }
  /** Replace the visible-models config. Pass `catalogIds: null` to revert
   * to "show all catalog models". */
  async putVisibleModelsConfig(config) {
    await this.http.request("/api/admin/settings/visible-models", {
      method: "PUT",
      body: config
    });
  }
  /** Public read of the resolved visible-models config — this is the seam
   * user-facing chat input pickers should use to render their model list. */
  async getVisibleModels() {
    const res = await this.http.get(
      "/api/platform/visible-models"
    );
    return res.data ?? { catalogIds: null, openrouterModels: [] };
  }
  // ===========================================================================
  // Workspace-scoped visible models (owner/admin curated subset)
  // ===========================================================================
  /** Read the resolved visible-models set for a workspace: the platform-visible
   * set narrowed by the workspace admin's allowlist. `allowedModelIds` is the
   * raw allowlist (`null` = inherit all platform-visible models). Use this in
   * the chat picker when a workspace is active; any member may read it. */
  async getWorkspaceVisibleModels(workspaceId) {
    const res = await this.http.get(
      `/api/workspaces/${encodeURIComponent(workspaceId)}/visible-models`
    );
    return res.data ?? { catalogIds: null, openrouterModels: [], allowedModelIds: null };
  }
  /** Replace the workspace's model allowlist (owner/admin only). Pass `null` or
   * `[]` to revert to "inherit all platform-visible models". Every id must be in
   * the platform-visible set or the server rejects the write. */
  async putWorkspaceVisibleModels(workspaceId, allowedModelIds) {
    const res = await this.http.request(
      `/api/workspaces/${encodeURIComponent(workspaceId)}/visible-models`,
      { method: "PUT", body: { allowedModelIds } }
    );
    return res.data ?? { ok: false, allowedModelIds: null };
  }
  // ===========================================================================
  // Admin: DB-defined model catalog (custom providers + models)
  // ===========================================================================
  /** List custom model providers. Keys are returned masked, never in plaintext. */
  async listModelProviders() {
    const res = await this.http.get("/api/admin/settings/model-providers");
    return res.data?.providers ?? [];
  }
  /** Create a custom model provider. `apiKey` is required and stored encrypted. */
  async createModelProvider(input) {
    const res = await this.http.request(
      "/api/admin/settings/model-providers",
      { method: "POST", body: input }
    );
    return res.data.provider;
  }
  /** Update a custom model provider. Omit `apiKey` to keep the existing key. */
  async updateModelProvider(id, input) {
    const res = await this.http.request(
      `/api/admin/settings/model-providers/${encodeURIComponent(id)}`,
      { method: "PUT", body: input }
    );
    return res.data.provider;
  }
  /** Delete a custom model provider. Linked models have their providerId nulled. */
  async deleteModelProvider(id) {
    await this.http.request(`/api/admin/settings/model-providers/${encodeURIComponent(id)}`, { method: "DELETE" });
  }
  /** List DB-defined models. */
  async listModels() {
    const res = await this.http.get("/api/admin/settings/models");
    return res.data?.models ?? [];
  }
  /** Create a DB-defined model. */
  async createModel(input) {
    const res = await this.http.request(
      "/api/admin/settings/models",
      { method: "POST", body: input }
    );
    return res.data.model;
  }
  /** Update a DB-defined model. */
  async updateModel(id, input) {
    const res = await this.http.request(
      `/api/admin/settings/models/${encodeURIComponent(id)}`,
      { method: "PUT", body: input }
    );
    return res.data.model;
  }
  /** Delete a DB-defined model. */
  async deleteModel(id) {
    await this.http.request(`/api/admin/settings/models/${encodeURIComponent(id)}`, { method: "DELETE" });
  }
  /** List the public `/v1/*` API model aliases (e.g. `hoshi-1.0`). */
  async getPublicModels() {
    const res = await this.http.get("/api/admin/settings/public-models");
    return res.data?.models ?? [];
  }
  /** Replace the public model alias map. Each `backingModelId` must resolve to
   * a known model or the server rejects the whole write. */
  async putPublicModels(models) {
    const res = await this.http.request(
      "/api/admin/settings/public-models",
      { method: "PUT", body: { models } }
    );
    return res.data?.models ?? [];
  }
  // ===========================================================================
  // Admin: Native provider keys + live model discovery
  // ===========================================================================
  /** Read masked native provider keys (anthropic / openai / google /
   * openrouter), with how each is configured (`db` = admin-entered & encrypted,
   * `env` = server env var). Keys are never returned in plaintext. */
  async getAdminProviderKeyMasks() {
    const res = await this.http.get("/api/admin/settings/provider-keys");
    return res.data?.keys ?? {};
  }
  /** Store native provider keys (encrypted at rest). Pass a map keyed by
   * provider id; `''`/`null` clears a stored key (env fallback then applies). */
  async putAdminProviderKeys(keys) {
    const res = await this.http.request("/api/admin/settings/provider-keys", { method: "PUT", body: keys });
    return res.data?.keys ?? {};
  }
  /** Live model list from a native provider (anthropic / openai / openrouter),
   * discovered with the configured key. OpenRouter's catalog is public, so it
   * lists even without a key. */
  async getProviderModels(provider) {
    const res = await this.http.get(`/api/admin/settings/providers/${encodeURIComponent(provider)}/models`);
    return res.data ?? { ok: false, models: [] };
  }
  /** Persist which discovered provider models are enabled. Each entry upserts a
   * DB model row (enable) or flips it disabled (enable: false). */
  async setProviderModelsEnabled(provider, models) {
    await this.http.request(
      `/api/admin/settings/providers/${encodeURIComponent(provider)}/models/enable`,
      { method: "POST", body: { models } }
    );
  }
  // ===========================================================================
  // Admin: Model pricing (LiteLLM-sourced)
  // ===========================================================================
  /** When token prices were last refreshed from LiteLLM, and whether the
   * daily TTL has elapsed. */
  async getPricingStatus() {
    const res = await this.http.get(
      "/api/admin/settings/pricing/status"
    );
    return res.data ?? { refreshedAt: null, stale: true };
  }
  /** Refresh per-token pricing + context window on DB models from LiteLLM.
   * `force` refreshes regardless of TTL; otherwise it no-ops when still fresh. */
  async refreshModelPricing(force = false) {
    const res = await this.http.request("/api/admin/settings/pricing/refresh", { method: "POST", body: { force } });
    return res.data ?? { ok: false, refreshedAt: null, updated: 0, total: 0 };
  }
  // ===========================================================================
  // Admin: Feature Flags
  // ===========================================================================
  /** Read super-admin feature flag overrides. `null` means "use platform default". */
  async getFeatureFlags() {
    const res = await this.http.get("/api/admin/settings/features");
    return res.data ?? { marketplace: null, ezMode: null, phoneChannel: null };
  }
  /** Update feature flag overrides. Pass `null` for a flag to reset to platform default. */
  async putFeatureFlags(patch) {
    const res = await this.http.request(
      "/api/admin/settings/features",
      { method: "PUT", body: patch }
    );
    return res.data ?? { ok: false, flags: { marketplace: null, ezMode: null, phoneChannel: null } };
  }
  // ===========================================================================
  // Local: Instance Info
  // ===========================================================================
  /** Get local instance registration info (machine name, tunnel status, etc.) */
  async getInstanceInfo() {
    const res = await this.http.get("/api/local/instance-info");
    return res.data;
  }
  /** Update the local instance display name. Restarts the tunnel to re-register. */
  async updateInstanceName(name) {
    const res = await this.http.request(
      "/api/local/instance-name",
      { method: "PUT", body: { name } }
    );
    return res.data ?? { ok: false };
  }
};

// src/machines/index.ts
var MachinesApi = class {
  constructor(http) {
    this.http = http;
  }
  /** List all paired machines (desktop + worker) for a workspace. */
  async list(opts) {
    const res = await this.http.get("/api/instances", {
      workspaceId: opts.workspaceId
    });
    return res.data?.instances ?? [];
  }
  /** List only currently-online machines (trimmed shape; used by pickers). */
  async listOnline(opts) {
    const res = await this.http.get("/api/instances/online", {
      workspaceId: opts.workspaceId
    });
    return res.data?.instances ?? [];
  }
  /** Fetch a single machine by ID. */
  async get(id) {
    const res = await this.http.get(`/api/instances/${encodeURIComponent(id)}`);
    return res.data ?? null;
  }
  /**
   * Read a project's current pin. Returns `preferredInstanceId: null`
   * when the project is cloud-routed (the default).
   */
  async getProjectPin(projectId) {
    const res = await this.http.get(
      `/api/projects/${encodeURIComponent(projectId)}/preferred-instance`
    );
    return res.data ?? {
      preferredInstanceId: null,
      preferredInstancePolicy: "pinned",
      instance: null
    };
  }
  /**
   * Pin a project to a specific paired machine. Future external-trigger
   * requests to `/api/projects/:projectId/agent-proxy/*` will be relayed
   * through that machine's outbound tunnel instead of a cloud pod.
   */
  async pinProject(projectId, opts) {
    const body = {
      instanceId: opts.instanceId
    };
    if (opts.policy) body.policy = opts.policy;
    const res = await this.http.request(
      `/api/projects/${encodeURIComponent(projectId)}/preferred-instance`,
      { method: "PUT", body }
    );
    if (!res.data) throw new Error("Failed to pin project");
    return res.data;
  }
  /** Clear the project's pin, restoring cloud-pod routing. */
  async unpinProject(projectId) {
    await this.http.delete(
      `/api/projects/${encodeURIComponent(projectId)}/preferred-instance`
    );
  }
};

// src/projects/index.ts
var ProjectsApi = class {
  /**
   * Note: the API key isn't accessible from {@link HttpClient} directly
   * (it uses a token getter). For the bulk transport routes we accept an
   * explicit `apiKey` on pull/push options, falling back to whatever the
   * HttpClient already has wired up.
   */
  constructor(http, resolveApiKey, resolveApiUrl) {
    this.http = http;
    this.resolveApiKey = resolveApiKey;
    this.resolveApiUrl = resolveApiUrl;
  }
  /** Fetch the workspace manifest (full tree, server-filtered). */
  async manifest(projectId) {
    const res = await this.http.get(
      `/api/projects/${encodeURIComponent(projectId)}/workspace/manifest`
    );
    return res.data?.files ?? [];
  }
  /** List files via the Studio-style endpoint (filtered to source extensions). */
  async listFiles(projectId) {
    const res = await this.http.get(
      `/api/projects/${encodeURIComponent(projectId)}/s3/files`
    );
    return res.data?.files ?? [];
  }
  /**
   * Read a single file's contents as a string. Convenience wrapper over
   * `GET /api/projects/:id/files/<path>` for ad-hoc reads.
   */
  async readFile(projectId, path) {
    const res = await this.http.get(
      `/api/projects/${encodeURIComponent(projectId)}/files/${path}`
    );
    return res.data?.content ?? "";
  }
  /** Write a file via `PUT /api/projects/:id/files/<path>`. */
  async writeFile(projectId, path, content) {
    await this.http.request(
      `/api/projects/${encodeURIComponent(projectId)}/files/${path}`,
      { method: "PUT", body: { content } }
    );
  }
  /** Delete a remote file. */
  async deleteFile(projectId, path) {
    await this.http.delete(
      `/api/projects/${encodeURIComponent(projectId)}/files/${path}`
    );
  }
  /**
   * Clone a project's workspace from cloud → local.
   *
   * Atomic: writes into `<into>.shogo-pull-tmp/` and renames over `<into>`
   * on success.
   */
  async pull(projectId, opts = {}) {
    const transport = this.makeTransport(projectId, opts.into ?? `./${projectId}`, opts);
    return transport.downloadAll();
  }
  /**
   * Push local edits back to staging. Mirrors deletes when
   * `opts.deleteRemote` is true.
   */
  async push(projectId, opts = {}) {
    const transport = this.makeTransport(projectId, opts.from ?? `./${projectId}`, opts);
    return transport.uploadAll({ deleteRemote: opts.deleteRemote });
  }
  /**
   * Build a {@link CloudFileTransport} for advanced workflows (e.g. the
   * worker's auto-pull hook or a long-running watcher). The transport
   * inherits the same API key + base URL resolution as `pull` / `push`.
   */
  transport(projectId, localDir, opts = {}) {
    return this.makeTransport(projectId, localDir, opts);
  }
  makeTransport(projectId, localDir, opts) {
    const apiKey = opts.apiKey ?? this.resolveApiKey();
    const apiUrl = opts.apiUrl ?? this.resolveApiUrl();
    if (!apiKey) {
      throw new Error(
        "ProjectsApi: no API key configured. Pass `apiKey` to pull/push, or sign in with a `shogo_sk_*` key first."
      );
    }
    if (!apiUrl) {
      throw new Error("ProjectsApi: no apiUrl configured on the client");
    }
    return new CloudFileTransport({
      apiUrl,
      apiKey,
      projectId,
      localDir,
      include: opts.include,
      fetchImpl: opts.fetchImpl,
      fs: opts.fs,
      onProgress: opts.onProgress
    });
  }
};

// src/storage/adapter.ts
function isBrowser() {
  return typeof window !== "undefined" && typeof window.localStorage !== "undefined";
}
function getDefaultStorageAdapter() {
  if (isBrowser()) {
    return new WebStorageAdapter();
  }
  return new NoOpStorageAdapter();
}
var WebStorageAdapter = class {
  constructor(prefix = "shogo_") {
    this.prefix = prefix;
  }
  getKey(key) {
    return `${this.prefix}${key}`;
  }
  getItem(key) {
    try {
      return localStorage.getItem(this.getKey(key));
    } catch {
      return null;
    }
  }
  setItem(key, value) {
    try {
      localStorage.setItem(this.getKey(key), value);
    } catch {
    }
  }
  removeItem(key) {
    try {
      localStorage.removeItem(this.getKey(key));
    } catch {
    }
  }
  clear() {
    try {
      const keysToRemove = [];
      for (let i = 0; i < localStorage.length; i++) {
        const key = localStorage.key(i);
        if (key?.startsWith(this.prefix)) {
          keysToRemove.push(key);
        }
      }
      keysToRemove.forEach((key) => localStorage.removeItem(key));
    } catch {
    }
  }
};
var AsyncStorageAdapter = class {
  constructor(asyncStorage, prefix = "shogo_") {
    this.asyncStorage = asyncStorage;
    this.prefix = prefix;
  }
  getKey(key) {
    return `${this.prefix}${key}`;
  }
  async getItem(key) {
    try {
      return await this.asyncStorage.getItem(this.getKey(key));
    } catch {
      return null;
    }
  }
  async setItem(key, value) {
    try {
      await this.asyncStorage.setItem(this.getKey(key), value);
    } catch {
    }
  }
  async removeItem(key) {
    try {
      await this.asyncStorage.removeItem(this.getKey(key));
    } catch {
    }
  }
  async clear() {
    if (this.asyncStorage.clear) {
      try {
        await this.asyncStorage.clear();
      } catch {
      }
    }
  }
};
var NoOpStorageAdapter = class {
  getItem(_key) {
    return null;
  }
  setItem(_key, _value) {
  }
  removeItem(_key) {
  }
  clear() {
  }
};
var MemoryStorageAdapter = class {
  constructor() {
    this.store = /* @__PURE__ */ new Map();
  }
  getItem(key) {
    return this.store.get(key) ?? null;
  }
  setItem(key, value) {
    this.store.set(key, value);
  }
  removeItem(key) {
    this.store.delete(key);
  }
  clear() {
    this.store.clear();
  }
};

// src/llm/index.ts
import { createOpenAICompatible } from "@ai-sdk/openai-compatible";
var DEFAULT_SHOGO_CLOUD_URL = "https://studio.shogo.ai";
function createShogoLlmProvider(opts) {
  const hasApiKey = Boolean(opts.apiKey);
  const hasRuntimeToken = Boolean(opts.runtimeToken);
  if (hasApiKey === hasRuntimeToken) {
    throw new Error(
      hasApiKey ? "createShogoLlmProvider: pass exactly one of `apiKey` or `runtimeToken`, not both." : "createShogoLlmProvider: one of `apiKey` or `runtimeToken` is required."
    );
  }
  const root = (opts.baseUrl ?? DEFAULT_SHOGO_CLOUD_URL).replace(/\/$/, "");
  return createOpenAICompatible({
    name: "shogo",
    baseURL: `${root}/api/ai/v1`,
    // Both credentials ride as `Authorization: Bearer <token>`. The Shogo
    // AI proxy distinguishes by prefix server-side: `shogo_sk_*` →
    // workspace API key, `rt_v1_*` → per-project runtime token.
    apiKey: opts.apiKey ?? opts.runtimeToken,
    headers: opts.headers,
    fetch: opts.fetch,
    includeUsage: opts.includeUsage ?? true
  });
}

// src/client.ts
import {
  DirectTelephonyClient,
  HostedRuntimeTokenClient,
  HostedTelephonyClient,
  MockTelephonyClient,
  isVoiceMockEnv
} from "@shogo-ai/voice";
var ShogoClientImpl = class {
  constructor(config) {
    this.config = config;
    const storage = config.storage ?? getDefaultStorageAdapter();
    this._http = new HttpClient({
      baseUrl: config.apiUrl,
      mcpPath: "/mcp",
      authPath: config.auth?.authPath ?? "/api/auth",
      projectId: config.projectId
    });
    this.auth = new ShogoAuth(this._http, storage, config.auth);
    this.platform = new PlatformApi(this._http);
    this.machines = new MachinesApi(this._http);
    this.shogoApiKey = config.shogoApiKey ?? null;
    this.projects = new ProjectsApi(
      this._http,
      () => this.shogoApiKey,
      () => config.apiUrl
    );
    this._http.setTokenGetter(() => this.auth.getToken());
    this.db = config.db;
    this.shogoCloudUrl = config.shogoCloudUrl;
    this.llm = this.buildLlm(config.shogoApiKey ?? null);
    this.voice = { telephony: this.buildTelephony(config.shogoApiKey) };
  }
  buildLlm(shogoApiKey) {
    const runtimeToken = typeof process !== "undefined" ? process.env?.RUNTIME_AUTH_SECRET : void 0;
    const hasRuntime = Boolean(runtimeToken);
    const hasApiKey = Boolean(shogoApiKey);
    if (hasRuntime && hasApiKey) {
      console.warn(
        "[shogo] createClient received both RUNTIME_AUTH_SECRET env + shogoApiKey for client.llm; using runtime-token (pod-native). Drop shogoApiKey to silence this warning."
      );
    }
    if (hasRuntime) {
      return createShogoLlmProvider({
        runtimeToken,
        baseUrl: this.shogoCloudUrl
      });
    }
    if (hasApiKey) {
      return createShogoLlmProvider({
        apiKey: shogoApiKey,
        baseUrl: this.shogoCloudUrl
      });
    }
    return null;
  }
  buildTelephony(shogoApiKey) {
    const { projectId, elevenlabs, twilio, apiUrl } = this.config;
    if (isVoiceMockEnv()) {
      return new MockTelephonyClient({ projectId: projectId ?? void 0 });
    }
    const runtimeToken = typeof process !== "undefined" ? process.env?.RUNTIME_AUTH_SECRET : void 0;
    const hasRuntime = Boolean(runtimeToken && projectId);
    const hasHosted = Boolean(shogoApiKey && projectId);
    const hasDirect = Boolean(elevenlabs && twilio);
    if (hasRuntime && hasHosted) {
      console.warn(
        "[shogo] createClient received both RUNTIME_AUTH_SECRET env + shogoApiKey; using runtime-token (pod-native). Drop shogoApiKey to silence this warning."
      );
    } else if (hasHosted && hasDirect) {
      console.warn(
        "[shogo] createClient received both shogoApiKey + direct elevenlabs/twilio creds; using hosted (Mode B). Drop shogoApiKey to use direct (Mode A)."
      );
    }
    if (hasRuntime) {
      return new HostedRuntimeTokenClient({
        runtimeToken,
        projectId,
        apiUrl
      });
    }
    if (hasHosted) {
      return new HostedTelephonyClient({
        shogoApiKey,
        projectId,
        apiUrl
      });
    }
    if (hasDirect) {
      return new DirectTelephonyClient({
        projectId,
        elevenlabs,
        twilio
      });
    }
    return null;
  }
  setShogoApiKey(key) {
    this.shogoApiKey = key;
    this.llm = this.buildLlm(key);
    this.voice = { telephony: this.buildTelephony(key) };
  }
};
var defaultClient = null;
function createClient(config) {
  const client = new ShogoClientImpl(config);
  defaultClient = client;
  return client;
}
function getDefaultClient() {
  if (!defaultClient) {
    throw new Error(
      "No default client. Call createClient() first."
    );
  }
  return defaultClient;
}

// src/attachments/data-url.ts
function buildDataUrlFromBase64(mimeType, base64) {
  const mime = mimeType.trim() || "application/octet-stream";
  return `data:${mime};base64,${base64}`;
}

// src/stores/optimistic-store.ts
import { makeAutoObservable, runInAction } from "mobx";
var OptimisticStore = class {
  constructor(config) {
    /** All items in the store */
    this.items = /* @__PURE__ */ new Map();
    /** Loading state */
    this.isLoading = false;
    /** Last error message */
    this.error = null;
    /** Items currently being created (temp ID -> optimistic item) */
    this.pendingCreates = /* @__PURE__ */ new Map();
    /** Items currently being updated (ID -> previous state for rollback) */
    this.pendingUpdates = /* @__PURE__ */ new Map();
    /** Items currently being deleted */
    this.pendingDeletes = /* @__PURE__ */ new Set();
    this.http = config.http;
    this.endpoint = config.endpoint;
    this.transform = config.transform ?? ((x) => x);
    makeAutoObservable(this, {
      http: false,
      endpoint: false,
      transform: false
    });
  }
  // ==========================================================================
  // Getters
  // ==========================================================================
  /** Get all items as array */
  get all() {
    return Array.from(this.items.values());
  }
  /** Get item by ID */
  get(id) {
    return this.items.get(id);
  }
  /** Check if an item has a pending operation */
  isPending(id) {
    return this.pendingCreates.has(id) || this.pendingUpdates.has(id) || this.pendingDeletes.has(id);
  }
  /** Check if item is being created */
  isCreating(id) {
    return this.pendingCreates.has(id);
  }
  /** Check if item is being updated */
  isUpdating(id) {
    return this.pendingUpdates.has(id);
  }
  /** Check if item is being deleted */
  isDeleting(id) {
    return this.pendingDeletes.has(id);
  }
  // ==========================================================================
  // Read Operations
  // ==========================================================================
  /**
   * Load all items from the API
   */
  async loadAll(params) {
    runInAction(() => {
      this.isLoading = true;
      this.error = null;
    });
    try {
      const { data } = await this.http.get(
        this.endpoint,
        params
      );
      if (!data?.ok || !data.items) {
        throw new Error("Invalid API response");
      }
      const items = data.items.map(this.transform);
      runInAction(() => {
        this.items.clear();
        for (const item of items) {
          this.items.set(item.id, item);
        }
        this.isLoading = false;
      });
      return items;
    } catch (error) {
      runInAction(() => {
        this.error = error instanceof Error ? error.message : "Failed to load";
        this.isLoading = false;
      });
      throw error;
    }
  }
  /**
   * Load a single item by ID
   */
  async loadById(id) {
    try {
      const { data } = await this.http.get(
        `${this.endpoint}/${id}`
      );
      if (!data?.ok || !data.data) {
        throw new Error("Item not found");
      }
      const item = this.transform(data.data);
      runInAction(() => {
        this.items.set(item.id, item);
      });
      return item;
    } catch (error) {
      runInAction(() => {
        this.error = error instanceof Error ? error.message : "Failed to load";
      });
      throw error;
    }
  }
  // ==========================================================================
  // Create (Optimistic)
  // ==========================================================================
  /**
   * Create a new item with optimistic update.
   * Item appears immediately, rolls back on failure.
   */
  async create(input) {
    const tempId = `temp-${crypto.randomUUID()}`;
    const now = /* @__PURE__ */ new Date();
    const optimistic = {
      ...input,
      id: tempId,
      createdAt: now,
      updatedAt: now
    };
    runInAction(() => {
      this.pendingCreates.set(tempId, optimistic);
      this.items.set(tempId, optimistic);
      this.error = null;
    });
    try {
      const { data } = await this.http.post(
        this.endpoint,
        input
      );
      if (!data?.ok || !data.data) {
        throw new Error("Create failed");
      }
      const item = this.transform(data.data);
      runInAction(() => {
        this.pendingCreates.delete(tempId);
        this.items.delete(tempId);
        this.items.set(item.id, item);
      });
      return item;
    } catch (error) {
      runInAction(() => {
        this.pendingCreates.delete(tempId);
        this.items.delete(tempId);
        this.error = error instanceof Error ? error.message : "Create failed";
      });
      throw error;
    }
  }
  // ==========================================================================
  // Update (Optimistic)
  // ==========================================================================
  /**
   * Update an item with optimistic update.
   * Changes appear immediately, roll back on failure.
   */
  async update(id, changes) {
    if (!id || typeof id !== "string") {
      console.error("[OptimisticStore] update called with invalid id:", id);
      throw new Error("Invalid ID");
    }
    const existing = this.items.get(id);
    if (!existing) {
      throw new Error("Item not found");
    }
    if (this.pendingUpdates.has(id)) {
      throw new Error("Update already in progress");
    }
    const previousState = { ...existing };
    runInAction(() => {
      this.pendingUpdates.set(id, previousState);
      this.items.set(id, {
        ...existing,
        ...changes,
        updatedAt: /* @__PURE__ */ new Date()
      });
      this.error = null;
    });
    try {
      const { data } = await this.http.patch(
        `${this.endpoint}/${id}`,
        changes
      );
      if (!data?.ok || !data.data) {
        throw new Error("Update failed");
      }
      const item = this.transform(data.data);
      runInAction(() => {
        this.pendingUpdates.delete(id);
        this.items.set(id, item);
      });
      return item;
    } catch (error) {
      runInAction(() => {
        this.pendingUpdates.delete(id);
        this.items.set(id, previousState);
        this.error = error instanceof Error ? error.message : "Update failed";
      });
      throw error;
    }
  }
  // ==========================================================================
  // Delete (Optimistic)
  // ==========================================================================
  /**
   * Delete an item with optimistic update.
   * Item disappears immediately, reappears on failure.
   */
  async delete(id) {
    if (!id || typeof id !== "string") {
      console.error("[OptimisticStore] delete called with invalid id:", id);
      throw new Error("Invalid ID");
    }
    const existing = this.items.get(id);
    if (!existing) {
      return;
    }
    if (this.pendingDeletes.has(id)) {
      return;
    }
    runInAction(() => {
      this.pendingDeletes.add(id);
      this.items.delete(id);
      this.error = null;
    });
    try {
      await this.http.delete(`${this.endpoint}/${id}`);
      runInAction(() => {
        this.pendingDeletes.delete(id);
      });
    } catch (error) {
      runInAction(() => {
        this.pendingDeletes.delete(id);
        this.items.set(id, existing);
        this.error = error instanceof Error ? error.message : "Delete failed";
      });
      throw error;
    }
  }
  // ==========================================================================
  // Utilities
  // ==========================================================================
  /** Clear error state */
  clearError() {
    this.error = null;
  }
  /** Clear all data */
  clear() {
    this.items.clear();
    this.pendingCreates.clear();
    this.pendingUpdates.clear();
    this.pendingDeletes.clear();
    this.error = null;
    this.isLoading = false;
  }
  /**
   * Add an item directly to the store (for external updates like WebSocket)
   */
  addItem(item) {
    this.items.set(item.id, item);
  }
  /**
   * Remove an item directly from the store (for external updates like WebSocket)
   */
  removeItem(id) {
    this.items.delete(id);
  }
  /**
   * Update an item directly in the store (for external updates like WebSocket)
   */
  updateItem(id, changes) {
    const existing = this.items.get(id);
    if (existing) {
      this.items.set(id, { ...existing, ...changes });
    }
  }
};

// src/persistence/APIPersistence.ts
function toRoutePath(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
var APIPersistence = class {
  constructor(config) {
    this.userId = null;
    this.http = config.http;
    this.basePath = config.basePath ?? "/api";
    this.routeMap = config.routeMap ?? {};
  }
  /**
   * Set the current user ID for user-scoped queries.
   */
  setUserId(userId) {
    if (this.userId !== userId) {
      this.http.clearCache();
    }
    this.userId = userId;
  }
  // =========================================================================
  // Collection Operations
  // =========================================================================
  /**
   * Load a collection via REST API.
   */
  async loadCollection(ctx) {
    try {
      const endpoint = this.getCollectionEndpoint(ctx);
      if (!endpoint) {
        console.warn(`[APIPersistence] No endpoint for model: ${ctx.modelName}`);
        return null;
      }
      const response = await this.http.get(endpoint);
      if (!response.data?.ok || !response.data.items) {
        return null;
      }
      const items = {};
      for (const item of response.data.items) {
        if (item.id) {
          items[item.id] = this.transformForMST(item);
        }
      }
      return { items };
    } catch (error) {
      console.error("[APIPersistence] loadCollection error:", error.message);
      return null;
    }
  }
  /**
   * Save a collection via REST API.
   */
  async saveCollection(ctx, snapshot) {
    const items = snapshot.items || {};
    for (const [id, entity] of Object.entries(items)) {
      try {
        await this.saveEntity({ ...ctx, entityId: id }, entity);
      } catch (error) {
        console.error(`[APIPersistence] saveCollection failed for ${id}:`, error.message);
      }
    }
  }
  // =========================================================================
  // Entity Operations
  // =========================================================================
  /**
   * Load a single entity via REST API.
   */
  async loadEntity(ctx) {
    try {
      const endpoint = this.getEntityEndpoint(ctx);
      if (!endpoint) {
        console.warn(`[APIPersistence] No endpoint for model: ${ctx.modelName}`);
        return null;
      }
      const response = await this.http.get(endpoint);
      if (!response.data?.ok || !response.data.data) {
        return null;
      }
      return this.transformForMST(response.data.data);
    } catch (error) {
      if (error.message?.includes("not found") || error.status === 404) {
        return null;
      }
      console.error("[APIPersistence] loadEntity error:", error.message);
      return null;
    }
  }
  /**
   * Save a single entity via REST API.
   */
  async saveEntity(ctx, snapshot) {
    try {
      const existing = await this.loadEntity(ctx);
      const baseEndpoint = this.getCollectionEndpoint(ctx);
      if (!baseEndpoint) {
        throw new Error(`No endpoint for model: ${ctx.modelName}`);
      }
      const apiData = this.transformForAPI(snapshot);
      if (existing) {
        await this.http.patch(`${baseEndpoint}/${ctx.entityId}`, apiData);
      } else {
        await this.http.post(baseEndpoint, { ...apiData, id: ctx.entityId });
      }
    } catch (error) {
      console.error("[APIPersistence] saveEntity error:", error.message);
      throw error;
    }
  }
  // =========================================================================
  // Schema Operations (not supported with Prisma)
  // =========================================================================
  async loadSchema(_name, _location) {
    console.debug("[APIPersistence] loadSchema not supported with Prisma backend");
    return null;
  }
  async listSchemas(_location) {
    console.debug("[APIPersistence] listSchemas not supported with Prisma backend");
    return [];
  }
  // =========================================================================
  // Endpoint Mapping
  // =========================================================================
  /**
   * Get route path for a model name
   */
  getRoutePath(modelName) {
    if (this.routeMap[modelName]) {
      return this.routeMap[modelName];
    }
    return toRoutePath(modelName);
  }
  /**
   * Get collection endpoint URL
   */
  getCollectionEndpoint(ctx) {
    const routePath = this.getRoutePath(ctx.modelName);
    if (!routePath) return null;
    const params = new URLSearchParams();
    if (ctx.filter) {
      for (const [key, value] of Object.entries(ctx.filter)) {
        if (value !== void 0 && value !== null) {
          params.set(key, String(value));
        }
      }
    }
    if (this.userId) {
      const userScopedModels = ["Workspace", "StarredProject", "Notification"];
      if (userScopedModels.includes(ctx.modelName)) {
        params.set("userId", this.userId);
      }
    }
    const queryString = params.toString();
    return `${this.basePath}/${routePath}${queryString ? `?${queryString}` : ""}`;
  }
  /**
   * Get entity endpoint URL
   */
  getEntityEndpoint(ctx) {
    const routePath = this.getRoutePath(ctx.modelName);
    if (!routePath) return null;
    return `${this.basePath}/${routePath}/${encodeURIComponent(ctx.entityId)}`;
  }
  // =========================================================================
  // Data Transformations
  // =========================================================================
  /**
   * Transform API response for MST compatibility
   */
  transformForMST(obj) {
    if (!obj || typeof obj !== "object") return obj;
    const dateFields = ["createdAt", "updatedAt", "expiresAt", "publishedAt", "readAt", "emailSentAt", "lastActiveAt"];
    const referenceFields = ["workspaceId", "projectId", "folderId", "parentId", "userId"];
    const referenceObjectFields = ["workspace", "project", "folder", "parent", "user"];
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
      if (value === null) continue;
      if (referenceFields.includes(key) && typeof value === "string") {
        const refName = key.replace(/Id$/, "");
        if (!(refName in obj)) {
          result[refName] = value;
        }
        result[key] = value;
      } else if (referenceObjectFields.includes(key) && value && typeof value === "object" && "id" in value) {
        result[key] = value.id;
      } else if (dateFields.includes(key) && typeof value === "string") {
        result[key] = new Date(value).getTime();
      } else if (value && typeof value === "object" && !Array.isArray(value)) {
        result[key] = this.transformForMST(value);
      } else {
        result[key] = value;
      }
    }
    return result;
  }
  /**
   * Transform MST snapshot for API compatibility
   */
  transformForAPI(obj) {
    if (!obj || typeof obj !== "object") return obj;
    const dateFields = ["createdAt", "updatedAt", "expiresAt", "publishedAt", "readAt", "emailSentAt", "lastActiveAt"];
    const referenceFields = ["workspace", "project", "folder", "parent", "user", "session", "chatSession"];
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
      if (value === void 0) continue;
      if (dateFields.includes(key) && typeof value === "number") {
        result[key] = new Date(value).toISOString();
      } else if (referenceFields.includes(key) && typeof value === "string") {
        result[`${key}Id`] = value;
      } else if (value && typeof value === "object" && !Array.isArray(value)) {
        result[key] = this.transformForAPI(value);
      } else {
        result[key] = value;
      }
    }
    return result;
  }
};
export {
  APIPersistence,
  AsyncStorageAdapter,
  AuthError,
  BYOK_PROVIDERS,
  CloudFileTransport,
  DEFAULT_SHOGO_CLOUD_URL,
  DatabaseError,
  DomainProvider,
  HttpClient,
  MachinesApi,
  MemoryStorageAdapter,
  NoOpStorageAdapter,
  OptimisticStore,
  PlatformApi,
  ProjectsApi,
  ShogoAuth,
  ShogoError,
  WebStorageAdapter,
  buildDataUrlFromBase64,
  createClient,
  createShogoLlmProvider,
  getDefaultClient,
  getDefaultStorageAdapter,
  isBrowser,
  useCollection,
  useDomain,
  useDomainReady,
  withDomain
};
