// @shogo-ai/cli/dist/chunk-L635SMF7.js
// src/pkg.ts
import { execSync as nodeExecSync, spawn } from "child_process";
import { existsSync, readlinkSync, lstatSync } from "fs";
import { isAbsolute, join, resolve as resolvePath, dirname } from "path";
var NodeMissingError = class extends Error {
  constructor() {
    super(
      "Shogo Desktop on Windows requires Node.js 20+ to run project sandboxes. npm.cmd was not found on PATH and is not present at C:\\Program Files\\nodejs\\. Install Node.js from https://nodejs.org/ (LTS recommended), then restart Shogo."
    );
    this.code = "NODE_NOT_INSTALLED";
    this.name = "NodeMissingError";
  }
};
var WINDOWS_NODE_DIR = "C:\\Program Files\\nodejs";
function isNodeAvailableOnWindows(pathEnv = process.env.PATH) {
  if (process.platform !== "win32") return true;
  if (existsSync(join(WINDOWS_NODE_DIR, "npm.cmd"))) return true;
  if (!pathEnv) return false;
  for (const dir of pathEnv.split(";")) {
    if (dir && existsSync(join(dir, "npm.cmd"))) return true;
  }
  return false;
}
var _unixNodeAvailable = null;
function isNodeAvailableOnUnix(pathEnv = process.env.PATH) {
  if (process.platform === "win32") return false;
  if (pathEnv === process.env.PATH && _unixNodeAvailable !== null) return _unixNodeAvailable;
  let found = false;
  if (pathEnv) {
    for (const dir of pathEnv.split(":")) {
      if (!dir) continue;
      if (existsSync(join(dir, "node"))) {
        found = true;
        break;
      }
    }
  }
  if (pathEnv === process.env.PATH) _unixNodeAvailable = found;
  return found;
}
function _resetUnixNodeCache() {
  _unixNodeAvailable = null;
}
function resolveBinInvocation(workspaceDir, binName) {
  const binDir = join(workspaceDir, "node_modules", ".bin");
  const isWindows = process.platform === "win32";
  const candidates = isWindows ? [join(binDir, `${binName}.CMD`), join(binDir, `${binName}.cmd`), join(binDir, `${binName}.exe`)] : [join(binDir, binName)];
  const shim = candidates.find((p) => existsSync(p));
  if (!shim) return null;
  if (isWindows) return { cmd: shim, argsPrefix: [] };
  if (isNodeAvailableOnUnix()) return { cmd: shim, argsPrefix: [] };
  try {
    const st = lstatSync(shim);
    if (!st.isSymbolicLink()) return { cmd: shim, argsPrefix: [] };
    const target = readlinkSync(shim);
    const jsEntry = isAbsolute(target) ? target : resolvePath(dirname(shim), target);
    if (!existsSync(jsEntry)) return { cmd: shim, argsPrefix: [] };
    const bun = process.env.SHOGO_BUN_PATH || "bun";
    return { cmd: bun, argsPrefix: [jsEntry] };
  } catch {
    return { cmd: shim, argsPrefix: [] };
  }
}
var IS_WINDOWS = process.platform === "win32";
var DEFAULT_INSTALL_TIMEOUT = parseInt(
  process.env.SHOGO_INSTALL_TIMEOUT_MS || (IS_WINDOWS ? "300000" : "300000"),
  10
);
var DEFAULT_EXEC_TIMEOUT = 6e4;
var PlatformPackageManager = class {
  constructor() {
    this.isWindows = IS_WINDOWS;
  }
  /** Resolved path to the bun binary — prefers SHOGO_BUN_PATH (set by desktop app) over bare `bun`. */
  get bunBinary() {
    return process.env.SHOGO_BUN_PATH || "bun";
  }
  shellOpt() {
    return IS_WINDOWS ? true : void 0;
  }
  spawnEnv(base) {
    const source = base ?? process.env;
    const env = {};
    let pathValue = "";
    for (const [k, v] of Object.entries(source)) {
      if (v === void 0) continue;
      if (IS_WINDOWS && k.toLowerCase() === "path") {
        if (v) pathValue = v;
        continue;
      }
      env[k] = v;
    }
    if (IS_WINDOWS) {
      const nodePath = "C:\\Program Files\\nodejs";
      env.PATH = pathValue.includes(nodePath) ? pathValue : `${nodePath};${pathValue}`;
    }
    return env;
  }
  // ---------------------------------------------------------------------------
  // Install — synchronous
  // ---------------------------------------------------------------------------
  installSync(cwd, opts) {
    const timeout = opts?.timeout ?? DEFAULT_INSTALL_TIMEOUT;
    const stdio = opts?.stdio ?? "pipe";
    const env = this.spawnEnv(opts?.env);
    if (IS_WINDOWS) {
      if (isNodeAvailableOnWindows(env.PATH)) {
        try {
          nodeExecSync("npm.cmd install --loglevel=error", {
            cwd,
            timeout,
            stdio,
            env,
            shell: "cmd.exe"
          });
          return;
        } catch (err) {
          const wrapped = wrapWindowsNpmError(err);
          if (wrapped) {
          } else if (isNpmEresolveError(err)) {
            console.warn(
              "[platform-pkg] npm install hit ERESOLVE \u2014 retrying with --legacy-peer-deps."
            );
            nodeExecSync("npm.cmd install --loglevel=error --legacy-peer-deps", {
              cwd,
              timeout,
              stdio,
              env,
              shell: "cmd.exe"
            });
            return;
          } else {
            throw err;
          }
        }
      }
      this.installSyncBunCopyfile(cwd, opts);
      return;
    }
    const bun = this.bunBinary;
    if (opts?.frozen) {
      try {
        nodeExecSync(`"${bun}" install --frozen-lockfile 2>&1`, { cwd, timeout, stdio, env });
        return;
      } catch {
      }
    }
    nodeExecSync(`"${bun}" install`, { cwd, timeout, stdio, env });
  }
  /**
   * Windows-only fallback path: `bun install [--frozen-lockfile]
   * --backend=copyfile`. Used when npm.cmd isn't available on PATH
   * (typically: Shogo Desktop bundles bun but the user didn't install
   * Node.js separately).
   *
   * The `--backend=copyfile` flag forces bun to copy package contents
   * into `node_modules/` instead of hardlinking from
   * `~/.bun/install/cache`. Hardlinks are bun's default for speed but
   * fail in confusing ways on Windows (the destination directory gets
   * created but ends up empty, and downstream `require('webidl-conversions')`
   * etc. then crash with MODULE_NOT_FOUND). See bun#10327 / #28653.
   */
  installSyncBunCopyfile(cwd, opts) {
    const timeout = opts?.timeout ?? DEFAULT_INSTALL_TIMEOUT;
    const stdio = opts?.stdio ?? "pipe";
    const env = this.spawnEnv(opts?.env);
    const bun = this.bunBinary;
    console.warn(
      "[platform-pkg] Node.js / npm.cmd not found on PATH \u2014 falling back to `bun install --backend=copyfile` (Shogo Desktop bundle path). This is slower than hardlink but avoids the Windows extract bug."
    );
    if (opts?.frozen) {
      try {
        nodeExecSync(`"${bun}" install --frozen-lockfile --backend=copyfile`, {
          cwd,
          timeout,
          stdio,
          env
        });
        return;
      } catch {
      }
    }
    nodeExecSync(`"${bun}" install --backend=copyfile`, { cwd, timeout, stdio, env });
  }
  // ---------------------------------------------------------------------------
  // Install — async with spawn (used by RuntimeManager)
  // ---------------------------------------------------------------------------
  installAsync(cwd, opts) {
    if (IS_WINDOWS) {
      const env = this.spawnEnv(opts?.env);
      if (isNodeAvailableOnWindows(env.PATH)) {
        return this.installAsyncWindowsNpm(cwd, opts).catch((err) => {
          if (err instanceof NodeMissingError) {
            return this.installAsyncBunCopyfile(cwd, opts);
          }
          throw err;
        });
      }
      return this.installAsyncBunCopyfile(cwd, opts);
    }
    return this.installAsyncBun(cwd, opts);
  }
  installAsyncWindowsNpm(cwd, opts) {
    const timeout = opts?.timeout ?? DEFAULT_INSTALL_TIMEOUT;
    const env = this.spawnEnv(opts?.env);
    const cmd = "npm.cmd";
    const runOnce = (extraArgs) => new Promise((resolvePromise, rejectPromise) => {
      const proc = spawn(cmd, ["install", "--loglevel=error", ...extraArgs], {
        cwd,
        stdio: "pipe",
        env,
        shell: true
      });
      const timer = setTimeout(() => {
        proc.kill();
        rejectPromise(new Error(`${cmd} install timed out after ${timeout / 1e3}s`));
      }, timeout);
      let stderr = "";
      proc.stderr?.on("data", (d) => {
        stderr += d.toString();
      });
      proc.on("exit", (code) => {
        clearTimeout(timer);
        resolvePromise({ code, stderr });
      });
      proc.on("error", (err) => {
        clearTimeout(timer);
        if (err.code === "ENOENT") return rejectPromise(new NodeMissingError());
        rejectPromise(err);
      });
    });
    return (async () => {
      const first = await runOnce([]);
      if (first.code === 0) return;
      if (/not recognized as an internal or external command/i.test(first.stderr)) {
        throw new NodeMissingError();
      }
      if (isEresolveStderr(first.stderr)) {
        console.warn(
          "[platform-pkg] npm install hit ERESOLVE \u2014 retrying with --legacy-peer-deps."
        );
        const second = await runOnce(["--legacy-peer-deps"]);
        if (second.code === 0) return;
        throw new Error(
          `${cmd} install --legacy-peer-deps exited with code ${second.code}
${second.stderr}`
        );
      }
      throw new Error(`${cmd} install exited with code ${first.code}
${first.stderr}`);
    })();
  }
  installAsyncBunCopyfile(cwd, opts) {
    console.warn(
      "[platform-pkg] Node.js / npm.cmd not found on PATH \u2014 falling back to `bun install --backend=copyfile` (Shogo Desktop bundle path)."
    );
    return this.installAsyncBun(cwd, opts, ["--backend=copyfile"]);
  }
  installAsyncBun(cwd, opts, extraArgs = []) {
    const timeout = opts?.timeout ?? DEFAULT_INSTALL_TIMEOUT;
    const env = this.spawnEnv(opts?.env);
    const bun = this.bunBinary;
    const baseArgs = ["install", ...extraArgs];
    const frozenArgs = opts?.frozen ? ["install", "--frozen-lockfile", ...extraArgs] : null;
    const runOnce = (args) => new Promise((resolve, reject) => {
      const proc = spawn(bun, args, {
        cwd,
        stdio: "pipe",
        env
      });
      const timer = setTimeout(() => {
        proc.kill();
        reject(new Error(`bun ${args.join(" ")} timed out after ${timeout / 1e3}s`));
      }, timeout);
      let stderr = "";
      proc.stderr?.on("data", (d) => {
        stderr += d.toString();
      });
      proc.stdout?.on("data", () => {
      });
      proc.on("exit", (code) => {
        clearTimeout(timer);
        resolve({ ok: code === 0, stderr });
      });
      proc.on("error", (err) => {
        clearTimeout(timer);
        reject(err);
      });
    });
    const run = async () => {
      if (frozenArgs) {
        const first = await runOnce(frozenArgs);
        if (first.ok) return;
      }
      const second = await runOnce(baseArgs);
      if (!second.ok) {
        throw new Error(`bun install failed: ${second.stderr}`);
      }
    };
    return run();
  }
  // ---------------------------------------------------------------------------
  // Exec — run an npx/bunx tool synchronously
  // ---------------------------------------------------------------------------
  execToolSync(tool, args, cwd, opts) {
    const timeout = opts?.timeout ?? DEFAULT_EXEC_TIMEOUT;
    const stdio = opts?.stdio ?? "pipe";
    const env = this.spawnEnv(opts?.env);
    const argStr = args.length > 0 ? ` ${args.join(" ")}` : "";
    const bun = this.bunBinary;
    const cmd = IS_WINDOWS ? `npx ${tool}${argStr}` : `"${bun}" x ${opts?.useBunFlag ? "--bun " : ""}${tool}${argStr}`;
    return nodeExecSync(cmd, {
      cwd,
      timeout,
      stdio,
      env,
      // shellOpt() returns `true` on Windows so Node/Bun pick the platform
      // default shell from `process.env.ComSpec`. ExecSyncOptions only
      // narrowly types `shell` as `string | undefined`, but the runtime
      // accepts a boolean and the cross-platform behavior depends on it
      // (see shellOpt() doc).
      shell: this.shellOpt(),
      encoding: "utf-8"
    });
  }
  // ---------------------------------------------------------------------------
  // Exec — async variant that DOES NOT block the JS event loop.
  //
  // The sync variant above is fine for one-shot project scaffolding
  // (template.copy.ts) where blocking is intentional. Inside the runtime's
  // hot path (PreviewManager.runPrismaIfNeeded → /pool/assign) it caused
  // a measured ~4.7s freeze of /pool/assign in staging because prisma's
  // child process held the event loop for the full duration. Always
  // prefer this variant in long-lived services.
  // ---------------------------------------------------------------------------
  async execToolAsync(tool, args, cwd, opts) {
    const timeout = opts?.timeout ?? DEFAULT_EXEC_TIMEOUT;
    const stdio = opts?.stdio ?? "pipe";
    const env = this.spawnEnv(opts?.env);
    let cmd;
    let argv;
    let useShell;
    if (IS_WINDOWS) {
      const argStr = args.length > 0 ? ` ${args.join(" ")}` : "";
      cmd = `npx ${tool}${argStr}`;
      argv = [];
      useShell = true;
    } else {
      cmd = this.bunBinary;
      argv = ["x"];
      if (opts?.useBunFlag) argv.push("--bun");
      argv.push(tool, ...args);
      useShell = false;
    }
    return new Promise((resolvePromise, rejectPromise) => {
      const proc = spawn(cmd, argv, {
        cwd,
        env,
        stdio,
        shell: useShell ? this.shellOpt() : false
      });
      let stdout = "";
      let stderr = "";
      proc.stdout?.on("data", (chunk) => {
        stdout += chunk.toString();
      });
      proc.stderr?.on("data", (chunk) => {
        stderr += chunk.toString();
      });
      const timer = setTimeout(() => {
        try {
          proc.kill("SIGKILL");
        } catch {
        }
        rejectPromise(
          new Error(`${tool} ${args.join(" ")} timed out after ${timeout}ms`)
        );
      }, timeout);
      proc.on("error", (err) => {
        clearTimeout(timer);
        rejectPromise(err);
      });
      proc.on("exit", (code) => {
        clearTimeout(timer);
        if (code === 0) {
          resolvePromise(stdout);
        } else {
          const errMsg = stderr.trim() || stdout.trim() || `exit ${code}`;
          rejectPromise(new Error(`${tool} ${args.join(" ")} failed: ${errMsg}`));
        }
      });
    });
  }
  // ---------------------------------------------------------------------------
  // Prisma convenience wrappers
  // ---------------------------------------------------------------------------
  prismaGenerate(cwd, opts) {
    this.execToolSync("prisma", ["generate"], cwd, {
      timeout: opts?.timeout ?? 3e4,
      stdio: opts?.stdio,
      env: opts?.env
    });
  }
  prismaDbPush(cwd, opts) {
    const args = ["db", "push"];
    if (opts?.acceptDataLoss) args.push("--accept-data-loss");
    this.execToolSync("prisma", args, cwd, {
      timeout: opts?.timeout ?? DEFAULT_EXEC_TIMEOUT,
      stdio: opts?.stdio,
      env: opts?.env
    });
  }
  // Async siblings of the prisma wrappers above. Use these from any
  // long-running service (runtime, agent gateway, preview manager) so a
  // 1-3s prisma child process doesn't freeze /health, /pool/assign, or
  // other concurrent HTTP work. Same arg shape, same timeout defaults.
  async prismaGenerateAsync(cwd, opts) {
    await this.execToolAsync("prisma", ["generate"], cwd, {
      timeout: opts?.timeout ?? 3e4,
      stdio: opts?.stdio,
      env: opts?.env
    });
  }
  async prismaDbPushAsync(cwd, opts) {
    const args = ["db", "push"];
    if (opts?.acceptDataLoss) args.push("--accept-data-loss");
    await this.execToolAsync("prisma", args, cwd, {
      timeout: opts?.timeout ?? DEFAULT_EXEC_TIMEOUT,
      stdio: opts?.stdio,
      env: opts?.env
    });
  }
};
function wrapWindowsNpmError(err) {
  if (!IS_WINDOWS) return null;
  const blob = `${err?.stderr?.toString?.() ?? ""}
${err?.stdout?.toString?.() ?? ""}
${err?.message ?? ""}`;
  if (/not recognized as an internal or external command/i.test(blob)) {
    return new NodeMissingError();
  }
  if (err?.code === "ENOENT") return new NodeMissingError();
  return null;
}
function isEresolveStderr(stderr) {
  return /ERESOLVE/i.test(stderr);
}
function isNpmEresolveError(err) {
  const blob = `${err?.stderr?.toString?.() ?? ""}
${err?.stdout?.toString?.() ?? ""}
${err?.message ?? ""}`;
  return isEresolveStderr(blob);
}
var pkg = new PlatformPackageManager();

export {
  NodeMissingError,
  isNodeAvailableOnWindows,
  isNodeAvailableOnUnix,
  _resetUnixNodeCache,
  resolveBinInvocation,
  PlatformPackageManager,
  isEresolveStderr,
  pkg
};
