// @shogo-ai/cli/dist/deploy.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/deploy.ts
var deploy_exports = {};
__export(deploy_exports, {
  AGENT_NAME_PATTERN: () => AGENT_NAME_PATTERN,
  buildSyncPayload: () => buildSyncPayload,
  normalizeToolEntry: () => normalizeToolEntry,
  runDeploy: () => runDeploy,
  validateManifest: () => validateManifest
});
module.exports = __toCommonJS(deploy_exports);
var AGENT_NAME_PATTERN = /^[a-z][a-z0-9_-]{0,63}$/;
var KNOWN_KEYS = /* @__PURE__ */ new Set([
  "systemPrompt",
  "tools",
  "characterName",
  "displayName",
  "voiceId",
  "firstMessage",
  "model"
]);
function validateManifest(raw) {
  const issues = [];
  if (raw == null) return { agents: {}, issues };
  if (typeof raw !== "object" || Array.isArray(raw)) {
    issues.push({
      path: "agents",
      message: "`agents` must be an object keyed by agent name"
    });
    return { agents: {}, issues };
  }
  const agents = {};
  for (const [name, value] of Object.entries(raw)) {
    if (!AGENT_NAME_PATTERN.test(name)) {
      issues.push({
        path: `agents.${name}`,
        message: `agent name must match ${AGENT_NAME_PATTERN}`
      });
      continue;
    }
    if (value == null || typeof value !== "object" || Array.isArray(value)) {
      issues.push({
        path: `agents.${name}`,
        message: "value must be an object"
      });
      continue;
    }
    const entry = {};
    for (const [k, v] of Object.entries(value)) {
      if (!KNOWN_KEYS.has(k)) {
        issues.push({
          path: `agents.${name}.${k}`,
          message: `unknown field; supported: ${[...KNOWN_KEYS].join(", ")}`
        });
        continue;
      }
      if (k === "tools") {
        if (!Array.isArray(v)) {
          issues.push({
            path: `agents.${name}.tools`,
            message: "must be an array of strings or { name, description?, inputSchema? } objects"
          });
          continue;
        }
        const tools = [];
        const seenNames = /* @__PURE__ */ new Set();
        let valid = true;
        for (let i = 0; i < v.length; i++) {
          const t = v[i];
          if (typeof t === "string") {
            if (t.length === 0) {
              issues.push({
                path: `agents.${name}.tools[${i}]`,
                message: "tool name must be a non-empty string"
              });
              valid = false;
              continue;
            }
            if (seenNames.has(t)) {
              issues.push({
                path: `agents.${name}.tools[${i}]`,
                message: `duplicate tool name '${t}'`
              });
              valid = false;
              continue;
            }
            seenNames.add(t);
            tools.push(t);
            continue;
          }
          if (t == null || typeof t !== "object" || Array.isArray(t)) {
            issues.push({
              path: `agents.${name}.tools[${i}]`,
              message: "tools entries must be a string or { name, description?, inputSchema? }"
            });
            valid = false;
            continue;
          }
          const obj = t;
          if (typeof obj.name !== "string" || obj.name.length === 0) {
            issues.push({
              path: `agents.${name}.tools[${i}].name`,
              message: "name is required and must be a non-empty string"
            });
            valid = false;
            continue;
          }
          if (seenNames.has(obj.name)) {
            issues.push({
              path: `agents.${name}.tools[${i}]`,
              message: `duplicate tool name '${obj.name}'`
            });
            valid = false;
            continue;
          }
          if (obj.description !== void 0 && typeof obj.description !== "string") {
            issues.push({
              path: `agents.${name}.tools[${i}].description`,
              message: "description must be a string"
            });
            valid = false;
            continue;
          }
          if (obj.inputSchema !== void 0 && (obj.inputSchema === null || typeof obj.inputSchema !== "object" || Array.isArray(obj.inputSchema))) {
            issues.push({
              path: `agents.${name}.tools[${i}].inputSchema`,
              message: "inputSchema must be a JSON Schema object"
            });
            valid = false;
            continue;
          }
          seenNames.add(obj.name);
          const descriptor = { name: obj.name };
          if (typeof obj.description === "string") {
            descriptor.description = obj.description;
          }
          if (obj.inputSchema !== void 0) {
            descriptor.inputSchema = obj.inputSchema;
          }
          tools.push(descriptor);
        }
        if (valid) entry.tools = tools;
        continue;
      }
      if (typeof v !== "string") {
        issues.push({
          path: `agents.${name}.${k}`,
          message: "must be a string"
        });
        continue;
      }
      ;
      entry[k] = v;
    }
    if (entry.voiceId && !entry.firstMessage) {
      issues.push({
        path: `agents.${name}.firstMessage`,
        message: "voiceId is set; consider providing firstMessage so the agent has a greeting"
      });
    }
    agents[name] = entry;
  }
  return { agents, issues };
}
function normalizeToolEntry(entry) {
  if (typeof entry === "string") return { name: entry };
  const out = { name: entry.name };
  if (entry.description !== void 0) out.description = entry.description;
  if (entry.inputSchema !== void 0) out.inputSchema = entry.inputSchema;
  return out;
}
function buildSyncPayload(opts) {
  const out = { agents: {}, prune: opts.prune, dryRun: opts.dryRun };
  for (const [name, entry] of Object.entries(opts.manifest)) {
    const a = {};
    if (entry.systemPrompt !== void 0) a.systemPrompt = entry.systemPrompt;
    if (entry.tools !== void 0) {
      a.tools = entry.tools.map(normalizeToolEntry);
    }
    if (entry.characterName !== void 0) a.characterName = entry.characterName;
    if (entry.displayName !== void 0) a.displayName = entry.displayName;
    if (entry.voiceId !== void 0) a.voiceId = entry.voiceId;
    if (entry.firstMessage !== void 0) a.firstMessage = entry.firstMessage;
    if (entry.model !== void 0) a.model = entry.model;
    out.agents[name] = a;
  }
  return out;
}
function resolveDeployHeaders(options) {
  const headers = {
    "content-type": "application/json"
  };
  const auth = options.auth ?? (options.shogoApiKey ? { kind: "apiKey", apiKey: options.shogoApiKey } : void 0);
  if (!auth) {
    throw new Error(
      'runDeploy: missing auth \u2014 pass `auth: { kind: "apiKey", apiKey }` or `auth: { kind: "runtimeToken", token }`'
    );
  }
  if (auth.kind === "apiKey") {
    headers.authorization = `Bearer ${auth.apiKey}`;
  } else {
    headers["x-runtime-token"] = auth.token;
  }
  return headers;
}
async function runDeploy(options) {
  const fetchImpl = options.fetchImpl ?? globalThis.fetch;
  if (typeof fetchImpl !== "function") {
    throw new Error("runDeploy: global fetch is unavailable; pass options.fetchImpl");
  }
  const url = `${options.apiUrl.replace(/\/+$/, "")}/api/projects/${encodeURIComponent(options.projectId)}/agents/sync`;
  const payload = buildSyncPayload({
    manifest: options.manifest,
    prune: options.prune,
    dryRun: options.dryRun
  });
  const res = await fetchImpl(url, {
    method: "POST",
    headers: resolveDeployHeaders(options),
    body: JSON.stringify(payload)
  });
  let body = null;
  try {
    body = await res.json();
  } catch {
  }
  return { status: res.status, body };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AGENT_NAME_PATTERN,
  buildSyncPayload,
  normalizeToolEntry,
  runDeploy,
  validateManifest
});
