// @shogo-ai/sdk/dist/generators/index.js
// src/generators/generated-file-license-header.ts
var GENERATED_FILE_LICENSE_HEADER = [
  "// SPDX-License-Identifier: MIT",
  "// Copyright (C) 2026 Shogo Technologies, Inc."
].join("\n");

// src/generators/types-generator.ts
function toFileName(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function generateTypes(models, enums) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Types",
    " *",
    " * TypeScript types for models and operations.",
    " * Generated by `shogo generate` - DO NOT EDIT",
    " */",
    ""
  ];
  const enumNames = new Set(enums.map((e) => e.name));
  if (enums.length > 0) {
    lines.push("// ============================================================================");
    lines.push("// Enums");
    lines.push("// ============================================================================");
    lines.push("");
    for (const enumDef of enums) {
      lines.push(`export type ${enumDef.name} = ${enumDef.values.map((v) => `'${v.name}'`).join(" | ")}`);
      lines.push("");
    }
  }
  lines.push("// ============================================================================");
  lines.push("// Model Types");
  lines.push("// ============================================================================");
  lines.push("");
  for (const model of models) {
    lines.push(...generateModelType(model, enumNames));
  }
  lines.push("// ============================================================================");
  lines.push("// Input Types");
  lines.push("// ============================================================================");
  lines.push("");
  for (const model of models) {
    lines.push(...generateInputTypes(model));
  }
  lines.push(...generateHookTypes(models, enumNames));
  return lines.join("\n");
}
function resolveModelTypeName(modelName, enumNames) {
  const candidate = `${modelName}Type`;
  return enumNames.has(candidate) ? modelName : candidate;
}
function mapPrismaType(field) {
  if (field.kind === "object") {
    return field.isList ? `${field.type}Type[]` : `${field.type}Type`;
  }
  if (field.kind === "enum") {
    return field.type;
  }
  switch (field.type) {
    case "String":
      return "string";
    case "Int":
    case "Float":
    case "Decimal":
      return "number";
    case "Boolean":
      return "boolean";
    case "DateTime":
      return "Date";
    case "Json":
      return "Record<string, unknown>";
    case "BigInt":
      return "bigint";
    case "Bytes":
      return "Buffer";
    default:
      return "unknown";
  }
}
function generateModelType(model, enumNames = /* @__PURE__ */ new Set()) {
  const typeName = resolveModelTypeName(model.name, enumNames);
  const lines = [
    `export interface ${typeName} {`
  ];
  for (const field of model.fields) {
    if (field.kind === "object") continue;
    const tsType = mapPrismaType(field);
    const optional = !field.isRequired && !field.isId ? "?" : "";
    const nullable = !field.isRequired ? " | null" : "";
    lines.push(`  ${field.name}${optional}: ${tsType}${!optional && nullable ? nullable : ""}`);
  }
  lines.push("}");
  lines.push("");
  return lines;
}
function generateInputTypes(model) {
  const lines = [];
  const scalarFields = getScalarFields(model);
  lines.push(`export interface ${model.name}CreateInput {`);
  for (const field of scalarFields) {
    if (field.isId && field.hasDefaultValue) continue;
    if (field.name === "createdAt" || field.name === "updatedAt") continue;
    const tsType = mapPrismaType(field);
    const optional = !field.isRequired || field.hasDefaultValue ? "?" : "";
    lines.push(`  ${field.name}${optional}: ${tsType}`);
  }
  lines.push("}");
  lines.push("");
  lines.push(`export interface ${model.name}UpdateInput {`);
  for (const field of scalarFields) {
    if (field.isId) continue;
    if (field.name === "createdAt" || field.name === "updatedAt") continue;
    const tsType = mapPrismaType(field);
    lines.push(`  ${field.name}?: ${tsType}`);
  }
  lines.push("}");
  lines.push("");
  return lines;
}
function generateHookTypes(models, enumNames = /* @__PURE__ */ new Set()) {
  const lines = [
    "// ============================================================================",
    "// Hook Types",
    "// ============================================================================",
    "",
    "export interface HookContext {",
    "  userId?: string",
    "}",
    "",
    "export interface ListHookResult {",
    "  where?: Record<string, unknown>",
    "}",
    "",
    "export interface GetHookResult {",
    "  deny?: boolean",
    "}",
    "",
    "export interface UpdateHookResult<T> {",
    "  deny?: boolean",
    "  input?: T",
    "}",
    "",
    "export interface DeleteHookResult {",
    "  deny?: boolean",
    "}",
    "",
    "export interface ModelHooks<TCreate, TUpdate, TModel> {",
    "  beforeList?: (ctx: HookContext) => Promise<ListHookResult | void>",
    "  beforeGet?: (id: string, ctx: HookContext) => Promise<GetHookResult | void>",
    "  beforeCreate?: (input: TCreate, ctx: HookContext) => Promise<TCreate>",
    "  afterCreate?: (item: TModel, ctx: HookContext) => Promise<void>",
    "  beforeUpdate?: (id: string, input: TUpdate, ctx: HookContext) => Promise<UpdateHookResult<TUpdate> | void>",
    "  afterUpdate?: (item: TModel, ctx: HookContext) => Promise<void>",
    "  beforeDelete?: (id: string, ctx: HookContext) => Promise<DeleteHookResult | void>",
    "  afterDelete?: (id: string, ctx: HookContext) => Promise<void>",
    "}",
    "",
    "export interface ServerFunctionHooks {"
  ];
  for (const model of models) {
    const typeName = resolveModelTypeName(model.name, enumNames);
    lines.push(`  ${model.name}?: ModelHooks<${model.name}CreateInput, ${model.name}UpdateInput, ${typeName}>`);
  }
  lines.push("}");
  lines.push("");
  return lines;
}
function generateModelTypes(model, enums = [], fileExtension = "tsx") {
  const modelName = model.name;
  const fileName = `${toFileName(modelName)}.types.${fileExtension}`;
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    ` * Auto-generated ${modelName} Types`,
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    ""
  ];
  const modelEnumTypes = new Set(
    model.fields.filter((f) => f.kind === "enum").map((f) => f.type)
  );
  const emittedEnumNames = /* @__PURE__ */ new Set();
  if (modelEnumTypes.size > 0) {
    const relevantEnums = enums.filter((e) => modelEnumTypes.has(e.name));
    for (const enumDef of relevantEnums) {
      lines.push(`export type ${enumDef.name} = ${enumDef.values.map((v) => `'${v.name}'`).join(" | ")}`);
      lines.push("");
      emittedEnumNames.add(enumDef.name);
    }
  }
  lines.push(...generateModelType(model, emittedEnumNames));
  lines.push(...generateInputTypes(model));
  return {
    modelName,
    fileName,
    code: lines.join("\n")
  };
}
function generateTypesPerModel(models, enums = [], fileExtension = "tsx") {
  return models.map((model) => generateModelTypes(model, enums, fileExtension));
}
function generateTypesIndex(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Types Index",
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    "// Re-export all types"
  ];
  for (const model of models) {
    const fileName = toFileName(model.name);
    lines.push(`export * from "./${fileName}.types"`);
  }
  lines.push("");
  return lines.join("\n");
}

// src/generators/routes-generator.ts
function toRoutePath(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function toFileName2(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function getWritableScalarFields(model) {
  return getScalarFields(model).filter((f) => !(f.isId && f.hasDefaultValue));
}
function tsStringArrayLiteral(names) {
  if (names.length === 0) return "[] as const";
  const inner = names.map((n) => `"${n}"`).join(", ");
  return `[${inner}] as const`;
}
function generateModelRoutes(model, config = {}) {
  const idField = getIdField(model);
  if (!idField) {
    return null;
  }
  const modelName = model.name;
  const modelLower = toCamelCase2(modelName);
  const ext = config.fileExtension || "tsx";
  const fileName = `${toFileName2(modelName)}.routes.${ext}`;
  const writableScalarFieldNames = getWritableScalarFields(model).map((f) => f.name);
  const relationFieldNames = getRelationFields2(model).map((f) => f.name);
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    ` * Auto-generated ${modelName} Routes`,
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { Hono } from "hono"',
    'import { PrismaClient } from "./prisma/client"',
    `import type { ${modelName}Hooks } from "./${toFileName2(modelName)}.hooks"`,
    "",
    "// JSON replacer that coerces native BigInt values to strings so they can",
    "// be serialized by JSON.stringify (and therefore by Hono's default json",
    "// helper, which throws on bare BigInt values). Strings preserve full",
    "// precision for very large counts (e.g. byte totals). Consumers that need",
    "// a numeric should coerce explicitly.",
    "const bigIntReplacer = (_key: string, value: unknown) =>",
    '  typeof value === "bigint" ? value.toString() : value',
    "",
    "// Drop-in replacement for c.json that uses bigIntReplacer. Keeps the same",
    "// (body, status?) ergonomics as c.json so call sites are mechanical.",
    "function sendJson(c: any, body: unknown, status: number = 200) {",
    "  return c.body(JSON.stringify(body, bigIntReplacer), status, {",
    '    "content-type": "application/json; charset=UTF-8",',
    "  })",
    "}",
    "",
    "// Prisma client instance (injected)",
    "let prisma: PrismaClient | null = null",
    "",
    `let hooks: ${modelName}Hooks = {}`,
    "",
    "/**",
    " * Set the Prisma client instance",
    " */",
    "export function setPrisma(client: PrismaClient) {",
    "  prisma = client",
    "}",
    "",
    "/**",
    ` * Set hooks for ${modelName} routes`,
    " */",
    `export function set${modelName}Hooks(h: ${modelName}Hooks) {`,
    "  hooks = h",
    "}",
    "",
    "/**",
    " * Get Prisma client (throws if not set)",
    " */",
    "function getPrisma(): PrismaClient {",
    "  if (!prisma) {",
    '    throw new Error("Prisma client not set. Call setPrisma() first.")',
    "  }",
    "  return prisma",
    "}",
    "",
    "/**",
    " * Build route context from Hono context",
    " */",
    "function buildContext(c: any, body?: any) {",
    '  const auth = c.get("auth")',
    "  return {",
    "    body: body || {},",
    "    params: c.req.param() || {},",
    "    query: Object.fromEntries(new URL(c.req.url).searchParams),",
    "    userId: auth?.userId,",
    "    tunnelAuthenticated: !!auth?.tunnelAuthenticated,",
    "    prisma: getPrisma(),",
    "  }",
    "}",
    "",
    "// Field allowlists derived from the Prisma schema at code-gen time. Used",
    "// by `pickWritableFields` to keep the create/update payload Prisma-safe.",
    `const WRITABLE_SCALAR_FIELDS = ${tsStringArrayLiteral(writableScalarFieldNames)}`,
    `const RELATION_FIELDS = ${tsStringArrayLiteral(relationFieldNames)}`,
    "",
    "type WriteBodyResult =",
    "  | { ok: true; data: Record<string, unknown> }",
    "  | { ok: false; error: { code: string; message: string } }",
    "",
    "/**",
    " * Build a Prisma-safe `data` payload from the request body:",
    " * - Scalar/enum fields are passed through (only fields the schema declares).",
    " * - Relation fields must use Prisma's nested-write shape (an object like",
    " *   `{ connect: { id: ... } }` or `{ create: { ... } }`), not a raw scalar",
    " *   or array. The common bug `{ scorecards: 1 }` would otherwise crash with",
    ' *   "Argument `scorecards`: Invalid value provided. Expected',
    ' *    XCreateNestedManyWithoutYInput, provided Int." We surface it as a 400.',
    " * - Unknown keys are silently dropped to prevent injection of fields the",
    " *   schema does not expose.",
    " */",
    "function pickWritableFields(body: any): WriteBodyResult {",
    '  if (body == null || typeof body !== "object" || Array.isArray(body)) {',
    '    return { ok: false, error: { code: "invalid_body", message: "Request body must be a JSON object." } }',
    "  }",
    "  const data: Record<string, unknown> = {}",
    "  for (const key of WRITABLE_SCALAR_FIELDS) {",
    "    if (body[key] !== undefined) data[key] = body[key]",
    "  }",
    "  for (const key of RELATION_FIELDS) {",
    "    const value = body[key]",
    "    if (value === undefined || value === null) continue",
    '    if (typeof value !== "object" || Array.isArray(value)) {',
    "      return {",
    "        ok: false,",
    "        error: {",
    '          code: "invalid_relation_shape",',
    `          message: \`Field "\${key}" is a relation. Provide it as { connect: { id: ... } } or { create: { ... } }, not a scalar/array.\`,`,
    "        },",
    "      }",
    "    }",
    "    data[key] = value",
    "  }",
    "  return { ok: true, data }",
    "}",
    "",
    "/**",
    ` * Create ${modelName} CRUD routes`,
    " */",
    `export function create${modelName}Routes(): Hono {`,
    "  const router = new Hono()",
    ""
  ];
  lines.push("  // GET / - List all");
  lines.push('  router.get("/", async (c) => {');
  lines.push("    try {");
  lines.push("      const ctx = buildContext(c)");
  lines.push("      const prisma = getPrisma()");
  lines.push("      const query = ctx.query");
  lines.push("      ");
  lines.push("      // Build initial where from query params (exclude pagination/meta params)");
  lines.push('      const reservedParams = ["limit", "offset", "userId", "include", "orderBy"]');
  lines.push("      let where: any = {}");
  lines.push("      ");
  lines.push("      for (const [key, value] of Object.entries(query)) {");
  lines.push('        if (!reservedParams.includes(key) && value !== undefined && value !== null && value !== "") {');
  lines.push("          // Try to parse as number or boolean");
  lines.push("          let parsedValue: any = value");
  lines.push('          if (value === "true") parsedValue = true');
  lines.push('          else if (value === "false") parsedValue = false');
  lines.push('          else if (!isNaN(Number(value)) && value !== "") parsedValue = Number(value)');
  lines.push("          ");
  lines.push("          where[key] = parsedValue");
  lines.push("        }");
  lines.push("      }");
  lines.push("      ");
  lines.push("      let include: any = undefined");
  lines.push("      let orderBy: any = undefined");
  lines.push("");
  lines.push("      // Apply beforeList hook (can override where/include/orderBy)");
  lines.push("      if (hooks.beforeList) {");
  lines.push("        const result = await hooks.beforeList(ctx)");
  lines.push("        if (result && !result.ok) {");
  lines.push("          return sendJson(c, { error: result.error }, 400)");
  lines.push("        }");
  lines.push("        if (result?.data) {");
  lines.push("          where = result.data.where || where");
  lines.push("          include = result.data.include || include");
  lines.push("          orderBy = result.data.orderBy || orderBy");
  lines.push("        }");
  lines.push("      }");
  lines.push("");
  lines.push(`      const [items, total] = await Promise.all([`);
  lines.push(`        prisma.${modelLower}.findMany({`);
  lines.push("          where,");
  lines.push("          include,");
  lines.push("          orderBy,");
  lines.push("          take: query.limit ? parseInt(query.limit) : undefined,");
  lines.push("          skip: query.offset ? parseInt(query.offset) : undefined,");
  lines.push("        }),");
  lines.push(`        prisma.${modelLower}.count({ where }),`);
  lines.push("      ])");
  lines.push("");
  lines.push("      return sendJson(c, { ok: true, items, total })");
  lines.push("    } catch (error: any) {");
  lines.push(`      console.error("[${modelName}] List error:", error)`);
  lines.push('      return sendJson(c, { error: { code: "list_failed", message: error.message } }, 500)');
  lines.push("    }");
  lines.push("  })");
  lines.push("");
  lines.push("  // GET /:id - Get by ID");
  lines.push('  router.get("/:id", async (c) => {');
  lines.push("    try {");
  lines.push('      const id = c.req.param("id")');
  lines.push("      const ctx = buildContext(c)");
  lines.push("      const prisma = getPrisma()");
  lines.push("");
  lines.push("      // Apply beforeGet hook");
  lines.push("      if (hooks.beforeGet) {");
  lines.push("        const result = await hooks.beforeGet(id, ctx)");
  lines.push("        if (result && !result.ok) {");
  lines.push('          return sendJson(c, { error: result.error }, result.error?.code === "not_found" ? 404 : 400)');
  lines.push("        }");
  lines.push("      }");
  lines.push("");
  lines.push(`      const item = await prisma.${modelLower}.findUnique({`);
  lines.push("        where: { id },");
  lines.push("      })");
  lines.push("");
  lines.push("      if (!item) {");
  lines.push(`        return sendJson(c, { error: { code: "not_found", message: "${modelName} not found" } }, 404)`);
  lines.push("      }");
  lines.push("");
  lines.push("      return sendJson(c, { ok: true, data: item })");
  lines.push("    } catch (error: any) {");
  lines.push(`      console.error("[${modelName}] Get error:", error)`);
  lines.push('      return sendJson(c, { error: { code: "get_failed", message: error.message } }, 500)');
  lines.push("    }");
  lines.push("  })");
  lines.push("");
  lines.push("  // POST / - Create");
  lines.push('  router.post("/", async (c) => {');
  lines.push("    try {");
  lines.push("      let body = await c.req.json()");
  lines.push("      const ctx = buildContext(c, body)");
  lines.push("      const prisma = getPrisma()");
  lines.push("");
  lines.push("      // Apply beforeCreate hook");
  lines.push("      if (hooks.beforeCreate) {");
  lines.push("        const result = await hooks.beforeCreate(body, ctx)");
  lines.push("        if (result && !result.ok) {");
  lines.push("          return sendJson(c, { error: result.error }, 400)");
  lines.push("        }");
  lines.push("        if (result?.data) {");
  lines.push("          body = result.data");
  lines.push("        }");
  lines.push("      }");
  lines.push("");
  lines.push("      // Filter the body to schema-known scalars + relation-shaped relations.");
  lines.push("      // Prevents accidental relation-as-scalar errors and unknown-key Prisma crashes.");
  lines.push("      const picked = pickWritableFields(body)");
  lines.push("      if (!picked.ok) {");
  lines.push("        return sendJson(c, { error: picked.error }, 400)");
  lines.push("      }");
  lines.push("");
  lines.push(`      const item = await prisma.${modelLower}.create({`);
  lines.push("        data: picked.data,");
  lines.push("      })");
  lines.push("");
  lines.push("      // Apply afterCreate hook");
  lines.push("      if (hooks.afterCreate) {");
  lines.push("        await hooks.afterCreate(item, ctx)");
  lines.push("      }");
  lines.push("");
  lines.push("      return sendJson(c, { ok: true, data: item }, 201)");
  lines.push("    } catch (error: any) {");
  lines.push(`      console.error("[${modelName}] Create error:", error)`);
  lines.push('      return sendJson(c, { error: { code: "create_failed", message: error.message } }, 500)');
  lines.push("    }");
  lines.push("  })");
  lines.push("");
  lines.push("  // PATCH /:id - Update");
  lines.push('  router.patch("/:id", async (c) => {');
  lines.push("    try {");
  lines.push('      const id = c.req.param("id")');
  lines.push("      let body = await c.req.json()");
  lines.push("      const ctx = buildContext(c, body)");
  lines.push("      const prisma = getPrisma()");
  lines.push("");
  lines.push("      // Apply beforeUpdate hook");
  lines.push("      if (hooks.beforeUpdate) {");
  lines.push("        const result = await hooks.beforeUpdate(id, body, ctx)");
  lines.push("        if (result && !result.ok) {");
  lines.push("          return sendJson(c, { error: result.error }, 400)");
  lines.push("        }");
  lines.push("        if (result?.data) {");
  lines.push("          body = result.data");
  lines.push("        }");
  lines.push("      }");
  lines.push("");
  lines.push("      // Same allowlist filter as on create \u2014 see pickWritableFields above.");
  lines.push("      const picked = pickWritableFields(body)");
  lines.push("      if (!picked.ok) {");
  lines.push("        return sendJson(c, { error: picked.error }, 400)");
  lines.push("      }");
  lines.push("");
  lines.push(`      const item = await prisma.${modelLower}.update({`);
  lines.push("        where: { id },");
  lines.push("        data: picked.data,");
  lines.push("      })");
  lines.push("");
  lines.push("      // Apply afterUpdate hook");
  lines.push("      if (hooks.afterUpdate) {");
  lines.push("        await hooks.afterUpdate(item, ctx)");
  lines.push("      }");
  lines.push("");
  lines.push("      return sendJson(c, { ok: true, data: item })");
  lines.push("    } catch (error: any) {");
  lines.push(`      console.error("[${modelName}] Update error:", error)`);
  lines.push('      return sendJson(c, { error: { code: "update_failed", message: error.message } }, 500)');
  lines.push("    }");
  lines.push("  })");
  lines.push("");
  lines.push("  // DELETE /:id - Delete");
  lines.push('  router.delete("/:id", async (c) => {');
  lines.push("    try {");
  lines.push('      const id = c.req.param("id")');
  lines.push("      const ctx = buildContext(c)");
  lines.push("      const prisma = getPrisma()");
  lines.push("");
  lines.push("      // Apply beforeDelete hook");
  lines.push("      if (hooks.beforeDelete) {");
  lines.push("        const result = await hooks.beforeDelete(id, ctx)");
  lines.push("        if (result && !result.ok) {");
  lines.push("          return sendJson(c, { error: result.error }, 400)");
  lines.push("        }");
  lines.push("      }");
  lines.push("");
  lines.push(`      await prisma.${modelLower}.delete({`);
  lines.push("        where: { id },");
  lines.push("      })");
  lines.push("");
  lines.push("      // Apply afterDelete hook");
  lines.push("      if (hooks.afterDelete) {");
  lines.push("        await hooks.afterDelete(id, ctx)");
  lines.push("      }");
  lines.push("");
  lines.push("      return sendJson(c, { ok: true })");
  lines.push("    } catch (error: any) {");
  lines.push(`      console.error("[${modelName}] Delete error:", error)`);
  lines.push('      return sendJson(c, { error: { code: "delete_failed", message: error.message } }, 500)');
  lines.push("    }");
  lines.push("  })");
  lines.push("");
  lines.push("  return router");
  lines.push("}");
  lines.push("");
  return {
    modelName,
    fileName,
    code: lines.join("\n")
  };
}
function generateModelHooks(model, config = {}) {
  const modelName = model.name;
  const ext = config.fileExtension || "tsx";
  const fileName = `${toFileName2(modelName)}.hooks.${ext}`;
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    ` * ${modelName} Hooks`,
    " *",
    " * Customize business logic for CRUD operations.",
    " * This file is safe to edit - it will not be overwritten.",
    " */",
    "",
    "/**",
    " * Result from a hook that can modify or reject the operation",
    " */",
    "export interface HookResult<T = any> {",
    "  ok: boolean",
    "  error?: { code: string; message: string }",
    "  data?: T",
    "}",
    "",
    "/**",
    ` * Hook context with Prisma client`,
    " */",
    "export interface HookContext {",
    "  body: any",
    "  params: Record<string, string>",
    "  query: Record<string, string>",
    "  userId?: string",
    "  tunnelAuthenticated: boolean",
    "  prisma: any",
    "}",
    "",
    "/**",
    ` * Hooks for ${modelName} routes`,
    " */",
    `export interface ${modelName}Hooks {`,
    "  /**",
    "   * Called before listing records. Can modify where/include/orderBy.",
    "   * Note: Query parameters (except limit, offset, userId, include, orderBy) are automatically",
    "   * added to the where clause. This hook receives them and can override/extend them.",
    "   */",
    "  beforeList?: (ctx: HookContext) => Promise<HookResult<{ where?: any; include?: any; orderBy?: any }> | void>",
    "  /** Called before getting a single record. Can reject access. */",
    "  beforeGet?: (id: string, ctx: HookContext) => Promise<HookResult | void>",
    "  /** Called before creating a record. Can modify input or reject. */",
    "  beforeCreate?: (input: any, ctx: HookContext) => Promise<HookResult<any> | void>",
    "  /** Called after creating a record. Can perform side effects. */",
    "  afterCreate?: (record: any, ctx: HookContext) => Promise<void>",
    "  /** Called before updating a record. Can modify input or reject. */",
    "  beforeUpdate?: (id: string, input: any, ctx: HookContext) => Promise<HookResult<any> | void>",
    "  /** Called after updating a record. Can perform side effects. */",
    "  afterUpdate?: (record: any, ctx: HookContext) => Promise<void>",
    "  /** Called before deleting a record. Can reject deletion. */",
    "  beforeDelete?: (id: string, ctx: HookContext) => Promise<HookResult | void>",
    "  /** Called after deleting a record. Can perform cleanup. */",
    "  afterDelete?: (id: string, ctx: HookContext) => Promise<void>",
    "}",
    "",
    "/**",
    ` * Default ${modelName} hooks (customize as needed)`,
    " */",
    `export const ${toCamelCase2(modelName)}Hooks: ${modelName}Hooks = {`,
    "  // beforeList: async (ctx) => {",
    "  //   // Query params are automatically added to where clause",
    '  //   // Example: GET /api/projects?workspaceId=123 => where: { workspaceId: "123" }',
    "  //   ",
    "  //   // You can override or extend the where clause:",
    "  //   // return { ok: true, data: { where: { ...ctx.query, userId: ctx.userId } } }",
    "  // },",
    "  // beforeCreate: async (input, ctx) => {",
    "  //   // Set userId on create",
    "  //   return { ok: true, data: { ...input, userId: ctx.userId } }",
    "  // },",
    "}",
    ""
  ];
  return {
    modelName,
    fileName,
    code: lines.join("\n")
  };
}
function generateRoutes(models, config = {}) {
  const routes = [];
  const hooks = [];
  for (const model of models) {
    const routeFile = generateModelRoutes(model, config);
    if (routeFile) {
      routes.push(routeFile);
      hooks.push(generateModelHooks(model, config));
    }
  }
  return { routes, hooks };
}
function generateRoutesIndex(models) {
  const routeModels = models.filter((model) => getIdField(model) !== void 0);
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Routes Index",
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { Hono } from "hono"',
    'import { PrismaClient } from "./prisma/client"',
    "",
    "// Route imports"
  ];
  for (const model of routeModels) {
    const fileName = toFileName2(model.name);
    lines.push(`import { create${model.name}Routes, setPrisma as setPrisma${model.name}, set${model.name}Hooks } from "./${fileName}.routes"`);
  }
  lines.push("");
  lines.push("// Hook imports");
  for (const model of routeModels) {
    const fileName = toFileName2(model.name);
    const hookVarName = toCamelCase2(model.name) + "Hooks";
    lines.push(`import { ${hookVarName} } from "./${fileName}.hooks"`);
  }
  lines.push("");
  lines.push("// Re-export route creators and setters");
  lines.push("export {");
  const routeExports = [];
  for (const model of routeModels) {
    routeExports.push(`  create${model.name}Routes`);
    routeExports.push(`  setPrisma${model.name}`);
    routeExports.push(`  set${model.name}Hooks`);
  }
  lines.push(routeExports.join(",\n"));
  lines.push("}");
  lines.push("");
  lines.push("// Re-export hooks (model-specific)");
  lines.push("export {");
  const hookExports = [];
  for (const model of routeModels) {
    const hookVarName = toCamelCase2(model.name) + "Hooks";
    hookExports.push(`  ${hookVarName}`);
  }
  lines.push(hookExports.join(",\n"));
  lines.push("}");
  lines.push("");
  lines.push("// Re-export hook types");
  for (const model of routeModels) {
    const fileName = toFileName2(model.name);
    lines.push(`export type { ${model.name}Hooks } from "./${fileName}.hooks"`);
  }
  lines.push("");
  lines.push("/**");
  lines.push(" * Create all routes and mount them on a single Hono app");
  lines.push(" */");
  lines.push("export function createAllRoutes(prisma: PrismaClient): Hono {");
  lines.push("  const app = new Hono()");
  lines.push("");
  lines.push("  // Set Prisma client for all routes");
  for (const model of routeModels) {
    lines.push(`  setPrisma${model.name}(prisma)`);
  }
  lines.push("");
  lines.push("  // Set hooks for all routes");
  for (const model of routeModels) {
    const hookVarName = toCamelCase2(model.name) + "Hooks";
    lines.push(`  set${model.name}Hooks(${hookVarName})`);
  }
  lines.push("");
  lines.push("  // Mount routes");
  for (const model of routeModels) {
    const routePath = toRoutePath(model.name);
    lines.push(`  app.route("/${routePath}", create${model.name}Routes())`);
  }
  lines.push("");
  lines.push("  return app");
  lines.push("}");
  lines.push("");
  return lines.join("\n");
}

// src/generators/stores-generator.ts
function toFileName3(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function toRoutePath2(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function generateModelStore(model, config = {}) {
  const idField = getIdField(model);
  if (!idField) {
    return null;
  }
  const basePath = config.basePath || "/api";
  const ext = config.fileExtension || "tsx";
  const modelName = model.name;
  const modelLower = toCamelCase2(modelName);
  const fileName = `${toFileName3(modelName)}.store.${ext}`;
  const routePath = toRoutePath2(modelName);
  const endpoint = `${basePath}/${routePath}`;
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    ` * Auto-generated ${modelName} Store`,
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { OptimisticStore } from "@shogo-ai/sdk"',
    'import type { HttpClient } from "@shogo-ai/sdk"',
    `import type { ${modelName}Type } from "./${toFileName3(modelName)}.types"`,
    "",
    "// Store instance (singleton)",
    `let _store: OptimisticStore<${modelName}Type> | null = null`,
    "",
    "/**",
    ` * Get or create the ${modelName} store`,
    " *",
    " * @param http - HttpClient instance (required on first call)",
    " */",
    `export function get${modelName}Store(http?: HttpClient): OptimisticStore<${modelName}Type> {`,
    "  if (!_store) {",
    "    if (!http) {",
    `      throw new Error("HttpClient required to initialize ${modelName}Store")`,
    "    }",
    `    _store = new OptimisticStore<${modelName}Type>({`,
    "      http,",
    `      endpoint: "${endpoint}",`,
    "    })",
    "  }",
    "  return _store",
    "}",
    "",
    "/**",
    ` * Reset the ${modelName} store (for testing)`,
    " */",
    `export function reset${modelName}Store(): void {`,
    "  _store = null",
    "}",
    "",
    `// Convenience export for direct import`,
    `export type { ${modelName}Type }`,
    ""
  ];
  return {
    modelName,
    fileName,
    code: lines.join("\n")
  };
}
function generateStores(models, config = {}) {
  const stores = [];
  for (const model of models) {
    const storeFile = generateModelStore(model, config);
    if (storeFile) {
      stores.push(storeFile);
    }
  }
  return stores;
}
function generateStoresIndex(models, config = {}) {
  const basePath = config.basePath || "/api";
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Stores Index",
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { OptimisticStore } from "@shogo-ai/sdk"',
    'import type { HttpClient } from "@shogo-ai/sdk"',
    "",
    "// Type imports"
  ];
  for (const model of models) {
    const fileName = toFileName3(model.name);
    lines.push(`import type { ${model.name}Type } from "./${fileName}.types"`);
  }
  lines.push("");
  lines.push("// Store imports");
  for (const model of models) {
    const fileName = toFileName3(model.name);
    lines.push(`export { get${model.name}Store, reset${model.name}Store } from "./${fileName}.store"`);
  }
  lines.push("");
  lines.push("// Re-export types");
  for (const model of models) {
    const fileName = toFileName3(model.name);
    lines.push(`export type { ${model.name}Type } from "./${fileName}.types"`);
  }
  lines.push("");
  lines.push("/**");
  lines.push(" * All stores as an object (for convenience)");
  lines.push(" */");
  lines.push("export interface AllStores {");
  for (const model of models) {
    lines.push(`  ${toCamelCase2(model.name)}: OptimisticStore<${model.name}Type>`);
  }
  lines.push("}");
  lines.push("");
  lines.push("/**");
  lines.push(" * Create all stores with the given HttpClient");
  lines.push(" */");
  lines.push("export function createAllStores(http: HttpClient): AllStores {");
  lines.push("  return {");
  for (const model of models) {
    const modelLower = toCamelCase2(model.name);
    lines.push(`    ${modelLower}: get${model.name}Store(http),`);
  }
  lines.push("  }");
  lines.push("}");
  lines.push("");
  return lines.join("\n");
}

// src/generators/mst-model-generator.ts
function toFileName4(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function mapPrismaToMSTType(field, enums) {
  if (field.kind === "enum") {
    const enumDef = enums.find((e) => e.name === field.type);
    const enumType = enumDef && enumDef.values.length > 0 ? `types.enumeration("${field.type}", [${enumDef.values.map((v) => `"${v.name}"`).join(", ")}])` : `types.string`;
    if (field.isList) {
      return `types.array(${enumType})`;
    }
    return enumType;
  }
  if (field.kind === "object") {
    if (field.isList) {
      return `types.array(types.safeReference(types.late(() => ${field.type}Model)))`;
    }
    return `types.safeReference(types.late(() => ${field.type}Model))`;
  }
  let baseType;
  switch (field.type) {
    case "String":
      baseType = "types.string";
      break;
    case "Int":
    case "Float":
    case "Decimal":
      baseType = "types.number";
      break;
    case "Boolean":
      baseType = "types.boolean";
      break;
    case "DateTime":
      baseType = "types.number";
      break;
    case "Json":
      baseType = "types.frozen()";
      break;
    case "BigInt":
      baseType = "types.string";
      break;
    case "Bytes":
      baseType = "types.string";
      break;
    default:
      baseType = "types.frozen()";
  }
  if (field.isList) {
    return `types.array(${baseType})`;
  }
  return baseType;
}
function wrapOptional(mstType, field) {
  if (field.isId) {
    return "types.identifier";
  }
  if (!field.isRequired) {
    return `types.maybe(${mstType})`;
  }
  return mstType;
}
function generateMSTModel(model, allModels, enums, includedModelNames, fileExtension = "tsx") {
  const modelName = model.name;
  const fileName = `${toFileName4(modelName)}.model.${fileExtension}`;
  const scalarFields = getScalarFields(model);
  const relationFields = getRelationFields2(model);
  const referencedModels = /* @__PURE__ */ new Set();
  const includedRelationFields = [];
  for (const field of relationFields) {
    const isIncluded = !includedModelNames || includedModelNames.has(field.type);
    if (field.type !== modelName && isIncluded) {
      referencedModels.add(field.type);
      includedRelationFields.push(field);
    }
  }
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    ` * Auto-generated ${modelName} MST Model`,
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { types, Instance, SnapshotIn, SnapshotOut } from "mobx-state-tree"'
  ];
  if (referencedModels.size > 0) {
    lines.push("");
    lines.push("// Referenced models (use types.late() to avoid circular imports)");
    for (const refModel of referencedModels) {
      const refFileName = toFileName4(refModel);
      lines.push(`import { ${refModel}Model } from "./${refFileName}.model"`);
    }
  }
  lines.push("");
  lines.push("// ============================================================================");
  lines.push(`// ${modelName} Model`);
  lines.push("// ============================================================================");
  lines.push("");
  lines.push(`export const ${modelName}Model = types`);
  lines.push(`  .model("${modelName}", {`);
  for (const field of scalarFields) {
    const mstType = mapPrismaToMSTType(field, enums);
    const wrappedType = wrapOptional(mstType, field);
    if (field.isList) {
      lines.push(`    ${field.name}: types.optional(${mstType}, []),`);
    } else if (field.hasDefaultValue && !field.isId) {
      lines.push(`    ${field.name}: types.optional(${mstType}, ${getDefaultValue(field)}),`);
    } else if (!field.isRequired && !field.isId) {
      if (field.kind === "enum") {
        lines.push(`    ${field.name}: types.maybeNull(${mstType}),`);
      } else {
        lines.push(`    ${field.name}: types.optional(${mstType}, ${getDefaultValue(field)}),`);
      }
    } else {
      lines.push(`    ${field.name}: ${wrappedType},`);
    }
  }
  for (const field of includedRelationFields) {
    const mstType = mapPrismaToMSTType(field, enums);
    if (field.isList) {
      lines.push(`    ${field.name}: types.optional(${mstType}, []),`);
    } else if (!field.isRequired) {
      lines.push(`    ${field.name}: ${mstType},`);
    } else {
      lines.push(`    ${field.name}: ${mstType},`);
    }
  }
  lines.push("  })");
  lines.push("  .views(self => ({");
  lines.push("    /** Check if this is a new/unsaved entity */");
  lines.push("    get isNew(): boolean {");
  lines.push('      return self.id.startsWith("temp-")');
  lines.push("    },");
  lines.push("  }))");
  lines.push("  .actions(self => ({");
  lines.push("    /** Update entity fields */");
  lines.push("    update(changes: Partial<SnapshotIn<typeof self>>) {");
  lines.push("      Object.assign(self, changes)");
  lines.push("    },");
  lines.push("  }))");
  lines.push("");
  lines.push("// Type exports");
  lines.push(`export interface I${modelName} extends Instance<typeof ${modelName}Model> {}`);
  lines.push(`export interface I${modelName}SnapshotIn extends SnapshotIn<typeof ${modelName}Model> {}`);
  lines.push(`export interface I${modelName}SnapshotOut extends SnapshotOut<typeof ${modelName}Model> {}`);
  lines.push("");
  return {
    modelName,
    fileName,
    code: lines.join("\n")
  };
}
function getDefaultValue(field) {
  const literal = getLiteralDefault(field);
  if (literal !== void 0) return literal;
  if (field.kind === "enum") {
    return "undefined";
  }
  switch (field.type) {
    case "String":
      return '""';
    case "Int":
    case "Float":
    case "Decimal":
      return "0";
    case "Boolean":
      return "false";
    case "DateTime":
      return "0";
    case "Json":
      return "{}";
    default:
      return "undefined";
  }
}
function getLiteralDefault(field) {
  if (!field.hasDefaultValue || field.default === void 0) return void 0;
  const d = field.default;
  if (typeof d === "string") return JSON.stringify(d);
  if (typeof d === "number") return String(d);
  if (typeof d === "boolean") return String(d);
  return void 0;
}
function generateMSTModels(models, allModels, enums = [], fileExtension = "tsx") {
  const includedModelNames = new Set(models.map((m) => m.name));
  const modelsForReference = allModels ?? models;
  return models.filter((model) => getIdField(model)).map((model) => generateMSTModel(model, modelsForReference, enums, includedModelNames, fileExtension));
}

// src/generators/mst-collection-generator.ts
function toFileName5(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function toRoutePath3(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function generateMSTCollection(model, fileExtension = "tsx") {
  const modelName = model.name;
  const modelLower = toCamelCase2(modelName);
  const fileName = `${toFileName5(modelName)}.collection.${fileExtension}`;
  const routePath = toRoutePath3(modelName);
  const relationFields = getRelationFields2(model);
  const relationFieldNames = relationFields.map((f) => f.name);
  const dateFieldNames = model.fields.filter((f) => f.kind === "scalar" && f.type === "DateTime" && !f.isList).map((f) => f.name);
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    ` * Auto-generated ${modelName} Collection`,
    " *",
    " * MST collection with CRUD operations, optimistic updates, and mixins.",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { types, flow, getEnv, getSnapshot, applySnapshot, Instance } from "mobx-state-tree"',
    `import { ${modelName}Model, type I${modelName}, type I${modelName}SnapshotIn } from "./${toFileName5(modelName)}.model"`,
    'import type { ISDKEnvironment } from "@shogo-ai/sdk"',
    "",
    "// ============================================================================",
    `// ${modelName} Collection`,
    "// ============================================================================",
    "",
    `const ENDPOINT = "/api/${routePath}"`,
    "",
    "/**",
    ` * ${modelName} Collection Model`,
    " *",
    " * Features:",
    " * - Items stored in Map<id, Entity>",
    " * - Loading/error state tracking",
    " * - CRUD with optimistic updates and rollback",
    " * - Persistence via SDK HttpClient",
    " */",
    `export const ${modelName}Collection = types`,
    `  .model("${modelName}Collection", {`,
    `    items: types.map(${modelName}Model),`,
    "    isLoading: types.optional(types.boolean, false),",
    "    isLoadingMore: types.optional(types.boolean, false),",
    "    error: types.maybeNull(types.string),",
    "    total: types.optional(types.number, 0),",
    "    hasMore: types.optional(types.boolean, false),",
    "  })",
    "",
    "  // =========================================================================",
    "  // Views",
    "  // =========================================================================",
    "  .views(self => ({",
    "    /** Get all items as array */",
    `    get all(): I${modelName}[] {`,
    "      return Array.from(self.items.values())",
    "    },",
    "",
    "    /** Get item by ID */",
    `    get(id: string): I${modelName} | undefined {`,
    "      return self.items.get(id)",
    "    },",
    "",
    "    /** Get item count */",
    "    get count(): number {",
    "      return self.items.size",
    "    },",
    "",
    "    /** Model name for persistence context */",
    "    get modelName(): string {",
    `      return "${modelName}"`,
    "    },",
    "",
    "    /** Filter items by predicate */",
    `    filter(predicate: (item: I${modelName}) => boolean): I${modelName}[] {`,
    "      return this.all.filter(predicate)",
    "    },",
    "",
    "    /** Find first item matching predicate */",
    `    find(predicate: (item: I${modelName}) => boolean): I${modelName} | undefined {`,
    "      return this.all.find(predicate)",
    "    },",
    "  }))",
    "",
    "  // =========================================================================",
    "  // Actions",
    "  // =========================================================================",
    "  .actions(self => {",
    "    // Track pending operations for optimistic updates",
    "    const pendingCreates = new Map<string, any>()",
    "    const pendingUpdates = new Map<string, any>()",
    "    const pendingDeletes = new Set<string>()",
    "",
    "    return {",
    "      /** Set loading state */",
    "      setLoading(loading: boolean) {",
    "        self.isLoading = loading",
    "      },",
    "",
    "      /** Set error state */",
    "      setError(error: string | null) {",
    "        self.error = error",
    "      },",
    "",
    "      /** Clear error state */",
    "      clearError() {",
    "        self.error = null",
    "      },",
    "",
    "      /** Add an item directly (for external updates like WebSocket) */",
    `      addItem(item: I${modelName}SnapshotIn) {`,
    "        self.items.put(item)",
    "      },",
    "",
    "      /** Remove an item directly */",
    "      removeItem(id: string) {",
    "        self.items.delete(id)",
    "      },",
    "",
    "      /** Check if item has pending operation */",
    "      isPending(id: string): boolean {",
    "        return pendingCreates.has(id) || pendingUpdates.has(id) || pendingDeletes.has(id)",
    "      },",
    "",
    "      // =====================================================================",
    "      // CRUD Operations with Optimistic Updates",
    "      // =====================================================================",
    "",
    "      /**",
    "       * Load all items from API",
    "       * Uses merge strategy to preserve existing MST node references",
    "       */",
    "      loadAll: flow(function* (filter?: Record<string, any>) {",
    "        self.isLoading = true",
    "        self.error = null",
    "",
    "        try {",
    "          const env = getEnv<ISDKEnvironment>(self)",
    '          const params = filter ? new URLSearchParams(filter as any).toString() : ""',
    "          const url = params ? `${ENDPOINT}?${params}` : ENDPOINT",
    "",
    "          const response = yield env.http.get<{ ok: boolean; items?: any[] }>(url)",
    "",
    "          if (response.data?.ok && response.data.items) {",
    "            // Use merge strategy to preserve existing node references",
    "            const newIds = new Set<string>()",
    "",
    "            for (const item of response.data.items) {",
    "              const transformed = transformForMST(item)",
    "              const id = transformed.id",
    "              newIds.add(id)",
    "",
    "              const existing = self.items.get(id)",
    "              if (existing) {",
    "                // Update existing item in place (preserves node reference)",
    "                applySnapshot(existing, transformed)",
    "              } else {",
    "                // Add new item",
    "                self.items.put(transformed)",
    "              }",
    "            }",
    "",
    "            // Remove items that are no longer in the response",
    "            for (const id of self.items.keys()) {",
    "              if (!newIds.has(id)) {",
    "                self.items.delete(id)",
    "              }",
    "            }",
    "          }",
    "",
    "          self.isLoading = false",
    "          return self.all",
    "        } catch (error: any) {",
    '          self.error = error.message || "Failed to load"',
    "          self.isLoading = false",
    "          throw error",
    "        }",
    "      }),",
    "",
    "      /**",
    "       * Load a page of items (append mode for infinite scroll).",
    "       * API returns items in desc order (newest first).",
    "       * On offset=0 (first page), clears the map to handle refresh.",
    "       */",
    "      loadPage: flow(function* (",
    "        filter?: Record<string, any>,",
    "        pagination?: { limit?: number; offset?: number },",
    "      ) {",
    "        const limit = pagination?.limit ?? 50",
    "        const offset = pagination?.offset ?? 0",
    "        const isFirstPage = offset === 0",
    "",
    "        if (isFirstPage) {",
    "          self.isLoading = true",
    "        } else {",
    "          self.isLoadingMore = true",
    "        }",
    "        self.error = null",
    "",
    "        try {",
    "          const env = getEnv<ISDKEnvironment>(self)",
    "          const params = new URLSearchParams({",
    "            ...(filter as any),",
    "            limit: String(limit),",
    "            offset: String(offset),",
    "          })",
    "          const url = `${ENDPOINT}?${params.toString()}`",
    "",
    "          const response = yield env.http.get<{",
    "            ok: boolean",
    "            items?: any[]",
    "            total?: number",
    "          }>(url)",
    "",
    "          if (response.data?.ok && response.data.items) {",
    "            if (isFirstPage) {",
    "              self.items.clear()",
    "            }",
    "",
    "            for (const item of response.data.items) {",
    "              const transformed = transformForMST(item)",
    "              const id = transformed.id",
    "              const existing = self.items.get(id)",
    "              if (existing) {",
    "                applySnapshot(existing, transformed)",
    "              } else {",
    "                self.items.put(transformed)",
    "              }",
    "            }",
    "",
    "            const serverTotal = response.data.total ?? 0",
    "            self.total = serverTotal",
    "            self.hasMore = offset + response.data.items.length < serverTotal",
    "          }",
    "",
    "          self.isLoading = false",
    "          self.isLoadingMore = false",
    "          return self.all",
    "        } catch (error: any) {",
    '          self.error = error.message || "Failed to load page"',
    "          self.isLoading = false",
    "          self.isLoadingMore = false",
    "          throw error",
    "        }",
    "      }),",
    "",
    "      /**",
    "       * Load single item by ID",
    "       */",
    "      loadById: flow(function* (id: string) {",
    "        try {",
    "          const env = getEnv<ISDKEnvironment>(self)",
    "          const response = yield env.http.get<{ ok: boolean; data?: any }>(`${ENDPOINT}/${id}`)",
    "",
    "          if (response.data?.ok && response.data.data) {",
    "            const transformed = transformForMST(response.data.data)",
    "            self.items.put(transformed)",
    "            return self.items.get(id)",
    "          }",
    "",
    "          return undefined",
    "        } catch (error: any) {",
    '          self.error = error.message || "Failed to load"',
    "          throw error",
    "        }",
    "      }),",
    "",
    "      /**",
    "       * Create new item with optimistic update",
    "       */",
    `      create: flow(function* (input: Partial<I${modelName}SnapshotIn>) {`,
    "        const tempId = `temp-${crypto.randomUUID()}`",
    "        const now = Date.now()",
    "",
    "        // Create optimistic item",
    "        const optimistic = {",
    "          id: tempId,",
    "          ...input,",
    "          createdAt: now,",
    "          updatedAt: now,",
    "        }",
    "",
    "        // Add optimistically",
    "        pendingCreates.set(tempId, optimistic)",
    "        self.items.put(optimistic as any)",
    "        self.error = null",
    "",
    "        try {",
    "          const env = getEnv<ISDKEnvironment>(self)",
    "          const response = yield env.http.post<{ ok: boolean; data?: any }>(ENDPOINT, input)",
    "",
    "          if (!response.data?.ok || !response.data.data) {",
    '            throw new Error("Create failed")',
    "          }",
    "",
    "          // Replace temp with real item",
    "          const realItem = transformForMST(response.data.data)",
    "          pendingCreates.delete(tempId)",
    "          self.items.delete(tempId)",
    "          self.items.put(realItem)",
    "",
    "          return self.items.get(realItem.id)",
    "        } catch (error: any) {",
    "          // Rollback",
    "          pendingCreates.delete(tempId)",
    "          self.items.delete(tempId)",
    '          self.error = error.message || "Create failed"',
    "          throw error",
    "        }",
    "      }),",
    "",
    "      /**",
    "       * Update item with optimistic update",
    "       */",
    `      update: flow(function* (id: string, changes: Partial<I${modelName}SnapshotIn>) {`,
    "        const existing = self.items.get(id)",
    "        if (!existing) {",
    '          throw new Error("Item not found")',
    "        }",
    "",
    "        // Don't allow concurrent updates",
    "        if (pendingUpdates.has(id)) {",
    '          throw new Error("Update already in progress")',
    "        }",
    "",
    "        // Save previous state for rollback",
    "        const previousState = getSnapshot(existing)",
    "        pendingUpdates.set(id, previousState)",
    "",
    "        // Apply optimistically",
    "        existing.update({ ...changes, updatedAt: Date.now() })",
    "        self.error = null",
    "",
    "        try {",
    "          const env = getEnv<ISDKEnvironment>(self)",
    "          const response = yield env.http.patch<{ ok: boolean; data?: any }>(`${ENDPOINT}/${id}`, changes)",
    "",
    "          if (!response.data?.ok || !response.data.data) {",
    '            throw new Error("Update failed")',
    "          }",
    "",
    "          // Apply server response",
    "          const serverData = transformForMST(response.data.data)",
    "          pendingUpdates.delete(id)",
    "          applySnapshot(existing, serverData)",
    "",
    "          return existing",
    "        } catch (error: any) {",
    "          // Rollback",
    "          const prev = pendingUpdates.get(id)",
    "          pendingUpdates.delete(id)",
    "          if (prev) {",
    "            applySnapshot(existing, prev)",
    "          }",
    '          self.error = error.message || "Update failed"',
    "          throw error",
    "        }",
    "      }),",
    "",
    "      /**",
    "       * Delete item with optimistic update",
    "       */",
    "      delete: flow(function* (id: string) {",
    "        const existing = self.items.get(id)",
    "        if (!existing) return",
    "",
    "        // Don't allow concurrent deletes",
    "        if (pendingDeletes.has(id)) return",
    "",
    "        // Save for rollback",
    "        const previousState = getSnapshot(existing)",
    "        pendingDeletes.add(id)",
    "",
    "        // Remove optimistically",
    "        self.items.delete(id)",
    "        self.error = null",
    "",
    "        try {",
    "          const env = getEnv<ISDKEnvironment>(self)",
    "          yield env.http.delete(`${ENDPOINT}/${id}`)",
    "",
    "          pendingDeletes.delete(id)",
    "        } catch (error: any) {",
    "          // Rollback",
    "          pendingDeletes.delete(id)",
    "          self.items.put(previousState as any)",
    '          self.error = error.message || "Delete failed"',
    "          throw error",
    "        }",
    "      }),",
    "",
    "      /** Clear all items */",
    "      clear() {",
    "        self.items.clear()",
    "        self.error = null",
    "        self.isLoading = false",
    "      },",
    "    }",
    "  })",
    "",
    "// ============================================================================",
    "// Helpers",
    "// ============================================================================",
    "",
    `// Relation fields that expect IDs (safeReference)`,
    `const relationFields = ${JSON.stringify(relationFieldNames)}`,
    "",
    "/**",
    " * Transform API response for MST compatibility:",
    " * - Convert ISO date strings to timestamps",
    " * - Extract IDs from nested relation objects (for safeReference)",
    " * - Handle null -> undefined for optional fields",
    " */",
    "function transformForMST(obj: any): any {",
    '  if (!obj || typeof obj !== "object") return obj',
    "",
    `  const dateFields = ${JSON.stringify(dateFieldNames)}`,
    "  const result: Record<string, any> = {}",
    "",
    "  for (const [key, value] of Object.entries(obj)) {",
    "    // Skip null values (MST uses undefined)",
    "    if (value === null) continue",
    "",
    "    // Convert date strings to timestamps",
    '    if (dateFields.includes(key) && typeof value === "string") {',
    "      result[key] = new Date(value).getTime()",
    "    }",
    "    // For relation fields, extract only the ID (for safeReference)",
    '    else if (relationFields.includes(key) && value && typeof value === "object" && !Array.isArray(value) && value.id) {',
    "      result[key] = value.id",
    "    }",
    "    // For array relation fields, extract IDs",
    "    else if (relationFields.includes(key) && Array.isArray(value)) {",
    '      result[key] = value.map((item: any) => (item && typeof item === "object" && item.id) ? item.id : item)',
    "    }",
    "    // Skip nested objects that are not relations (they might be JSON fields)",
    '    else if (value && typeof value === "object" && !Array.isArray(value)) {',
    "      // Only include if it has an id (likely a reference) or is a plain data object",
    "      if (value.id && !relationFields.includes(key)) {",
    "        result[key] = value // Keep as-is for JSON fields",
    "      } else if (!value.id) {",
    "        result[key] = value // Keep plain objects (like metadata)",
    "      }",
    "      // If it has an id but is a relation field, we already handled it above",
    "    }",
    "    else {",
    "      result[key] = value",
    "    }",
    "  }",
    "",
    "  return result",
    "}",
    "",
    "// Type exports",
    `export type I${modelName}Collection = Instance<typeof ${modelName}Collection>`,
    ""
  ];
  return {
    modelName,
    fileName,
    code: lines.join("\n")
  };
}
function generateMSTCollections(models, fileExtension = "tsx") {
  return models.filter((model) => getIdField(model)).map((model) => generateMSTCollection(model, fileExtension));
}

// src/generators/mst-domain-generator.ts
function toFileName6(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function generateMSTDomain(models, fileExtension = "tsx") {
  const validModels = models.filter((model) => getIdField(model));
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Domain Store",
    " *",
    " * Root MST store combining all collections.",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { types, Instance, SnapshotIn, SnapshotOut } from "mobx-state-tree"',
    'import type { HttpClient } from "@shogo-ai/sdk"',
    "",
    "// Collection imports"
  ];
  for (const model of validModels) {
    const fileName = toFileName6(model.name);
    lines.push(`import { ${model.name}Collection, type I${model.name}Collection } from "./${fileName}.collection"`);
  }
  lines.push("");
  lines.push("// Model imports (for type re-exports)");
  for (const model of validModels) {
    const fileName = toFileName6(model.name);
    lines.push(`export { ${model.name}Model, type I${model.name}, type I${model.name}SnapshotIn, type I${model.name}SnapshotOut } from "./${fileName}.model"`);
  }
  lines.push("");
  lines.push("// Collection type re-exports");
  for (const model of validModels) {
    lines.push(`export { ${model.name}Collection, type I${model.name}Collection }`);
  }
  lines.push("");
  lines.push("// ============================================================================");
  lines.push("// SDK Environment Type");
  lines.push("// ============================================================================");
  lines.push("");
  lines.push("/**");
  lines.push(" * Environment interface for SDK MST stores.");
  lines.push(" * Inject this when creating the store.");
  lines.push(" */");
  lines.push("export interface ISDKEnvironment {");
  lines.push("  /** HTTP client for API calls */");
  lines.push("  http: HttpClient");
  lines.push("  /** Optional context for authorization */");
  lines.push("  context?: {");
  lines.push("    userId?: string");
  lines.push("    workspaceId?: string");
  lines.push("  }");
  lines.push("}");
  lines.push("");
  lines.push("// ============================================================================");
  lines.push("// Root Store");
  lines.push("// ============================================================================");
  lines.push("");
  lines.push("/**");
  lines.push(" * Root domain store combining all collections.");
  lines.push(" */");
  lines.push("export const DomainStore = types");
  lines.push('  .model("DomainStore", {');
  for (const model of validModels) {
    const propName = toCamelCase2(model.name) + "Collection";
    lines.push(`    ${propName}: types.optional(${model.name}Collection, { items: {} }),`);
  }
  lines.push("  })");
  lines.push("");
  lines.push("  .views(self => ({");
  lines.push("    /** Get all collection names */");
  lines.push("    get collectionNames(): string[] {");
  lines.push("      return [");
  for (const model of validModels) {
    lines.push(`        "${toCamelCase2(model.name)}Collection",`);
  }
  lines.push("      ]");
  lines.push("    },");
  lines.push("  }))");
  lines.push("");
  lines.push("  .actions(self => ({");
  lines.push("    /** Clear all collections */");
  lines.push("    clearAll() {");
  for (const model of validModels) {
    const propName = toCamelCase2(model.name) + "Collection";
    lines.push(`      self.${propName}.clear()`);
  }
  lines.push("    },");
  lines.push("");
  lines.push("    /** Clear all errors */");
  lines.push("    clearAllErrors() {");
  for (const model of validModels) {
    const propName = toCamelCase2(model.name) + "Collection";
    lines.push(`      self.${propName}.clearError()`);
  }
  lines.push("    },");
  lines.push("  }))");
  lines.push("");
  lines.push("// Type exports");
  lines.push("export type IDomainStore = Instance<typeof DomainStore>");
  lines.push("export type IDomainStoreSnapshotIn = SnapshotIn<typeof DomainStore>");
  lines.push("export type IDomainStoreSnapshotOut = SnapshotOut<typeof DomainStore>");
  lines.push("");
  lines.push("// ============================================================================");
  lines.push("// Factory Functions");
  lines.push("// ============================================================================");
  lines.push("");
  lines.push("/** Singleton store instance */");
  lines.push("let _store: IDomainStore | null = null");
  lines.push("");
  lines.push("/**");
  lines.push(" * Create a new domain store instance.");
  lines.push(" *");
  lines.push(" * @param env - SDK environment with HttpClient");
  lines.push(" * @returns New domain store instance");
  lines.push(" *");
  lines.push(" * @example");
  lines.push(" * ```typescript");
  lines.push(' * import { createDomainStore, type ISDKEnvironment } from "./generated/domain"');
  lines.push(' * import { HttpClient } from "@shogo-ai/sdk"');
  lines.push(" *");
  lines.push(' * const http = new HttpClient({ baseUrl: "http://localhost:3000" })');
  lines.push(" * const store = createDomainStore({ http })");
  lines.push(" *");
  lines.push(" * // Load data");
  lines.push(" * await store.workspaceCollection.loadAll()");
  lines.push(" * console.log(store.workspaceCollection.all)");
  lines.push(" * ```");
  lines.push(" */");
  lines.push("export function createDomainStore(env: ISDKEnvironment): IDomainStore {");
  lines.push("  return DomainStore.create({}, env)");
  lines.push("}");
  lines.push("");
  lines.push("/**");
  lines.push(" * Get or create singleton domain store.");
  lines.push(" *");
  lines.push(" * @param env - SDK environment (required on first call)");
  lines.push(" * @returns Singleton domain store instance");
  lines.push(" */");
  lines.push("export function getDomainStore(env?: ISDKEnvironment): IDomainStore {");
  lines.push("  if (!_store) {");
  lines.push("    if (!env) {");
  lines.push('      throw new Error("SDK environment required to initialize domain store")');
  lines.push("    }");
  lines.push("    _store = createDomainStore(env)");
  lines.push("  }");
  lines.push("  return _store");
  lines.push("}");
  lines.push("");
  lines.push("/**");
  lines.push(" * Reset singleton store (for testing or auth changes)");
  lines.push(" */");
  lines.push("export function resetDomainStore(): void {");
  lines.push("  _store = null");
  lines.push("}");
  lines.push("");
  return {
    fileName: `domain.${fileExtension}`,
    code: lines.join("\n")
  };
}

// src/generators/server-generator.ts
function generateServer(config = {}) {
  const {
    port = 3e3,
    cors = true,
    corsOrigins = ["http://localhost:5173", "http://localhost:3000"],
    routesPath = "./src/generated",
    dbPath = "./src/lib/db",
    apiBasePath = "/api",
    staticDir = "./dist",
    skipStatic = false,
    customRoutesPath,
    dynamicCrudImport = false,
    bunServe = false,
    tools = true
  } = config;
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Hono Server",
    " *",
    " * Auto-generated by @shogo-ai/sdk",
    " * This file can be customized - it will not be overwritten if it exists.",
    " */",
    "",
    "import { Hono } from 'hono'"
  ];
  if (!skipStatic) {
    lines.push("import { serveStatic } from 'hono/bun'");
  }
  if (!dynamicCrudImport) {
    lines.push(`import { createAllRoutes } from '${routesPath}'`);
    lines.push(`import { prisma } from '${dbPath}'`);
  }
  if (customRoutesPath) {
    lines.push(`import customRoutes from '${customRoutesPath}'`);
  }
  if (tools) {
    lines.push("import { createToolsHandlers } from '@shogo-ai/sdk/tools/server'");
  }
  lines.push("");
  lines.push("const app = new Hono()");
  lines.push("");
  if (cors) {
    lines.push("// CORS \u2014 manual middleware so the wildcard always propagates");
    lines.push("app.use('*', async (c, next) => {");
    lines.push("  c.res.headers.set('Access-Control-Allow-Origin', '*')");
    lines.push("  c.res.headers.set('Access-Control-Allow-Methods', 'GET,POST,PUT,PATCH,DELETE,OPTIONS')");
    lines.push("  c.res.headers.set('Access-Control-Allow-Headers', 'Content-Type,Authorization')");
    lines.push("  if (c.req.method === 'OPTIONS') return c.text('', 204)");
    lines.push("  await next()");
    lines.push("})");
    lines.push("");
  }
  lines.push("// Health check endpoint");
  lines.push("app.get('/health', (c) => c.json({ ok: true, timestamp: new Date().toISOString() }))");
  lines.push("");
  if (dynamicCrudImport) {
    lines.push("// CRUD routes (available after schema.prisma has models)");
    lines.push("try {");
    lines.push(`  const { createAllRoutes } = await import('${routesPath}')`);
    lines.push(`  const { prisma } = await import('${dbPath}')`);
    lines.push(`  app.route('${apiBasePath}', createAllRoutes(prisma))`);
    lines.push("} catch {");
    lines.push("  // No generated routes yet \u2014 CRUD routes will appear after schema.prisma is written");
    lines.push("}");
  } else {
    lines.push("// Mount SDK-generated API routes");
    lines.push(`app.route('${apiBasePath}', createAllRoutes(prisma))`);
  }
  lines.push("");
  if (customRoutesPath) {
    lines.push("// Custom API routes (always mounted)");
    lines.push(`app.route('${apiBasePath}', customRoutes)`);
    lines.push("");
  }
  if (tools) {
    lines.push("// Installed-integration tools proxy \u2192 forwards to the agent runtime");
    lines.push("// using RUNTIME_AUTH_SECRET + RUNTIME_PORT (auto-detected from env).");
    lines.push("const tools = createToolsHandlers({})");
    lines.push(`app.post('${apiBasePath}/tools/execute', (c) => tools.execute(c.req.raw))`);
    lines.push(`app.get('${apiBasePath}/tools/schemas', (c) => tools.list(c.req.raw))`);
    lines.push("");
  }
  if (!skipStatic) {
    lines.push("// Serve static files in production");
    lines.push(`app.use('/*', serveStatic({ root: '${staticDir}' }))`);
    lines.push(`app.get('*', serveStatic({ path: '${staticDir}/index.html' }))`);
    lines.push("");
  }
  lines.push(`const port = Number(process.env.PORT) || ${port}`);
  lines.push("console.log(`\u{1F680} Server running on http://localhost:${port}`)");
  lines.push("");
  if (bunServe) {
    lines.push("Bun.serve({ port, fetch: app.fetch })");
  } else {
    lines.push("export default {");
    lines.push("  port,");
    lines.push("  fetch: app.fetch,");
    lines.push("}");
  }
  lines.push("");
  return lines.join("\n");
}
function generateDbModule() {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Prisma Database Client",
    " *",
    " * Auto-generated by @shogo-ai/sdk",
    " * This file can be customized - it will not be overwritten if it exists.",
    " */",
    "",
    "import { PrismaPg } from '@prisma/adapter-pg'",
    "import { PrismaClient } from '../generated/prisma/client'",
    "",
    "const globalForPrisma = globalThis as unknown as {",
    "  prisma: PrismaClient | undefined",
    "}",
    "",
    "const adapter = new PrismaPg({",
    "  connectionString: process.env.DATABASE_URL,",
    "})",
    "",
    "export const prisma =",
    "  globalForPrisma.prisma ??",
    "  new PrismaClient({",
    "    adapter,",
    "    log: process.env.NODE_ENV === 'development' ? ['error', 'warn'] : ['error'],",
    "  })",
    "",
    "if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma",
    ""
  ];
  return lines.join("\n");
}
function generateSqliteDbModule() {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Prisma Database Client (SQLite via Bun adapter)",
    " *",
    " * Auto-generated by @shogo-ai/sdk",
    " * This file can be customized - it will not be overwritten if it exists.",
    " */",
    "",
    "import { PrismaBunSqlite } from 'prisma-adapter-bun-sqlite'",
    "import { PrismaClient } from './generated/prisma/client'",
    "",
    "const globalForPrisma = globalThis as unknown as {",
    "  prisma: PrismaClient | undefined",
    "}",
    "",
    "const adapter = new PrismaBunSqlite({",
    "  url: process.env.DATABASE_URL ?? 'file:./skill.db',",
    "})",
    "",
    "export const prisma =",
    "  globalForPrisma.prisma ??",
    "  new PrismaClient({",
    "    adapter,",
    "    log: process.env.NODE_ENV === 'development' ? ['error', 'warn'] : ['error'],",
    "  })",
    "",
    "if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma",
    ""
  ];
  return lines.join("\n");
}

// src/generators/api-client.ts
function toRoutePath4(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function generateApiClient(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated API Client",
    " *",
    " * Typed HTTP client for CRUD operations.",
    " * Generated by `shogo generate` - DO NOT EDIT",
    " */",
    "",
    "import type {"
  ];
  for (const model of models) {
    lines.push(`  ${model.name}Type,`);
    lines.push(`  ${model.name}CreateInput,`);
    lines.push(`  ${model.name}UpdateInput,`);
  }
  lines.push("} from './types'");
  lines.push("");
  lines.push("// ============================================================================");
  lines.push("// Response Types");
  lines.push("// ============================================================================");
  lines.push("");
  lines.push("export interface ApiResponse<T> {");
  lines.push("  ok: boolean");
  lines.push("  data?: T");
  lines.push("  error?: { code: string; message: string }");
  lines.push("}");
  lines.push("");
  lines.push("export interface ApiListResponse<T> {");
  lines.push("  ok: boolean");
  lines.push("  items?: T[]");
  lines.push("  error?: { code: string; message: string }");
  lines.push("}");
  lines.push("");
  lines.push("// ============================================================================");
  lines.push("// Client Configuration");
  lines.push("// ============================================================================");
  lines.push("");
  lines.push("export interface ApiClientConfig {");
  lines.push('  /** Base URL for API requests (e.g., "http://localhost:3000/api") */');
  lines.push("  baseUrl: string");
  lines.push("  /** Optional auth token to include in requests */");
  lines.push("  token?: string");
  lines.push("  /** Optional user ID to include in requests */");
  lines.push("  userId?: string");
  lines.push("}");
  lines.push("");
  lines.push("let config: ApiClientConfig = { baseUrl: '/api' }");
  lines.push("");
  lines.push("/**");
  lines.push(" * Configure the API client");
  lines.push(" */");
  lines.push("export function configureApiClient(newConfig: Partial<ApiClientConfig>) {");
  lines.push("  config = { ...config, ...newConfig }");
  lines.push("}");
  lines.push("");
  lines.push("/**");
  lines.push(" * Get current API client configuration");
  lines.push(" */");
  lines.push("export function getApiConfig(): ApiClientConfig {");
  lines.push("  return { ...config }");
  lines.push("}");
  lines.push("");
  lines.push("// ============================================================================");
  lines.push("// HTTP Helpers");
  lines.push("// ============================================================================");
  lines.push("");
  lines.push("async function request<T>(");
  lines.push("  method: string,");
  lines.push("  path: string,");
  lines.push("  body?: unknown");
  lines.push("): Promise<ApiResponse<T>> {");
  lines.push("  const headers: Record<string, string> = {");
  lines.push("    'Content-Type': 'application/json',");
  lines.push("  }");
  lines.push("");
  lines.push("  if (config.token) {");
  lines.push("    headers['Authorization'] = `Bearer ${config.token}`");
  lines.push("  }");
  lines.push("");
  lines.push("  const url = `${config.baseUrl}${path}`");
  lines.push("");
  lines.push("  try {");
  lines.push("    const response = await fetch(url, {");
  lines.push("      method,");
  lines.push("      headers,");
  lines.push("      body: body ? JSON.stringify(body) : undefined,");
  lines.push("    })");
  lines.push("");
  lines.push("    const data = await response.json()");
  lines.push("");
  lines.push("    if (!response.ok) {");
  lines.push("      return {");
  lines.push("        ok: false,");
  lines.push("        error: data.error || { code: 'request_failed', message: `HTTP ${response.status}` },");
  lines.push("      }");
  lines.push("    }");
  lines.push("");
  lines.push("    return data");
  lines.push("  } catch (error) {");
  lines.push("    return {");
  lines.push("      ok: false,");
  lines.push("      error: {");
  lines.push("        code: 'network_error',");
  lines.push("        message: error instanceof Error ? error.message : 'Network error',");
  lines.push("      },");
  lines.push("    }");
  lines.push("  }");
  lines.push("}");
  lines.push("");
  for (const model of models) {
    lines.push(...generateModelClient(model));
  }
  lines.push(...generateCombinedClient(models));
  return lines.join("\n");
}
function generateModelClient(model) {
  const name = model.name;
  const nameLower = toCamelCase2(name);
  const routePath = toRoutePath4(name);
  const idField = getIdField(model);
  if (!idField) {
    return [`// Skipped ${name} - no @id field found`, ""];
  }
  return [
    `// ============================================================================`,
    `// ${name} API Client`,
    `// ============================================================================`,
    "",
    `export const ${nameLower}Api = {`,
    `  /**`,
    `   * List all ${name} records`,
    `   * @param options - Query options including where filters, pagination, and custom query params`,
    `   * @param options.where - Filter conditions (passed as query params to the API)`,
    `   * @param options.limit - Maximum number of records to return`,
    `   * @param options.offset - Number of records to skip`,
    `   * @param options.params - Additional query parameters to append to the request`,
    `   */`,
    `  async list(options?: { where?: Record<string, unknown>; limit?: number; offset?: number; params?: Record<string, string | number | boolean> }): Promise<ApiListResponse<${name}Type>> {`,
    `    const params = new URLSearchParams()`,
    `    `,
    `    // Add where filters as query params`,
    `    if (options?.where) {`,
    `      for (const [key, value] of Object.entries(options.where)) {`,
    `        if (value !== undefined && value !== null) {`,
    `          params.set(key, String(value))`,
    `        }`,
    `      }`,
    `    }`,
    `    `,
    `    // Add additional params`,
    `    if (options?.params) {`,
    `      for (const [key, value] of Object.entries(options.params)) {`,
    `        if (value !== undefined && value !== null) {`,
    `          params.set(key, String(value))`,
    `        }`,
    `      }`,
    `    }`,
    `    `,
    `    // Add pagination`,
    `    if (options?.limit) params.set('limit', String(options.limit))`,
    `    if (options?.offset) params.set('offset', String(options.offset))`,
    `    if (config.userId) params.set('userId', config.userId)`,
    `    `,
    `    const query = params.toString() ? \`?\${params.toString()}\` : ''`,
    `    return request<${name}Type>('GET', \`/${routePath}\${query}\`) as Promise<ApiListResponse<${name}Type>>`,
    `  },`,
    "",
    `  /**`,
    `   * Get a single ${name} by ID`,
    `   */`,
    `  async get(id: string): Promise<ApiResponse<${name}Type>> {`,
    `    return request<${name}Type>('GET', \`/${routePath}/\${id}\`)`,
    `  },`,
    "",
    `  /**`,
    `   * Create a new ${name}`,
    `   */`,
    `  async create(input: ${name}CreateInput): Promise<ApiResponse<${name}Type>> {`,
    `    const body = config.userId ? { ...input, userId: config.userId } : input`,
    `    return request<${name}Type>('POST', \`/${routePath}\`, body)`,
    `  },`,
    "",
    `  /**`,
    `   * Update an existing ${name}`,
    `   */`,
    `  async update(id: string, input: ${name}UpdateInput): Promise<ApiResponse<${name}Type>> {`,
    `    return request<${name}Type>('PATCH', \`/${routePath}/\${id}\`, input)`,
    `  },`,
    "",
    `  /**`,
    `   * Delete a ${name}`,
    `   */`,
    `  async delete(id: string): Promise<ApiResponse<void>> {`,
    `    return request<void>('DELETE', \`/${routePath}/\${id}\`)`,
    `  },`,
    `}`,
    ""
  ];
}
function generateCombinedClient(models) {
  const lines = [
    "// ============================================================================",
    "// Combined API Client",
    "// ============================================================================",
    "",
    "/**",
    " * Combined API client with all model endpoints",
    " */",
    "export const api = {"
  ];
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    const idField = getIdField(model);
    if (!idField) continue;
    lines.push(`  ${nameLower}: ${nameLower}Api,`);
  }
  lines.push("}");
  lines.push("");
  lines.push("export default api");
  lines.push("");
  return lines;
}

// src/generators/auth-store-generator.ts
function findEmailField(model) {
  return model.fields.find(
    (f) => f.name === "email" || f.name === "emailAddress" || f.type === "String" && f.name.toLowerCase().includes("email")
  );
}
function findNameField(model) {
  return model.fields.find(
    (f) => f.name === "name" || f.name === "displayName" || f.name === "fullName" || f.name === "username"
  );
}
function generateAuthStore(options) {
  const { userModel, storageKey = "shogo-auth-user", apiBase = "/api" } = options;
  const emailField = findEmailField(userModel);
  const nameField = findNameField(userModel);
  if (!emailField) {
    throw new Error(`User model "${userModel.name}" must have an email field for auth generation`);
  }
  const emailFieldName = emailField.name;
  const nameFieldName = nameField?.name || null;
  const userModelName = userModel.name;
  const userEndpoint = userModelName.toLowerCase() + "s";
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auth Store",
    " *",
    " * MobX authentication store with localStorage persistence.",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    "import { makeAutoObservable, runInAction } from 'mobx'",
    "",
    `const API_BASE = '${apiBase}'`,
    `const STORAGE_KEY = '${storageKey}'`,
    "",
    "// ============================================================================",
    "// Types",
    "// ============================================================================",
    "",
    "export interface AuthUser {",
    "  id: string",
    `  ${emailFieldName}: string`
  ];
  if (nameFieldName) {
    const nameRequired = nameField?.isRequired;
    lines.push(`  ${nameFieldName}: string${nameRequired ? "" : " | null"}`);
  }
  lines.push(
    "  createdAt: string | Date",
    "  updatedAt: string | Date",
    "}",
    "",
    "export interface SignUpInput {",
    `  ${emailFieldName}: string`,
    "  password?: string"
  );
  if (nameFieldName) {
    lines.push(`  ${nameFieldName}?: string`);
  }
  lines.push(
    "}",
    "",
    "export interface SignInInput {",
    `  ${emailFieldName}: string`,
    "  password?: string",
    "}",
    "",
    "// ============================================================================",
    "// AuthStore",
    "// ============================================================================",
    "",
    "export class AuthStore {",
    "  user: AuthUser | null = null",
    "  isLoading = false",
    "  error: string | null = null",
    "",
    "  constructor() {",
    "    makeAutoObservable(this)",
    "    this.loadFromStorage()",
    "  }",
    "",
    "  // === Getters ===",
    "",
    "  get isAuthenticated(): boolean {",
    "    return this.user !== null",
    "  }",
    "",
    "  get userId(): string | null {",
    "    return this.user?.id ?? null",
    "  }",
    "",
    "  // === Storage ===",
    "",
    "  private loadFromStorage() {",
    "    try {",
    '      if (typeof window === "undefined") return',
    "      const stored = localStorage.getItem(STORAGE_KEY)",
    "      if (stored) {",
    "        this.user = JSON.parse(stored)",
    "      }",
    "    } catch {",
    "      // Ignore storage errors",
    "    }",
    "  }",
    "",
    "  private saveToStorage(user: AuthUser | null) {",
    "    try {",
    '      if (typeof window === "undefined") return',
    "      if (user) {",
    "        localStorage.setItem(STORAGE_KEY, JSON.stringify(user))",
    "      } else {",
    "        localStorage.removeItem(STORAGE_KEY)",
    "      }",
    "    } catch {",
    "      // Ignore storage errors",
    "    }",
    "  }",
    "",
    "  // === Actions ===",
    "",
    "  /**",
    "   * Sign in with email",
    `   * Looks up user by ${emailFieldName} in the database`,
    "   */",
    "  async signIn(input: SignInInput): Promise<AuthUser> {",
    "    this.isLoading = true",
    "    this.error = null",
    "",
    "    try {",
    `      // Find user by ${emailFieldName}`,
    `      const res = await fetch(\`\${API_BASE}/${userEndpoint}?${emailFieldName}=\${encodeURIComponent(input.${emailFieldName})}\`)`,
    "      const data = await res.json()",
    "",
    "      if (!data.ok) {",
    "        throw new Error(data.error?.message || 'Failed to sign in')",
    "      }",
    "",
    "      const users = data.items || []",
    "      if (users.length === 0) {",
    "        throw new Error('No account found with this email. Please sign up first.')",
    "      }",
    "",
    "      const user = users[0] as AuthUser",
    "",
    "      runInAction(() => {",
    "        this.user = user",
    "        this.isLoading = false",
    "        this.saveToStorage(user)",
    "      })",
    "",
    "      return user",
    "    } catch (err: unknown) {",
    "      const message = err instanceof Error ? err.message : 'Sign in failed'",
    "      runInAction(() => {",
    "        this.error = message",
    "        this.isLoading = false",
    "      })",
    "      throw err",
    "    }",
    "  }",
    "",
    "  /**",
    "   * Sign up with email",
    "   * Creates a new user in the database",
    "   */",
    "  async signUp(input: SignUpInput): Promise<AuthUser> {",
    "    this.isLoading = true",
    "    this.error = null",
    "",
    "    try {",
    "      // Check if user already exists",
    `      const checkRes = await fetch(\`\${API_BASE}/${userEndpoint}?${emailFieldName}=\${encodeURIComponent(input.${emailFieldName})}\`)`,
    "      const checkData = await checkRes.json()",
    "",
    "      if (checkData.ok && checkData.items?.length > 0) {",
    "        throw new Error('An account with this email already exists. Please sign in.')",
    "      }",
    "",
    "      // Create new user",
    `      const res = await fetch(\`\${API_BASE}/${userEndpoint}\`, {`,
    "        method: 'POST',",
    "        headers: { 'Content-Type': 'application/json' },",
    "        body: JSON.stringify({",
    `          ${emailFieldName}: input.${emailFieldName},`
  );
  if (nameFieldName) {
    lines.push(`          ${nameFieldName}: input.${nameFieldName} ?? null,`);
  }
  lines.push(
    "        }),",
    "      })",
    "      const data = await res.json()",
    "",
    "      if (!data.ok) {",
    "        throw new Error(data.error?.message || 'Failed to create account')",
    "      }",
    "",
    "      const user = data.data as AuthUser",
    "",
    "      runInAction(() => {",
    "        this.user = user",
    "        this.isLoading = false",
    "        this.saveToStorage(user)",
    "      })",
    "",
    "      return user",
    "    } catch (err: unknown) {",
    "      const message = err instanceof Error ? err.message : 'Sign up failed'",
    "      runInAction(() => {",
    "        this.error = message",
    "        this.isLoading = false",
    "      })",
    "      throw err",
    "    }",
    "  }",
    "",
    "  /**",
    "   * Sign out",
    "   * Clears user from state and localStorage",
    "   */",
    "  signOut() {",
    "    this.user = null",
    "    this.error = null",
    "    this.saveToStorage(null)",
    "  }",
    "",
    "  /**",
    "   * Clear error state",
    "   */",
    "  clearError() {",
    "    this.error = null",
    "  }",
    "",
    "  /**",
    "   * Refresh user data from server",
    "   */",
    "  async refresh(): Promise<AuthUser | null> {",
    "    if (!this.user) return null",
    "",
    "    try {",
    `      const res = await fetch(\`\${API_BASE}/${userEndpoint}/\${this.user.id}\`)`,
    "      const data = await res.json()",
    "",
    "      if (data.ok && data.data) {",
    "        runInAction(() => {",
    "          this.user = data.data",
    "          this.saveToStorage(data.data)",
    "        })",
    "        return data.data",
    "      }",
    "",
    "      // User not found, sign out",
    "      this.signOut()",
    "      return null",
    "    } catch {",
    "      return this.user",
    "    }",
    "  }",
    "}",
    "",
    "// ============================================================================",
    "// Singleton & Factory",
    "// ============================================================================",
    "",
    "let authStore: AuthStore | null = null",
    "",
    "/**",
    " * Get or create the auth store singleton",
    " */",
    "export function getAuthStore(): AuthStore {",
    "  if (!authStore) {",
    "    authStore = new AuthStore()",
    "  }",
    "  return authStore",
    "}",
    "",
    "/**",
    " * Create a new auth store (for testing or SSR)",
    " */",
    "export function createAuthStore(): AuthStore {",
    "  return new AuthStore()",
    "}",
    "",
    "/**",
    " * Reset the singleton auth store",
    " */",
    "export function resetAuthStore(): void {",
    "  authStore = null",
    "}",
    ""
  );
  return lines.join("\n");
}
function hasUserModel(models) {
  const userModel = models.find(
    (m) => m.name === "User" || m.name === "Account" || m.name.toLowerCase() === "user"
  );
  if (!userModel) return false;
  return !!findEmailField(userModel);
}
function getUserModel(models) {
  return models.find(
    (m) => m.name === "User" || m.name === "Account" || m.name.toLowerCase() === "user"
  );
}

// src/generators/docs-generator.ts
function toRoutePath5(name) {
  const kebab = toKebabCase(name);
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function mapFieldType(field) {
  if (field.kind === "enum") return field.type;
  if (field.kind === "object") return field.isList ? `${field.type}[]` : field.type;
  switch (field.type) {
    case "String":
      return "string";
    case "Int":
    case "Float":
    case "Decimal":
      return "number";
    case "Boolean":
      return "boolean";
    case "DateTime":
      return "Date";
    case "Json":
      return "Record<string, unknown>";
    case "BigInt":
      return "bigint";
    case "Bytes":
      return "Buffer";
    default:
      return "unknown";
  }
}
function describeField(field) {
  const parts = [];
  if (field.isId) parts.push("Primary key");
  if (field.isUnique) parts.push("Unique");
  if (!field.isRequired) parts.push("Nullable");
  if (field.hasDefaultValue) parts.push("Auto-generated");
  if (field.kind === "enum") parts.push(`Enum`);
  if (field.relationName) parts.push(`Relation`);
  return parts.join(", ") || "\u2014";
}
function generateModelDoc(model, allModels, enums, config = {}) {
  const name = model.name;
  const kebab = toKebabCase(name);
  const camel = toCamelCase2(name);
  const routePath = toRoutePath5(name);
  const apiBase = config.apiBasePath || "/api";
  const scalarFields = getScalarFields(model);
  const relationFields = getRelationFields2(model);
  const idField = getIdField(model);
  const lines = [];
  lines.push("---");
  lines.push(`title: ${name}`);
  lines.push(`sidebar_label: ${name}`);
  lines.push(`description: API reference and usage guide for the ${name} model.`);
  lines.push("---");
  lines.push("");
  lines.push(`# ${name}`);
  lines.push("");
  const fieldCount = scalarFields.length;
  const relationCount = relationFields.length;
  lines.push(`The \`${name}\` model has **${fieldCount}** fields${relationCount > 0 ? ` and **${relationCount}** relation${relationCount !== 1 ? "s" : ""}` : ""}.`);
  lines.push("");
  lines.push("## Fields");
  lines.push("");
  lines.push("| Field | Type | Required | Attributes |");
  lines.push("|-------|------|----------|------------|");
  for (const field of scalarFields) {
    const tsType = mapFieldType(field);
    const required = field.isRequired ? "Yes" : "No";
    const attrs = describeField(field);
    lines.push(`| \`${field.name}\` | \`${tsType}\` | ${required} | ${attrs} |`);
  }
  lines.push("");
  const modelEnumTypes = new Set(
    model.fields.filter((f) => f.kind === "enum").map((f) => f.type)
  );
  if (modelEnumTypes.size > 0) {
    const relevantEnums = enums.filter((e) => modelEnumTypes.has(e.name));
    if (relevantEnums.length > 0) {
      lines.push("## Enums");
      lines.push("");
      for (const enumDef of relevantEnums) {
        lines.push(`### ${enumDef.name}`);
        lines.push("");
        lines.push("| Value |");
        lines.push("|-------|");
        for (const val of enumDef.values) {
          lines.push(`| \`${val.name}\` |`);
        }
        lines.push("");
      }
    }
  }
  if (relationFields.length > 0) {
    lines.push("## Relations");
    lines.push("");
    lines.push("| Relation | Type | Kind |");
    lines.push("|----------|------|------|");
    for (const rel of relationFields) {
      const kind = rel.isList ? "Has many" : "Belongs to";
      const type = rel.isList ? `${rel.type}[]` : rel.type;
      const relatedModel = allModels.find((m) => m.name === rel.type);
      const typeDisplay = relatedModel ? `[\`${type}\`](../models/${toKebabCase(rel.type)})` : `\`${type}\``;
      lines.push(`| \`${rel.name}\` | ${typeDisplay} | ${kind} |`);
    }
    lines.push("");
  }
  lines.push("## TypeScript Types");
  lines.push("");
  lines.push("The following types are auto-generated:");
  lines.push("");
  lines.push("```typescript");
  lines.push(`interface ${name}Type {`);
  for (const field of scalarFields) {
    const tsType = mapFieldType(field);
    const opt = !field.isRequired && !field.isId ? "?" : "";
    const nullable = !field.isRequired ? " | null" : "";
    lines.push(`  ${field.name}${opt}: ${tsType}${!opt && nullable ? nullable : ""}`);
  }
  lines.push("}");
  lines.push("");
  lines.push(`interface ${name}CreateInput {`);
  for (const field of scalarFields) {
    if (field.isId && field.hasDefaultValue) continue;
    if (field.name === "createdAt" || field.name === "updatedAt") continue;
    const tsType = mapFieldType(field);
    const opt = !field.isRequired || field.hasDefaultValue ? "?" : "";
    lines.push(`  ${field.name}${opt}: ${tsType}`);
  }
  lines.push("}");
  lines.push("");
  lines.push(`interface ${name}UpdateInput {`);
  for (const field of scalarFields) {
    if (field.isId) continue;
    if (field.name === "createdAt" || field.name === "updatedAt") continue;
    const tsType = mapFieldType(field);
    lines.push(`  ${field.name}?: ${tsType}`);
  }
  lines.push("}");
  lines.push("```");
  lines.push("");
  if (idField) {
    lines.push("## API Endpoints");
    lines.push("");
    lines.push(`All endpoints are relative to \`${apiBase}/${routePath}\`.`);
    lines.push("");
    lines.push("### List all");
    lines.push("");
    lines.push("```http");
    lines.push(`GET ${apiBase}/${routePath}`);
    lines.push("```");
    lines.push("");
    lines.push("**Query parameters:**");
    lines.push("");
    lines.push("| Parameter | Type | Description |");
    lines.push("|-----------|------|-------------|");
    lines.push("| `limit` | `number` | Max items to return |");
    lines.push("| `offset` | `number` | Items to skip (pagination) |");
    for (const field of scalarFields) {
      if (field.isId) continue;
      if (field.name === "createdAt" || field.name === "updatedAt") continue;
      lines.push(`| \`${field.name}\` | \`${mapFieldType(field)}\` | Filter by ${field.name} |`);
    }
    lines.push("");
    lines.push("**Response:** `200 OK`");
    lines.push("");
    lines.push("```json");
    lines.push(`[{ ${scalarFields.slice(0, 3).map((f) => `"${f.name}": ...`).join(", ")}, ... }]`);
    lines.push("```");
    lines.push("");
    lines.push("### Get by ID");
    lines.push("");
    lines.push("```http");
    lines.push(`GET ${apiBase}/${routePath}/:${idField.name}`);
    lines.push("```");
    lines.push("");
    lines.push("**Response:** `200 OK` | `404 Not Found`");
    lines.push("");
    lines.push("### Create");
    lines.push("");
    lines.push("```http");
    lines.push(`POST ${apiBase}/${routePath}`);
    lines.push("```");
    lines.push("");
    lines.push("**Request body:**");
    lines.push("");
    lines.push("```json");
    lines.push("{");
    const createFields2 = scalarFields.filter((f) => {
      if (f.isId && f.hasDefaultValue) return false;
      if (f.name === "createdAt" || f.name === "updatedAt") return false;
      return true;
    });
    for (let i = 0; i < createFields2.length; i++) {
      const f = createFields2[i];
      const example = getExampleValue(f);
      const comma = i < createFields2.length - 1 ? "," : "";
      const optional = !f.isRequired || f.hasDefaultValue ? "  // optional" : "";
      lines.push(`  "${f.name}": ${example}${comma}${optional}`);
    }
    lines.push("}");
    lines.push("```");
    lines.push("");
    lines.push("**Response:** `201 Created`");
    lines.push("");
    lines.push("### Update");
    lines.push("");
    lines.push("```http");
    lines.push(`PATCH ${apiBase}/${routePath}/:${idField.name}`);
    lines.push("```");
    lines.push("");
    lines.push("**Request body:** Any subset of fields from `" + name + "UpdateInput`.");
    lines.push("");
    lines.push("**Response:** `200 OK` | `404 Not Found`");
    lines.push("");
    lines.push("### Delete");
    lines.push("");
    lines.push("```http");
    lines.push(`DELETE ${apiBase}/${routePath}/:${idField.name}`);
    lines.push("```");
    lines.push("");
    lines.push("**Response:** `200 OK` | `404 Not Found`");
    lines.push("");
  }
  lines.push("## Store Usage");
  lines.push("");
  lines.push(`Import the generated store and types to interact with \`${name}\` records:`);
  lines.push("");
  lines.push("```typescript");
  lines.push(`import { ${camel}Store } from './generated/${kebab}.store'`);
  lines.push(`import type { ${name}Type, ${name}CreateInput } from './generated/${kebab}.types'`);
  lines.push("");
  lines.push("// List all");
  lines.push(`const items = await ${camel}Store.fetchAll()`);
  lines.push("");
  lines.push("// Get by ID");
  lines.push(`const item = await ${camel}Store.fetchById('some-id')`);
  lines.push("");
  lines.push("// Create");
  lines.push(`const newItem = await ${camel}Store.create({`);
  for (const f of createFields(scalarFields)) {
    lines.push(`  ${f.name}: ${getExampleValueTS(f)},`);
  }
  lines.push("})");
  lines.push("");
  lines.push("// Update");
  lines.push(`await ${camel}Store.update('some-id', { /* fields to update */ })`);
  lines.push("");
  lines.push("// Delete");
  lines.push(`await ${camel}Store.delete('some-id')`);
  lines.push("```");
  lines.push("");
  if (idField) {
    lines.push("## Hooks");
    lines.push("");
    lines.push(`You can customize server-side behavior for \`${name}\` by editing the hooks file:`);
    lines.push("");
    lines.push("```typescript");
    lines.push(`// ${kebab}.hooks.ts`);
    lines.push(`import type { ${name}Hooks } from './${kebab}.hooks'`);
    lines.push("");
    lines.push(`export const hooks: ${name}Hooks = {`);
    lines.push("  beforeList: async (ctx) => {");
    lines.push("    // Add custom filtering");
    lines.push("    return { ok: true, data: { where: { /* ... */ } } }");
    lines.push("  },");
    lines.push("  beforeCreate: async (ctx) => {");
    lines.push("    // Validate or transform input before creation");
    lines.push("    return { ok: true, data: ctx.body }");
    lines.push("  },");
    lines.push("  afterCreate: async (item, ctx) => {");
    lines.push("    // Side effects after creation (notifications, logging, etc.)");
    lines.push("  },");
    lines.push("}");
    lines.push("```");
    lines.push("");
    lines.push("### Available Hook Points");
    lines.push("");
    lines.push("| Hook | Trigger | Use Case |");
    lines.push("|------|---------|----------|");
    lines.push("| `beforeList` | Before fetching list | Add filters, access control |");
    lines.push("| `beforeGet` | Before fetching by ID | Access control |");
    lines.push("| `beforeCreate` | Before inserting | Validate, transform, set defaults |");
    lines.push("| `afterCreate` | After inserting | Notifications, audit logging |");
    lines.push("| `beforeUpdate` | Before updating | Validate, prevent field changes |");
    lines.push("| `afterUpdate` | After updating | Sync, notifications |");
    lines.push("| `beforeDelete` | Before deleting | Access control, soft-delete |");
    lines.push("| `afterDelete` | After deleting | Cleanup, audit logging |");
    lines.push("");
  }
  return {
    path: `docs/models/${kebab}.md`,
    content: lines.join("\n")
  };
}
function generateModelsIndex(models, config = {}) {
  const projectName = config.projectName || "My App";
  const lines = [];
  lines.push("---");
  lines.push("title: Models Overview");
  lines.push("sidebar_label: Overview");
  lines.push("sidebar_position: 1");
  lines.push("---");
  lines.push("");
  lines.push("# Models Overview");
  lines.push("");
  lines.push(`This section documents all **${models.length}** models generated from the Prisma schema.`);
  lines.push("");
  lines.push("## Models");
  lines.push("");
  lines.push("| Model | Fields | Relations | Description |");
  lines.push("|-------|--------|-----------|-------------|");
  for (const model of models) {
    const scalarCount = getScalarFields(model).length;
    const relationCount = getRelationFields2(model).length;
    const kebab = toKebabCase(model.name);
    lines.push(`| [**${model.name}**](./models/${kebab}) | ${scalarCount} | ${relationCount} | ${model.name} model |`);
  }
  lines.push("");
  lines.push("## Quick Start");
  lines.push("");
  lines.push("Each model page includes:");
  lines.push("");
  lines.push("- **Fields** \u2014 Complete field reference with types and attributes");
  lines.push("- **Relations** \u2014 How this model connects to others");
  lines.push("- **TypeScript Types** \u2014 Generated `Type`, `CreateInput`, and `UpdateInput` interfaces");
  lines.push("- **API Endpoints** \u2014 REST endpoints with request/response examples");
  lines.push("- **Store Usage** \u2014 How to use the generated client-side store");
  lines.push("- **Hooks** \u2014 How to customize server-side CRUD behavior");
  lines.push("");
  return {
    path: "docs/models-overview.md",
    content: lines.join("\n")
  };
}
function generateApiOverview(models, config = {}) {
  const apiBase = config.apiBasePath || "/api";
  const lines = [];
  lines.push("---");
  lines.push("title: API Reference");
  lines.push("sidebar_label: API Reference");
  lines.push("sidebar_position: 2");
  lines.push("---");
  lines.push("");
  lines.push("# API Reference");
  lines.push("");
  lines.push("Complete REST API reference for all generated endpoints.");
  lines.push("");
  lines.push("## Endpoints");
  lines.push("");
  lines.push("| Method | Path | Description |");
  lines.push("|--------|------|-------------|");
  for (const model of models) {
    const idField = getIdField(model);
    if (!idField) continue;
    const routePath = toRoutePath5(model.name);
    const kebab = toKebabCase(model.name);
    lines.push(`| \`GET\` | \`${apiBase}/${routePath}\` | [List all ${model.name}s](./models/${kebab}#list-all) |`);
    lines.push(`| \`GET\` | \`${apiBase}/${routePath}/:${idField.name}\` | [Get ${model.name} by ID](./models/${kebab}#get-by-id) |`);
    lines.push(`| \`POST\` | \`${apiBase}/${routePath}\` | [Create ${model.name}](./models/${kebab}#create) |`);
    lines.push(`| \`PATCH\` | \`${apiBase}/${routePath}/:${idField.name}\` | [Update ${model.name}](./models/${kebab}#update) |`);
    lines.push(`| \`DELETE\` | \`${apiBase}/${routePath}/:${idField.name}\` | [Delete ${model.name}](./models/${kebab}#delete) |`);
  }
  lines.push("");
  lines.push("## Authentication");
  lines.push("");
  lines.push("All endpoints support authentication via the `auth` context. When a user is authenticated,");
  lines.push("`ctx.userId` is available in hooks for access control.");
  lines.push("");
  lines.push("## Query Parameters");
  lines.push("");
  lines.push("List endpoints accept the following common query parameters:");
  lines.push("");
  lines.push("| Parameter | Type | Description |");
  lines.push("|-----------|------|-------------|");
  lines.push("| `limit` | `number` | Maximum number of items to return |");
  lines.push("| `offset` | `number` | Number of items to skip (for pagination) |");
  lines.push("");
  lines.push("Additionally, any model field can be used as a filter parameter.");
  lines.push("");
  lines.push("## Error Responses");
  lines.push("");
  lines.push("| Status | Description |");
  lines.push("|--------|-------------|");
  lines.push("| `400 Bad Request` | Invalid input or hook rejection |");
  lines.push("| `404 Not Found` | Resource not found |");
  lines.push("| `500 Internal Server Error` | Server error |");
  lines.push("");
  return {
    path: "docs/api-reference.md",
    content: lines.join("\n")
  };
}
function generateModelsCategoryMeta() {
  return {
    path: "docs/models/_category_.json",
    content: JSON.stringify({
      label: "Models",
      position: 3,
      collapsed: false,
      link: null
    }, null, 2) + "\n"
  };
}
function generateDocs(models, enums, config = {}) {
  const files = [];
  files.push(generateModelsIndex(models, config));
  files.push(generateApiOverview(models, config));
  files.push(generateModelsCategoryMeta());
  for (const model of models) {
    files.push(generateModelDoc(model, models, enums, config));
  }
  return files;
}
function getExampleValue(field) {
  if (field.kind === "enum") return `"VALUE"`;
  switch (field.type) {
    case "String":
      return field.name.includes("email") ? '"user@example.com"' : field.name.includes("url") || field.name.includes("Url") ? '"https://example.com"' : field.name === "name" ? `"Example ${field.name}"` : `"string"`;
    case "Int":
      return "1";
    case "Float":
    case "Decimal":
      return "1.0";
    case "Boolean":
      return "true";
    case "DateTime":
      return '"2025-01-01T00:00:00Z"';
    case "Json":
      return "{}";
    default:
      return "null";
  }
}
function getExampleValueTS(field) {
  if (field.kind === "enum") return `'VALUE'`;
  switch (field.type) {
    case "String":
      return field.name.includes("email") ? "'user@example.com'" : field.name.includes("url") || field.name.includes("Url") ? "'https://example.com'" : field.name === "name" ? `'Example'` : `'...'`;
    case "Int":
      return "1";
    case "Float":
    case "Decimal":
      return "1.0";
    case "Boolean":
      return "true";
    case "DateTime":
      return "new Date()";
    case "Json":
      return "{}";
    default:
      return "null";
  }
}
function createFields(scalarFields) {
  return scalarFields.filter((f) => {
    if (f.isId && f.hasDefaultValue) return false;
    if (f.name === "createdAt" || f.name === "updatedAt") return false;
    return f.isRequired && !f.hasDefaultValue;
  });
}

// src/generators/docs-site-generator.ts
function generateDocsSiteScaffold(config = {}) {
  const files = [];
  files.push(generatePackageJson(config));
  files.push(generateDocusaurusConfig(config));
  files.push(generateSidebars());
  files.push(generateIntroPage(config));
  files.push(generateLandingPage(config));
  return files;
}
function generatePackageJson(config) {
  const projectName = config.projectName || "my-app";
  const slug = projectName.toLowerCase().replace(/\s+/g, "-");
  const pkg = {
    name: `${slug}-dev-docs`,
    version: "0.0.0",
    private: true,
    scripts: {
      docusaurus: "docusaurus",
      start: "docusaurus start",
      build: "docusaurus build",
      swizzle: "docusaurus swizzle",
      deploy: "docusaurus deploy",
      clear: "docusaurus clear",
      serve: "docusaurus serve",
      "write-translations": "docusaurus write-translations",
      "write-heading-ids": "docusaurus write-heading-ids"
    },
    dependencies: {
      "@docusaurus/core": "3.9.0",
      "@docusaurus/preset-classic": "3.9.0",
      "@mdx-js/react": "^3.0.0",
      "clsx": "^2.0.0",
      "prism-react-renderer": "^2.3.0",
      "react": "^19.0.0",
      "react-dom": "^19.0.0"
    },
    devDependencies: {
      "@docusaurus/module-type-aliases": "3.9.0",
      "@docusaurus/tsconfig": "3.9.0",
      "typescript": "~5.6.0"
    },
    browserslist: {
      production: [">0.5%", "not dead", "not op_mini all"],
      development: [
        "last 3 chrome version",
        "last 3 firefox version",
        "last 5 safari version"
      ]
    },
    engines: {
      node: ">=20.0"
    }
  };
  return {
    path: "package.json",
    content: JSON.stringify(pkg, null, 2) + "\n",
    skipIfExists: true
  };
}
function generateDocusaurusConfig(config) {
  const projectName = config.projectName || "My App";
  const tagline = config.tagline || "Auto-generated developer reference";
  const baseUrl = config.baseUrl || "/";
  const url = config.url || "https://your-docs-site.example.com";
  const content = `${GENERATED_FILE_LICENSE_HEADER}

import { themes as prismThemes } from 'prism-react-renderer'
import type { Config } from '@docusaurus/types'
import type * as Preset from '@docusaurus/preset-classic'

const config: Config = {
  title: '${projectName} Developer Docs',
  tagline: '${tagline}',
  favicon: 'img/favicon.ico',

  url: '${url}',
  baseUrl: '${baseUrl}',

  onBrokenLinks: 'warn',
  onBrokenMarkdownLinks: 'warn',

  i18n: {
    defaultLocale: 'en',
    locales: ['en'],
  },

  presets: [
    [
      'classic',
      {
        docs: {
          sidebarPath: './sidebars.ts',
          routeBasePath: '/',
        },
        blog: false,
        theme: {
          customCss: './src/css/custom.css',
        },
      } satisfies Preset.Options,
    ],
  ],

  themeConfig: {
    navbar: {
      title: '${projectName} Developer Docs',
      items: [
        {
          type: 'docSidebar',
          sidebarId: 'docsSidebar',
          position: 'left',
          label: 'Developer Reference',
        },
      ],
    },
    footer: {
      style: 'dark',
      copyright: \`Built with Docusaurus. Generated by Shogo SDK.\`,
    },
    prism: {
      theme: prismThemes.github,
      darkTheme: prismThemes.dracula,
      additionalLanguages: ['bash', 'json', 'http'],
    },
  } satisfies Preset.ThemeConfig,
}

export default config
`;
  return {
    path: "docusaurus.config.ts",
    content,
    skipIfExists: true
  };
}
function generateSidebars() {
  const content = `${GENERATED_FILE_LICENSE_HEADER}

import type { SidebarsConfig } from '@docusaurus/plugin-content-docs'

const sidebars: SidebarsConfig = {
  docsSidebar: [
    'intro',
    'models-overview',
    'api-reference',
    {
      type: 'autogenerated',
      dirName: 'models',
    },
  ],
}

export default sidebars
`;
  return {
    path: "sidebars.ts",
    content,
    skipIfExists: true
  };
}
function generateIntroPage(config) {
  const projectName = config.projectName || "My App";
  const content = `---
title: Introduction
sidebar_label: Introduction
sidebar_position: 0
slug: /
---

# ${projectName} Developer Docs

Welcome to the auto-generated developer documentation for **${projectName}**.

This documentation is generated from your Prisma schema by the [Shogo SDK](https://github.com/shogo-ai/shogo-ai) and includes:

- **Models** \u2014 Field definitions, types, and relationships
- **API Reference** \u2014 REST endpoints with request/response examples
- **TypeScript Types** \u2014 Generated interfaces for type-safe development
- **Stores** \u2014 Client-side data store usage examples
- **Hooks** \u2014 Server-side customization points

## Getting Started

Browse the sidebar to explore your models, or check the [Models Overview](./models-overview) for a summary.

## Regenerating Docs

These docs are generated automatically. To update them after changing your Prisma schema:

\`\`\`bash
shogo generate
\`\`\`

> **Note:** Files marked with \`skipIfExists\` (like this intro page) won't be overwritten, so your edits are safe.
`;
  return {
    path: "docs/intro.md",
    content,
    skipIfExists: true
  };
}
function generateLandingPage(config) {
  return {
    path: "src/css/custom.css",
    content: `:root {
  --ifm-color-primary: #2e8555;
  --ifm-color-primary-dark: #29784c;
  --ifm-color-primary-darker: #277148;
  --ifm-color-primary-darkest: #205d3b;
  --ifm-color-primary-light: #33925d;
  --ifm-color-primary-lighter: #359962;
  --ifm-color-primary-lightest: #3cad6e;
  --ifm-code-font-size: 95%;
  --docusaurus-highlighted-code-line-bg: rgba(0, 0, 0, 0.1);
}

[data-theme='dark'] {
  --ifm-color-primary: #25c2a0;
  --ifm-color-primary-dark: #21af90;
  --ifm-color-primary-darker: #1fa588;
  --ifm-color-primary-darkest: #1a8870;
  --ifm-color-primary-lighter: #29d5b0;
  --ifm-color-primary-lightest: #32d8b4;
  --docusaurus-highlighted-code-line-bg: rgba(0, 0, 0, 0.3);
}
`,
    skipIfExists: true
  };
}
function generateDocsTsConfig() {
  const content = {
    extends: "@docusaurus/tsconfig",
    compilerOptions: {
      baseUrl: "."
    }
  };
  return {
    path: "tsconfig.json",
    content: JSON.stringify(content, null, 2) + "\n",
    skipIfExists: true
  };
}

// src/generators/admin-routes-generator.ts
function toRoutePath6(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function getSearchableFields(model) {
  return getScalarFields(model).filter((f) => f.type === "String" && !f.isId && !f.isList);
}
function getListRelations(model) {
  return getRelationFields2(model).filter((f) => f.isList);
}
function getSingleRelations(model) {
  return getRelationFields2(model).filter((f) => !f.isList);
}
function generateAdminRoutes(models, config = {}) {
  const ext = config.fileExtension || "ts";
  const fileName = `admin-routes.${ext}`;
  const routeModels = models.filter((m) => getIdField(m));
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Admin Routes",
    " *",
    " * Unrestricted CRUD for all models with pagination, search, and sorting.",
    " * Protect these routes by passing middleware in the config.",
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'import { Hono } from "hono"',
    "",
    "// JSON replacer that coerces native BigInt values to strings so they can",
    "// be serialized by JSON.stringify (and therefore by Hono's default json",
    "// helper, which throws on bare BigInt values). Strings preserve full",
    "// precision for very large counts (e.g. byte totals). Consumers that need",
    "// a numeric should coerce explicitly.",
    "const bigIntReplacer = (_key: string, value: unknown) =>",
    '  typeof value === "bigint" ? value.toString() : value',
    "",
    "// Drop-in replacement for c.json that uses bigIntReplacer. Keeps the same",
    "// (body, status?) ergonomics as c.json so call sites are mechanical.",
    "function sendJson(c: any, body: unknown, status: number = 200) {",
    "  return c.body(JSON.stringify(body, bigIntReplacer), status, {",
    '    "content-type": "application/json; charset=UTF-8",',
    "  })",
    "}",
    "",
    "export interface AdminRoutesConfig {",
    "  /** Prisma client instance */",
    "  prisma: any",
    "  /** Middleware to apply to all admin routes (e.g. auth, role check) */",
    "  middleware?: Array<(c: any, next: () => Promise<void>) => Promise<any>>",
    "}",
    "",
    "/**",
    " * Create admin CRUD routes for all models.",
    " *",
    " * Usage:",
    " * ```ts",
    ' * app.route("/api/admin", createAdminRoutes({',
    " *   prisma,",
    " *   middleware: [authMiddleware, requireAuth, requireSuperAdmin],",
    " * }))",
    " * ```",
    " */",
    "export function createAdminRoutes(config: AdminRoutesConfig): Hono {",
    "  const { prisma, middleware = [] } = config",
    "  const router = new Hono()",
    "",
    '  // SQLite does not support `mode: "insensitive"` on string filters.',
    "  // Detect the provider at startup so search clauses can adapt.",
    "  const isSQLite = (() => {",
    "    try {",
    '      const url = process.env.DATABASE_URL || ""',
    '      return url.startsWith("file:") || url.endsWith(".db") || url.includes("sqlite")',
    "    } catch { return false }",
    "  })()",
    "",
    "  // Apply middleware stack",
    "  for (const mw of middleware) {",
    '    router.use("*", mw)',
    "  }",
    ""
  ];
  for (const model of routeModels) {
    const modelName = model.name;
    const modelLower = toCamelCase2(modelName);
    const routePath = toRoutePath6(modelName);
    const searchFields = getSearchableFields(model);
    const listRelations = getListRelations(model);
    const singleRelations = getSingleRelations(model);
    const scalarFields = getScalarFields(model);
    lines.push(`  // ========== ${modelName} ==========`);
    lines.push("");
    lines.push(`  // LIST ${routePath}`);
    lines.push(`  router.get("/${routePath}", async (c) => {`);
    lines.push("    try {");
    lines.push("      const url = new URL(c.req.url)");
    lines.push('      const page = Math.max(1, parseInt(url.searchParams.get("page") || "1"))');
    lines.push('      const limit = Math.min(Math.max(1, parseInt(url.searchParams.get("limit") || "20")), 100)');
    lines.push('      const search = url.searchParams.get("search") || ""');
    lines.push('      const orderBy = url.searchParams.get("orderBy") || "createdAt"');
    lines.push('      const order = (url.searchParams.get("order") || "desc") as "asc" | "desc"');
    lines.push("");
    if (searchFields.length > 0) {
      lines.push("      let where: any = {}");
      lines.push("      if (search) {");
      lines.push("        const searchFilter = isSQLite");
      lines.push("          ? { contains: search }");
      lines.push('          : { contains: search, mode: "insensitive" as const }');
      lines.push("        where = {");
      lines.push("          OR: [");
      for (const field of searchFields) {
        lines.push(`            { ${field.name}: searchFilter },`);
      }
      lines.push("          ]");
      lines.push("        }");
      lines.push("      }");
    } else {
      lines.push("      const where: any = {}");
    }
    lines.push("");
    lines.push("      // Additional query param filters");
    lines.push('      const reserved = new Set(["page", "limit", "search", "orderBy", "order"])');
    lines.push("      for (const [key, value] of url.searchParams.entries()) {");
    lines.push('        if (!reserved.has(key) && value !== "") {');
    lines.push('          if (value === "true") where[key] = true');
    lines.push('          else if (value === "false") where[key] = false');
    lines.push("          else where[key] = value");
    lines.push("        }");
    lines.push("      }");
    lines.push("");
    const countSelect = listRelations.map((r) => `${r.name}: true`);
    lines.push("      const [items, total] = await Promise.all([");
    lines.push(`        prisma.${modelLower}.findMany({`);
    lines.push("          where,");
    lines.push("          orderBy: { [orderBy]: order },");
    lines.push("          skip: (page - 1) * limit,");
    lines.push("          take: limit,");
    if (countSelect.length > 0) {
      lines.push("          include: {");
      lines.push(`            _count: { select: { ${countSelect.join(", ")} } },`);
      lines.push("          },");
    }
    lines.push("        }),");
    lines.push(`        prisma.${modelLower}.count({ where }),`);
    lines.push("      ])");
    lines.push("");
    const dataKey = modelLower + "s";
    lines.push(`      return sendJson(c, { ok: true, data: { ${dataKey}: items, total, page, limit } })`);
    lines.push("    } catch (error: any) {");
    lines.push(`      console.error("[Admin] List ${modelName} error:", error)`);
    lines.push('      return sendJson(c, { error: { code: "list_failed", message: error.message } }, 500)');
    lines.push("    }");
    lines.push("  })");
    lines.push("");
    lines.push(`  // POST ${routePath}`);
    lines.push(`  router.post("/${routePath}", async (c) => {`);
    lines.push("    try {");
    lines.push("      const body = await c.req.json()");
    lines.push("");
    lines.push(`      const item = await prisma.${modelLower}.create({`);
    lines.push("        data: body,");
    lines.push("      })");
    lines.push("");
    lines.push("      return sendJson(c, { ok: true, data: item }, 201)");
    lines.push("    } catch (error: any) {");
    lines.push(`      console.error("[Admin] Create ${modelName} error:", error)`);
    lines.push('      return sendJson(c, { error: { code: "create_failed", message: error.message } }, 500)');
    lines.push("    }");
    lines.push("  })");
    lines.push("");
    lines.push(`  // GET ${routePath}/:id`);
    lines.push(`  router.get("/${routePath}/:id", async (c) => {`);
    lines.push("    try {");
    lines.push('      const id = c.req.param("id")');
    const hasRelations = singleRelations.length > 0 || listRelations.length > 0;
    if (hasRelations) {
      lines.push(`      const item = await prisma.${modelLower}.findUnique({`);
      lines.push("        where: { id },");
      lines.push("        include: {");
      for (const rel of singleRelations) {
        lines.push(`          ${rel.name}: true,`);
      }
      for (const rel of listRelations) {
        lines.push(`          ${rel.name}: { take: 50 },`);
      }
      lines.push("        },");
      lines.push("      })");
    } else {
      lines.push(`      const item = await prisma.${modelLower}.findUnique({`);
      lines.push("        where: { id },");
      lines.push("      })");
    }
    lines.push("");
    lines.push("      if (!item) {");
    lines.push(`        return sendJson(c, { error: { code: "not_found", message: "${modelName} not found" } }, 404)`);
    lines.push("      }");
    lines.push("");
    lines.push("      return sendJson(c, { ok: true, data: item })");
    lines.push("    } catch (error: any) {");
    lines.push(`      console.error("[Admin] Get ${modelName} error:", error)`);
    lines.push('      return sendJson(c, { error: { code: "get_failed", message: error.message } }, 500)');
    lines.push("    }");
    lines.push("  })");
    lines.push("");
    lines.push(`  // PATCH ${routePath}/:id`);
    lines.push(`  router.patch("/${routePath}/:id", async (c) => {`);
    lines.push("    try {");
    lines.push('      const id = c.req.param("id")');
    lines.push("      const body = await c.req.json()");
    lines.push("");
    lines.push(`      const item = await prisma.${modelLower}.update({`);
    lines.push("        where: { id },");
    lines.push("        data: body,");
    lines.push("      })");
    lines.push("");
    lines.push("      return sendJson(c, { ok: true, data: item })");
    lines.push("    } catch (error: any) {");
    lines.push(`      console.error("[Admin] Update ${modelName} error:", error)`);
    lines.push('      return sendJson(c, { error: { code: "update_failed", message: error.message } }, 500)');
    lines.push("    }");
    lines.push("  })");
    lines.push("");
    lines.push(`  // DELETE ${routePath}/:id`);
    lines.push(`  router.delete("/${routePath}/:id", async (c) => {`);
    lines.push("    try {");
    lines.push('      const id = c.req.param("id")');
    lines.push(`      await prisma.${modelLower}.delete({ where: { id } })`);
    lines.push("      return sendJson(c, { ok: true })");
    lines.push("    } catch (error: any) {");
    lines.push(`      console.error("[Admin] Delete ${modelName} error:", error)`);
    lines.push('      return sendJson(c, { error: { code: "delete_failed", message: error.message } }, 500)');
    lines.push("    }");
    lines.push("  })");
    lines.push("");
  }
  lines.push("  // ========== Stats ==========");
  lines.push("");
  lines.push("  // GET /stats - aggregate counts for all models");
  lines.push('  router.get("/stats", async (c) => {');
  lines.push("    try {");
  lines.push("      const [");
  for (let i = 0; i < routeModels.length; i++) {
    const modelLower = toCamelCase2(routeModels[i].name);
    lines.push(`        ${modelLower}Count${i < routeModels.length - 1 ? "," : ""}`);
  }
  lines.push("      ] = await Promise.all([");
  for (const model of routeModels) {
    const modelLower = toCamelCase2(model.name);
    lines.push(`        prisma.${modelLower}.count(),`);
  }
  lines.push("      ])");
  lines.push("");
  lines.push("      return sendJson(c, {");
  lines.push("        ok: true,");
  lines.push("        data: {");
  for (const model of routeModels) {
    const modelLower = toCamelCase2(model.name);
    lines.push(`          ${modelLower}: ${modelLower}Count,`);
  }
  lines.push("        }");
  lines.push("      })");
  lines.push("    } catch (error: any) {");
  lines.push('      console.error("[Admin] Stats error:", error)');
  lines.push('      return sendJson(c, { error: { code: "stats_failed", message: error.message } }, 500)');
  lines.push("    }");
  lines.push("  })");
  lines.push("");
  lines.push("  return router");
  lines.push("}");
  lines.push("");
  return {
    fileName,
    code: lines.join("\n")
  };
}

// src/generators/shogo-client-generator.ts
function generateShogoClient(options = {}) {
  const {
    dbImportPath = "./db",
    defaultApiUrl = "http://localhost:8002",
    fileExtension = "ts"
  } = options;
  const code = `${GENERATED_FILE_LICENSE_HEADER}
/**
 * Zero-config Shogo SDK client for generated pod apps.
 *
 * This module is produced by \`shogo generate\` under the
 * \`shogo-client\` output kind. Do not edit by hand \u2014 your changes will
 * be overwritten on the next generate. To customise behavior, construct
 * your own client elsewhere in the app.
 *
 * The Shogo-managed pod runtime injects:
 *   - PROJECT_ID           \u2192 the project this pod serves
 *   - RUNTIME_AUTH_SECRET  \u2192 per-project HMAC capability (picked up by
 *                            the SDK automatically; never hard-code)
 *   - SHOGO_API_URL        \u2192 base URL of the Shogo API to call
 *                            (falls back to \`${defaultApiUrl}\`)
 *
 * Usage:
 *   \`\`\`ts
 *   import { shogo, PROJECT_ID } from '@/lib/shogo'
 *   const res = await shogo.voice.telephony!.outboundCall({ to })
 *   \`\`\`
 */

import { createClient } from '@shogo-ai/sdk'
import { prisma } from '${dbImportPath}'

/**
 * The project this pod is scoped to. Guaranteed to be set in production;
 * in local dev a missing value is a configuration bug worth failing loudly.
 */
export const PROJECT_ID = process.env.PROJECT_ID ?? ''
if (!PROJECT_ID && process.env.NODE_ENV === 'production') {
  throw new Error(
    '[shogo] PROJECT_ID is not set. Generated pods should have this injected by the runtime.',
  )
}

/**
 * Singleton Shogo client. Voice telephony auto-detects
 * \`RUNTIME_AUTH_SECRET\` in env and authenticates every outbound call
 * with \`x-runtime-token\`. No API key is required in pod mode.
 */
export const shogo = createClient({
  apiUrl: process.env.SHOGO_API_URL ?? '${defaultApiUrl}',
  db: prisma,
  projectId: PROJECT_ID,
})

export type ShogoClient = typeof shogo
`;
  return {
    fileName: `shogo.${fileExtension}`,
    code
  };
}

// src/generators/voice-components-generator.ts
var HEADER = GENERATED_FILE_LICENSE_HEADER.trim();
function voiceButtonCode(sdkReactImport) {
  return `${HEADER}
/**
 * VoiceButton \u2014 click-to-talk Shogo voice widget.
 *
 * REQUIRES <ShogoVoiceProvider> ABOVE IT IN THE TREE
 * --------------------------------------------------
 * This component calls \`useShogoVoice()\`, which delegates to
 * \`@elevenlabs/react\`'s \`useConversation\` (\u2265 1.1). The provider
 * owns the convai session context, and a missing provider throws:
 *
 *     useRegisterCallbacks must be used within a ConversationProvider
 *
 * Wrap your app once near the root and you're set:
 *
 *     // App.tsx / layout.tsx / etc.
 *     import { ShogoVoiceProvider } from '@/components/shogo'
 *
 *     export default function Root({ children }) {
 *       return <ShogoVoiceProvider>{children}</ShogoVoiceProvider>
 *     }
 *
 * Don't wrap each voice component individually \u2014 \`<VoiceButton>\` and
 * \`<VoiceSphere>\` only share session state when they live under the
 * SAME provider, so a per-component wrap will give the sphere a
 * separate (disconnected) session and break the visualizer.
 *
 * Otherwise zero-config in a generated pod: the underlying
 * \`useShogoVoice()\` auto-detects \`RUNTIME_AUTH_SECRET\` +
 * \`PROJECT_ID\` (injected by the runtime) and talks to the pod's
 * \`/api/voice/signed-url\` route, which proxies through the Shogo
 * API with an \`x-runtime-token\`.
 *
 * Customize freely \u2014 this file is NOT regenerated after the first
 * \`shogo generate\` run.
 */
import * as React from 'react'
import { useShogoVoice } from '${sdkReactImport}'

export interface VoiceButtonProps {
  /** Label shown when idle. Defaults to "Talk to Shogo". */
  idleLabel?: string
  /** Label shown while a session is active. Defaults to "End call". */
  activeLabel?: string
  /** Optional className for layout / theming. */
  className?: string
  /** Fired when the session transitions to connected. */
  onConnected?: () => void
  /** Fired when the session disconnects. */
  onDisconnected?: () => void
}

export function VoiceButton(props: VoiceButtonProps = {}): React.ReactElement {
  const {
    idleLabel = 'Talk to Shogo',
    activeLabel = 'End call',
    className,
    onConnected,
    onDisconnected,
  } = props
  const { start, end, status, isSpeaking } = useShogoVoice()

  React.useEffect(() => {
    if (status === 'connected') onConnected?.()
    if (status === 'disconnected') onDisconnected?.()
  }, [status, onConnected, onDisconnected])

  const active = status !== 'disconnected'

  return (
    <button
      type="button"
      onClick={active ? end : () => void start()}
      className={className}
      data-shogo-voice-status={status}
      data-shogo-voice-speaking={isSpeaking ? 'true' : 'false'}
      aria-pressed={active}
    >
      {active ? activeLabel : idleLabel}
    </button>
  )
}

export default VoiceButton
`;
}
function voiceSphereCode(sdkReactImport) {
  return `${HEADER}
/**
 * VoiceSphere \u2014 a minimal animated sphere that pulses while the agent
 * is speaking. Drives its animation off of the \`isSpeaking\` flag from
 * \`useShogoVoice()\`, so it mirrors the actual audio stream rather
 * than a guess based on time.
 *
 * REQUIRES <ShogoVoiceProvider> ABOVE IT IN THE TREE
 * --------------------------------------------------
 * Same provider requirement as \`<VoiceButton>\` \u2014 the underlying
 * \`useShogoVoice()\` hook needs a \`@elevenlabs/react\`
 * \`<ConversationProvider>\` ancestor (re-exported as
 * \`<ShogoVoiceProvider>\`). Mount one provider at the app root and
 * place the sphere and button under it together so they share a
 * single voice session; per-component providers give isolated
 * sessions and the sphere will never see the button's audio.
 *
 * Customize freely \u2014 this file is NOT regenerated after the first
 * \`shogo generate\` run.
 */
import * as React from 'react'
import { useShogoVoice } from '${sdkReactImport}'

export interface VoiceSphereProps {
  /** Diameter in pixels. Defaults to 96. */
  size?: number
  /** Optional className for additional styling. */
  className?: string
}

export function VoiceSphere(props: VoiceSphereProps = {}): React.ReactElement {
  const { size = 96, className } = props
  const { status, isSpeaking } = useShogoVoice()
  const scale = isSpeaking ? 1.12 : status === 'connected' ? 1.03 : 1
  return (
    <div
      className={className}
      data-shogo-voice-status={status}
      data-shogo-voice-speaking={isSpeaking ? 'true' : 'false'}
      style={{
        width: size,
        height: size,
        borderRadius: '50%',
        background:
          'radial-gradient(circle at 30% 30%, #93c5fd 0%, #3b82f6 50%, #1e3a8a 100%)',
        transform: \`scale(\${scale})\`,
        transition: 'transform 180ms ease-out',
        boxShadow: isSpeaking
          ? '0 0 36px rgba(59, 130, 246, 0.55)'
          : '0 0 12px rgba(59, 130, 246, 0.2)',
      }}
    />
  )
}

export default VoiceSphere
`;
}
function phoneButtonCode(sdkReactImport) {
  return `${HEADER}
/**
 * PhoneButton \u2014 starts an outbound call from the project's
 * provisioned Twilio number to \`phoneNumber\`.
 *
 * Requires voice phone provisioning for the project (see
 * \`shogo enable voice.phoneNumber\` or POST
 * \`/api/voice/twilio/provision-number/:projectId\`). Until a number
 * is provisioned, \`start()\` will 400 and this button surfaces the
 * error via its \`onError\` prop.
 *
 * Customize freely \u2014 this file is NOT regenerated after the first
 * \`shogo generate\` run.
 */
import * as React from 'react'
import { shogo, PROJECT_ID } from '../../lib/shogo'

export interface PhoneButtonProps {
  /** E.164 number to dial (e.g. "+15551234567"). Required. */
  phoneNumber: string
  /** Button label. Defaults to "Call". */
  label?: string
  /** Optional className for styling. */
  className?: string
  /** Fired once the call is initiated. */
  onInitiated?: (callId: string) => void
  /** Fired when the SDK reports an error. */
  onError?: (err: Error) => void
}

export function PhoneButton(props: PhoneButtonProps): React.ReactElement {
  const { phoneNumber, label = 'Call', className, onInitiated, onError } = props
  const [pending, setPending] = React.useState(false)

  const onClick = React.useCallback(async () => {
    if (!shogo.voice.telephony) {
      onError?.(
        new Error(
          'shogo.voice.telephony is not configured \u2014 ensure RUNTIME_AUTH_SECRET + PROJECT_ID are set.',
        ),
      )
      return
    }
    setPending(true)
    try {
      const result = await shogo.voice.telephony.outboundCall({
        to: phoneNumber,
        projectId: PROJECT_ID,
      })
      onInitiated?.(result.callId ?? '')
    } catch (err) {
      onError?.(err instanceof Error ? err : new Error(String(err)))
    } finally {
      setPending(false)
    }
  }, [phoneNumber, onInitiated, onError])

  return (
    <button
      type="button"
      onClick={onClick}
      disabled={pending}
      className={className}
    >
      {pending ? 'Calling\u2026' : label}
    </button>
  )
}

export default PhoneButton
`;
}
function indexCode(sdkReactImport) {
  return `${HEADER}
/**
 * Shogo voice components \u2014 drop-in UI for the pod-native voice flow.
 *
 * Setup (one-time, near the root of your app):
 *
 *     import { ShogoVoiceProvider } from '@/components/shogo'
 *
 *     export default function Root({ children }) {
 *       return <ShogoVoiceProvider>{children}</ShogoVoiceProvider>
 *     }
 *
 * Usage (anywhere under that provider):
 *
 *     import { VoiceButton, VoiceSphere } from '@/components/shogo'
 *
 *     export default function App() {
 *       return (
 *         <div className="flex items-center gap-4">
 *           <VoiceSphere />
 *           <VoiceButton />
 *         </div>
 *       )
 *     }
 *
 * Why the provider? \`@elevenlabs/react\` \u2265 1.1 requires every
 * \`useConversation\` caller (which includes \`useShogoVoice\`,
 * \`useVoiceConversation\`, \`VoiceButton\`, and \`VoiceSphere\`) to live
 * under a \`<ConversationProvider>\`. \`<ShogoVoiceProvider>\` is a
 * thin re-export that keeps that import owned by the SDK rather than
 * leaking into every consumer file.
 *
 * Customize freely \u2014 this file is NOT regenerated after the first
 * \`shogo generate\` run.
 */
export { ShogoVoiceProvider, type ShogoVoiceProviderProps } from '${sdkReactImport}'
export { VoiceButton, type VoiceButtonProps } from './VoiceButton'
export { VoiceSphere, type VoiceSphereProps } from './VoiceSphere'
export { PhoneButton, type PhoneButtonProps } from './PhoneButton'
`;
}
function generateVoiceComponents(opts = {}) {
  const ext = opts.fileExtension ?? "tsx";
  const sdk = opts.sdkReactImport ?? "@shogo-ai/sdk/voice/react";
  return [
    {
      fileName: `VoiceButton.${ext}`,
      code: voiceButtonCode(sdk),
      skipIfExists: true
    },
    {
      fileName: `VoiceSphere.${ext}`,
      code: voiceSphereCode(sdk),
      skipIfExists: true
    },
    {
      fileName: `PhoneButton.${ext}`,
      code: phoneButtonCode(sdk),
      skipIfExists: true
    },
    {
      fileName: "index.ts",
      code: indexCode(sdk),
      skipIfExists: true
    }
  ];
}

// src/generators/prisma-generator.ts
async function parsePrismaSchema(schemaPath) {
  const { getDMMF } = await import("@prisma/internals");
  const { readFileSync: readFileSync2, existsSync: existsSync3 } = await import("fs");
  const { dirname, join: join2, resolve: resolve2 } = await import("path");
  const schemaString = readFileSync2(schemaPath, "utf-8");
  const schemaDir = dirname(schemaPath);
  const projectRoot = resolve2(schemaDir, "..");
  const possibleConfigPaths = [
    join2(projectRoot, "prisma.config.ts"),
    join2(schemaDir, "prisma.config.ts")
  ];
  const configPath = possibleConfigPaths.find((p) => existsSync3(p));
  return await getDMMF({
    datamodel: schemaString,
    ...configPath && { prismaConfigPath: configPath }
  });
}
function getScalarFields(model) {
  return model.fields.filter((f) => f.kind === "scalar" || f.kind === "enum");
}
function getRelationFields2(model) {
  return model.fields.filter((f) => f.kind === "object");
}
function getIdField(model) {
  return model.fields.find((f) => f.isId);
}
function toCamelCase2(name) {
  return name.charAt(0).toLowerCase() + name.slice(1);
}
function toKebabCase(name) {
  return name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
async function generateFromPrisma(options) {
  const { schemaPath, outputDir, models: includeModels, excludeModels = [], outputs } = options;
  const warnings = [];
  const files = [];
  const dmmf = await parsePrismaSchema(schemaPath);
  let models = dmmf.datamodel.models;
  if (includeModels && includeModels.length > 0) {
    models = models.filter((m) => includeModels.includes(m.name));
  }
  models = models.filter((m) => !excludeModels.includes(m.name));
  const enums = dmmf.datamodel.enums;
  const hasModels = models.length > 0;
  if (outputs && outputs.length > 0) {
    for (const output of outputs) {
      const dir = output.dir;
      const perModel = output.perModel !== false;
      const ext = output.fileExtension || "tsx";
      const indexFile = `${dir}/index.${ext}`;
      if (hasModels && output.generate.includes("routes")) {
        if (perModel) {
          const { routes, hooks } = generateRoutes(models, { fileExtension: ext });
          for (const route of routes) {
            files.push({
              path: `${dir}/${route.fileName}`,
              content: route.code
            });
          }
          if (output.generate.includes("hooks")) {
            for (const hook of hooks) {
              files.push({
                path: `${dir}/${hook.fileName}`,
                content: hook.code,
                skipIfExists: true
                // Don't overwrite user customizations
              });
            }
          }
          files.push({
            path: indexFile,
            content: generateRoutesIndex(models)
          });
        }
      }
      if (hasModels && output.generate.includes("types")) {
        if (perModel) {
          const typeFiles = generateTypesPerModel(models, enums, ext);
          for (const typeFile of typeFiles) {
            files.push({
              path: `${dir}/${typeFile.fileName}`,
              content: typeFile.code
            });
          }
          files.push({
            path: `${dir}/types.${ext}`,
            content: generateTypesIndex(models)
          });
          const existingIndex = files.find((f) => f.path === indexFile);
          if (existingIndex) {
            existingIndex.content += "\n" + generateTypesIndex(models);
          } else {
            files.push({
              path: indexFile,
              content: generateTypesIndex(models)
            });
          }
        } else {
          files.push({
            path: `${dir}/types.${ext}`,
            content: generateTypes(models, enums)
          });
        }
      }
      if (hasModels && output.generate.includes("stores")) {
        if (perModel) {
          const storeFiles = generateStores(models, { fileExtension: ext });
          for (const storeFile of storeFiles) {
            files.push({
              path: `${dir}/${storeFile.fileName}`,
              content: storeFile.code
            });
          }
          const existingStoresIndex = files.find((f) => f.path === indexFile);
          if (existingStoresIndex) {
            existingStoresIndex.content += "\n" + generateStoresIndex(models);
          } else {
            files.push({
              path: indexFile,
              content: generateStoresIndex(models)
            });
          }
        }
      }
      if (hasModels && output.generate.includes("mst")) {
        if (perModel) {
          const mstModelFiles = generateMSTModels(models, models, enums, ext);
          for (const modelFile of mstModelFiles) {
            files.push({
              path: `${dir}/${modelFile.fileName}`,
              content: modelFile.code
            });
          }
          const mstCollectionFiles = generateMSTCollections(models, ext);
          for (const collectionFile of mstCollectionFiles) {
            files.push({
              path: `${dir}/${collectionFile.fileName}`,
              content: collectionFile.code
            });
          }
          const domainFile = generateMSTDomain(models, ext);
          files.push({
            path: `${dir}/${domainFile.fileName}`,
            content: domainFile.code
          });
          const existingMstIndex = files.find((f) => f.path === indexFile);
          const mstExports = generateMSTIndex(models);
          if (existingMstIndex) {
            existingMstIndex.content += "\n" + mstExports;
          } else {
            files.push({
              path: indexFile,
              content: mstExports
            });
          }
        }
      }
      if (hasModels && output.generate.includes("api-client")) {
        files.push({
          path: `${dir}/api-client.${ext}`,
          content: generateApiClient(models)
        });
      }
      if (hasModels && output.generate.includes("auth")) {
        if (hasUserModel(models)) {
          const userModel = getUserModel(models);
          files.push({
            path: `${dir}/auth.${ext}`,
            content: generateAuthStore({ userModel })
          });
        } else {
          warnings.push("Auth store generation skipped: No User model with email field found");
        }
      }
      if (output.generate.includes("server")) {
        files.push({
          path: `${dir}/server.${ext}`,
          content: generateServer({
            ...output.serverConfig
          }),
          skipIfExists: true
          // Don't overwrite user customizations
        });
      }
      if (hasModels && output.generate.includes("db")) {
        files.push({
          path: `${dir}/db.${ext}`,
          content: output.dbProvider === "sqlite" ? generateSqliteDbModule() : generateDbModule(),
          skipIfExists: true
          // Don't overwrite user customizations
        });
      }
      if (hasModels && output.generate.includes("docs")) {
        const scaffoldFiles = generateDocsSiteScaffold({
          projectName: "My App"
        });
        for (const sf of scaffoldFiles) {
          files.push({
            path: `${dir}/${sf.path}`,
            content: sf.content,
            skipIfExists: sf.skipIfExists
          });
        }
        const tsConfig = generateDocsTsConfig();
        files.push({
          path: `${dir}/${tsConfig.path}`,
          content: tsConfig.content,
          skipIfExists: true
        });
        const docFiles = generateDocs(models, enums, {
          apiBasePath: "/api"
        });
        for (const df of docFiles) {
          files.push({
            path: `${dir}/${df.path}`,
            content: df.content,
            skipIfExists: df.skipIfExists
          });
        }
      }
      if (hasModels && output.generate.includes("admin-routes")) {
        const adminFile = generateAdminRoutes(models, { fileExtension: ext });
        files.push({
          path: `${dir}/${adminFile.fileName}`,
          content: adminFile.code
        });
      }
      if (output.generate.includes("shogo-client")) {
        const shogoFile = generateShogoClient({
          fileExtension: ext === "tsx" ? "ts" : ext,
          ...output.shogoClient ?? {}
        });
        files.push({
          path: `${dir}/${shogoFile.fileName}`,
          content: shogoFile.code
        });
      }
      if (output.generate.includes("voice-components")) {
        const voiceFiles = generateVoiceComponents({
          fileExtension: ext,
          ...output.voiceComponents ?? {}
        });
        for (const f of voiceFiles) {
          files.push({
            path: `${dir}/${f.fileName}`,
            content: f.code,
            skipIfExists: f.skipIfExists
          });
        }
      }
    }
    return {
      files,
      models: models.map((m) => m.name),
      warnings
    };
  }
  if (!outputDir) {
    throw new Error("Either outputDir or outputs[] must be provided");
  }
  if (!hasModels) {
    throw new Error("No models found to generate");
  }
  const typesCode = generateTypes(models, enums);
  files.push({
    path: `${outputDir}/types.tsx`,
    content: typesCode
  });
  const hooksCode = generateHooksTemplate(models);
  files.push({
    path: `${outputDir}/hooks.tsx`,
    content: hooksCode
  });
  const indexCode2 = generateIndexFile(models);
  files.push({
    path: `${outputDir}/index.tsx`,
    content: indexCode2
  });
  return {
    files,
    models: models.map((m) => m.name),
    warnings
  };
}
function generateHooksTemplate(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Server Function Hooks",
    " *",
    " * Customize CRUD behavior with before/after hooks.",
    " * This file is safe to edit - it will not be overwritten.",
    " */",
    "",
    "import type { ServerFunctionHooks } from './types'",
    "",
    "export const hooks: ServerFunctionHooks = {"
  ];
  for (const model of models) {
    const name = model.name;
    lines.push(`  ${name}: {`);
    lines.push(`    // beforeList: async (ctx) => {`);
    lines.push(`    //   // Filter by user ownership`);
    lines.push(`    //   return { where: { userId: ctx.userId } }`);
    lines.push(`    // },`);
    lines.push(`    // beforeCreate: async (input, ctx) => {`);
    lines.push(`    //   // Set userId on create`);
    lines.push(`    //   return { ...input, userId: ctx.userId }`);
    lines.push(`    // },`);
    lines.push(`  },`);
  }
  lines.push("}");
  lines.push("");
  lines.push("export default hooks");
  lines.push("");
  return lines.join("\n");
}
function generateIndexFile(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Generated Shogo SDK Code",
    " *",
    " * DO NOT EDIT - regenerate with `shogo generate`",
    " */",
    "",
    "// Types",
    "export * from './types'",
    "",
    "// Server Functions",
    "export * from './server-functions'",
    "",
    "// Domain Store (MST/MobX)",
    "export * from './domain'",
    "",
    "// Hooks (customizable)",
    "export { hooks } from './hooks'",
    ""
  ];
  return lines.join("\n");
}
function generateMSTIndex(_models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * MST Domain Exports",
    " *",
    " * Generated by @shogo-ai/sdk - DO NOT EDIT DIRECTLY",
    " */",
    "",
    'export * from "./domain"',
    ""
  ];
  return lines.join("\n");
}

// src/generators/custom-routes.ts
import { existsSync, writeFileSync } from "fs";
import { join } from "path";
var CUSTOM_ROUTES_SCAFFOLD = `// SPDX-License-Identifier: MIT
// Copyright (C) 2026 Shogo Technologies, Inc.
/**
 * Custom API Routes
 *
 * Mount custom non-CRUD routes (proxies, aggregations, webhooks, etc.)
 * on the exported Hono \`app\`. They show up at \`/api/...\` alongside
 * the generated CRUD routes.
 *
 * Examples:
 *   app.get('/dashboard', async (c) => c.json({ ... }))
 *   app.post('/webhooks/stripe', async (c) => { ... })
 *
 * The Prisma client is available \u2014 import it from \`./src/lib/db\`:
 *   import { prisma } from './src/lib/db'
 *
 * Saving this file triggers a fast restart of server.tsx so changes
 * are live in under a second. Do NOT edit server.tsx; it is
 * auto-generated by \`bun x shogo generate\`.
 */

import { Hono } from 'hono'

const app = new Hono()

// Add your custom routes here. Example:
//   app.get('/hello', (c) => c.json({ message: 'Hello!' }))

export default app
`;
function ensureCustomRoutes(projectDir) {
  const tsPath = join(projectDir, "custom-routes.ts");
  const tsxPath = join(projectDir, "custom-routes.tsx");
  if (existsSync(tsPath)) {
    return { path: "./custom-routes", created: false, absolutePath: tsPath };
  }
  if (existsSync(tsxPath)) {
    return { path: "./custom-routes", created: false, absolutePath: tsxPath };
  }
  writeFileSync(tsPath, CUSTOM_ROUTES_SCAFFOLD, "utf-8");
  return { path: "./custom-routes", created: true, absolutePath: tsPath };
}

// src/generators/server-custom-regions.ts
var CUSTOM_REGION_START = "// SHOGO:CUSTOM-START";
var CUSTOM_REGION_END = "// SHOGO:CUSTOM-END";
function extractProtectedRegions(content) {
  const lines = content.split("\n");
  const regions = [];
  let current = null;
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.startsWith(CUSTOM_REGION_START)) {
      const id = trimmed.slice(CUSTOM_REGION_START.length).trim();
      current = { id, body: [] };
      continue;
    }
    if (trimmed.startsWith(CUSTOM_REGION_END)) {
      if (current) {
        regions.push({ id: current.id, body: current.body.join("\n") });
        current = null;
      }
      continue;
    }
    if (current) current.body.push(line);
  }
  return regions;
}
function mergeProtectedRegions(existing, generated) {
  const regions = extractProtectedRegions(existing);
  if (regions.length === 0) return generated;
  const generatedRegionIds = new Set(extractProtectedRegions(generated).map((r) => r.id));
  let out = generated;
  const toAppend = [];
  for (const region of regions) {
    const block = renderRegion(region);
    if (region.id && generatedRegionIds.has(region.id)) {
      out = replaceRegionBody(out, region);
    } else {
      toAppend.push(block);
    }
  }
  if (toAppend.length === 0) return out;
  return insertBeforeDefaultExport(out, toAppend.join("\n\n"));
}
function renderRegion(region) {
  const start = region.id ? `${CUSTOM_REGION_START} ${region.id}` : CUSTOM_REGION_START;
  return [start, region.body, CUSTOM_REGION_END].join("\n");
}
function replaceRegionBody(content, region) {
  const lines = content.split("\n");
  const out = [];
  let inTarget = false;
  for (const line of lines) {
    const trimmed = line.trim();
    if (!inTarget && trimmed.startsWith(CUSTOM_REGION_START) && trimmed.slice(CUSTOM_REGION_START.length).trim() === region.id) {
      inTarget = true;
      out.push(line);
      out.push(region.body);
      continue;
    }
    if (inTarget) {
      if (trimmed.startsWith(CUSTOM_REGION_END)) {
        inTarget = false;
        out.push(line);
      }
      continue;
    }
    out.push(line);
  }
  return out.join("\n");
}
function insertBeforeDefaultExport(content, block) {
  const lines = content.split("\n");
  for (let i = lines.length - 1; i >= 0; i--) {
    if (/^\s*export\s+default\b/.test(lines[i])) {
      lines.splice(i, 0, block, "");
      return lines.join("\n");
    }
  }
  return `${content}

${block}
`;
}

// src/generators/prisma-schema-guard.ts
var SCHEMA_HEADER_REGION_ID = "prisma-header";
var SCHEMA_HEADER_MANAGED_COMMENT = "// Managed by Shogo. Do not add a datasource `url` or change the generator `provider` \u2014 the database URL is configured in prisma.config.ts (Prisma 7+).";
var DEFAULT_PRISMA_HEADER = [
  "generator client {",
  '  provider = "prisma-client"',
  '  output   = "../src/generated/prisma"',
  "}",
  "",
  "datasource db {",
  '  provider = "sqlite"',
  "}"
].join("\n");
var HEADER_BLOCK_RE = /^[ \t]*(generator|datasource)[ \t]+\w+[ \t]*\{[\s\S]*?^[ \t]*\}[ \t]*$/gm;
function findBlocks(source) {
  const blocks = [];
  for (const m of source.matchAll(HEADER_BLOCK_RE)) {
    blocks.push({ kind: m[1], text: m[0] });
  }
  return blocks;
}
function firstBlock(source, kind) {
  return findBlocks(source).find((b) => b.kind === kind)?.text ?? null;
}
function stripBlocks(source) {
  return collapseBlankRuns(source.replace(HEADER_BLOCK_RE, ""));
}
function collapseBlankRuns(s) {
  return s.replace(/\n{3,}/g, "\n\n");
}
function sanitizeGenerator(block) {
  return block.replace(/(provider\s*=\s*["'])prisma-client-js(["'])/g, "$1prisma-client$2");
}
function generatorHasOutput(block) {
  return /^\s*output\s*=/m.test(block);
}
function isClientGenerator(block) {
  return /provider\s*=\s*["']prisma-client(-js)?["']/.test(block);
}
function sanitizeDatasource(block) {
  return block.split("\n").filter((line) => !/^\s*url\s*=/.test(line)).join("\n");
}
function headerIsDowngraded(schema) {
  for (const block of findBlocks(schema)) {
    if (block.kind === "datasource" && /^\s*url\s*=/m.test(block.text)) return true;
    if (block.kind === "generator") {
      if (/prisma-client-js/.test(block.text)) return true;
      if (isClientGenerator(block.text) && !generatorHasOutput(block.text)) return true;
    }
  }
  return false;
}
function hasMarkedSchemaHeader(schema) {
  return splitHeaderRegion(schema).regionBody !== null;
}
function splitHeaderRegion(schema) {
  const lines = schema.split("\n");
  let startIdx = -1;
  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i].trim();
    if (trimmed.startsWith(CUSTOM_REGION_START) && trimmed.slice(CUSTOM_REGION_START.length).trim() === SCHEMA_HEADER_REGION_ID) {
      startIdx = i;
      break;
    }
  }
  if (startIdx === -1) return { regionBody: null, rest: schema };
  let endIdx = -1;
  for (let i = startIdx + 1; i < lines.length; i++) {
    if (lines[i].trim().startsWith(CUSTOM_REGION_END)) {
      endIdx = i;
      break;
    }
  }
  if (endIdx === -1) return { regionBody: null, rest: schema };
  const regionBody = lines.slice(startIdx + 1, endIdx).join("\n");
  const rest = [...lines.slice(0, startIdx), ...lines.slice(endIdx + 1)].join("\n");
  return { regionBody, rest };
}
function enforceSchemaHeader(schema, fallbackHeader = DEFAULT_PRISMA_HEADER) {
  const { regionBody, rest } = splitHeaderRegion(schema);
  const headerSource = regionBody ?? "";
  const canonicalGenerator = firstBlock(fallbackHeader, "generator") ?? firstBlock(DEFAULT_PRISMA_HEADER, "generator");
  const candidateGenerator = firstBlock(headerSource, "generator") ?? firstBlock(rest, "generator");
  const generator = candidateGenerator && generatorHasOutput(candidateGenerator) ? candidateGenerator : canonicalGenerator;
  const datasource = firstBlock(headerSource, "datasource") ?? firstBlock(rest, "datasource") ?? firstBlock(fallbackHeader, "datasource") ?? firstBlock(DEFAULT_PRISMA_HEADER, "datasource");
  const headerBody = [
    SCHEMA_HEADER_MANAGED_COMMENT,
    sanitizeGenerator(generator).trim(),
    "",
    sanitizeDatasource(datasource).trim()
  ].join("\n");
  const region = [`${CUSTOM_REGION_START} ${SCHEMA_HEADER_REGION_ID}`, headerBody, CUSTOM_REGION_END].join("\n");
  const body = stripBlocks(rest).trim();
  return body ? `${region}

${body}
` : `${region}
`;
}

// src/generators/prisma-config-guard.ts
var PRISMA_CONFIG_MANAGED_COMMENT = "// Managed by Shogo. `prisma db push` (Prisma 7) requires `datasource.url`; do not move the URL under `migrate` or an `async url()` resolver.";
var DEFAULT_PRISMA_CONFIG = [
  PRISMA_CONFIG_MANAGED_COMMENT,
  "import { defineConfig } from 'prisma/config'",
  "",
  "export default defineConfig({",
  "  schema: 'prisma/schema.prisma',",
  "  datasource: {",
  "    url: process.env.DATABASE_URL ?? 'file:./dev.db',",
  "  },",
  "})",
  ""
].join("\n");
function datasourceBody(source) {
  const m = source.match(/\bdatasource\s*:\s*\{([^}]*)\}/s);
  return m ? m[1] : null;
}
function configIsDowngraded(source) {
  if (!source || !source.trim()) return true;
  const ds = datasourceBody(source);
  if (ds === null) return true;
  return !/\burl\s*:/.test(ds);
}
function enforcePrismaConfig(source) {
  return configIsDowngraded(source) ? DEFAULT_PRISMA_CONFIG : source;
}

// src/generators/deps-doctor.ts
import { existsSync as existsSync2, readFileSync, writeFileSync as writeFileSync2 } from "fs";
import { resolve } from "path";
var FEATURE_DEPS = {
  voice: {
    // 1.1+ is required because `useConversation` now lives behind
    // `<ConversationProvider>` (which the SDK surfaces as
    // `<ShogoVoiceProvider>`). Older 0.x versions don't ship the
    // provider symbol and would crash the generated `VoiceButton` /
    // `VoiceSphere` at mount time.
    "@elevenlabs/react": "^1.1.1",
    "@elevenlabs/client": "^1.1.1"
  }
};
function ensureFeatureDeps(opts) {
  const packageJsonPath = resolve(opts.cwd, "package.json");
  const missing = {};
  const added = {};
  const warnings = [];
  if (!existsSync2(packageJsonPath)) {
    warnings.push(`package.json not found at ${packageJsonPath}`);
    return { packageJsonPath, missing, added, modified: false, warnings };
  }
  let raw;
  let pkg;
  try {
    raw = readFileSync(packageJsonPath, "utf-8");
    pkg = JSON.parse(raw);
  } catch (err) {
    warnings.push(
      `Failed to parse package.json: ${err instanceof Error ? err.message : String(err)}`
    );
    return { packageJsonPath, missing, added, modified: false, warnings };
  }
  const deps = pkg.dependencies ?? (pkg.dependencies = {});
  const enabledFeatures = [];
  if (opts.features?.voice) enabledFeatures.push("voice");
  for (const feature of enabledFeatures) {
    const required = FEATURE_DEPS[feature];
    if (!required) continue;
    const featMissing = {};
    for (const [name, range] of Object.entries(required)) {
      if (!deps[name]) {
        featMissing[name] = range;
        added[name] = range;
        deps[name] = range;
      }
    }
    if (Object.keys(featMissing).length > 0) missing[feature] = featMissing;
  }
  const modified = Object.keys(added).length > 0;
  if (modified && !opts.dryRun) {
    const hasTrailingNewline = raw.endsWith("\n");
    const serialized = JSON.stringify(pkg, null, 2) + (hasTrailingNewline ? "\n" : "");
    writeFileSync2(packageJsonPath, serialized);
  }
  return { packageJsonPath, missing, added, modified, warnings };
}

// src/generators/server-functions.ts
function toRoutePath7(name) {
  const kebab = name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  if (kebab.endsWith("y")) return kebab.slice(0, -1) + "ies";
  if (kebab.endsWith("s") || kebab.endsWith("x") || kebab.endsWith("ch") || kebab.endsWith("sh")) {
    return kebab + "es";
  }
  return kebab + "s";
}
function generateServerFunctions(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Client Functions",
    " *",
    " * Fetch-based CRUD operations that call the Hono REST API.",
    " * No server-side imports \u2014 safe for browser bundles.",
    " * Generated by `shogo generate` - DO NOT EDIT",
    " */",
    "",
    "import type {"
  ];
  for (const model of models) {
    lines.push(`  ${model.name}Type,`);
    lines.push(`  ${model.name}CreateInput,`);
    lines.push(`  ${model.name}UpdateInput,`);
  }
  lines.push("} from './types'");
  lines.push("");
  lines.push("/** Get the API base URL */");
  lines.push("function getApiBase(): string {");
  lines.push("  if (typeof window !== 'undefined') {");
  lines.push("    return window.location.origin");
  lines.push("  }");
  lines.push("  return process.env.API_URL || 'http://localhost:3001'");
  lines.push("}");
  lines.push("");
  for (const model of models) {
    lines.push(...generateModelFunctions(model));
  }
  return lines.join("\n");
}
function generateModelFunctions(model) {
  const name = model.name;
  const nameLower = toCamelCase2(name);
  const routePath = toRoutePath7(name);
  const idField = getIdField(model);
  if (!idField) {
    return [`// Skipped ${name} - no @id field found`, ""];
  }
  const lines = [
    `// ============================================================================`,
    `// ${name} Client Functions`,
    `// ============================================================================`,
    ""
  ];
  lines.push(...generateListFunction(model, routePath));
  lines.push(...generateGetFunction(model, routePath));
  lines.push(...generateCreateFunction(model, routePath));
  lines.push(...generateUpdateFunction(model, routePath));
  lines.push(...generateDeleteFunction(model, routePath));
  return lines;
}
function generateListFunction(model, routePath) {
  const name = model.name;
  return [
    `/**`,
    ` * List all ${name} records`,
    ` */`,
    `export async function get${name}List(args: { data: { userId?: string; where?: Record<string, unknown> } }): Promise<${name}Type[]> {`,
    `  const params = new URLSearchParams()`,
    `  if (args.data.userId) params.set('userId', args.data.userId)`,
    `  if (args.data.where) {`,
    `    for (const [key, value] of Object.entries(args.data.where)) {`,
    `      if (value !== undefined && value !== null) params.set(key, String(value))`,
    `    }`,
    `  }`,
    `  const qs = params.toString()`,
    `  const url = \`\${getApiBase()}/api/${routePath}\${qs ? \`?\${qs}\` : ''}\``,
    `  const response = await fetch(url)`,
    `  if (!response.ok) {`,
    `    const err = await response.json().catch(() => ({ error: { message: response.statusText } }))`,
    `    throw new Error(err.error?.message || 'Failed to list ${name}')`,
    `  }`,
    `  const json = await response.json()`,
    `  return (json.items || []) as ${name}Type[]`,
    `}`,
    ""
  ];
}
function generateGetFunction(model, routePath) {
  const name = model.name;
  return [
    `/**`,
    ` * Get a single ${name} by ID`,
    ` */`,
    `export async function get${name}ById(args: { data: { id: string; userId?: string } }): Promise<${name}Type> {`,
    `  const url = \`\${getApiBase()}/api/${routePath}/\${args.data.id}\``,
    `  const response = await fetch(url)`,
    `  if (!response.ok) {`,
    `    const err = await response.json().catch(() => ({ error: { message: response.statusText } }))`,
    `    throw new Error(err.error?.message || '${name} not found')`,
    `  }`,
    `  const json = await response.json()`,
    `  return json.data as ${name}Type`,
    `}`,
    ""
  ];
}
function generateCreateFunction(model, routePath) {
  const name = model.name;
  return [
    `/**`,
    ` * Create a new ${name}`,
    ` */`,
    `export async function create${name}(args: { data: { input: ${name}CreateInput; userId?: string } }): Promise<${name}Type> {`,
    `  const url = \`\${getApiBase()}/api/${routePath}\``,
    `  const response = await fetch(url, {`,
    `    method: 'POST',`,
    `    headers: { 'Content-Type': 'application/json' },`,
    `    body: JSON.stringify(args.data.input),`,
    `  })`,
    `  if (!response.ok) {`,
    `    const err = await response.json().catch(() => ({ error: { message: response.statusText } }))`,
    `    throw new Error(err.error?.message || 'Failed to create ${name}')`,
    `  }`,
    `  const json = await response.json()`,
    `  return json.data as ${name}Type`,
    `}`,
    ""
  ];
}
function generateUpdateFunction(model, routePath) {
  const name = model.name;
  return [
    `/**`,
    ` * Update an existing ${name}`,
    ` */`,
    `export async function update${name}(args: { data: { id: string; input: ${name}UpdateInput; userId?: string } }): Promise<${name}Type> {`,
    `  const url = \`\${getApiBase()}/api/${routePath}/\${args.data.id}\``,
    `  const response = await fetch(url, {`,
    `    method: 'PATCH',`,
    `    headers: { 'Content-Type': 'application/json' },`,
    `    body: JSON.stringify(args.data.input),`,
    `  })`,
    `  if (!response.ok) {`,
    `    const err = await response.json().catch(() => ({ error: { message: response.statusText } }))`,
    `    throw new Error(err.error?.message || 'Failed to update ${name}')`,
    `  }`,
    `  const json = await response.json()`,
    `  return json.data as ${name}Type`,
    `}`,
    ""
  ];
}
function generateDeleteFunction(model, routePath) {
  const name = model.name;
  return [
    `/**`,
    ` * Delete a ${name}`,
    ` */`,
    `export async function delete${name}(args: { data: { id: string; userId?: string } }): Promise<{ success: boolean }> {`,
    `  const url = \`\${getApiBase()}/api/${routePath}/\${args.data.id}\``,
    `  const response = await fetch(url, {`,
    `    method: 'DELETE',`,
    `  })`,
    `  if (!response.ok) {`,
    `    const err = await response.json().catch(() => ({ error: { message: response.statusText } }))`,
    `    throw new Error(err.error?.message || 'Failed to delete ${name}')`,
    `  }`,
    `  return { success: true }`,
    `}`,
    ""
  ];
}

// src/generators/domain-store.ts
function generateDomainStore(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Domain Store",
    " *",
    " * MobX-State-Tree store with collections for each model.",
    " * Generated by `shogo generate` - DO NOT EDIT",
    " */",
    "",
    "import { makeAutoObservable, runInAction } from 'mobx'",
    "import {"
  ];
  for (const model of models) {
    const name = model.name;
    lines.push(`  get${name}List,`);
    lines.push(`  get${name}ById,`);
    lines.push(`  create${name},`);
    lines.push(`  update${name},`);
    lines.push(`  delete${name},`);
  }
  lines.push("} from './server-functions'");
  lines.push("");
  lines.push("import type {");
  for (const model of models) {
    lines.push(`  ${model.name}Type,`);
    lines.push(`  ${model.name}CreateInput,`);
    lines.push(`  ${model.name}UpdateInput,`);
  }
  lines.push("} from './types'");
  lines.push("");
  for (const model of models) {
    lines.push(...generateModelStore2(model));
  }
  lines.push(...generateRootStore(models));
  lines.push(...generateFactory(models));
  return lines.join("\n");
}
function generateModelStore2(model) {
  const name = model.name;
  const nameLower = toCamelCase2(name);
  return [
    `// ============================================================================`,
    `// ${name} Store`,
    `// ============================================================================`,
    "",
    `export class ${name}Store {`,
    `  items: Map<string, ${name}Type> = new Map()`,
    `  isLoading = false`,
    `  error: string | null = null`,
    ``,
    `  // Track pending operations for optimistic updates`,
    `  private pendingDeletes = new Set<string>()`,
    `  private pendingUpdates = new Set<string>()`,
    ``,
    `  constructor() {`,
    `    makeAutoObservable(this, {`,
    `      pendingDeletes: false,`,
    `      pendingUpdates: false,`,
    `    })`,
    `  }`,
    ``,
    `  // === Getters ===`,
    ``,
    `  /** Get all items as array */`,
    `  get all(): ${name}Type[] {`,
    `    return Array.from(this.items.values())`,
    `  }`,
    ``,
    `  /** Get item by ID */`,
    `  get(id: string): ${name}Type | undefined {`,
    `    return this.items.get(id)`,
    `  }`,
    ``,
    `  /** Check if item has pending operation */`,
    `  isPending(id: string): boolean {`,
    `    return this.pendingDeletes.has(id) || this.pendingUpdates.has(id)`,
    `  }`,
    ``,
    `  // === Actions ===`,
    ``,
    `  /** Load all items from server */`,
    `  async loadAll(userId?: string, where?: Record<string, unknown>) {`,
    `    runInAction(() => {`,
    `      this.isLoading = true`,
    `      this.error = null`,
    `    })`,
    ``,
    `    try {`,
    `      const items = await get${name}List({ data: { userId, where } })`,
    ``,
    `      runInAction(() => {`,
    `        this.items.clear()`,
    `        for (const item of items) {`,
    `          this.items.set(item.id, item)`,
    `        }`,
    `        this.isLoading = false`,
    `      })`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        this.error = e instanceof Error ? e.message : 'Failed to load'`,
    `        this.isLoading = false`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Load single item by ID */`,
    `  async loadById(id: string, userId?: string) {`,
    `    try {`,
    `      const item = await get${name}ById({ data: { id, userId } })`,
    ``,
    `      runInAction(() => {`,
    `        this.items.set(item.id, item)`,
    `      })`,
    ``,
    `      return item`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        this.error = e instanceof Error ? e.message : 'Failed to load'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Create new item with optimistic update */`,
    `  async create(input: ${name}CreateInput, userId?: string) {`,
    `    // Create optimistic item`,
    `    const tempId = \`temp-\${crypto.randomUUID()}\``,
    `    const optimisticItem: ${name}Type = {`,
    `      id: tempId,`,
    `      ...input,`,
    `      createdAt: new Date(),`,
    `      updatedAt: new Date(),`,
    `    } as ${name}Type`,
    ``,
    `    // Add optimistically`,
    `    runInAction(() => {`,
    `      this.items.set(tempId, optimisticItem)`,
    `    })`,
    ``,
    `    try {`,
    `      const item = await create${name}({ data: { input, userId } })`,
    ``,
    `      runInAction(() => {`,
    `        // Remove temp, add real`,
    `        this.items.delete(tempId)`,
    `        this.items.set(item.id, item)`,
    `      })`,
    ``,
    `      return item`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        // Rollback`,
    `        this.items.delete(tempId)`,
    `        this.error = e instanceof Error ? e.message : 'Failed to create'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Update item with optimistic update */`,
    `  async update(id: string, input: ${name}UpdateInput, userId?: string) {`,
    `    // Validate id to prevent undefined from reaching the API`,
    `    if (!id || typeof id !== 'string') {`,
    `      console.error('[${name}Store] update called with invalid id:', id)`,
    `      return`,
    `    }`,
    `    const existing = this.items.get(id)`,
    `    if (!existing || this.pendingUpdates.has(id)) return`,
    ``,
    `    const previousState = { ...existing }`,
    `    this.pendingUpdates.add(id)`,
    ``,
    `    // Optimistically update`,
    `    runInAction(() => {`,
    `      this.items.set(id, {`,
    `        ...existing,`,
    `        ...input,`,
    `        updatedAt: new Date(),`,
    `      } as ${name}Type)`,
    `    })`,
    ``,
    `    try {`,
    `      const item = await update${name}({ data: { id, input, userId } })`,
    ``,
    `      runInAction(() => {`,
    `        this.items.set(id, item)`,
    `        this.pendingUpdates.delete(id)`,
    `      })`,
    ``,
    `      return item`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        // Rollback`,
    `        this.items.set(id, previousState)`,
    `        this.pendingUpdates.delete(id)`,
    `        this.error = e instanceof Error ? e.message : 'Failed to update'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Delete item with optimistic update */`,
    `  async delete(id: string, userId?: string) {`,
    `    // Validate id to prevent undefined from reaching the API`,
    `    if (!id || typeof id !== 'string') {`,
    `      console.error('[${name}Store] delete called with invalid id:', id)`,
    `      return`,
    `    }`,
    `    const existing = this.items.get(id)`,
    `    if (!existing || this.pendingDeletes.has(id)) return`,
    ``,
    `    this.pendingDeletes.add(id)`,
    ``,
    `    // Optimistically remove`,
    `    runInAction(() => {`,
    `      this.items.delete(id)`,
    `    })`,
    ``,
    `    try {`,
    `      await delete${name}({ data: { id, userId } })`,
    ``,
    `      runInAction(() => {`,
    `        this.pendingDeletes.delete(id)`,
    `      })`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        // Rollback`,
    `        this.items.set(id, existing)`,
    `        this.pendingDeletes.delete(id)`,
    `        this.error = e instanceof Error ? e.message : 'Failed to delete'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Clear error state */`,
    `  clearError() {`,
    `    this.error = null`,
    `  }`,
    ``,
    `  /** Clear all data */`,
    `  clear() {`,
    `    this.items.clear()`,
    `    this.error = null`,
    `    this.isLoading = false`,
    `  }`,
    `}`,
    ""
  ];
}
function generateRootStore(models) {
  const lines = [
    "// ============================================================================",
    "// Root Store",
    "// ============================================================================",
    "",
    "export class RootStore {"
  ];
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`  ${nameLower}: ${model.name}Store`);
  }
  lines.push("");
  lines.push("  constructor() {");
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`    this.${nameLower} = new ${model.name}Store()`);
  }
  lines.push("    makeAutoObservable(this)");
  lines.push("  }");
  lines.push("");
  lines.push("  /** Clear all stores */"), lines.push("  clearAll() {");
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`    this.${nameLower}.clear()`);
  }
  lines.push("  }");
  lines.push("}");
  lines.push("");
  return lines;
}
function generateFactory(models) {
  return [
    "// ============================================================================",
    "// Factory & Singleton",
    "// ============================================================================",
    "",
    "let rootStore: RootStore | null = null",
    "",
    "/**",
    " * Get or create the root store singleton",
    " */",
    "export function getStore(): RootStore {",
    "  if (!rootStore) {",
    "    rootStore = new RootStore()",
    "  }",
    "  return rootStore",
    "}",
    "",
    "/**",
    " * Create a new root store (for testing or SSR)",
    " */",
    "export function createStore(): RootStore {",
    "  return new RootStore()",
    "}",
    "",
    "/**",
    " * Reset the singleton store",
    " */",
    "export function resetStore(): void {",
    "  rootStore = null",
    "}",
    ""
  ];
}

// src/generators/api-domain-store.ts
function generateApiDomainStore(models) {
  const lines = [
    GENERATED_FILE_LICENSE_HEADER,
    "/**",
    " * Auto-generated Domain Store (API-based)",
    " *",
    " * MobX store with collections for each model.",
    " * Uses HTTP API client instead of server functions.",
    " * Generated by `shogo generate` - DO NOT EDIT",
    " */",
    "",
    "import { makeAutoObservable, runInAction } from 'mobx'",
    "import {"
  ];
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`  ${nameLower}Api,`);
  }
  lines.push("} from './api-client'");
  lines.push("");
  lines.push("import type {");
  for (const model of models) {
    lines.push(`  ${model.name}Type,`);
    lines.push(`  ${model.name}CreateInput,`);
    lines.push(`  ${model.name}UpdateInput,`);
  }
  lines.push("} from './types'");
  lines.push("");
  for (const model of models) {
    lines.push(...generateApiModelStore(model));
  }
  lines.push(...generateApiRootStore(models));
  lines.push(...generateApiFactory(models));
  return lines.join("\n");
}
function generateApiModelStore(model) {
  const name = model.name;
  const nameLower = toCamelCase2(name);
  return [
    `// ============================================================================`,
    `// ${name} Store (API-based)`,
    `// ============================================================================`,
    "",
    `export class ${name}Store {`,
    `  items: Map<string, ${name}Type> = new Map()`,
    `  isLoading = false`,
    `  error: string | null = null`,
    ``,
    `  // Track pending operations for optimistic updates`,
    `  private pendingDeletes = new Set<string>()`,
    `  private pendingUpdates = new Set<string>()`,
    ``,
    `  constructor() {`,
    `    makeAutoObservable(this, {`,
    `      pendingDeletes: false,`,
    `      pendingUpdates: false,`,
    `    })`,
    `  }`,
    ``,
    `  // === Getters ===`,
    ``,
    `  /** Get all items as array */`,
    `  get all(): ${name}Type[] {`,
    `    return Array.from(this.items.values())`,
    `  }`,
    ``,
    `  /** Get item by ID */`,
    `  get(id: string): ${name}Type | undefined {`,
    `    return this.items.get(id)`,
    `  }`,
    ``,
    `  /** Check if item has pending operation */`,
    `  isPending(id: string): boolean {`,
    `    return this.pendingDeletes.has(id) || this.pendingUpdates.has(id)`,
    `  }`,
    ``,
    `  // === Actions ===`,
    ``,
    `  /** Load all items from API */`,
    `  async loadAll(where?: Record<string, unknown>) {`,
    `    runInAction(() => {`,
    `      this.isLoading = true`,
    `      this.error = null`,
    `    })`,
    ``,
    `    try {`,
    `      const result = await ${nameLower}Api.list({ where })`,
    ``,
    `      if (!result.ok || !result.items) {`,
    `        throw new Error(result.error?.message || 'Failed to load')`,
    `      }`,
    ``,
    `      runInAction(() => {`,
    `        this.items.clear()`,
    `        for (const item of result.items!) {`,
    `          this.items.set(item.id, item)`,
    `        }`,
    `        this.isLoading = false`,
    `      })`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        this.error = e instanceof Error ? e.message : 'Failed to load'`,
    `        this.isLoading = false`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Load single item by ID */`,
    `  async loadById(id: string) {`,
    `    try {`,
    `      const result = await ${nameLower}Api.get(id)`,
    ``,
    `      if (!result.ok || !result.data) {`,
    `        throw new Error(result.error?.message || '${name} not found')`,
    `      }`,
    ``,
    `      runInAction(() => {`,
    `        this.items.set(result.data!.id, result.data!)`,
    `      })`,
    ``,
    `      return result.data`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        this.error = e instanceof Error ? e.message : 'Failed to load'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Create new item with optimistic update */`,
    `  async create(input: ${name}CreateInput) {`,
    `    // Create optimistic item`,
    `    const tempId = \`temp-\${crypto.randomUUID()}\``,
    `    const optimisticItem: ${name}Type = {`,
    `      id: tempId,`,
    `      ...input,`,
    `      createdAt: new Date(),`,
    `      updatedAt: new Date(),`,
    `    } as ${name}Type`,
    ``,
    `    // Add optimistically`,
    `    runInAction(() => {`,
    `      this.items.set(tempId, optimisticItem)`,
    `    })`,
    ``,
    `    try {`,
    `      const result = await ${nameLower}Api.create(input)`,
    ``,
    `      if (!result.ok || !result.data) {`,
    `        throw new Error(result.error?.message || 'Failed to create')`,
    `      }`,
    ``,
    `      runInAction(() => {`,
    `        // Remove temp, add real`,
    `        this.items.delete(tempId)`,
    `        this.items.set(result.data!.id, result.data!)`,
    `      })`,
    ``,
    `      return result.data`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        // Rollback`,
    `        this.items.delete(tempId)`,
    `        this.error = e instanceof Error ? e.message : 'Failed to create'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Update item with optimistic update */`,
    `  async update(id: string, input: ${name}UpdateInput) {`,
    `    // Validate id to prevent undefined from reaching the API`,
    `    if (!id || typeof id !== 'string') {`,
    `      console.error('[${name}Store] update called with invalid id:', id)`,
    `      return`,
    `    }`,
    `    const existing = this.items.get(id)`,
    `    if (!existing || this.pendingUpdates.has(id)) return`,
    ``,
    `    const previousState = { ...existing }`,
    `    this.pendingUpdates.add(id)`,
    ``,
    `    // Optimistically update`,
    `    runInAction(() => {`,
    `      this.items.set(id, {`,
    `        ...existing,`,
    `        ...input,`,
    `        updatedAt: new Date(),`,
    `      } as ${name}Type)`,
    `    })`,
    ``,
    `    try {`,
    `      const result = await ${nameLower}Api.update(id, input)`,
    ``,
    `      if (!result.ok || !result.data) {`,
    `        throw new Error(result.error?.message || 'Failed to update')`,
    `      }`,
    ``,
    `      runInAction(() => {`,
    `        this.items.set(id, result.data!)`,
    `        this.pendingUpdates.delete(id)`,
    `      })`,
    ``,
    `      return result.data`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        // Rollback`,
    `        this.items.set(id, previousState)`,
    `        this.pendingUpdates.delete(id)`,
    `        this.error = e instanceof Error ? e.message : 'Failed to update'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Delete item with optimistic update */`,
    `  async delete(id: string) {`,
    `    // Validate id to prevent undefined from reaching the API`,
    `    if (!id || typeof id !== 'string') {`,
    `      console.error('[${name}Store] delete called with invalid id:', id)`,
    `      return`,
    `    }`,
    `    const existing = this.items.get(id)`,
    `    if (!existing || this.pendingDeletes.has(id)) return`,
    ``,
    `    this.pendingDeletes.add(id)`,
    ``,
    `    // Optimistically remove`,
    `    runInAction(() => {`,
    `      this.items.delete(id)`,
    `    })`,
    ``,
    `    try {`,
    `      const result = await ${nameLower}Api.delete(id)`,
    ``,
    `      if (!result.ok) {`,
    `        throw new Error(result.error?.message || 'Failed to delete')`,
    `      }`,
    ``,
    `      runInAction(() => {`,
    `        this.pendingDeletes.delete(id)`,
    `      })`,
    `    } catch (e) {`,
    `      runInAction(() => {`,
    `        // Rollback`,
    `        this.items.set(id, existing)`,
    `        this.pendingDeletes.delete(id)`,
    `        this.error = e instanceof Error ? e.message : 'Failed to delete'`,
    `      })`,
    `      throw e`,
    `    }`,
    `  }`,
    ``,
    `  /** Clear error state */`,
    `  clearError() {`,
    `    this.error = null`,
    `  }`,
    ``,
    `  /** Clear all data */`,
    `  clear() {`,
    `    this.items.clear()`,
    `    this.error = null`,
    `    this.isLoading = false`,
    `  }`,
    `}`,
    ""
  ];
}
function generateApiRootStore(models) {
  const lines = [
    "// ============================================================================",
    "// Root Store",
    "// ============================================================================",
    "",
    "export class RootStore {"
  ];
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`  ${nameLower}: ${model.name}Store`);
  }
  lines.push("");
  lines.push("  constructor() {");
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`    this.${nameLower} = new ${model.name}Store()`);
  }
  lines.push("    makeAutoObservable(this)");
  lines.push("  }");
  lines.push("");
  lines.push("  /** Clear all stores */");
  lines.push("  clearAll() {");
  for (const model of models) {
    const nameLower = toCamelCase2(model.name);
    lines.push(`    this.${nameLower}.clear()`);
  }
  lines.push("  }");
  lines.push("}");
  lines.push("");
  return lines;
}
function generateApiFactory(models) {
  return [
    "// ============================================================================",
    "// Factory & Singleton",
    "// ============================================================================",
    "",
    "let rootStore: RootStore | null = null",
    "",
    "/**",
    " * Get or create the root store singleton",
    " */",
    "export function getStore(): RootStore {",
    "  if (!rootStore) {",
    "    rootStore = new RootStore()",
    "  }",
    "  return rootStore",
    "}",
    "",
    "/**",
    " * Create a new root store (for testing or SSR)",
    " */",
    "export function createStore(): RootStore {",
    "  return new RootStore()",
    "}",
    "",
    "/**",
    " * Reset the singleton store",
    " */",
    "export function resetStore(): void {",
    "  rootStore = null",
    "}",
    ""
  ];
}
export {
  CUSTOM_REGION_END,
  CUSTOM_REGION_START,
  CUSTOM_ROUTES_SCAFFOLD,
  DEFAULT_PRISMA_CONFIG,
  DEFAULT_PRISMA_HEADER,
  FEATURE_DEPS,
  GENERATED_FILE_LICENSE_HEADER,
  PRISMA_CONFIG_MANAGED_COMMENT,
  SCHEMA_HEADER_MANAGED_COMMENT,
  SCHEMA_HEADER_REGION_ID,
  configIsDowngraded,
  enforcePrismaConfig,
  enforceSchemaHeader,
  ensureCustomRoutes,
  ensureFeatureDeps,
  extractProtectedRegions,
  generateAdminRoutes,
  generateApiClient,
  generateApiDomainStore,
  generateApiOverview,
  generateAuthStore,
  generateDbModule,
  generateDocs,
  generateDocsSiteScaffold,
  generateDocsTsConfig,
  generateDomainStore,
  generateFromPrisma,
  generateMSTCollection,
  generateMSTCollections,
  generateMSTDomain,
  generateMSTModel,
  generateMSTModels,
  generateModelDoc,
  generateModelHooks,
  generateModelRoutes,
  generateModelStore,
  generateModelTypes,
  generateModelsIndex,
  generateRoutes,
  generateRoutesIndex,
  generateServer,
  generateServerFunctions,
  generateShogoClient,
  generateSqliteDbModule,
  generateStores,
  generateStoresIndex,
  generateTypes,
  generateTypesIndex,
  generateTypesPerModel,
  generateVoiceComponents,
  getIdField,
  getRelationFields2 as getRelationFields,
  getScalarFields,
  getUserModel,
  hasMarkedSchemaHeader,
  hasUserModel,
  headerIsDowngraded,
  mergeProtectedRegions,
  parsePrismaSchema,
  toCamelCase2 as toCamelCase,
  toKebabCase
};
