// @shogo-ai/db/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  createAdapterSync: () => createAdapterSync,
  createDatabaseAdapter: () => createDatabaseAdapter,
  createPrismaClient: () => createPrismaClient,
  createPrismaClientSync: () => createPrismaClientSync,
  detectProvider: () => detectProvider,
  detectSchemaProvider: () => detectSchemaProvider,
  getCurrentProvider: () => getCurrentProvider,
  getTestDatabaseUrl: () => getTestDatabaseUrl,
  isPostgres: () => isPostgres,
  isTestMode: () => isTestMode,
  restoreSchema: () => restoreSchema,
  transformSchema: () => transformSchema,
  transformSchemaFile: () => transformSchemaFile
});
module.exports = __toCommonJS(index_exports);

// src/adapters.ts
function detectProvider(url) {
  const envProvider = process.env.DATABASE_PROVIDER?.toLowerCase();
  if (envProvider === "sqlite") return "sqlite";
  if (envProvider === "postgresql" || envProvider === "postgres") return "postgresql";
  const connectionUrl = url ?? process.env.DATABASE_URL ?? "";
  if (connectionUrl.startsWith("file:") || connectionUrl.endsWith(".db")) {
    return "sqlite";
  }
  if (connectionUrl.startsWith("postgres://") || connectionUrl.startsWith("postgresql://")) {
    return "postgresql";
  }
  return "postgresql";
}
async function createDatabaseAdapter(config = "auto") {
  const options = typeof config === "string" ? { provider: config } : config;
  const url = options.url ?? process.env.DATABASE_URL ?? "";
  const provider = options.provider === "auto" || !options.provider ? detectProvider(url) : options.provider;
  console.log(`[shogo/db] Creating ${provider} adapter`);
  if (provider === "sqlite") {
    return createSqliteAdapter(url || "file:./dev.db");
  } else {
    return createPostgresAdapter(url);
  }
}
async function createPostgresAdapter(connectionString) {
  try {
    const module2 = await import("@prisma/adapter-pg");
    const PrismaPg = module2.PrismaPg || module2.default?.PrismaPg;
    return new PrismaPg({ connectionString });
  } catch (error) {
    throw new Error(
      "Failed to create PostgreSQL adapter. Make sure @prisma/adapter-pg is installed.\nRun: bun add @prisma/adapter-pg"
    );
  }
}
async function createSqliteAdapter(url) {
  try {
    const module2 = await import("@prisma/adapter-libsql");
    const PrismaLibSql = module2.PrismaLibSql || module2.default?.PrismaLibSql;
    return new PrismaLibSql({ url });
  } catch (error) {
    throw new Error(
      "Failed to create SQLite adapter. Make sure @prisma/adapter-libsql is installed.\nRun: bun add @prisma/adapter-libsql"
    );
  }
}
var globalForPrisma = globalThis;
async function createPrismaClient(PrismaClientClass, config = {}) {
  if (globalForPrisma.__shogo_prisma) {
    return globalForPrisma.__shogo_prisma;
  }
  const adapter = await createDatabaseAdapter(config);
  const log = config.log ?? (process.env.NODE_ENV === "development" ? ["warn", "error"] : ["error"]);
  const client = new PrismaClientClass({
    adapter,
    log
  });
  if (process.env.NODE_ENV !== "production") {
    globalForPrisma.__shogo_prisma = client;
  }
  return client;
}
function createAdapterSync(config = {}) {
  const url = config.url ?? process.env.DATABASE_URL ?? "";
  const provider = config.provider === "auto" || !config.provider ? detectProvider(url) : config.provider;
  console.log(`[shogo/db] Creating ${provider} adapter (sync)`);
  if (provider === "sqlite") {
    try {
      const module2 = require("@prisma/adapter-libsql");
      const PrismaLibSql = module2.PrismaLibSql || module2.default?.PrismaLibSql;
      return new PrismaLibSql({ url: url || "file:./dev.db" });
    } catch {
      throw new Error(
        "SQLite adapter not found. Install it with: bun add @prisma/adapter-libsql"
      );
    }
  } else {
    try {
      const module2 = require("@prisma/adapter-pg");
      const PrismaPg = module2.PrismaPg || module2.default?.PrismaPg;
      return new PrismaPg({ connectionString: url });
    } catch {
      throw new Error(
        "PostgreSQL adapter not found. Install it with: bun add @prisma/adapter-pg"
      );
    }
  }
}
function createPrismaClientSync(PrismaClientClass, config = {}) {
  if (globalForPrisma.__shogo_prisma) {
    return globalForPrisma.__shogo_prisma;
  }
  const adapter = createAdapterSync(config);
  const log = config.log ?? (process.env.NODE_ENV === "development" ? ["warn", "error"] : ["error"]);
  const client = new PrismaClientClass({
    adapter,
    log
  });
  if (process.env.NODE_ENV !== "production") {
    globalForPrisma.__shogo_prisma = client;
  }
  return client;
}
function isTestMode() {
  return detectProvider() === "sqlite";
}
function isPostgres() {
  return detectProvider() === "postgresql";
}
function getTestDatabaseUrl() {
  return "file:./test.db";
}
function getCurrentProvider() {
  return detectProvider();
}

// src/schema-transformer.ts
var import_fs = require("fs");
var import_child_process = require("child_process");
var import_path = require("path");
var POSTGRES_TO_SQLITE_TRANSFORMS = [
  // uuid() is PostgreSQL-specific, use cuid() for compatibility
  { pattern: /@default\(uuid\(\)\)/g, replacement: "@default(cuid())" },
  // dbgenerated() with PostgreSQL-specific functions
  { pattern: /@default\(dbgenerated\("gen_random_uuid\(\)"\)\)/g, replacement: "@default(cuid())" },
  // auto() is PostgreSQL-specific for autoincrement
  { pattern: /@default\(auto\(\)\)/g, replacement: "@default(autoincrement())" },
  // Remove @db.* native type annotations (SQLite doesn't support these)
  { pattern: /@db\.\w+(\([^)]*\))?/g, replacement: "" }
];
var SQLITE_TO_POSTGRES_TRANSFORMS = [
  // autoincrement() is SQLite-specific, use auto() for PostgreSQL
  { pattern: /@default\(autoincrement\(\)\)/g, replacement: "@default(auto())" }
];
function detectSchemaProvider(schemaContent) {
  if (schemaContent.includes('provider = "postgresql"') || schemaContent.includes("provider = 'postgresql'")) {
    return "postgresql";
  }
  if (schemaContent.includes('provider = "sqlite"') || schemaContent.includes("provider = 'sqlite'")) {
    return "sqlite";
  }
  return null;
}
function transformSchema(schemaContent, targetProvider) {
  const currentProvider = detectSchemaProvider(schemaContent);
  const warnings = [];
  if (currentProvider === targetProvider) {
    return { content: schemaContent, warnings: [] };
  }
  let result = schemaContent;
  if (targetProvider === "sqlite") {
    result = result.replace(/provider\s*=\s*"postgresql"/g, 'provider = "sqlite"');
  } else {
    result = result.replace(/provider\s*=\s*"sqlite"/g, 'provider = "postgresql"');
  }
  const transforms = targetProvider === "sqlite" ? POSTGRES_TO_SQLITE_TRANSFORMS : SQLITE_TO_POSTGRES_TRANSFORMS;
  for (const { pattern, replacement } of transforms) {
    const matches = result.match(pattern);
    if (matches && matches.length > 0) {
      if (pattern.source.includes("db\\.")) {
        warnings.push(`Removed PostgreSQL-specific @db.* annotations for SQLite`);
      }
      result = result.replace(pattern, replacement);
    }
  }
  result = result.replace(/  +/g, " ");
  return { content: result, warnings };
}
async function transformSchemaFile(options) {
  const {
    schemaPath,
    targetProvider,
    generate = true,
    push = false,
    databaseUrl,
    verbose = false
  } = options;
  const absolutePath = (0, import_path.resolve)(process.cwd(), schemaPath);
  if (!(0, import_fs.existsSync)(absolutePath)) {
    throw new Error(`Schema file not found: ${absolutePath}`);
  }
  const originalContent = (0, import_fs.readFileSync)(absolutePath, "utf-8");
  const originalProvider = detectSchemaProvider(originalContent);
  if (!originalProvider) {
    throw new Error("Could not detect database provider in schema");
  }
  const { content: newContent, warnings } = transformSchema(originalContent, targetProvider);
  const modified = newContent !== originalContent;
  if (modified) {
    (0, import_fs.writeFileSync)(absolutePath, newContent, "utf-8");
    if (verbose) {
      console.log(`Transformed schema from ${originalProvider} to ${targetProvider}`);
    }
  } else if (verbose) {
    console.log(`Schema already using ${targetProvider} provider`);
  }
  const env = { ...process.env };
  if (targetProvider === "sqlite") {
    env.DATABASE_URL = databaseUrl || "file:./test.db";
  } else if (databaseUrl) {
    env.DATABASE_URL = databaseUrl;
  }
  const schemaDir = (0, import_path.dirname)(absolutePath);
  if (generate) {
    if (verbose) {
      console.log("Running prisma generate...");
    }
    try {
      (0, import_child_process.execSync)("bun x --bun prisma generate", {
        cwd: schemaDir,
        env,
        stdio: verbose ? "inherit" : "pipe"
      });
    } catch (error) {
      throw new Error(`prisma generate failed: ${error}`);
    }
  }
  if (push) {
    if (verbose) {
      console.log("Running prisma db push...");
    }
    try {
      (0, import_child_process.execSync)("bun x --bun prisma db push", {
        cwd: schemaDir,
        env,
        stdio: verbose ? "inherit" : "pipe"
      });
    } catch (error) {
      throw new Error(`prisma db push failed: ${error}`);
    }
  }
  return {
    modified,
    originalProvider,
    newProvider: targetProvider,
    schemaPath: absolutePath,
    warnings
  };
}
async function restoreSchema(schemaPath, verbose = false) {
  await transformSchemaFile({
    schemaPath,
    targetProvider: "postgresql",
    generate: true,
    push: false,
    verbose
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createAdapterSync,
  createDatabaseAdapter,
  createPrismaClient,
  createPrismaClientSync,
  detectProvider,
  detectSchemaProvider,
  getCurrentProvider,
  getTestDatabaseUrl,
  isPostgres,
  isTestMode,
  restoreSchema,
  transformSchema,
  transformSchemaFile
});
