// @shogo-ai/sdk/dist/tools/index.js
// src/tools/client.ts
var BROWSER_PATHS = {
  schemas: "/api/tools/schemas",
  execute: "/api/tools/execute"
};
var RUNTIME_PATHS = {
  schemas: "/agent/tools/schemas",
  execute: "/agent/tools/execute"
};
var ToolsClient = class {
  constructor(config) {
    this.baseUrl = config?.baseUrl?.replace(/\/$/, "") ?? "";
    this.headers = { ...config?.headers };
    this.doFetch = config?.fetch ?? globalThis.fetch.bind(globalThis);
    this.paths = config?.paths ?? BROWSER_PATHS;
  }
  url(path) {
    return `${this.baseUrl}${path}`;
  }
  /**
   * List all installed tools with their schemas.
   */
  async listTools() {
    const res = await this.doFetch(this.url(this.paths.schemas), {
      headers: this.headers
    });
    if (!res.ok) {
      throw new Error(`Tools list failed (${res.status}): ${await res.text().catch(() => res.statusText)}`);
    }
    const data = await res.json();
    return data.tools ?? [];
  }
  /**
   * Execute an installed tool by name.
   *
   * The runtime tool-execution path always JSON.stringifies the tool's
   * response into `data`, so `data` arrives as a string from the wire.
   * This method auto-parses it: every JSON-encoded payload (object,
   * array, primitive) is rehydrated to its natural JS shape. Tools that
   * return raw text (markdown, etc.) leave `data` as the original
   * string — `JSON.parse` failures are caught and ignored. Error
   * payloads (`ok: false`) are passed through untouched.
   *
   * @param toolName - Tool slug (e.g. `METAADS_GET_INSIGHTS`, `GMAIL_SEND_EMAIL`)
   * @param args - Arguments matching the tool's parameter schema
   */
  async execute(toolName, args = {}) {
    const res = await this.doFetch(this.url(this.paths.execute), {
      method: "POST",
      headers: { "Content-Type": "application/json", ...this.headers },
      body: JSON.stringify({ tool: toolName, args })
    });
    if (!res.ok && res.status !== 404) {
      const text = await res.text().catch(() => res.statusText);
      return { ok: false, error: `Request failed (${res.status}): ${text}` };
    }
    const result = await res.json();
    if (result.ok && typeof result.data === "string") {
      try {
        ;
        result.data = JSON.parse(result.data);
      } catch {
      }
    }
    return result;
  }
};
var defaultClient = null;
function getToolsClient(config) {
  if (!defaultClient || config) {
    defaultClient = new ToolsClient(config);
  }
  return defaultClient;
}
var defaultServerClient = null;
function getServerToolsClient(config) {
  if (defaultServerClient && !config) return defaultServerClient;
  const runtimeToken = process.env.RUNTIME_AUTH_SECRET;
  const portRaw = process.env.RUNTIME_PORT;
  if (!runtimeToken || !portRaw) {
    throw new Error(
      "getServerToolsClient: RUNTIME_AUTH_SECRET and RUNTIME_PORT must be set. These are injected automatically inside a Shogo-managed pod. For browser code, use `getToolsClient()` or `useTools()` instead."
    );
  }
  const port = parseInt(portRaw, 10);
  if (!Number.isFinite(port) || port <= 0) {
    throw new Error(`getServerToolsClient: invalid RUNTIME_PORT "${portRaw}"`);
  }
  const merged = {
    baseUrl: config?.baseUrl ?? `http://127.0.0.1:${port}`,
    paths: RUNTIME_PATHS,
    headers: { "x-runtime-token": runtimeToken, ...config?.headers ?? {} },
    ...config?.fetch ? { fetch: config.fetch } : {}
  };
  const client = new ToolsClient(merged);
  if (!config) defaultServerClient = client;
  return client;
}

// src/tools/hooks.ts
import { useState, useEffect, useCallback, useRef } from "react";
function useToolsClient(config) {
  const ref = useRef(null);
  if (!ref.current) {
    ref.current = config ? new ToolsClient(config) : getToolsClient();
  }
  return ref.current;
}
function useTools(options) {
  const client = useToolsClient(options?.config);
  const [tools, setTools] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const fetchSchemas = useCallback(async () => {
    setLoading(true);
    try {
      const schemas = await client.listTools();
      setTools(schemas);
      setError(null);
    } catch (e) {
      setError(e instanceof Error ? e : new Error(String(e)));
    } finally {
      setLoading(false);
    }
  }, [client]);
  useEffect(() => {
    if (options?.autoLoad !== false) {
      fetchSchemas();
    } else {
      setLoading(false);
    }
  }, [fetchSchemas, options?.autoLoad]);
  const execute = useCallback(
    (toolName, args) => client.execute(toolName, args),
    [client]
  );
  return { tools, loading, error, refresh: fetchSchemas, execute };
}
export {
  ToolsClient,
  getServerToolsClient,
  getToolsClient,
  useTools
};
