// @shogo-ai/sdk/dist/tools/server.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/tools/server.ts
var server_exports = {};
__export(server_exports, {
  createToolsHandlers: () => createToolsHandlers
});
module.exports = __toCommonJS(server_exports);
function json(data, status = 200, extraHeaders = {}) {
  return new Response(JSON.stringify(data), {
    status,
    headers: { "Content-Type": "application/json", ...extraHeaders }
  });
}
var NOOP_LOGGER = () => {
};
function resolveProxyOptions(config) {
  if (config.proxy) {
    if (!config.proxy.runtimeToken || !config.proxy.runtimePort) {
      throw new Error("createToolsHandlers: proxy requires runtimeToken and runtimePort");
    }
    return config.proxy;
  }
  if (typeof process === "undefined" || !process.env) return null;
  const runtimeToken = process.env.RUNTIME_AUTH_SECRET;
  const portRaw = process.env.RUNTIME_PORT;
  if (!runtimeToken || !portRaw) return null;
  const runtimePort = parseInt(portRaw, 10);
  if (!Number.isFinite(runtimePort) || runtimePort <= 0) return null;
  return {
    runtimeToken,
    runtimePort,
    ...config.fetch ? { fetch: config.fetch } : {}
  };
}
function createProxyHandlers(proxy, log) {
  const fetchImpl = proxy.fetch ?? globalThis.fetch;
  if (typeof fetchImpl !== "function") {
    throw new Error(
      "createToolsHandlers proxy: global fetch is unavailable; pass proxy.fetch"
    );
  }
  const apiBase = `http://127.0.0.1:${proxy.runtimePort}`;
  async function passThrough(method, path, req, label) {
    if (req.method !== method) return json({ error: "Method Not Allowed" }, 405);
    try {
      const init = {
        method,
        headers: { "x-runtime-token": proxy.runtimeToken }
      };
      if (method === "POST") {
        const body = await req.text();
        init.body = body;
        init.headers = {
          ...init.headers,
          "content-type": req.headers.get("content-type") ?? "application/json"
        };
      }
      const upstream = await fetchImpl(`${apiBase}${path}`, init);
      const text = await upstream.text();
      const headers = { "content-type": "application/json" };
      const ct = upstream.headers.get("content-type");
      if (ct) headers["content-type"] = ct;
      return new Response(text, { status: upstream.status, headers });
    } catch (err) {
      log("error", `proxy ${label} failed`, {
        error: err instanceof Error ? err.message : String(err)
      });
      return json({ error: `${label} proxy failed` }, 502);
    }
  }
  return {
    execute: (req) => passThrough("POST", "/agent/tools/execute", req, "execute"),
    list: (req) => passThrough("GET", "/agent/tools/schemas", req, "list")
  };
}
function notImplemented(label) {
  return async () => json(
    {
      error: "Not Implemented",
      detail: `${label} requires runtime-token proxy mode. Set RUNTIME_AUTH_SECRET and RUNTIME_PORT in env, or pass an explicit proxy option. In a Shogo-managed pod these are injected automatically.`
    },
    501
  );
}
function createToolsHandlers(config = {}) {
  const log = config.logger ?? NOOP_LOGGER;
  const proxy = resolveProxyOptions(config);
  if (proxy) return createProxyHandlers(proxy, log);
  return {
    execute: notImplemented("tools.execute"),
    list: notImplemented("tools.list")
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createToolsHandlers
});
