// @shogo-ai/sdk/dist/memory/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/memory/index.ts
var memory_exports = {};
__export(memory_exports, {
  MemorySearchEngine: () => MemorySearchEngine,
  MemoryStore: () => MemoryStore,
  createLlmSummarizer: () => createLlmSummarizer,
  createSqliteDriver: () => createSqliteDriver,
  detectDriver: () => detectDriver,
  identitySummarizer: () => identitySummarizer,
  sanitizeUserId: () => sanitizeUserId
});
module.exports = __toCommonJS(memory_exports);

// src/memory/engine.ts
var import_node_fs = require("fs");
var import_node_path = require("path");

// src/memory/drivers/bun.ts
var import_node_module = require("module");
var import_meta = {};
var req = (0, import_node_module.createRequire)(
  typeof __filename === "string" ? __filename : import_meta.url
);
function isBunRuntime() {
  return typeof globalThis.Bun !== "undefined";
}
var _bunRuntimeGuardForTests = { check: isBunRuntime };
var createBunDriver = (dbPath) => {
  if (!_bunRuntimeGuardForTests.check()) {
    throw new Error("[@shogo-ai/sdk/memory] bun:sqlite driver requires Bun runtime");
  }
  const mod = req("bun:sqlite");
  const db = new mod.Database(dbPath);
  return wrap(db);
};
function wrap(db) {
  return {
    exec: (sql) => db.exec(sql),
    prepare: (sql) => {
      const stmt = db.prepare(sql);
      return {
        run: (...args) => stmt.run(...args),
        get: (...args) => stmt.get(...args),
        all: (...args) => stmt.all(...args)
      };
    },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    transaction: (fn) => db.transaction(fn),
    close: () => db.close()
  };
}

// src/memory/drivers/node.ts
var import_node_module2 = require("module");
var import_meta2 = {};
var req2 = (0, import_node_module2.createRequire)(
  typeof __filename === "string" ? __filename : import_meta2.url
);
var createNodeDriver = (dbPath) => {
  let Ctor;
  try {
    const mod = req2("better-sqlite3");
    Ctor = mod.default ?? mod;
  } catch (e) {
    const detail = e instanceof Error ? e.message : String(e);
    throw new Error(
      `[@shogo-ai/sdk/memory] Missing optional peer dependency 'better-sqlite3'. Install it: npm install better-sqlite3 (${detail})`
    );
  }
  const db = new Ctor(dbPath);
  return wrap2(db);
};
function wrap2(db) {
  return {
    exec: (sql) => db.exec(sql),
    prepare: (sql) => {
      const stmt = db.prepare(sql);
      return {
        run: (...args) => stmt.run(...args),
        get: (...args) => stmt.get(...args),
        all: (...args) => stmt.all(...args)
      };
    },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    transaction: (fn) => db.transaction(fn),
    close: () => db.close()
  };
}

// src/memory/drivers/index.ts
function detectDriver() {
  return isBunRuntime() ? createBunDriver : createNodeDriver;
}
var createSqliteDriver = (dbPath) => {
  return detectDriver()(dbPath);
};

// src/memory/engine.ts
var CHUNK_SIZE = 6;
var CHUNK_OVERLAP = 2;
var MemorySearchEngine = class {
  /**
   * @param workspaceDir - Directory containing `MEMORY.md` and optional `memory/*.md`
   * @param options - Optional `createDriver` to open `.memory-index.db` (default: better-sqlite3 or bun:sqlite)
   */
  constructor(workspaceDir, options) {
    this.workspaceDir = workspaceDir;
    this.options = { createDriver: options?.createDriver ?? createSqliteDriver };
    const dbPath = (0, import_node_path.join)(workspaceDir, ".memory-index.db");
    this.db = this.options.createDriver(dbPath);
    this.db.exec("PRAGMA journal_mode = WAL");
    this.init();
  }
  init() {
    this.db.exec(`
      CREATE VIRTUAL TABLE IF NOT EXISTS memory_fts USING fts5(
        file, chunk, line_start, line_end,
        tokenize='porter unicode61'
      );

      CREATE TABLE IF NOT EXISTS memory_meta (
        file TEXT PRIMARY KEY,
        last_indexed_mtime INTEGER NOT NULL,
        chunk_count INTEGER NOT NULL DEFAULT 0
      );

      CREATE TABLE IF NOT EXISTS memory_terms (
        term TEXT PRIMARY KEY,
        doc_freq INTEGER NOT NULL DEFAULT 0
      );

      CREATE TABLE IF NOT EXISTS memory_vectors (
        rowid INTEGER PRIMARY KEY,
        file TEXT NOT NULL,
        chunk TEXT NOT NULL,
        line_start INTEGER NOT NULL,
        line_end INTEGER NOT NULL,
        vector_json TEXT NOT NULL
      );
    `);
  }
  /**
   * Index all memory files that have changed since last index.
   * Called automatically before search, or can be called after writes.
   */
  /**
   * Force the next reindex() to re-process this file regardless of its mtime.
   * Used by MemoryStore right after it rewrites MEMORY.md, so the FTS index
   * picks up the new content even when the write lands inside the same
   * millisecond as the previous write (atomic-rename mtime equality bug
   * surfaced on overlay / tmpfs filesystems).
   */
  invalidateMeta(relativePath) {
    this.db.prepare("DELETE FROM memory_meta WHERE file = ?").run(relativePath);
  }
  reindex() {
    const files = this.getMemoryFiles();
    const staleFiles = [];
    const getMeta = this.db.prepare("SELECT last_indexed_mtime FROM memory_meta WHERE file = ?");
    const onDisk = new Set(files.map((f) => f.relativePath));
    for (const { relativePath, absolutePath } of files) {
      const mtime = (0, import_node_fs.statSync)(absolutePath).mtimeMs;
      const meta = getMeta.get(relativePath);
      if (!meta || meta.last_indexed_mtime < mtime) {
        staleFiles.push(relativePath);
      }
    }
    const tracked = this.db.prepare("SELECT file FROM memory_meta").all();
    for (const row of tracked) {
      if (!onDisk.has(row.file)) {
        staleFiles.push(row.file);
      }
    }
    if (staleFiles.length === 0) return;
    const indexFile = this.db.transaction((relPath) => {
      const absPath = (0, import_node_path.join)(this.workspaceDir, relPath);
      if (!(0, import_node_fs.existsSync)(absPath)) {
        this.db.prepare("DELETE FROM memory_fts WHERE file = ?").run(relPath);
        this.db.prepare("DELETE FROM memory_vectors WHERE file = ?").run(relPath);
        this.db.prepare("DELETE FROM memory_meta WHERE file = ?").run(relPath);
        return;
      }
      const content = (0, import_node_fs.readFileSync)(absPath, "utf-8");
      const chunks = this.chunkText(content, relPath);
      this.db.prepare("DELETE FROM memory_fts WHERE file = ?").run(relPath);
      this.db.prepare("DELETE FROM memory_vectors WHERE file = ?").run(relPath);
      const insertFts = this.db.prepare(
        "INSERT INTO memory_fts (file, chunk, line_start, line_end) VALUES (?, ?, ?, ?)"
      );
      const insertVec = this.db.prepare(
        "INSERT INTO memory_vectors (file, chunk, line_start, line_end, vector_json) VALUES (?, ?, ?, ?, ?)"
      );
      for (const chunk of chunks) {
        insertFts.run(chunk.file, chunk.chunk, chunk.lineStart, chunk.lineEnd);
        const vector = this.computeTfIdf(chunk.chunk);
        insertVec.run(chunk.file, chunk.chunk, chunk.lineStart, chunk.lineEnd, JSON.stringify(vector));
      }
      const mtime = (0, import_node_fs.statSync)(absPath).mtimeMs;
      this.db.prepare(
        "INSERT OR REPLACE INTO memory_meta (file, last_indexed_mtime, chunk_count) VALUES (?, ?, ?)"
      ).run(relPath, mtime, chunks.length);
    });
    for (const file of staleFiles) {
      indexFile(file);
    }
    this.rebuildTermFrequencies();
  }
  /**
   * Search memory with hybrid keyword + semantic ranking.
   * Returns top-k results sorted by combined score.
   */
  search(query, limit = 10) {
    this.reindex();
    const keywordResults = this.keywordSearch(query, limit * 2);
    const semanticResults = this.semanticSearch(query, limit * 2);
    const merged = /* @__PURE__ */ new Map();
    for (const r of keywordResults) {
      const key = `${r.file}:${r.lineStart}`;
      merged.set(key, r);
    }
    for (const r of semanticResults) {
      const key = `${r.file}:${r.lineStart}`;
      const existing = merged.get(key);
      if (existing) {
        existing.score = existing.score * 0.6 + r.score * 0.4;
        existing.matchType = "hybrid";
      } else {
        merged.set(key, r);
      }
    }
    return [...merged.values()].sort((a, b) => b.score - a.score).slice(0, limit);
  }
  /** Close the database connection */
  close() {
    this.db.close();
  }
  keywordSearch(query, limit) {
    try {
      const escaped = query.replace(/['"]/g, " ").trim();
      if (!escaped) return [];
      const terms = escaped.split(/\s+/).filter(Boolean);
      const ftsQuery = terms.map((t) => `"${t}"`).join(" OR ");
      const rows = this.db.prepare(`
        SELECT file, chunk, line_start, line_end, rank
        FROM memory_fts
        WHERE memory_fts MATCH ?
        ORDER BY rank
        LIMIT ?
      `).all(ftsQuery, limit);
      if (rows.length === 0) return [];
      const maxRank = Math.max(...rows.map((r) => Math.abs(r.rank)));
      return rows.map((r) => ({
        file: r.file,
        chunk: r.chunk,
        lineStart: r.line_start,
        lineEnd: r.line_end,
        score: maxRank > 0 ? 1 - Math.abs(r.rank) / maxRank : 0.5,
        matchType: "keyword"
      }));
    } catch {
      return this.fallbackKeywordSearch(query, limit);
    }
  }
  fallbackKeywordSearch(query, limit) {
    const lower = query.toLowerCase();
    const rows = this.db.prepare(
      "SELECT file, chunk, line_start, line_end FROM memory_vectors"
    ).all();
    return rows.filter((r) => r.chunk.toLowerCase().includes(lower)).slice(0, limit).map((r, i, arr) => ({
      file: r.file,
      chunk: r.chunk,
      lineStart: r.line_start,
      lineEnd: r.line_end,
      score: 1 - i / Math.max(arr.length, 1),
      matchType: "keyword"
    }));
  }
  semanticSearch(query, limit) {
    const queryVec = this.computeTfIdf(query);
    if (Object.keys(queryVec).length === 0) return [];
    const rows = this.db.prepare(
      "SELECT file, chunk, line_start, line_end, vector_json FROM memory_vectors"
    ).all();
    const scored = [];
    for (const r of rows) {
      let docVec;
      try {
        docVec = JSON.parse(r.vector_json);
      } catch {
        continue;
      }
      const sim = cosineSimilarity(queryVec, docVec);
      if (sim > 0.05) {
        scored.push({
          file: r.file,
          chunk: r.chunk,
          lineStart: r.line_start,
          lineEnd: r.line_end,
          score: sim,
          matchType: "semantic"
        });
      }
    }
    return scored.sort((a, b) => b.score - a.score).slice(0, limit);
  }
  computeTfIdf(text) {
    const tokens = tokenize(text);
    if (tokens.length === 0) return {};
    const tf = {};
    for (const t of tokens) {
      tf[t] = (tf[t] || 0) + 1;
    }
    for (const t of Object.keys(tf)) {
      tf[t] = tf[t] / tokens.length;
    }
    const totalDocs = this.getTotalDocCount();
    const vec = {};
    for (const [term, termFreq] of Object.entries(tf)) {
      const docFreq = this.getDocFreq(term);
      const idf = Math.log((totalDocs + 1) / (docFreq + 1)) + 1;
      vec[term] = termFreq * idf;
    }
    return vec;
  }
  getTotalDocCount() {
    const row = this.db.prepare("SELECT COUNT(*) as cnt FROM memory_vectors").get();
    return row.cnt || 1;
  }
  getDocFreq(term) {
    const row = this.db.prepare("SELECT doc_freq FROM memory_terms WHERE term = ?").get(term);
    return row?.doc_freq ?? 0;
  }
  rebuildTermFrequencies() {
    const rows = this.db.prepare("SELECT chunk FROM memory_vectors").all();
    const docFreqs = {};
    for (const row of rows) {
      const uniqueTerms = new Set(tokenize(row.chunk));
      for (const term of uniqueTerms) {
        docFreqs[term] = (docFreqs[term] || 0) + 1;
      }
    }
    this.db.exec("DELETE FROM memory_terms");
    const insert = this.db.prepare("INSERT INTO memory_terms (term, doc_freq) VALUES (?, ?)");
    const insertAll = this.db.transaction(() => {
      for (const [term, freq] of Object.entries(docFreqs)) {
        insert.run(term, freq);
      }
    });
    insertAll();
  }
  chunkText(content, file) {
    const lines = content.split("\n");
    if (lines.length <= CHUNK_SIZE) {
      return [{
        file,
        chunk: content.trim(),
        lineStart: 1,
        lineEnd: lines.length,
        timestamp: Date.now()
      }];
    }
    const chunks = [];
    for (let i = 0; i < lines.length; i += CHUNK_SIZE - CHUNK_OVERLAP) {
      const end = Math.min(i + CHUNK_SIZE, lines.length);
      const chunkLines = lines.slice(i, end);
      const text = chunkLines.join("\n").trim();
      if (text.length > 10) {
        chunks.push({
          file,
          chunk: text,
          lineStart: i + 1,
          lineEnd: end,
          timestamp: Date.now()
        });
      }
      if (end >= lines.length) break;
    }
    return chunks;
  }
  getMemoryFiles() {
    const files = [];
    const memoryMd = (0, import_node_path.join)(this.workspaceDir, "MEMORY.md");
    if ((0, import_node_fs.existsSync)(memoryMd)) {
      files.push({ relativePath: "MEMORY.md", absolutePath: memoryMd });
    }
    const memoryDir = (0, import_node_path.join)(this.workspaceDir, "memory");
    if ((0, import_node_fs.existsSync)(memoryDir)) {
      for (const f of (0, import_node_fs.readdirSync)(memoryDir)) {
        if (f.endsWith(".md")) {
          files.push({
            relativePath: `memory/${f}`,
            absolutePath: (0, import_node_path.join)(memoryDir, f)
          });
        }
      }
    }
    return files;
  }
};
var STOP_WORDS = /* @__PURE__ */ new Set([
  "a",
  "an",
  "the",
  "is",
  "are",
  "was",
  "were",
  "be",
  "been",
  "being",
  "have",
  "has",
  "had",
  "do",
  "does",
  "did",
  "will",
  "would",
  "could",
  "should",
  "may",
  "might",
  "shall",
  "can",
  "need",
  "dare",
  "ought",
  "used",
  "to",
  "of",
  "in",
  "for",
  "on",
  "with",
  "at",
  "by",
  "from",
  "as",
  "into",
  "through",
  "during",
  "before",
  "after",
  "above",
  "below",
  "between",
  "out",
  "off",
  "over",
  "under",
  "again",
  "further",
  "then",
  "once",
  "and",
  "but",
  "or",
  "nor",
  "not",
  "so",
  "yet",
  "both",
  "each",
  "few",
  "more",
  "most",
  "other",
  "some",
  "such",
  "no",
  "only",
  "own",
  "same",
  "than",
  "too",
  "very",
  "just",
  "because",
  "if",
  "when",
  "while",
  "where",
  "how",
  "what",
  "which",
  "who",
  "this",
  "that",
  "these",
  "those",
  "i",
  "me",
  "my",
  "we",
  "our",
  "you",
  "your",
  "he",
  "him",
  "his",
  "she",
  "her",
  "it",
  "its",
  "they",
  "them",
  "their",
  "all",
  "any",
  "about",
  "up"
]);
function tokenize(text) {
  return text.toLowerCase().replace(/[^a-z0-9\s_-]/g, " ").split(/\s+/).filter((t) => t.length > 1 && !STOP_WORDS.has(t)).map(stem);
}
function stem(word) {
  if (word.length < 4) return word;
  if (word.endsWith("ies") && word.length > 4) return word.slice(0, -3) + "y";
  if (word.endsWith("ing") && word.length > 5) return word.slice(0, -3);
  if (word.endsWith("tion")) return word.slice(0, -4);
  if (word.endsWith("ness")) return word.slice(0, -4);
  if (word.endsWith("ment")) return word.slice(0, -4);
  if (word.endsWith("able")) return word.slice(0, -4);
  if (word.endsWith("ful")) return word.slice(0, -3);
  if (word.endsWith("ly") && word.length > 4) return word.slice(0, -2);
  if (word.endsWith("ed") && word.length > 4) return word.slice(0, -2);
  if (word.endsWith("er") && word.length > 4) return word.slice(0, -2);
  if (word.endsWith("es") && word.length > 4) return word.slice(0, -2);
  if (word.endsWith("s") && !word.endsWith("ss") && word.length > 3) return word.slice(0, -1);
  return word;
}
function cosineSimilarity(a, b) {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  const allKeys = /* @__PURE__ */ new Set([...Object.keys(a), ...Object.keys(b)]);
  for (const key of allKeys) {
    const va = a[key] ?? 0;
    const vb = b[key] ?? 0;
    dot += va * vb;
    normA += va * va;
    normB += vb * vb;
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

// src/memory/store.ts
var import_node_fs2 = require("fs");
var import_node_path2 = require("path");

// src/memory/summarizer.ts
var identitySummarizer = {
  summarize(text) {
    return Promise.resolve(text);
  },
  consolidate(input) {
    const lines = [];
    for (const b of input.existingBullets) {
      const trimmed = b.trim();
      if (trimmed) lines.push(`- ${trimmed}`);
    }
    const transcript = input.transcript.trim();
    if (transcript) lines.push(`- ${transcript}`);
    return Promise.resolve(lines.join("\n"));
  }
};
var DEFAULT_PROMPT = (transcript) => `You extract durable user facts from a conversation transcript for long-term memory.

Rules:
- Output ONLY bullet lines starting with "- " (markdown bullets).
- Each bullet is ONE canonical fact (preferences, decisions, constraints, names, dates).
- No preamble, no closing, no JSON.
- Merge duplicates; prefer stable phrasing (e.g. "prefers_window_seat: true").

Transcript:
---
${transcript}
---

Bullets:`;
var DEFAULT_CONSOLIDATION_PROMPT = (input) => {
  const existing = input.existingBullets.length ? input.existingBullets.map((b) => `- ${b}`).join("\n") : "(none)";
  return `You maintain a long-term memory about a user. You are given (a) the existing memory as bullet lines and (b) a new conversation transcript. Produce the UPDATED COMPLETE set of memory bullets that should persist going forward.

Rules:
- Merge duplicates and near-duplicates into a single bullet.
- When facts conflict, keep ONLY the most recent value (e.g. if favorite color changed from cerulean to turquoise, keep turquoise).
- Drop small talk, greetings, agent replies, and anything transient.
- Keep durable facts: preferences, personal details, relationships, decisions, open follow-ups.
- Use canonical short forms: "favorite color: turquoise", "lives in Honolulu", "friend: Lindsay".
- Output ONLY the bullets, one per line, each prefixed with "- ". No preamble, no headers, no commentary.
- If there is nothing durable to remember, output nothing.

# Existing memory
${existing}

# New conversation transcript
${input.transcript}

# Task
Produce the updated complete memory bullet list.`;
};
function createLlmSummarizer(options) {
  const buildExtract = options.buildPrompt ?? DEFAULT_PROMPT;
  const buildConsolidate = options.buildConsolidationPrompt ?? DEFAULT_CONSOLIDATION_PROMPT;
  return {
    async summarize(text) {
      return options.complete(buildExtract(text));
    },
    async consolidate(input) {
      return options.complete(buildConsolidate(input));
    }
  };
}

// src/memory/store.ts
function sanitizeUserId(userId) {
  const s = userId.trim();
  if (!s) throw new Error("MemoryStore: userId must be non-empty");
  const cleaned = s.replace(/[^a-zA-Z0-9._@-]+/g, "_");
  const safe = cleaned.replace(/^\.+/, "_");
  return safe.slice(0, 256);
}
var BULLET_LINE = /^-\s*(?:\(\d{4}-\d{2}-\d{2}T[^)]+\)\s*)?(.+?)\s*$/;
function parseBullets(text) {
  return text.split("\n").map((l) => l.trim()).filter((l) => l.startsWith("-")).map((l) => l.replace(/^-\s*/, "").trim()).filter(Boolean);
}
var MemoryStore = class {
  constructor(config) {
    this.engine = null;
    const safe = sanitizeUserId(config.userId);
    this.workspaceDir = (0, import_node_path2.join)(config.dir, safe);
    this.createDriver = config.createDriver;
    this.summarizer = config.summarizer ?? identitySummarizer;
  }
  ensureWorkspace() {
    (0, import_node_fs2.mkdirSync)(this.workspaceDir, { recursive: true });
    (0, import_node_fs2.mkdirSync)((0, import_node_path2.join)(this.workspaceDir, "memory"), { recursive: true });
  }
  getEngine() {
    if (!this.engine) {
      this.engine = new MemorySearchEngine(this.workspaceDir, {
        createDriver: this.createDriver
      });
    }
    return this.engine;
  }
  /** Append a bullet to MEMORY.md with an ISO timestamp prefix */
  add(fact) {
    this.ensureWorkspace();
    const memoryMd = (0, import_node_path2.join)(this.workspaceDir, "MEMORY.md");
    const line = `- (${(/* @__PURE__ */ new Date()).toISOString()}) ${fact.trim()}
`;
    if (!(0, import_node_fs2.existsSync)(memoryMd)) {
      (0, import_node_fs2.writeFileSync)(memoryMd, `# Memory

${line}`, "utf-8");
    } else {
      (0, import_node_fs2.appendFileSync)(memoryMd, line, "utf-8");
    }
  }
  /** Append an entry to memory/YYYY-MM-DD.md */
  addDaily(entry, date) {
    this.ensureWorkspace();
    const d = date ?? (/* @__PURE__ */ new Date()).toISOString().slice(0, 10);
    const path = (0, import_node_path2.join)(this.workspaceDir, "memory", `${d}.md`);
    const line = `- (${(/* @__PURE__ */ new Date()).toISOString()}) ${entry.trim()}
`;
    if (!(0, import_node_fs2.existsSync)(path)) {
      (0, import_node_fs2.writeFileSync)(path, `# Daily log ${d}

${line}`, "utf-8");
    } else {
      (0, import_node_fs2.appendFileSync)(path, line, "utf-8");
    }
  }
  /** Hybrid search across MEMORY.md and memory/*.md */
  search(query, options) {
    const limit = options?.limit ?? 10;
    return this.getEngine().search(query, limit);
  }
  /**
   * Parse current MEMORY.md into a flat list of bullet strings, stripping any
   * `(ISO)` timestamp prefix written by {@link MemoryStore.add}. Useful for passing
   * existing memory into a consolidator or inspecting state.
   */
  readMemoryBullets() {
    const memoryMd = (0, import_node_path2.join)(this.workspaceDir, "MEMORY.md");
    if (!(0, import_node_fs2.existsSync)(memoryMd)) return [];
    const raw = (0, import_node_fs2.readFileSync)(memoryMd, "utf-8");
    const bullets = [];
    for (const line of raw.split("\n")) {
      const m = BULLET_LINE.exec(line);
      if (m && m[1]) bullets.push(m[1]);
    }
    return bullets;
  }
  /**
   * Store a transcript. Three modes:
   * - default / `{ summarize: false }`: append the raw transcript as a single bullet.
   * - `{ summarize: true }`: route through {@link Summarizer.summarize} and append each `- ` bullet.
   * - `{ consolidate: true }`: read existing bullets, merge with transcript via
   *   {@link Summarizer.consolidate} (or a fallback that concatenates into `summarize`),
   *   then atomically rewrite MEMORY.md and reindex.
   */
  async ingestTranscript(text, options) {
    const body = text.trim();
    if (!body) return { bullets: 0, previous: this.readMemoryBullets().length, unchanged: true };
    if (options?.consolidate) {
      return this.consolidateTranscript(body);
    }
    const summarize = options?.summarize ?? false;
    let content = body;
    if (summarize) {
      content = (await this.summarizer.summarize(content)).trim();
    }
    const previous = this.readMemoryBullets().length;
    const bullets = parseBullets(content);
    if (bullets.length === 0) {
      this.add(content);
      return { bullets: 1, previous, unchanged: false };
    }
    for (const b of bullets) this.add(b);
    return { bullets: bullets.length, previous, unchanged: false };
  }
  async consolidateTranscript(transcript) {
    const existing = this.readMemoryBullets();
    const previous = existing.length;
    let raw;
    if (this.summarizer.consolidate) {
      raw = await this.summarizer.consolidate({ existingBullets: existing, transcript });
    } else {
      const fallbackPrompt = [
        "# Existing memory",
        existing.length ? existing.map((b) => `- ${b}`).join("\n") : "(none)",
        "",
        "# New conversation transcript",
        transcript,
        "",
        "# Task",
        "Produce the updated complete memory bullet list."
      ].join("\n");
      raw = await this.summarizer.summarize(fallbackPrompt);
    }
    const bullets = parseBullets(raw);
    if (bullets.length === 0) {
      return { bullets: 0, previous, unchanged: true };
    }
    this.rewriteMemoryMd(bullets);
    this.getEngine().invalidateMeta("MEMORY.md");
    this.getEngine().reindex();
    return { bullets: bullets.length, previous, unchanged: false };
  }
  rewriteMemoryMd(bullets) {
    this.ensureWorkspace();
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const lines = bullets.map((b) => `- (${now}) ${b.trim()}`);
    const content = `# Memory

${lines.join("\n")}
`;
    const file = (0, import_node_path2.join)(this.workspaceDir, "MEMORY.md");
    const tmp = `${file}.tmp`;
    (0, import_node_fs2.writeFileSync)(tmp, content, "utf-8");
    (0, import_node_fs2.renameSync)(tmp, file);
  }
  /** Flush index (mainly for tests); normally runs automatically before search */
  reindex() {
    this.getEngine().reindex();
  }
  close() {
    this.engine?.close();
    this.engine = null;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MemorySearchEngine,
  MemoryStore,
  createLlmSummarizer,
  createSqliteDriver,
  detectDriver,
  identitySummarizer,
  sanitizeUserId
});
