// @shogo-ai/core/dist/chunk-KTDDFWXM.js
// src/tech-stack-registry.ts
var TECH_STACK_REGISTRY = {
  // Web (Vite-based) — share a bundled Vite template, so apps/api seeds them.
  "react-app": { id: "react-app", target: "web" },
  "threejs-game": { id: "threejs-game", target: "web" },
  "phaser-game": { id: "phaser-game", target: "web" },
  // Mobile (Metro/Expo) — agent-runtime owns the seed.
  "expo-app": { id: "expo-app", target: "mobile", metro: true, seedsOwnTemplate: true },
  "expo-three": { id: "expo-three", target: "mobile", metro: true, seedsOwnTemplate: true },
  "react-native": { id: "react-native", target: "mobile", metro: true, seedsOwnTemplate: true },
  // Data / scripting
  "python-data": { id: "python-data", target: "data", seedsOwnTemplate: true },
  // Native (full game engines)
  "unity-game": { id: "unity-game", target: "native", seedsOwnTemplate: true },
  // Bare / no-stack
  none: { id: "none", target: "none", seedsOwnTemplate: true }
};
function getStackEntry(techStackId) {
  if (!techStackId) return null;
  return TECH_STACK_REGISTRY[techStackId] ?? null;
}
function isMobileTechStack(techStackId) {
  return getStackEntry(techStackId)?.target === "mobile";
}
function usesMetroBundler(techStackId) {
  return getStackEntry(techStackId)?.metro === true;
}
function stackSeedsItself(techStackId) {
  return getStackEntry(techStackId)?.seedsOwnTemplate === true;
}

export {
  TECH_STACK_REGISTRY,
  getStackEntry,
  isMobileTechStack,
  usesMetroBundler,
  stackSeedsItself
};
