// @shogo-ai/core/dist/chunk-24AL3V6R.js
// src/logger.ts
var LOG_LEVELS = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3
};
var isProduction = process.env.NODE_ENV === "production" || process.env.NODE_ENV === "staging";
var minLevel = LOG_LEVELS[process.env.LOG_LEVEL || "info"] ?? 1;
function shouldLog(level) {
  return LOG_LEVELS[level] >= minLevel;
}
function formatEntry(entry) {
  if (isProduction) {
    return JSON.stringify(entry);
  }
  const { level: _level, msg, service, timestamp: _timestamp, ...extra } = entry;
  const extraStr = Object.keys(extra).length > 0 ? " " + JSON.stringify(extra) : "";
  return `[${service}] ${msg}${extraStr}`;
}
function createLogger(service, defaultExtra) {
  function log(level, msg, extra) {
    if (!shouldLog(level)) return;
    const entry = {
      level,
      msg,
      service,
      timestamp: (/* @__PURE__ */ new Date()).toISOString(),
      ...defaultExtra,
      ...extra
    };
    const formatted = formatEntry(entry);
    switch (level) {
      case "error":
        console.error(formatted);
        break;
      case "warn":
        console.warn(formatted);
        break;
      default:
        console.log(formatted);
        break;
    }
  }
  return {
    debug: (msg, extra) => log("debug", msg, extra),
    info: (msg, extra) => log("info", msg, extra),
    warn: (msg, extra) => log("warn", msg, extra),
    error: (msg, extra) => log("error", msg, extra),
    child: (extra) => createLogger(service, { ...defaultExtra, ...extra })
  };
}

export {
  createLogger
};
