// @shogo-ai/core/dist/chunk-IDGVRGAL.js
// src/instrumentation.ts
import { diag, DiagConsoleLogger, DiagLogLevel, trace, SpanStatusCode } from "@opentelemetry/api";
import { NodeSDK } from "@opentelemetry/sdk-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { BatchSpanProcessor, SimpleSpanProcessor } from "@opentelemetry/sdk-trace-base";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from "@opentelemetry/semantic-conventions";
var sdk = null;
function initInstrumentation(config) {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  if (!endpoint) {
    console.log(`[OTEL] Tracing disabled for ${config.serviceName} (OTEL_EXPORTER_OTLP_ENDPOINT not set)`);
    return;
  }
  const diagLevel = process.env.OTEL_LOG_LEVEL === "debug" ? DiagLogLevel.DEBUG : process.env.OTEL_LOG_LEVEL === "info" ? DiagLogLevel.INFO : DiagLogLevel.WARN;
  diag.setLogger(new DiagConsoleLogger(), diagLevel);
  const headers = {};
  if (process.env.SIGNOZ_INGESTION_KEY) {
    headers["signoz-ingestion-key"] = process.env.SIGNOZ_INGESTION_KEY;
  }
  const serviceName = process.env.OTEL_SERVICE_NAME || config.serviceName;
  console.log(`[OTEL] Configuring ${serviceName} \u2192 ${endpoint}/v1/traces`);
  let traceExporter;
  try {
    traceExporter = new OTLPTraceExporter({
      url: `${endpoint}/v1/traces`,
      headers
    });
  } catch (err) {
    console.error(`[OTEL] Failed to create exporter: ${err.message}`);
    return;
  }
  const isProduction = process.env.NODE_ENV === "production" || process.env.NODE_ENV === "staging";
  const spanProcessor = isProduction ? new BatchSpanProcessor(traceExporter) : new SimpleSpanProcessor(traceExporter);
  sdk = new NodeSDK({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: serviceName,
      [ATTR_SERVICE_VERSION]: config.serviceVersion || "0.1.0",
      "deployment.environment.name": process.env.NODE_ENV || "development"
    }),
    spanProcessors: [spanProcessor],
    instrumentations: []
  });
  try {
    sdk.start();
    console.log(`[OTEL] Tracing enabled for ${serviceName}`);
  } catch (err) {
    console.error(`[OTEL] SDK failed to start: ${err.message}`);
    sdk = null;
    return;
  }
  try {
    const tracer = trace.getTracer(serviceName);
    const testSpan = tracer.startSpan("otel.startup-test", {
      attributes: {
        "test.type": "startup-verification",
        "service.name": serviceName,
        // `process.versions.bun` is present under Bun (always undefined
        // under plain Node) — avoids referencing the `Bun` global so
        // this file typechecks without `@types/bun` in tsconfig.
        "runtime": process.versions.bun ? "bun" : "node"
      }
    });
    testSpan.setStatus({ code: SpanStatusCode.OK });
    testSpan.end();
    console.log(`[OTEL] Startup test span emitted for ${serviceName}`);
  } catch (err) {
    console.error(`[OTEL] Failed to emit test span: ${err.message}`);
  }
}
async function shutdownInstrumentation() {
  if (sdk) {
    try {
      await sdk.shutdown();
      console.log("[OTEL] Tracing shut down");
    } catch (err) {
      console.error("[OTEL] Error shutting down:", err.message);
    }
  }
}
function traceOperation(tracerName, spanName, attributes, fn) {
  const tracer = trace.getTracer(tracerName);
  return tracer.startActiveSpan(spanName, { attributes }, async (span) => {
    try {
      const result = await fn(span);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (err) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
      span.recordException(err);
      throw err;
    } finally {
      span.end();
    }
  });
}

export {
  initInstrumentation,
  shutdownInstrumentation,
  traceOperation
};
