// @shogo-ai/core/dist/instrumentation.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/instrumentation.ts
var instrumentation_exports = {};
__export(instrumentation_exports, {
  initInstrumentation: () => initInstrumentation,
  shutdownInstrumentation: () => shutdownInstrumentation,
  traceOperation: () => traceOperation
});
module.exports = __toCommonJS(instrumentation_exports);
var import_api = require("@opentelemetry/api");
var import_sdk_node = require("@opentelemetry/sdk-node");
var import_exporter_trace_otlp_http = require("@opentelemetry/exporter-trace-otlp-http");
var import_sdk_trace_base = require("@opentelemetry/sdk-trace-base");
var import_resources = require("@opentelemetry/resources");
var import_semantic_conventions = require("@opentelemetry/semantic-conventions");
var sdk = null;
function initInstrumentation(config) {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  if (!endpoint) {
    console.log(`[OTEL] Tracing disabled for ${config.serviceName} (OTEL_EXPORTER_OTLP_ENDPOINT not set)`);
    return;
  }
  const diagLevel = process.env.OTEL_LOG_LEVEL === "debug" ? import_api.DiagLogLevel.DEBUG : process.env.OTEL_LOG_LEVEL === "info" ? import_api.DiagLogLevel.INFO : import_api.DiagLogLevel.WARN;
  import_api.diag.setLogger(new import_api.DiagConsoleLogger(), diagLevel);
  const headers = {};
  if (process.env.SIGNOZ_INGESTION_KEY) {
    headers["signoz-ingestion-key"] = process.env.SIGNOZ_INGESTION_KEY;
  }
  const serviceName = process.env.OTEL_SERVICE_NAME || config.serviceName;
  console.log(`[OTEL] Configuring ${serviceName} \u2192 ${endpoint}/v1/traces`);
  let traceExporter;
  try {
    traceExporter = new import_exporter_trace_otlp_http.OTLPTraceExporter({
      url: `${endpoint}/v1/traces`,
      headers
    });
  } catch (err) {
    console.error(`[OTEL] Failed to create exporter: ${err.message}`);
    return;
  }
  const isProduction = process.env.NODE_ENV === "production" || process.env.NODE_ENV === "staging";
  const spanProcessor = isProduction ? new import_sdk_trace_base.BatchSpanProcessor(traceExporter) : new import_sdk_trace_base.SimpleSpanProcessor(traceExporter);
  sdk = new import_sdk_node.NodeSDK({
    resource: (0, import_resources.resourceFromAttributes)({
      [import_semantic_conventions.ATTR_SERVICE_NAME]: serviceName,
      [import_semantic_conventions.ATTR_SERVICE_VERSION]: config.serviceVersion || "0.1.0",
      "deployment.environment.name": process.env.NODE_ENV || "development"
    }),
    spanProcessors: [spanProcessor],
    instrumentations: []
  });
  try {
    sdk.start();
    console.log(`[OTEL] Tracing enabled for ${serviceName}`);
  } catch (err) {
    console.error(`[OTEL] SDK failed to start: ${err.message}`);
    sdk = null;
    return;
  }
  try {
    const tracer = import_api.trace.getTracer(serviceName);
    const testSpan = tracer.startSpan("otel.startup-test", {
      attributes: {
        "test.type": "startup-verification",
        "service.name": serviceName,
        // `process.versions.bun` is present under Bun (always undefined
        // under plain Node) — avoids referencing the `Bun` global so
        // this file typechecks without `@types/bun` in tsconfig.
        "runtime": process.versions.bun ? "bun" : "node"
      }
    });
    testSpan.setStatus({ code: import_api.SpanStatusCode.OK });
    testSpan.end();
    console.log(`[OTEL] Startup test span emitted for ${serviceName}`);
  } catch (err) {
    console.error(`[OTEL] Failed to emit test span: ${err.message}`);
  }
}
async function shutdownInstrumentation() {
  if (sdk) {
    try {
      await sdk.shutdown();
      console.log("[OTEL] Tracing shut down");
    } catch (err) {
      console.error("[OTEL] Error shutting down:", err.message);
    }
  }
}
function traceOperation(tracerName, spanName, attributes, fn) {
  const tracer = import_api.trace.getTracer(tracerName);
  return tracer.startActiveSpan(spanName, { attributes }, async (span) => {
    try {
      const result = await fn(span);
      span.setStatus({ code: import_api.SpanStatusCode.OK });
      return result;
    } catch (err) {
      span.setStatus({ code: import_api.SpanStatusCode.ERROR, message: err.message });
      span.recordException(err);
      throw err;
    } finally {
      span.end();
    }
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  initInstrumentation,
  shutdownInstrumentation,
  traceOperation
});
