// @shogo-ai/core/dist/logger.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/logger.ts
var logger_exports = {};
__export(logger_exports, {
  createLogger: () => createLogger
});
module.exports = __toCommonJS(logger_exports);
var LOG_LEVELS = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3
};
var isProduction = process.env.NODE_ENV === "production" || process.env.NODE_ENV === "staging";
var minLevel = LOG_LEVELS[process.env.LOG_LEVEL || "info"] ?? 1;
function shouldLog(level) {
  return LOG_LEVELS[level] >= minLevel;
}
function formatEntry(entry) {
  if (isProduction) {
    return JSON.stringify(entry);
  }
  const { level: _level, msg, service, timestamp: _timestamp, ...extra } = entry;
  const extraStr = Object.keys(extra).length > 0 ? " " + JSON.stringify(extra) : "";
  return `[${service}] ${msg}${extraStr}`;
}
function createLogger(service, defaultExtra) {
  function log(level, msg, extra) {
    if (!shouldLog(level)) return;
    const entry = {
      level,
      msg,
      service,
      timestamp: (/* @__PURE__ */ new Date()).toISOString(),
      ...defaultExtra,
      ...extra
    };
    const formatted = formatEntry(entry);
    switch (level) {
      case "error":
        console.error(formatted);
        break;
      case "warn":
        console.warn(formatted);
        break;
      default:
        console.log(formatted);
        break;
    }
  }
  return {
    debug: (msg, extra) => log("debug", msg, extra),
    info: (msg, extra) => log("info", msg, extra),
    warn: (msg, extra) => log("warn", msg, extra),
    error: (msg, extra) => log("error", msg, extra),
    child: (extra) => createLogger(service, { ...defaultExtra, ...extra })
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createLogger
});
