// @shogo-ai/core/dist/stream-buffer.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/stream-buffer.ts
var stream_buffer_exports = {};
__export(stream_buffer_exports, {
  StreamBufferStore: () => StreamBufferStore,
  createBufferingTransform: () => createBufferingTransform
});
module.exports = __toCommonJS(stream_buffer_exports);
var CLEANUP_INTERVAL_MS = 6e4;
var MAX_BUFFER_AGE_MS = 30 * 6e4;
var COMPLETED_GRACE_MS = 3e4;
var StreamBufferStore = class {
  constructor() {
    this.buffers = /* @__PURE__ */ new Map();
    this.cleanupTimer = null;
    this.cleanupTimer = setInterval(() => this.cleanup(), CLEANUP_INTERVAL_MS);
    const handle = this.cleanupTimer;
    if (handle && typeof handle.unref === "function") {
      handle.unref();
    }
  }
  /**
   * Create (or replace) a buffer for the given key. Any existing active
   * buffer is completed and discarded. Returns a writer bound to this
   * specific buffer instance, including the new `turnId`.
   */
  create(key, opts) {
    const existing = this.buffers.get(key);
    if (existing && existing.status === "active") {
      this.completeBuffer(existing, "replaced");
    }
    const now = Date.now();
    const turnId = opts?.turnId || generateTurnId();
    const buf = {
      turnId,
      frames: [],
      subscribers: /* @__PURE__ */ new Set(),
      status: "active",
      terminal: null,
      createdAt: now,
      completedAt: null,
      lastEventAt: now,
      nextSeq: 1
    };
    this.buffers.set(key, buf);
    return {
      get turnId() {
        return buf.turnId;
      },
      get lastSeq() {
        return buf.nextSeq - 1;
      },
      append: (chunk) => {
        if (buf.status !== "active") return -1;
        const seq = buf.nextSeq++;
        buf.frames.push({ seq, chunk });
        buf.lastEventAt = Date.now();
        for (const ctrl of buf.subscribers) {
          try {
            ctrl.enqueue(chunk);
          } catch {
            buf.subscribers.delete(ctrl);
          }
        }
        return seq;
      },
      complete: (reason) => {
        this.completeBuffer(buf, reason);
      },
      fail: (error) => {
        this.failBuffer(buf, error);
      }
    };
  }
  /**
   * Append a chunk by key (loose, untyped writer). Returns the seq assigned,
   * or -1 if the buffer is unknown/inactive.
   */
  append(key, chunk) {
    const buf = this.buffers.get(key);
    if (!buf || buf.status !== "active") return -1;
    const seq = buf.nextSeq++;
    buf.frames.push({ seq, chunk });
    buf.lastEventAt = Date.now();
    for (const ctrl of buf.subscribers) {
      try {
        ctrl.enqueue(chunk);
      } catch {
        buf.subscribers.delete(ctrl);
      }
    }
    return seq;
  }
  /** Mark a buffer as completed by key. Optional terminal reason. */
  complete(key, reason) {
    const buf = this.buffers.get(key);
    if (!buf) return;
    this.completeBuffer(buf, reason);
  }
  /**
   * Abort a stream: complete subscribers and remove the buffer entirely.
   * Future resume/replay requests for this key will return null (→ 204).
   */
  abort(key) {
    const buf = this.buffers.get(key);
    if (!buf) return;
    this.completeBuffer(buf, "aborted");
    buf.status = "aborted";
    this.buffers.delete(key);
  }
  has(key) {
    return this.buffers.has(key);
  }
  /** Read-only metadata about a buffered turn. Useful for resume responses. */
  snapshot(key) {
    const buf = this.buffers.get(key);
    if (!buf) return null;
    return {
      turnId: buf.turnId,
      status: buf.status,
      lastSeq: buf.nextSeq - 1,
      terminal: buf.terminal,
      createdAt: buf.createdAt,
      completedAt: buf.completedAt,
      lastEventAt: buf.lastEventAt
    };
  }
  /**
   * Create a ReadableStream that first replays all buffered chunks past
   * `opts.fromSeq` (default: 0 → replay everything), then subscribes for
   * any further live chunks until the stream completes.
   *
   * Returns null if no buffer exists for the key.
   */
  createReplayStream(key, opts = {}) {
    const buf = this.buffers.get(key);
    if (!buf) return null;
    const fromSeq = Math.max(0, opts.fromSeq ?? 0);
    let subscribedController = null;
    return new ReadableStream({
      start(controller) {
        for (const frame of buf.frames) {
          if (frame.seq <= fromSeq) continue;
          try {
            controller.enqueue(frame.chunk);
          } catch {
            return;
          }
        }
        if (buf.status !== "active") {
          try {
            controller.close();
          } catch {
          }
          return;
        }
        subscribedController = controller;
        buf.subscribers.add(controller);
      },
      cancel() {
        if (subscribedController) {
          buf.subscribers.delete(subscribedController);
          subscribedController = null;
        }
      }
    });
  }
  cleanup() {
    const now = Date.now();
    for (const [key, buf] of this.buffers) {
      if (buf.status !== "active" && buf.completedAt && now - buf.completedAt > COMPLETED_GRACE_MS) {
        this.buffers.delete(key);
      } else if (now - buf.createdAt > MAX_BUFFER_AGE_MS && buf.subscribers.size === 0) {
        this.completeBuffer(buf, "expired");
        this.buffers.delete(key);
      }
    }
  }
  dispose() {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
    for (const buf of this.buffers.values()) {
      this.completeBuffer(buf, "disposed");
    }
    this.buffers.clear();
  }
  completeBuffer(buf, reason) {
    if (buf.status !== "active") return;
    buf.status = "completed";
    buf.terminal = { reason };
    buf.completedAt = Date.now();
    buf.lastEventAt = buf.completedAt;
    for (const ctrl of buf.subscribers) {
      try {
        ctrl.close();
      } catch {
      }
    }
    buf.subscribers.clear();
  }
  failBuffer(buf, error) {
    if (buf.status !== "active") return;
    buf.status = "failed";
    buf.terminal = { error };
    buf.completedAt = Date.now();
    buf.lastEventAt = buf.completedAt;
    for (const ctrl of buf.subscribers) {
      try {
        ctrl.close();
      } catch {
      }
    }
    buf.subscribers.clear();
  }
};
function generateTurnId() {
  try {
    if (typeof globalThis.crypto?.randomUUID === "function") {
      return globalThis.crypto.randomUUID();
    }
  } catch {
  }
  return `turn_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 10)}`;
}
function createBufferingTransform(store, key) {
  return new TransformStream({
    transform(chunk, controller) {
      store.append(key, chunk);
      controller.enqueue(chunk);
    },
    flush() {
      store.complete(key);
    }
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StreamBufferStore,
  createBufferingTransform
});
