// @shogo-ai/core/dist/tech-stack-registry.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/tech-stack-registry.ts
var tech_stack_registry_exports = {};
__export(tech_stack_registry_exports, {
  TECH_STACK_REGISTRY: () => TECH_STACK_REGISTRY,
  getStackEntry: () => getStackEntry,
  isMobileTechStack: () => isMobileTechStack,
  stackSeedsItself: () => stackSeedsItself,
  usesMetroBundler: () => usesMetroBundler
});
module.exports = __toCommonJS(tech_stack_registry_exports);
var TECH_STACK_REGISTRY = {
  // Web (Vite-based) — share a bundled Vite template, so apps/api seeds them.
  "react-app": { id: "react-app", target: "web" },
  "threejs-game": { id: "threejs-game", target: "web" },
  "phaser-game": { id: "phaser-game", target: "web" },
  // Mobile (Metro/Expo) — agent-runtime owns the seed.
  "expo-app": { id: "expo-app", target: "mobile", metro: true, seedsOwnTemplate: true },
  "expo-three": { id: "expo-three", target: "mobile", metro: true, seedsOwnTemplate: true },
  "react-native": { id: "react-native", target: "mobile", metro: true, seedsOwnTemplate: true },
  // Data / scripting
  "python-data": { id: "python-data", target: "data", seedsOwnTemplate: true },
  // Native (full game engines)
  "unity-game": { id: "unity-game", target: "native", seedsOwnTemplate: true },
  // Bare / no-stack
  none: { id: "none", target: "none", seedsOwnTemplate: true }
};
function getStackEntry(techStackId) {
  if (!techStackId) return null;
  return TECH_STACK_REGISTRY[techStackId] ?? null;
}
function isMobileTechStack(techStackId) {
  return getStackEntry(techStackId)?.target === "mobile";
}
function usesMetroBundler(techStackId) {
  return getStackEntry(techStackId)?.metro === true;
}
function stackSeedsItself(techStackId) {
  return getStackEntry(techStackId)?.seedsOwnTemplate === true;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TECH_STACK_REGISTRY,
  getStackEntry,
  isMobileTechStack,
  stackSeedsItself,
  usesMetroBundler
});
