// @shogo-ai/voice/dist/chunk-Q3TJAJ54.js
// src/audioTags.ts
var AUDIO_TAGS = [
  // Emotions
  { tag: "happy", group: "emotion", label: "Happy", description: "Warm, upbeat delivery.", example: "[happy] Oh, I\u2019m so glad you stopped by." },
  { tag: "sad", group: "emotion", label: "Sad", description: "Subdued, heavy-hearted tone.", example: "[sad] I wish things had gone differently." },
  { tag: "excited", group: "emotion", label: "Excited", description: "Energized, eager delivery.", example: "[excited] Wait until you hear this!" },
  { tag: "curious", group: "emotion", label: "Curious", description: "Inquisitive, rising inflection.", example: "[curious] Hmm, what did they look like?" },
  { tag: "sarcastic", group: "emotion", label: "Sarcastic", description: "Dry, ironic tone.", example: "[sarcastic] Oh, fantastic. Another Monday." },
  { tag: "angry", group: "emotion", label: "Angry", description: "Sharp, irritated edge.", example: "[angry] I told you not to touch that." },
  { tag: "nervous", group: "emotion", label: "Nervous", description: "Uneasy, faltering delivery.", example: "[nervous] I\u2014 I didn\u2019t think you\u2019d actually come." },
  { tag: "confident", group: "emotion", label: "Confident", description: "Grounded, assured delivery.", example: "[confident] Trust me. I\u2019ve done this before." },
  // Delivery
  { tag: "whispers", group: "delivery", label: "Whispers", description: "Hushed, close-mic delivery.", example: "[whispers] Don\u2019t say it too loud." },
  { tag: "shouts", group: "delivery", label: "Shouts", description: "Loud, projected delivery.", example: "[shouts] Look out behind you!" },
  { tag: "softly", group: "delivery", label: "Softly", description: "Gentle, quiet tone.", example: "[softly] It\u2019s okay. I\u2019m here." },
  { tag: "slowly", group: "delivery", label: "Slowly", description: "Deliberate pacing.", example: "[slowly] Tell me\u2026 exactly\u2026 what happened." },
  { tag: "quickly", group: "delivery", label: "Quickly", description: "Rapid, urgent pacing.", example: "[quickly] We need to go, right now." },
  // Reactions (non-verbal sounds)
  { tag: "laughs", group: "reaction", label: "Laughs", description: "A genuine laugh.", example: "That\u2019s ridiculous [laughs]." },
  { tag: "chuckles", group: "reaction", label: "Chuckles", description: "A small, warm chuckle.", example: "[chuckles] You always say that." },
  { tag: "sighs", group: "reaction", label: "Sighs", description: "An audible sigh.", example: "[sighs] Alright, fine. Let\u2019s do it." },
  { tag: "gasps", group: "reaction", label: "Gasps", description: "A sharp inhale of surprise.", example: "[gasps] You didn\u2019t!" },
  { tag: "hesitates", group: "reaction", label: "Hesitates", description: "A pause with uncertainty.", example: "[hesitates] I\u2019m not sure I should say." },
  { tag: "clears throat", group: "reaction", label: "Clears throat", description: "Throat-clearing beat.", example: "[clears throat] As I was saying\u2026" }
];
var AUDIO_TAG_GROUPS = [
  { id: "emotion", label: "Emotions", description: "Overall feeling carried through the line." },
  { id: "delivery", label: "Delivery", description: "How the words are spoken (pace, volume)." },
  { id: "reaction", label: "Reactions", description: "Non-verbal sounds mixed into speech." }
];
var EXPRESSIVITY_OPTIONS = [
  { id: "off", label: "Off", description: "No audio tags. Plain, natural delivery." },
  { id: "subtle", label: "Subtle", description: "Occasional tag when it genuinely fits. Default." },
  { id: "full", label: "Full", description: "Freely expressive \u2014 tags sprinkled to carry emotion and reactions." }
];
var DEFAULT_ALLOWED_TAGS = [
  "whispers",
  "laughs",
  "sighs",
  "excited",
  "curious",
  "softly"
];
var DEFAULT_VOICE_SETTINGS = {
  stability: 0.5,
  similarity_boost: 0.8,
  style: 0,
  use_speaker_boost: true
};
var KNOWN_TAG_SET = new Set(AUDIO_TAGS.map((t) => t.tag.toLowerCase()));
function stripAudioTags(text) {
  if (!text) return "";
  return text.replace(/\[([a-z][a-z ]{0,24})\]/gi, (match, inner) => {
    const key = inner.trim().toLowerCase();
    return KNOWN_TAG_SET.has(key) ? "" : match;
  }).replace(/[ \t]{2,}/g, " ").replace(/\s+([,.!?;:])/g, "$1").trim();
}
var EXPRESSIVITY_BLOCK_OPEN = "<!--expressivity-->";
var EXPRESSIVITY_BLOCK_CLOSE = "<!--/expressivity-->";
function stripExpressivityBlock(prompt) {
  if (!prompt) return prompt;
  const open = EXPRESSIVITY_BLOCK_OPEN;
  const close = EXPRESSIVITY_BLOCK_CLOSE;
  const openIdx = prompt.indexOf(open);
  if (openIdx === -1) return prompt;
  const closeIdx = prompt.indexOf(close, openIdx);
  if (closeIdx === -1) return prompt;
  const before = prompt.slice(0, openIdx).replace(/\s+$/, "");
  const after = prompt.slice(closeIdx + close.length).replace(/^\s+/, "");
  return [before, after].filter(Boolean).join("\n\n");
}
function composeExpressivityBlock(expressivity, allowedTags) {
  if (expressivity === "off") return "";
  const allowed = (allowedTags && allowedTags.length > 0 ? allowedTags : DEFAULT_ALLOWED_TAGS).filter((t) => AUDIO_TAGS.some((a) => a.tag === t));
  if (allowed.length === 0) return "";
  const intensity = expressivity === "full" ? "Use audio tags liberally to color your speech with emotion, delivery, and non-verbal reactions. Most substantive replies should include at least one tag. Never add so many tags that a line becomes unnatural." : "Use audio tags sparingly \u2014 only when one genuinely fits the emotion or reaction of the moment. Most lines will have no tag. At most one or two tags per reply.";
  const tagList = allowed.map((t) => {
    const entry = AUDIO_TAGS.find((a) => a.tag === t);
    return `  - [${entry.tag}] \u2014 ${entry.description} Example: ${entry.example}`;
  }).join("\n");
  const body = [
    "You speak through an expressive TTS engine that understands inline audio tags.",
    'Audio tags are lowercase words in square brackets placed directly in your spoken text \u2014 e.g. "[whispers] don\u2019t tell anyone" or "that was wild [laughs]".',
    'Tags are performance cues, not words. Never describe a tag (do not say "asterisk laughs asterisk" or "*laughs*"). Never read a tag aloud. Never put tags inside quoted speech from another character.',
    intensity,
    "You may only use tags from this allow-list:",
    tagList,
    "If no tag fits, use none. Tags are optional flavor, not filler."
  ].join("\n");
  return `${EXPRESSIVITY_BLOCK_OPEN}
${body}
${EXPRESSIVITY_BLOCK_CLOSE}`;
}
function buildPreviewLine(allowedTags, characterName = "your companion") {
  const tags = (allowedTags ?? []).filter((t) => AUDIO_TAGS.some((a) => a.tag === t));
  if (tags.length === 0) {
    return `Hi there. This is a quick preview of how ${characterName} sounds.`;
  }
  const pick = tags.slice(0, 3);
  const lead = pick[0];
  const mid = pick[1];
  const tail = pick[2];
  if (pick.length === 1) {
    return `[${lead}] Hey \u2014 this is how ${characterName} sounds when the mood takes over.`;
  }
  if (pick.length === 2) {
    return `[${lead}] Oh, this is fun. [${mid}] Let me know what you think.`;
  }
  return `[${lead}] Okay, listen. [${mid}] This is the new voice. [${tail}] How does it feel?`;
}
function normalizeAudioTags(raw) {
  if (!Array.isArray(raw)) return null;
  const cleaned = raw.filter((x) => typeof x === "string").map((x) => x.trim().toLowerCase()).filter((x) => AUDIO_TAGS.some((a) => a.tag === x));
  return Array.from(new Set(cleaned));
}
function normalizeExpressivity(raw) {
  if (raw === "off" || raw === "subtle" || raw === "full") return raw;
  return void 0;
}
function normalizeVoiceSettings(raw) {
  if (!raw || typeof raw !== "object") return void 0;
  const src = raw;
  const out = {};
  const clamp = (v) => typeof v === "number" && Number.isFinite(v) ? Math.min(1, Math.max(0, v)) : void 0;
  const stability = clamp(src.stability);
  const similarity = clamp(src.similarity_boost ?? src.similarityBoost);
  const style = clamp(src.style);
  const boost = src.use_speaker_boost ?? src.useSpeakerBoost;
  if (stability !== void 0) out.stability = stability;
  if (similarity !== void 0) out.similarity_boost = similarity;
  if (style !== void 0) out.style = style;
  if (typeof boost === "boolean") out.use_speaker_boost = boost;
  return Object.keys(out).length ? out : void 0;
}
function readAudioTags(raw) {
  const norm = normalizeAudioTags(raw);
  return norm && norm.length > 0 ? norm : DEFAULT_ALLOWED_TAGS;
}
function readExpressivity(raw) {
  return normalizeExpressivity(raw) ?? "subtle";
}
function readVoiceSettings(raw) {
  return normalizeVoiceSettings(raw) ?? { ...DEFAULT_VOICE_SETTINGS };
}

export {
  AUDIO_TAGS,
  AUDIO_TAG_GROUPS,
  EXPRESSIVITY_OPTIONS,
  DEFAULT_ALLOWED_TAGS,
  DEFAULT_VOICE_SETTINGS,
  stripAudioTags,
  EXPRESSIVITY_BLOCK_OPEN,
  EXPRESSIVITY_BLOCK_CLOSE,
  stripExpressivityBlock,
  composeExpressivityBlock,
  buildPreviewLine,
  normalizeAudioTags,
  normalizeExpressivity,
  normalizeVoiceSettings,
  readAudioTags,
  readExpressivity,
  readVoiceSettings
};
