// @shogo-ai/voice/dist/chunk-B4JC3DK2.js
import {
  composeExpressivityBlock,
  stripExpressivityBlock
} from "./chunk-Q3TJAJ54.js";

// src/prompt.ts
var DEFAULT_MEMORY_BLOCK = `# Memory

Known context about this user from past conversations:
{{user_context}}

Memory retrieval is automatic \u2014 the top matching facts are already injected above. Never say "let me check my memory" or "searching memories."

You have one tool:
- add_memory(fact): save a concise, canonical fact about the user \u2014 preferences, decisions, personal details, follow-ups. Use short normalized phrasings like "favorite color: green" or "lives in Honolulu". Always call it when the user explicitly asks you to remember something.`;
function stripMemoryBlock(prompt) {
  if (!prompt) return prompt;
  return prompt.replace(/\n*<!-- zix-memory-v\d+ -->[\s\S]*$/m, "").replace(/\n*# Memory\s*\n\s*Known context about this user[\s\S]*$/m, "").trimEnd();
}
function extractBasePrompt(prompt) {
  return stripExpressivityBlock(stripMemoryBlock(prompt));
}
function composeAgentPrompt(basePrompt, options = {}) {
  const { expressivity = "subtle", audioTags = null, memoryBlock = DEFAULT_MEMORY_BLOCK } = options;
  const stripped = extractBasePrompt(basePrompt);
  const expr = composeExpressivityBlock(expressivity, audioTags);
  const parts = [stripped];
  if (expr) parts.push(expr);
  if (memoryBlock) parts.push(memoryBlock);
  return parts.join("\n\n");
}

// src/elevenlabs.ts
var DEFAULT_ELEVENLABS_BASE_URL = "https://api.elevenlabs.io";
var CONVAI_SUPPORTED_TTS_MODELS = /* @__PURE__ */ new Set([
  "eleven_turbo_v2",
  "eleven_turbo_v2_5",
  "eleven_flash_v2",
  "eleven_flash_v2_5",
  "eleven_multilingual_v2"
]);
var CONVAI_TTS_MODEL_FALLBACK = "eleven_turbo_v2_5";
function resolveConvaiTtsModel(requested) {
  if (requested && CONVAI_SUPPORTED_TTS_MODELS.has(requested)) return requested;
  return CONVAI_TTS_MODEL_FALLBACK;
}
var DEFAULT_MUSIC_MODEL = "music_v1";
var ElevenLabsApiError = class extends Error {
  constructor(message, status, body) {
    super(message);
    this.status = status;
    this.body = body;
    this.name = "ElevenLabsApiError";
  }
};
var ElevenLabsClient = class {
  constructor(config) {
    this.config = config;
    if (!config.apiKey) throw new Error("ElevenLabsClient: apiKey is required");
    this.baseUrl = (config.baseUrl ?? DEFAULT_ELEVENLABS_BASE_URL).replace(/\/+$/, "");
    this.fetchImpl = config.fetch ?? globalThis.fetch;
    if (typeof this.fetchImpl !== "function") {
      throw new Error("ElevenLabsClient: global fetch is unavailable; pass config.fetch");
    }
  }
  headers(extra = {}) {
    return {
      "xi-api-key": this.config.apiKey,
      ...extra
    };
  }
  /** Create a new convai agent. Returns the new `agent_id`. */
  async createAgent(params) {
    const prompt = composeAgentPrompt(params.systemPrompt, {
      expressivity: params.expressivity ?? "subtle",
      audioTags: params.audioTags ?? null,
      memoryBlock: params.memoryBlock
    });
    const ttsConfig = {
      voice_id: params.voiceId,
      model_id: resolveConvaiTtsModel(params.ttsModelId)
    };
    if (params.voiceSettings) ttsConfig.voice_settings = params.voiceSettings;
    const body = {
      name: `Companion-${params.characterName}-for-${params.displayName}`,
      conversation_config: {
        agent: {
          prompt: {
            prompt,
            ...params.tools ? { tools: params.tools } : {}
          },
          first_message: params.firstMessage,
          language: params.language ?? "en"
        },
        tts: ttsConfig
      },
      platform_settings: {
        overrides: {
          conversation_config_override: {
            agent: { prompt: { prompt: true }, first_message: true, language: true },
            tts: { voice_id: true },
            conversation: { text_only: true }
          }
        },
        dynamic_variables: {
          dynamic_variable_placeholders: {
            user_context: "No prior memories yet.",
            character_name: "",
            ...params.dynamicVariablePlaceholders ?? {}
          }
        }
      }
    };
    const res = await this.fetchImpl(`${this.baseUrl}/v1/convai/agents/create`, {
      method: "POST",
      headers: this.headers({ "content-type": "application/json" }),
      body: JSON.stringify(body)
    });
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(`createAgent failed: ${res.status}`, res.status, text);
    }
    const data = await res.json();
    return data.agent_id;
  }
  /** PATCH any subset of agent settings. Idempotent; only sends the keys you supply. */
  async patchAgent(agentId, params) {
    const body = {};
    if (params.displayName !== void 0 || params.characterName !== void 0) {
      body.name = `Companion-${params.characterName ?? ""}-for-${params.displayName ?? ""}`.replace(
        /-for-$/,
        ""
      );
    }
    const conv = {};
    const agent = {};
    const prompt = {};
    if (params.systemPrompt !== void 0) {
      prompt.prompt = composeAgentPrompt(params.systemPrompt, {
        expressivity: params.expressivity ?? "subtle",
        audioTags: params.audioTags ?? null,
        memoryBlock: params.memoryBlock
      });
    }
    if (params.tools) prompt.tools = params.tools;
    if (Object.keys(prompt).length) agent.prompt = prompt;
    if (params.firstMessage !== void 0) agent.first_message = params.firstMessage;
    if (Object.keys(agent).length) conv.agent = agent;
    const tts = {};
    if (params.voiceId !== void 0) tts.voice_id = params.voiceId;
    if (params.voiceSettings) tts.voice_settings = params.voiceSettings;
    if (params.ttsModelId !== void 0) tts.model_id = resolveConvaiTtsModel(params.ttsModelId);
    if (Object.keys(tts).length) conv.tts = tts;
    if (Object.keys(conv).length) body.conversation_config = conv;
    if (params.enableUserContextOverride) {
      body.platform_settings = {
        overrides: {
          conversation_config_override: {
            agent: { prompt: { prompt: true }, first_message: true, language: true },
            tts: { voice_id: true },
            conversation: { text_only: true }
          },
          custom_llm_extra_body: false,
          enable_conversation_initiation_client_data_from_webhook: false
        },
        dynamic_variables: {
          dynamic_variable_placeholders: {
            user_context: "No prior memories yet.",
            character_name: ""
          }
        }
      };
    }
    if (!Object.keys(body).length) return;
    const res = await this.fetchImpl(`${this.baseUrl}/v1/convai/agents/${encodeURIComponent(agentId)}`, {
      method: "PATCH",
      headers: this.headers({ "content-type": "application/json" }),
      body: JSON.stringify(body)
    });
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(`patchAgent failed: ${res.status}`, res.status, text);
    }
  }
  /**
   * Delete a convai agent. Best-effort: swallow errors so teardown during a
   * 500 response doesn't cascade into a second failure.
   */
  async deleteAgent(agentId) {
    try {
      await this.fetchImpl(`${this.baseUrl}/v1/convai/agents/${encodeURIComponent(agentId)}`, {
        method: "DELETE",
        headers: this.headers()
      });
    } catch {
    }
  }
  /** Fetch a short-lived signed URL the browser can use to start a session. */
  async getSignedUrl(agentId) {
    const res = await this.fetchImpl(
      `${this.baseUrl}/v1/convai/conversation/get-signed-url?agent_id=${encodeURIComponent(agentId)}`,
      { headers: this.headers() }
    );
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(`getSignedUrl failed: ${res.status}`, res.status, text);
    }
    const data = await res.json();
    return data.signed_url;
  }
  /**
   * One-shot TTS. Returns the raw audio bytes plus the model id that was
   * actually used (some models aren't available for every account, so we
   * fall back to `eleven_turbo_v2_5` on the first failure).
   */
  async textToSpeech(params) {
    const modelId = params.modelId && params.modelId.length > 0 ? params.modelId : "eleven_v3";
    const accept = params.accept ?? "audio/mpeg";
    const attempt = async (m) => {
      const res2 = await this.fetchImpl(
        `${this.baseUrl}/v1/text-to-speech/${encodeURIComponent(params.voiceId)}`,
        {
          method: "POST",
          headers: this.headers({ "content-type": "application/json", accept }),
          body: JSON.stringify({
            text: params.text,
            model_id: m,
            voice_settings: params.voiceSettings
          })
        }
      );
      return res2;
    };
    let res = await attempt(modelId);
    let used = modelId;
    if (!res.ok && modelId !== "eleven_turbo_v2_5") {
      res = await attempt("eleven_turbo_v2_5");
      used = "eleven_turbo_v2_5";
    }
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(`textToSpeech failed: ${res.status}`, res.status, text);
    }
    const audio = await res.arrayBuffer();
    const contentType = res.headers.get("content-type") ?? accept;
    return { audio, modelId: used, contentType };
  }
  /**
   * Low-level escape hatch for any ElevenLabs endpoint not covered by a
   * dedicated method (music, sound-generation, speech-to-text, dubbing,
   * voice-design, …). Handles auth, query encoding, JSON body encoding,
   * and JSON-vs-binary response parsing; non-2xx responses throw
   * {@link ElevenLabsApiError} like the typed methods do.
   *
   * @example
   * ```ts
   * const { json } = await el.request({ method: 'GET', path: '/v1/models' })
   * const { audio } = await el.request({
   *   method: 'POST',
   *   path: '/v1/sound-generation',
   *   body: { text: 'door creak' },
   *   accept: 'audio/mpeg',
   * })
   * ```
   */
  async request(opts) {
    const accept = opts.accept ?? "application/json";
    let url = `${this.baseUrl}${opts.path.startsWith("/") ? "" : "/"}${opts.path}`;
    if (opts.query) {
      const params = new URLSearchParams();
      for (const [k, v] of Object.entries(opts.query)) {
        if (v !== null && v !== void 0) params.set(k, String(v));
      }
      const qs = params.toString();
      if (qs) url += `${url.includes("?") ? "&" : "?"}${qs}`;
    }
    const headers = this.headers({ accept, ...opts.headers ?? {} });
    const init = { method: opts.method, headers };
    if (opts.body !== void 0 && opts.body !== null) {
      const isPlainPayload = typeof opts.body === "object" && !(opts.body instanceof ArrayBuffer) && !(opts.body instanceof Uint8Array) && !(typeof Blob !== "undefined" && opts.body instanceof Blob) && !(typeof FormData !== "undefined" && opts.body instanceof FormData);
      if (isPlainPayload) {
        headers["content-type"] = "application/json";
        init.body = JSON.stringify(opts.body);
      } else {
        init.body = opts.body;
      }
    }
    const res = await this.fetchImpl(url, init);
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(
        `request ${opts.method} ${opts.path} failed: ${res.status}`,
        res.status,
        text
      );
    }
    const contentType = res.headers.get("content-type") ?? accept;
    if (contentType.includes("application/json")) {
      return { status: res.status, contentType, json: await res.json() };
    }
    return { status: res.status, contentType, audio: await res.arrayBuffer() };
  }
  /**
   * Generate music via `POST /v1/music`. Provide exactly one of `prompt`
   * or `compositionPlan` (ElevenLabs returns 422 otherwise). Returns the
   * raw audio bytes plus the response content type.
   *
   * @example
   * ```ts
   * const { audio } = await el.composeMusic({
   *   prompt: 'Upbeat synthwave with driving bass',
   *   musicLengthMs: 30_000,
   * })
   * ```
   */
  async composeMusic(params) {
    const hasPrompt = typeof params.prompt === "string" && params.prompt.length > 0;
    const hasPlan = params.compositionPlan != null;
    if (hasPrompt === hasPlan) {
      throw new Error(
        "composeMusic: provide exactly one of `prompt` or `compositionPlan`"
      );
    }
    const body = {
      model_id: params.modelId ?? DEFAULT_MUSIC_MODEL
    };
    if (hasPrompt) {
      body.prompt = params.prompt;
      if (params.musicLengthMs !== void 0) body.music_length_ms = params.musicLengthMs;
      if (params.forceInstrumental !== void 0) {
        body.force_instrumental = params.forceInstrumental;
      }
    } else {
      body.composition_plan = params.compositionPlan;
    }
    const { audio, contentType } = await this.request({
      method: "POST",
      path: "/v1/music",
      body,
      accept: "audio/mpeg",
      ...params.outputFormat ? { query: { output_format: params.outputFormat } } : {}
    });
    if (!audio) {
      throw new ElevenLabsApiError(
        "composeMusic: expected audio bytes but got a JSON response",
        500,
        ""
      );
    }
    return { audio, contentType };
  }
  /** Probe whether a given voice id still exists on the caller's account. */
  async voiceExists(voiceId) {
    const res = await this.fetchImpl(
      `${this.baseUrl}/v1/voices/${encodeURIComponent(voiceId)}`,
      { headers: this.headers() }
    );
    return res.ok;
  }
  /**
   * Register a Twilio phone number with an ElevenLabs agent. EL becomes
   * the SIP target for inbound calls to that number — Twilio bridges
   * the call via EL's native integration and we never have to run a
   * Media Streams ↔ convai WebSocket bridge ourselves.
   *
   * Returns the ElevenLabs `phone_number_id` (distinct from the Twilio
   * `sid`) which is the handle used by `outboundCall` + `deletePhoneNumber`.
   */
  async createPhoneNumberTwilio(params) {
    const body = {
      phone_number: params.phoneNumber,
      label: params.label ?? params.phoneNumber,
      provider: "twilio",
      sid: params.twilioAccountSid,
      token: params.twilioAuthToken
    };
    const res = await this.fetchImpl(
      `${this.baseUrl}/v1/convai/phone-numbers`,
      {
        method: "POST",
        headers: this.headers({ "content-type": "application/json" }),
        body: JSON.stringify(body)
      }
    );
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(
        `createPhoneNumberTwilio failed: ${res.status}`,
        res.status,
        text
      );
    }
    const data = await res.json();
    const id = data.phone_number_id ?? data.id;
    if (!id) {
      throw new ElevenLabsApiError(
        "createPhoneNumberTwilio returned no id",
        500,
        JSON.stringify(data)
      );
    }
    const patchRes = await this.fetchImpl(
      `${this.baseUrl}/v1/convai/phone-numbers/${encodeURIComponent(id)}`,
      {
        method: "PATCH",
        headers: this.headers({ "content-type": "application/json" }),
        body: JSON.stringify({ agent_id: params.agentId })
      }
    );
    if (!patchRes.ok) {
      const text = await patchRes.text();
      try {
        await this.deletePhoneNumber(id);
      } catch {
      }
      throw new ElevenLabsApiError(
        `assignAgentToPhoneNumber failed: ${patchRes.status}`,
        patchRes.status,
        text
      );
    }
    return { phoneNumberId: id };
  }
  /** Detach a previously-registered phone number from ElevenLabs. */
  async deletePhoneNumber(phoneNumberId) {
    const res = await this.fetchImpl(
      `${this.baseUrl}/v1/convai/phone-numbers/${encodeURIComponent(phoneNumberId)}`,
      { method: "DELETE", headers: this.headers() }
    );
    if (!res.ok && res.status !== 404) {
      const text = await res.text();
      throw new ElevenLabsApiError(
        `deletePhoneNumber failed: ${res.status}`,
        res.status,
        text
      );
    }
  }
  /**
   * Place an outbound PSTN call via a previously-registered EL phone
   * number. EL drives Twilio (or a SIP trunk) under the hood — we just
   * hand it the destination number and the agent to bridge to. Returns
   * the Twilio `callSid` + EL `conversationId` for metering / tracing.
   *
   * As of 2026-Q2 the EL API uses provider-scoped routes:
   *   - Twilio numbers → POST /v1/convai/twilio/outbound-call
   *   - SIP numbers    → POST /v1/convai/sip-trunk/outbound-call
   * and `agent_phone_number_id` is a body field (no longer a path
   * segment). The previous `/v1/convai/phone-numbers/{id}/outbound-call`
   * route now returns 404. Callers that don't know the provider can
   * look it up from `GET /v1/convai/phone-numbers/{id}` (`provider:
   * "twilio" | "sip_trunk"`); most hosted setups are Twilio, so that's
   * the default here.
   */
  async outboundCall(params) {
    const provider = params.provider ?? "twilio";
    const body = {
      agent_id: params.agentId,
      agent_phone_number_id: params.phoneNumberId,
      to_number: params.toNumber
    };
    if (params.dynamicVariables) {
      body.conversation_initiation_client_data = {
        dynamic_variables: params.dynamicVariables
      };
    }
    const res = await this.fetchImpl(
      `${this.baseUrl}/v1/convai/${provider}/outbound-call`,
      {
        method: "POST",
        headers: this.headers({ "content-type": "application/json" }),
        body: JSON.stringify(body)
      }
    );
    if (!res.ok) {
      const text = await res.text();
      throw new ElevenLabsApiError(
        `outboundCall failed: ${res.status}`,
        res.status,
        text
      );
    }
    const data = await res.json();
    const callSid = data.call_sid ?? data.callSid;
    const conversationId = data.conversation_id ?? data.conversationId;
    if (!callSid || !conversationId) {
      throw new ElevenLabsApiError(
        "outboundCall missing callSid or conversationId",
        500,
        JSON.stringify(data)
      );
    }
    return { callSid, conversationId };
  }
};
var MEMORY_CLIENT_TOOLS = [
  {
    type: "client",
    name: "add_memory",
    description: "Save a concise canonical fact about the user (preferences, decisions, personal details, follow-ups). Use whenever the user shares something worth remembering.",
    expects_response: true,
    parameters: {
      type: "object",
      properties: {
        fact: {
          type: "string",
          description: 'A short normalized fact, e.g. "prefers window seats", "favorite color: green". Max 500 chars.'
        }
      },
      required: ["fact"]
    }
  }
];

export {
  DEFAULT_MEMORY_BLOCK,
  stripMemoryBlock,
  extractBasePrompt,
  composeAgentPrompt,
  DEFAULT_ELEVENLABS_BASE_URL,
  CONVAI_SUPPORTED_TTS_MODELS,
  CONVAI_TTS_MODEL_FALLBACK,
  resolveConvaiTtsModel,
  DEFAULT_MUSIC_MODEL,
  ElevenLabsApiError,
  ElevenLabsClient,
  MEMORY_CLIENT_TOOLS
};
