// @shogo-ai/voice/dist/chunk-O362EJQR.js
import {
  CONVAI_TTS_MODEL_FALLBACK,
  ElevenLabsApiError,
  ElevenLabsClient,
  MEMORY_CLIENT_TOOLS,
  extractBasePrompt,
  resolveConvaiTtsModel
} from "./chunk-B4JC3DK2.js";
import {
  AUDIO_TAGS,
  AUDIO_TAG_GROUPS,
  DEFAULT_ALLOWED_TAGS,
  DEFAULT_VOICE_SETTINGS,
  EXPRESSIVITY_OPTIONS,
  buildPreviewLine,
  normalizeAudioTags,
  normalizeExpressivity,
  normalizeVoiceSettings,
  readAudioTags,
  readExpressivity,
  readVoiceSettings
} from "./chunk-Q3TJAJ54.js";

// src/server.ts
function json(data, status = 200, extraHeaders = {}) {
  return new Response(JSON.stringify(data), {
    status,
    headers: { "Content-Type": "application/json", ...extraHeaders }
  });
}
async function readJson(req) {
  try {
    const text = await req.text();
    if (!text) return {};
    return JSON.parse(text);
  } catch {
    return null;
  }
}
var NOOP_LOGGER = () => {
};
function parseMusicBody(raw) {
  if (raw === null || typeof raw !== "object") return { error: "Invalid JSON body" };
  const b = raw;
  const hasPrompt = typeof b.prompt === "string" && b.prompt.trim().length > 0;
  const hasPlan = b.compositionPlan != null && typeof b.compositionPlan === "object";
  if (hasPrompt === hasPlan) {
    return { error: "Provide exactly one of `prompt` or `compositionPlan`" };
  }
  const params = {};
  if (hasPrompt) {
    params.prompt = b.prompt.trim();
    if (typeof b.musicLengthMs === "number") params.musicLengthMs = b.musicLengthMs;
    if (typeof b.forceInstrumental === "boolean") params.forceInstrumental = b.forceInstrumental;
  } else {
    params.compositionPlan = b.compositionPlan;
  }
  if (typeof b.modelId === "string" && b.modelId.length > 0) params.modelId = b.modelId;
  if (typeof b.outputFormat === "string" && b.outputFormat.length > 0) {
    params.outputFormat = b.outputFormat;
  }
  return { params };
}
function resolveProxyOptions(config) {
  if (config.proxy) {
    if (!config.proxy.runtimeToken || !config.proxy.projectId || !config.proxy.apiUrl) {
      throw new Error(
        "createVoiceHandlers: proxy requires runtimeToken, projectId, and apiUrl"
      );
    }
    return config.proxy;
  }
  if (typeof process === "undefined" || !process.env) return null;
  const runtimeToken = process.env.RUNTIME_AUTH_SECRET;
  const projectId = process.env.PROJECT_ID;
  if (!runtimeToken || !projectId) return null;
  const apiUrl = process.env.SHOGO_API_URL ?? process.env.SHOGO_CLOUD_URL ?? "http://localhost:8002";
  return {
    runtimeToken,
    projectId,
    apiUrl,
    ...config.fetch ? { fetch: config.fetch } : {}
  };
}
function createProxyHandlers(proxy, log) {
  const fetchImpl = proxy.fetch ?? globalThis.fetch;
  if (typeof fetchImpl !== "function") {
    throw new Error(
      "createVoiceHandlers proxy: global fetch is unavailable; pass proxy.fetch"
    );
  }
  const apiBase = proxy.apiUrl.replace(/\/+$/, "");
  function proxyUrl(path, forward) {
    const tail = [`projectId=${encodeURIComponent(proxy.projectId)}`];
    for (const [k, v] of Object.entries(forward)) {
      if (typeof v === "string" && v.length > 0) {
        tail.push(`${k}=${encodeURIComponent(v)}`);
      }
    }
    const sep = path.includes("?") ? "&" : "?";
    return `${apiBase}${path}${sep}${tail.join("&")}`;
  }
  const FORWARDED_SIGNED_URL_PARAMS = ["agentName"];
  async function doSignedUrl(req) {
    if (req.method !== "GET") return json({ error: "Method Not Allowed" }, 405);
    try {
      let incomingUrl = null;
      try {
        incomingUrl = new URL(req.url);
      } catch {
      }
      const forward = {};
      if (incomingUrl) {
        for (const key of FORWARDED_SIGNED_URL_PARAMS) {
          const v = incomingUrl.searchParams.get(key);
          if (v != null) forward[key] = v;
        }
      }
      const upstream = await fetchImpl(proxyUrl("/api/voice/signed-url", forward), {
        method: "GET",
        headers: { "x-runtime-token": proxy.runtimeToken },
        credentials: "omit"
      });
      const body = await upstream.text();
      const headers = { "content-type": "application/json" };
      const ct = upstream.headers.get("content-type");
      if (ct) headers["content-type"] = ct;
      return new Response(body, { status: upstream.status, headers });
    } catch (err) {
      log("error", "proxy signedUrl failed", {
        error: err instanceof Error ? err.message : String(err)
      });
      return json({ error: "signedUrl proxy failed" }, 502);
    }
  }
  async function doAudioTags(req) {
    if (req.method !== "GET") return json({ error: "Method Not Allowed" }, 405);
    return json({
      tags: AUDIO_TAGS,
      groups: AUDIO_TAG_GROUPS,
      expressivity: EXPRESSIVITY_OPTIONS,
      defaults: {
        allowedTags: DEFAULT_ALLOWED_TAGS,
        voiceSettings: DEFAULT_VOICE_SETTINGS,
        ttsModelId: CONVAI_TTS_MODEL_FALLBACK
      }
    });
  }
  async function doMusic(req) {
    if (req.method !== "POST") return json({ error: "Method Not Allowed" }, 405);
    const raw = await readJson(req);
    const parsed = parseMusicBody(raw);
    if ("error" in parsed) return json({ error: parsed.error }, 400);
    try {
      const upstream = await fetchImpl(proxyUrl("/api/voice/music", {}), {
        method: "POST",
        headers: {
          "x-runtime-token": proxy.runtimeToken,
          "content-type": "application/json",
          accept: "audio/mpeg"
        },
        body: JSON.stringify(parsed.params),
        credentials: "omit"
      });
      const buf = await upstream.arrayBuffer();
      const headers = {};
      const ct = upstream.headers.get("content-type");
      if (ct) headers["content-type"] = ct;
      return new Response(buf, { status: upstream.status, headers });
    } catch (err) {
      log("error", "proxy music failed", {
        error: err instanceof Error ? err.message : String(err)
      });
      return json({ error: "music proxy failed" }, 502);
    }
  }
  function notImplemented(label) {
    return async () => json(
      {
        error: "Not Implemented",
        detail: `${label} is not available in runtime-token proxy mode; use BYO-EL or the hosted bearer SDK path.`
      },
      501
    );
  }
  return {
    signedUrl: doSignedUrl,
    tts: notImplemented("tts"),
    music: doMusic,
    agent: {
      create: notImplemented("agent.create"),
      patch: notImplemented("agent.patch"),
      delete: notImplemented("agent.delete")
    },
    audioTags: doAudioTags,
    // In proxy mode we don't own an EL client. Expose a throwing stub so
    // accidental use is loud instead of silently passing around a
    // half-configured client.
    elevenLabs: new Proxy({}, {
      get(_t, prop) {
        throw new Error(
          `createVoiceHandlers: elevenLabs.${String(prop)} is not available in runtime-token proxy mode`
        );
      }
    })
  };
}
function createVoiceHandlers(config) {
  const log = config.logger ?? NOOP_LOGGER;
  const proxy = resolveProxyOptions(config);
  if (proxy) {
    if (config.apiKey) {
      log("warn", "createVoiceHandlers: apiKey provided but runtime-token proxy mode is active; apiKey is ignored");
    }
    return createProxyHandlers(proxy, log);
  }
  if (!config.apiKey) throw new Error("createVoiceHandlers: apiKey is required");
  if (!config.getUser) throw new Error("createVoiceHandlers: getUser is required");
  if (!config.companionStore) throw new Error("createVoiceHandlers: companionStore is required");
  const getUser = config.getUser;
  const companionStore = config.companionStore;
  const includeMemoryTools = config.includeMemoryTools ?? true;
  const tools = includeMemoryTools ? MEMORY_CLIENT_TOOLS : void 0;
  const elConfig = {
    apiKey: config.apiKey,
    ...config.elevenLabsBaseUrl ? { baseUrl: config.elevenLabsBaseUrl } : {},
    ...config.fetch ? { fetch: config.fetch } : {}
  };
  const el = new ElevenLabsClient(elConfig);
  async function resolveUser(req) {
    const u = await getUser(req);
    if (!u) return json({ error: "Unauthorized" }, 401);
    return u;
  }
  function handleElError(e, fallbackLabel) {
    if (e instanceof ElevenLabsApiError) {
      log("warn", `${fallbackLabel}: 11Labs ${e.status}`, { body: e.body });
      return json({ error: fallbackLabel, detail: e.body }, 502);
    }
    const message = e instanceof Error ? e.message : String(e);
    log("error", `${fallbackLabel}: ${message}`);
    return json({ error: fallbackLabel, detail: message }, 500);
  }
  async function doSignedUrl(req) {
    if (req.method !== "GET") return json({ error: "Method Not Allowed" }, 405);
    const u = await resolveUser(req);
    if (u instanceof Response) return u;
    const companion = await companionStore.findByUserId(u.id);
    if (!companion || !companion.agentId) return json({ error: "No companion" }, 404);
    const basePrompt = extractBasePrompt(companion.systemPrompt || "");
    try {
      await el.patchAgent(companion.agentId, {
        systemPrompt: basePrompt,
        tools,
        enableUserContextOverride: includeMemoryTools,
        voiceId: companion.voiceId,
        expressivity: readExpressivity(companion.expressivity),
        audioTags: readAudioTags(companion.audioTags),
        voiceSettings: readVoiceSettings(companion.voiceSettings),
        ttsModelId: companion.ttsModelId ?? null,
        ...includeMemoryTools ? {} : { memoryBlock: null }
      });
      if (basePrompt !== companion.systemPrompt) {
        await companionStore.update(u.id, { systemPrompt: basePrompt });
      }
    } catch (e) {
      log("warn", "signedUrl: pre-sync patch failed; continuing", {
        error: e instanceof Error ? e.message : String(e)
      });
    }
    let signedUrl;
    try {
      signedUrl = await el.getSignedUrl(companion.agentId);
    } catch (e) {
      return handleElError(e, "Failed to get signed url");
    }
    let userContext = "";
    if (config.memoryStore) {
      try {
        const store = config.memoryStore(u.id);
        const hits = store.search(companion.characterName || "recent context", { limit: 5 });
        if (hits.length) {
          userContext = hits.map(
            (h) => `- ${h.matchType ? `[${h.matchType}] ` : ""}${h.chunk.trim().replace(/\s+/g, " ").slice(0, 220)}`
          ).join("\n");
        }
      } catch (e) {
        log("warn", "signedUrl: memory preload failed", {
          error: e instanceof Error ? e.message : String(e)
        });
      }
    }
    return json({ signedUrl, agentId: companion.agentId, userContext });
  }
  async function doTts(req) {
    if (req.method !== "POST") return json({ error: "Method Not Allowed" }, 405);
    const u = await resolveUser(req);
    if (u instanceof Response) return u;
    const raw = await readJson(req);
    if (raw === null || typeof raw !== "object") return json({ error: "Invalid JSON body" }, 400);
    const body = raw;
    const voiceId = typeof body.voiceId === "string" ? body.voiceId.trim() : "";
    if (!voiceId) return json({ error: "Missing voiceId" }, 400);
    const voiceSettings = normalizeVoiceSettings(body.voiceSettings) ?? { stability: 0.5, similarity_boost: 0.8, style: 0 };
    let text;
    if (typeof body.text === "string" && body.text.trim().length > 0) {
      text = body.text.slice(0, 400);
    } else {
      const tags = normalizeAudioTags(body.audioTags) ?? [];
      text = buildPreviewLine(tags, body.characterName ?? "your companion").slice(0, 400);
    }
    try {
      const { audio, modelId, contentType } = await el.textToSpeech({
        voiceId,
        text,
        ...body.modelId ? { modelId: body.modelId } : {},
        voiceSettings
      });
      return new Response(audio, {
        headers: {
          "content-type": contentType,
          "cache-control": "private, max-age=3600",
          "x-tts-model-used": modelId
        }
      });
    } catch (e) {
      return handleElError(e, "TTS failed");
    }
  }
  async function doMusic(req) {
    if (req.method !== "POST") return json({ error: "Method Not Allowed" }, 405);
    const u = await resolveUser(req);
    if (u instanceof Response) return u;
    const raw = await readJson(req);
    const parsed = parseMusicBody(raw);
    if ("error" in parsed) return json({ error: parsed.error }, 400);
    try {
      const { audio, contentType } = await el.composeMusic(parsed.params);
      return new Response(audio, {
        headers: {
          "content-type": contentType,
          "cache-control": "private, max-age=3600"
        }
      });
    } catch (e) {
      return handleElError(e, "Music generation failed");
    }
  }
  async function doAgentCreate(req) {
    if (req.method !== "POST") return json({ error: "Method Not Allowed" }, 405);
    const u = await resolveUser(req);
    if (u instanceof Response) return u;
    const existing = await companionStore.findByUserId(u.id);
    if (existing) return json({ error: "Companion already exists" }, 409);
    const raw = await readJson(req);
    if (raw === null || typeof raw !== "object") return json({ error: "Invalid JSON body" }, 400);
    const body = raw;
    if (typeof body.displayName !== "string" || typeof body.characterName !== "string" || typeof body.voiceId !== "string" || typeof body.systemPrompt !== "string" || typeof body.firstMessage !== "string") {
      return json({ error: "Missing required fields" }, 400);
    }
    const expressivity = readExpressivity(body.expressivity);
    const audioTags = readAudioTags(body.audioTags);
    const voiceSettings = readVoiceSettings(body.voiceSettings);
    const ttsModelId = resolveConvaiTtsModel(body.ttsModelId);
    let agentId;
    try {
      agentId = await el.createAgent({
        displayName: body.displayName,
        characterName: body.characterName,
        voiceId: body.voiceId,
        systemPrompt: body.systemPrompt,
        firstMessage: body.firstMessage,
        expressivity,
        audioTags,
        voiceSettings,
        ttsModelId,
        ...tools ? { tools } : {},
        ...includeMemoryTools ? {} : { memoryBlock: null }
      });
    } catch (e) {
      return handleElError(e, "Failed to create agent");
    }
    const cleanedPrompt = extractBasePrompt(body.systemPrompt);
    const companion = await companionStore.create({
      userId: u.id,
      agentId,
      displayName: body.displayName,
      characterName: body.characterName,
      voiceId: body.voiceId,
      systemPrompt: cleanedPrompt,
      firstMessage: body.firstMessage,
      expressivity,
      audioTags,
      voiceSettings,
      ttsModelId
    });
    return json(companion);
  }
  async function doAgentPatch(req) {
    if (req.method !== "PATCH") return json({ error: "Method Not Allowed" }, 405);
    const u = await resolveUser(req);
    if (u instanceof Response) return u;
    const existing = await companionStore.findByUserId(u.id);
    if (!existing) return json({ error: "No companion" }, 404);
    const raw = await readJson(req);
    if (raw === null || typeof raw !== "object") return json({ error: "Invalid JSON body" }, 400);
    const body = raw;
    const patch = {};
    if (body.displayName !== void 0) patch.displayName = body.displayName;
    if (body.characterName !== void 0) patch.characterName = body.characterName;
    if (body.voiceId !== void 0) patch.voiceId = body.voiceId;
    if (body.firstMessage !== void 0) patch.firstMessage = body.firstMessage;
    const expressivity = normalizeExpressivity(body.expressivity);
    if (expressivity !== void 0) patch.expressivity = expressivity;
    let audioTagsPatch;
    if (body.audioTags !== void 0) {
      audioTagsPatch = normalizeAudioTags(body.audioTags) ?? [];
      patch.audioTags = audioTagsPatch;
    }
    if (body.voiceSettings !== void 0) {
      const vs = normalizeVoiceSettings(body.voiceSettings);
      patch.voiceSettings = vs ?? null;
    }
    if (body.ttsModelId !== void 0) patch.ttsModelId = body.ttsModelId ?? null;
    const cleanedPrompt = body.systemPrompt !== void 0 ? extractBasePrompt(body.systemPrompt) : void 0;
    if (cleanedPrompt !== void 0) patch.systemPrompt = cleanedPrompt;
    const needs11Labs = body.displayName !== void 0 || body.characterName !== void 0 || body.voiceId !== void 0 || body.systemPrompt !== void 0 || body.firstMessage !== void 0 || body.expressivity !== void 0 || body.audioTags !== void 0 || body.voiceSettings !== void 0 || body.ttsModelId !== void 0;
    if (existing.agentId && needs11Labs) {
      const finalExpressivity = patch.expressivity ?? readExpressivity(existing.expressivity);
      const finalTags = audioTagsPatch !== void 0 ? audioTagsPatch : readAudioTags(existing.audioTags);
      const finalVoiceSettings = body.voiceSettings !== void 0 ? normalizeVoiceSettings(body.voiceSettings) ?? void 0 : readVoiceSettings(existing.voiceSettings);
      const finalTtsModelId = body.ttsModelId !== void 0 ? body.ttsModelId ?? null : existing.ttsModelId ?? null;
      try {
        await el.patchAgent(existing.agentId, {
          ...body.displayName !== void 0 ? { displayName: body.displayName } : {},
          ...body.characterName !== void 0 ? { characterName: body.characterName } : {},
          ...body.voiceId !== void 0 ? { voiceId: body.voiceId } : {},
          systemPrompt: cleanedPrompt !== void 0 ? cleanedPrompt : body.expressivity !== void 0 || body.audioTags !== void 0 ? existing.systemPrompt : void 0,
          ...body.firstMessage !== void 0 ? { firstMessage: body.firstMessage } : {},
          expressivity: finalExpressivity,
          audioTags: finalTags,
          ...finalVoiceSettings ? { voiceSettings: finalVoiceSettings } : {},
          ttsModelId: finalTtsModelId,
          ...includeMemoryTools ? {} : { memoryBlock: null }
        });
      } catch (e) {
        return handleElError(e, "Failed to patch agent");
      }
    }
    const companion = await companionStore.update(u.id, patch);
    return json(companion);
  }
  async function doAgentDelete(req) {
    if (req.method !== "DELETE") return json({ error: "Method Not Allowed" }, 405);
    const u = await resolveUser(req);
    if (u instanceof Response) return u;
    const existing = await companionStore.findByUserId(u.id);
    if (!existing) return json({ ok: true });
    if (existing.agentId) await el.deleteAgent(existing.agentId);
    await companionStore.delete(u.id);
    return json({ ok: true });
  }
  async function doAudioTags(req) {
    if (req.method !== "GET") return json({ error: "Method Not Allowed" }, 405);
    return json({
      tags: AUDIO_TAGS,
      groups: AUDIO_TAG_GROUPS,
      expressivity: EXPRESSIVITY_OPTIONS,
      defaults: {
        allowedTags: DEFAULT_ALLOWED_TAGS,
        voiceSettings: DEFAULT_VOICE_SETTINGS,
        ttsModelId: CONVAI_TTS_MODEL_FALLBACK
      }
    });
  }
  return {
    signedUrl: doSignedUrl,
    tts: doTts,
    music: doMusic,
    agent: {
      create: doAgentCreate,
      patch: doAgentPatch,
      delete: doAgentDelete
    },
    audioTags: doAudioTags,
    elevenLabs: el
  };
}
function toNodeListener(handler) {
  return (nodeReq, nodeRes) => {
    void (async () => {
      const url = `http://${nodeReq.headers.host ?? "localhost"}${nodeReq.url ?? "/"}`;
      const chunks = [];
      for await (const chunk of nodeReq) {
        chunks.push(chunk);
      }
      const body = Buffer.concat(chunks);
      const init = {
        method: nodeReq.method,
        headers: nodeReq.headers
      };
      if (body.length > 0) {
        init.body = body;
      }
      const req = new Request(url, init);
      const res = await handler(req);
      nodeRes.statusCode = res.status;
      res.headers.forEach((value, key) => {
        nodeRes.setHeader(key, value);
      });
      const buf = Buffer.from(await res.arrayBuffer());
      nodeRes.end(buf);
    })().catch((err) => {
      nodeRes.statusCode = 500;
      nodeRes.end(err instanceof Error ? err.message : String(err));
    });
  };
}

export {
  createVoiceHandlers,
  toNodeListener
};
