// @shogo-ai/voice/dist/index.js
import {
  CONVAI_SUPPORTED_TTS_MODELS,
  CONVAI_TTS_MODEL_FALLBACK,
  DEFAULT_ELEVENLABS_BASE_URL,
  DEFAULT_MEMORY_BLOCK,
  DEFAULT_MUSIC_MODEL,
  ElevenLabsApiError,
  ElevenLabsClient,
  MEMORY_CLIENT_TOOLS,
  composeAgentPrompt,
  extractBasePrompt,
  resolveConvaiTtsModel,
  stripMemoryBlock
} from "./chunk-B4JC3DK2.js";
import {
  AUDIO_TAGS,
  AUDIO_TAG_GROUPS,
  DEFAULT_ALLOWED_TAGS,
  DEFAULT_VOICE_SETTINGS,
  EXPRESSIVITY_BLOCK_CLOSE,
  EXPRESSIVITY_BLOCK_OPEN,
  EXPRESSIVITY_OPTIONS,
  buildPreviewLine,
  composeExpressivityBlock,
  normalizeAudioTags,
  normalizeExpressivity,
  normalizeVoiceSettings,
  readAudioTags,
  readExpressivity,
  readVoiceSettings,
  stripAudioTags,
  stripExpressivityBlock
} from "./chunk-Q3TJAJ54.js";

// src/twilio.ts
var DEFAULT_TWILIO_BASE_URL = "https://api.twilio.com";
var TwilioApiError = class extends Error {
  constructor(message, status, body) {
    super(message);
    this.status = status;
    this.body = body;
    this.name = "TwilioApiError";
  }
};
function toBase64(s) {
  if (typeof btoa === "function") return btoa(s);
  const nodeBuffer = globalThis.Buffer;
  if (nodeBuffer) return nodeBuffer.from(s, "utf-8").toString("base64");
  throw new Error("twilio.ts: no base64 encoder available in this runtime");
}
var TwilioClient = class {
  constructor(config) {
    this.config = config;
    if (!config.accountSid) throw new Error("TwilioClient: accountSid is required");
    if (!config.authToken) throw new Error("TwilioClient: authToken is required");
    this.baseUrl = (config.baseUrl ?? DEFAULT_TWILIO_BASE_URL).replace(/\/+$/, "");
    this.fetchImpl = config.fetch ?? globalThis.fetch;
    if (typeof this.fetchImpl !== "function") {
      throw new Error("TwilioClient: global fetch is unavailable; pass config.fetch");
    }
    this.basicAuth = `Basic ${toBase64(`${config.accountSid}:${config.authToken}`)}`;
  }
  accountUrl(path) {
    return `${this.baseUrl}/2010-04-01/Accounts/${encodeURIComponent(
      this.config.accountSid
    )}${path}`;
  }
  headers(extra = {}) {
    return { authorization: this.basicAuth, ...extra };
  }
  async searchAvailable(params = {}) {
    const country = (params.country ?? "US").toUpperCase();
    const search = new URLSearchParams();
    if (params.areaCode) search.set("AreaCode", params.areaCode);
    if (params.contains) search.set("Contains", params.contains);
    search.set("PageSize", String(params.limit ?? 10));
    const url = `${this.accountUrl(
      `/AvailablePhoneNumbers/${country}/Local.json`
    )}?${search.toString()}`;
    const res = await this.fetchImpl(url, { headers: this.headers() });
    if (!res.ok) {
      throw new TwilioApiError(
        `searchAvailable failed: ${res.status}`,
        res.status,
        await res.text()
      );
    }
    const data = await res.json();
    return (data.available_phone_numbers ?? []).map((n) => ({
      phoneNumber: n.phone_number,
      friendlyName: n.friendly_name ?? n.phone_number,
      region: n.region,
      locality: n.locality,
      isoCountry: n.iso_country
    }));
  }
  async purchaseNumber(params) {
    const body = new URLSearchParams();
    body.set("PhoneNumber", params.phoneNumber);
    if (params.friendlyName) body.set("FriendlyName", params.friendlyName);
    if (params.voiceUrl) body.set("VoiceUrl", params.voiceUrl);
    const res = await this.fetchImpl(this.accountUrl("/IncomingPhoneNumbers.json"), {
      method: "POST",
      headers: this.headers({ "content-type": "application/x-www-form-urlencoded" }),
      body: body.toString()
    });
    if (!res.ok) {
      throw new TwilioApiError(
        `purchaseNumber failed: ${res.status}`,
        res.status,
        await res.text()
      );
    }
    const data = await res.json();
    return {
      sid: data.sid,
      phoneNumber: data.phone_number,
      friendlyName: data.friendly_name ?? data.phone_number,
      voiceUrl: data.voice_url
    };
  }
  async releaseNumber(sid) {
    const res = await this.fetchImpl(
      this.accountUrl(`/IncomingPhoneNumbers/${encodeURIComponent(sid)}.json`),
      { method: "DELETE", headers: this.headers() }
    );
    if (!res.ok && res.status !== 404) {
      throw new TwilioApiError(
        `releaseNumber failed: ${res.status}`,
        res.status,
        await res.text()
      );
    }
  }
};

// src/telephony.ts
var TelephonyConfigError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "TelephonyConfigError";
  }
};
var TelephonyApiError = class extends Error {
  constructor(message, status, body) {
    super(message);
    this.status = status;
    this.body = body;
    this.name = "TelephonyApiError";
  }
};
var HostedTelephonyClient = class {
  constructor(opts) {
    this.opts = opts;
    this.mode = "hosted";
    if (!opts.shogoApiKey) {
      throw new TelephonyConfigError("Hosted telephony requires shogoApiKey");
    }
    if (!opts.projectId) {
      throw new TelephonyConfigError("Hosted telephony requires projectId");
    }
    if (!opts.apiUrl) {
      throw new TelephonyConfigError("Hosted telephony requires apiUrl");
    }
    this.fetchImpl = opts.fetch ?? globalThis.fetch;
    if (typeof this.fetchImpl !== "function") {
      throw new TelephonyConfigError("global fetch is unavailable; pass opts.fetch");
    }
  }
  url(path) {
    return `${this.opts.apiUrl.replace(/\/+$/, "")}${path}`;
  }
  async request(path, init = {}) {
    const headers = {
      authorization: `Bearer ${this.opts.shogoApiKey}`,
      ...init.headers
    };
    let body = init.body;
    if (init.json !== void 0) {
      headers["content-type"] = "application/json";
      body = JSON.stringify(init.json);
    }
    const res = await this.fetchImpl(this.url(path), {
      ...init,
      headers,
      body,
      credentials: "omit"
    });
    const text = await res.text();
    let parsed = void 0;
    if (text) {
      try {
        parsed = JSON.parse(text);
      } catch {
        parsed = text;
      }
    }
    if (!res.ok) {
      const message = parsed?.error ?? parsed?.message ?? `Shogo API error: ${res.status}`;
      throw new TelephonyApiError(message, res.status, parsed);
    }
    return parsed;
  }
  async provisionNumber(opts = {}) {
    const data = await this.request(
      `/api/voice/twilio/provision-number/${encodeURIComponent(
        this.opts.projectId
      )}`,
      {
        method: "POST",
        json: {
          areaCode: opts.areaCode,
          country: opts.country,
          friendlyName: opts.friendlyName
        }
      }
    );
    return {
      phoneNumber: data.phoneNumber,
      twilioPhoneSid: data.twilioPhoneSid,
      elevenlabsPhoneId: data.elevenlabsPhoneId,
      setupBilledUsd: data.setupBilledUsd,
      monthlyBilledUsd: data.monthlyBilledUsd,
      usageDebited: data.usageDebited
    };
  }
  async outboundCall(opts) {
    const data = await this.request(
      `/api/voice/twilio/outbound/${encodeURIComponent(this.opts.projectId)}`,
      {
        method: "POST",
        json: { to: opts.to, dynamicVariables: opts.dynamicVariables }
      }
    );
    return {
      callSid: data.callSid,
      conversationId: data.conversationId,
      estimatedBilledUsd: data.estimatedBilledUsd,
      billedUsdPerMinute: data.billedUsdPerMinute
    };
  }
  async releaseNumber() {
    await this.request(
      `/api/voice/twilio/number/${encodeURIComponent(this.opts.projectId)}`,
      { method: "DELETE" }
    );
    return { released: true };
  }
  async getUsage(range = {}) {
    const q = new URLSearchParams();
    if (range.from) q.set("from", range.from);
    if (range.to) q.set("to", range.to);
    const path = `/api/voice/usage/${encodeURIComponent(this.opts.projectId)}${q.toString() ? `?${q.toString()}` : ""}`;
    return this.request(path, { method: "GET" });
  }
  async listCalls(opts = {}) {
    const q = new URLSearchParams();
    if (opts.limit) q.set("limit", String(opts.limit));
    if (opts.includeTranscript) q.set("includeTranscript", "1");
    const path = `/api/voice/calls/${encodeURIComponent(this.opts.projectId)}${q.toString() ? `?${q.toString()}` : ""}`;
    const res = await this.request(
      path,
      { method: "GET" }
    );
    return res.calls;
  }
  async getCall(callId) {
    const path = `/api/voice/calls/${encodeURIComponent(
      this.opts.projectId
    )}/${encodeURIComponent(callId)}`;
    return this.request(path, { method: "GET" });
  }
};
var HostedRuntimeTokenClient = class {
  constructor(opts) {
    this.opts = opts;
    this.mode = "hosted";
    if (!opts.runtimeToken) {
      throw new TelephonyConfigError(
        "Runtime-token telephony requires runtimeToken (process.env.RUNTIME_AUTH_SECRET)"
      );
    }
    if (!opts.projectId) {
      throw new TelephonyConfigError("Runtime-token telephony requires projectId");
    }
    if (!opts.apiUrl) {
      throw new TelephonyConfigError("Runtime-token telephony requires apiUrl");
    }
    this.fetchImpl = opts.fetch ?? globalThis.fetch;
    if (typeof this.fetchImpl !== "function") {
      throw new TelephonyConfigError("global fetch is unavailable; pass opts.fetch");
    }
  }
  url(path) {
    const base = this.opts.apiUrl.replace(/\/+$/, "");
    const sep = path.includes("?") ? "&" : "?";
    return `${base}${path}${sep}projectId=${encodeURIComponent(this.opts.projectId)}`;
  }
  async request(path, init = {}) {
    const headers = {
      "x-runtime-token": this.opts.runtimeToken,
      ...init.headers
    };
    let body = init.body;
    if (init.json !== void 0) {
      headers["content-type"] = "application/json";
      body = JSON.stringify(init.json);
    }
    const res = await this.fetchImpl(this.url(path), {
      ...init,
      headers,
      body,
      credentials: "omit"
    });
    const text = await res.text();
    let parsed = void 0;
    if (text) {
      try {
        parsed = JSON.parse(text);
      } catch {
        parsed = text;
      }
    }
    if (!res.ok) {
      const message = parsed?.error ?? parsed?.message ?? `Shogo API error: ${res.status}`;
      throw new TelephonyApiError(message, res.status, parsed);
    }
    return parsed;
  }
  async provisionNumber(opts = {}) {
    const data = await this.request(
      `/api/voice/twilio/provision-number/${encodeURIComponent(
        this.opts.projectId
      )}`,
      {
        method: "POST",
        json: {
          areaCode: opts.areaCode,
          country: opts.country,
          friendlyName: opts.friendlyName
        }
      }
    );
    return {
      phoneNumber: data.phoneNumber,
      twilioPhoneSid: data.twilioPhoneSid,
      elevenlabsPhoneId: data.elevenlabsPhoneId,
      setupBilledUsd: data.setupBilledUsd,
      monthlyBilledUsd: data.monthlyBilledUsd,
      usageDebited: data.usageDebited
    };
  }
  async outboundCall(opts) {
    const data = await this.request(
      `/api/voice/twilio/outbound/${encodeURIComponent(this.opts.projectId)}`,
      {
        method: "POST",
        json: { to: opts.to, dynamicVariables: opts.dynamicVariables }
      }
    );
    return {
      callSid: data.callSid,
      conversationId: data.conversationId,
      estimatedBilledUsd: data.estimatedBilledUsd,
      billedUsdPerMinute: data.billedUsdPerMinute
    };
  }
  async releaseNumber() {
    await this.request(
      `/api/voice/twilio/number/${encodeURIComponent(this.opts.projectId)}`,
      { method: "DELETE" }
    );
    return { released: true };
  }
  async getUsage(range = {}) {
    const q = new URLSearchParams();
    if (range.from) q.set("from", range.from);
    if (range.to) q.set("to", range.to);
    const path = `/api/voice/usage/${encodeURIComponent(this.opts.projectId)}${q.toString() ? `?${q.toString()}` : ""}`;
    return this.request(path, { method: "GET" });
  }
  async listCalls(opts = {}) {
    const q = new URLSearchParams();
    if (opts.limit) q.set("limit", String(opts.limit));
    if (opts.includeTranscript) q.set("includeTranscript", "1");
    const path = `/api/voice/calls/${encodeURIComponent(this.opts.projectId)}${q.toString() ? `?${q.toString()}` : ""}`;
    const res = await this.request(
      path,
      { method: "GET" }
    );
    return res.calls;
  }
  async getCall(callId) {
    const path = `/api/voice/calls/${encodeURIComponent(
      this.opts.projectId
    )}/${encodeURIComponent(callId)}`;
    return this.request(path, { method: "GET" });
  }
};
var DirectTelephonyClient = class {
  constructor(opts) {
    this.opts = opts;
    this.mode = "direct";
    if (!opts.elevenlabs?.apiKey) {
      throw new TelephonyConfigError("Direct telephony requires elevenlabs.apiKey");
    }
    if (!opts.elevenlabs?.agentId) {
      throw new TelephonyConfigError("Direct telephony requires elevenlabs.agentId");
    }
    if (!opts.twilio?.accountSid || !opts.twilio?.authToken) {
      throw new TelephonyConfigError(
        "Direct telephony requires twilio.accountSid + twilio.authToken"
      );
    }
    this.el = new ElevenLabsClient({
      apiKey: opts.elevenlabs.apiKey,
      baseUrl: opts.elevenlabs.baseUrl,
      fetch: opts.fetch
    });
    this.tw = new TwilioClient({
      accountSid: opts.twilio.accountSid,
      authToken: opts.twilio.authToken,
      baseUrl: opts.twilio.baseUrl,
      fetch: opts.fetch
    });
    this.state = {
      phoneNumber: opts.twilio.fromNumber,
      twilioSid: opts.twilio.phoneSid,
      elevenlabsPhoneId: opts.elevenlabs.phoneNumberId
    };
  }
  async provisionNumber(opts = {}) {
    const candidates = await this.tw.searchAvailable({
      country: opts.country,
      areaCode: opts.areaCode,
      limit: 1
    });
    if (candidates.length === 0) {
      throw new TelephonyApiError(
        "No Twilio numbers available matching criteria",
        404,
        { areaCode: opts.areaCode, country: opts.country }
      );
    }
    const purchased = await this.tw.purchaseNumber({
      phoneNumber: candidates[0].phoneNumber,
      friendlyName: opts.friendlyName
    });
    try {
      const { phoneNumberId } = await this.el.createPhoneNumberTwilio({
        phoneNumber: purchased.phoneNumber,
        label: opts.friendlyName ?? purchased.phoneNumber,
        agentId: this.opts.elevenlabs.agentId,
        twilioAccountSid: this.opts.twilio.accountSid,
        twilioAuthToken: this.opts.twilio.authToken
      });
      this.state = {
        phoneNumber: purchased.phoneNumber,
        twilioSid: purchased.sid,
        elevenlabsPhoneId: phoneNumberId
      };
      return {
        phoneNumber: purchased.phoneNumber,
        twilioPhoneSid: purchased.sid,
        elevenlabsPhoneId: phoneNumberId
      };
    } catch (err) {
      try {
        await this.tw.releaseNumber(purchased.sid);
      } catch {
      }
      throw err;
    }
  }
  async outboundCall(opts) {
    if (!this.state.elevenlabsPhoneId) {
      throw new TelephonyConfigError(
        "outboundCall requires an ElevenLabs phone id; call provisionNumber() or pass elevenlabs.phoneNumberId"
      );
    }
    const res = await this.el.outboundCall({
      phoneNumberId: this.state.elevenlabsPhoneId,
      agentId: this.opts.elevenlabs.agentId,
      toNumber: opts.to,
      dynamicVariables: opts.dynamicVariables
    });
    return { callSid: res.callSid, conversationId: res.conversationId };
  }
  async releaseNumber() {
    let released = false;
    if (this.state.elevenlabsPhoneId) {
      await this.el.deletePhoneNumber(this.state.elevenlabsPhoneId);
      released = true;
    }
    if (this.state.twilioSid) {
      await this.tw.releaseNumber(this.state.twilioSid);
      released = true;
    }
    this.state = {};
    return { released };
  }
  async getUsage() {
    throw new TelephonyConfigError(
      "getUsage() is unavailable in Mode A (self-hosted). Query Twilio + ElevenLabs directly."
    );
  }
  async listCalls() {
    throw new TelephonyConfigError(
      "listCalls() is unavailable in Mode A (self-hosted). Query ElevenLabs conversations directly."
    );
  }
  async getCall() {
    throw new TelephonyConfigError(
      "getCall() is unavailable in Mode A (self-hosted). Query ElevenLabs conversations directly."
    );
  }
};
function createTelephonyClient(opts) {
  if (opts.mode === "hosted") return new HostedTelephonyClient(opts);
  if (opts.mode === "runtime-token") return new HostedRuntimeTokenClient(opts);
  return new DirectTelephonyClient(opts);
}

// src/mock-telephony.ts
var DEFAULT_PHONE = "+14155550143";
var DEFAULT_TWILIO_SID = "PNmock0000000000000000000000000001";
var DEFAULT_EL_PHONE_ID = "phn_mock_0000000001";
var DEFAULT_CALL_FIXTURES = [
  {
    id: "mock_call_001",
    conversationId: "mock_conv_001",
    callSid: "CAmock00000000000000000000000000001",
    direction: "outbound",
    durationSeconds: 184,
    billedMinutes: 4,
    startedAt: "2026-05-06T22:01:00.000Z",
    endedAt: "2026-05-06T22:04:04.000Z",
    createdAt: "2026-05-06T22:01:00.000Z",
    billed: true,
    hasTranscript: true,
    transcriptSummary: "Prospect agreed to a 30-min demo on Thursday at 2pm PT. Booked via Calendly link."
  },
  {
    id: "mock_call_002",
    conversationId: "mock_conv_002",
    callSid: "CAmock00000000000000000000000000002",
    direction: "outbound",
    durationSeconds: 41,
    billedMinutes: 1,
    startedAt: "2026-05-06T22:05:00.000Z",
    endedAt: "2026-05-06T22:05:41.000Z",
    createdAt: "2026-05-06T22:05:00.000Z",
    billed: true,
    hasTranscript: true,
    transcriptSummary: "Voicemail. Left a 25-second AI-disclosed message with callback number."
  },
  {
    id: "mock_call_003",
    conversationId: "mock_conv_003",
    callSid: "CAmock00000000000000000000000000003",
    direction: "outbound",
    durationSeconds: 22,
    billedMinutes: 1,
    startedAt: "2026-05-06T22:06:30.000Z",
    endedAt: "2026-05-06T22:06:52.000Z",
    createdAt: "2026-05-06T22:06:30.000Z",
    billed: true,
    hasTranscript: true,
    transcriptSummary: 'Prospect said "remove me" 8 seconds in. Honored, persisted to do-not-call list.'
  }
];
var DEFAULT_TRANSCRIPTS = {
  mock_conv_001: [
    { role: "agent", message: "Hi, this is an AI assistant calling on behalf of Russell at Shogo. Do you have 30 seconds?", time_in_call_secs: 1 },
    { role: "user", message: "Sure, what is this about?", time_in_call_secs: 6 },
    { role: "agent", message: "Shogo is a universal AI agent that runs every job-to-be-done with real interfaces, not chat. We help teams like yours replace point tools with one platform.", time_in_call_secs: 10 },
    { role: "user", message: "Interesting. We've been looking at workflow automation. How is this different from Zapier?", time_in_call_secs: 24 },
    { role: "agent", message: "Zapier connects steps. Shogo runs the whole workflow with an agent that can use tools, build interfaces, and learn from your data.", time_in_call_secs: 30 },
    { role: "user", message: "OK I'd like to see it. Send me a demo invite.", time_in_call_secs: 52 },
    { role: "agent", message: "Great. I'm sending the Calendly link now. Does Thursday at 2pm Pacific work?", time_in_call_secs: 58 },
    { role: "user", message: "Yes, perfect.", time_in_call_secs: 66 },
    { role: "agent", message: "Booked. You'll get a confirmation email in 30 seconds. Thanks for your time.", time_in_call_secs: 70 }
  ],
  mock_conv_002: [
    { role: "agent", message: "Hi, this is an AI assistant calling on behalf of Russell at Shogo. Leaving a voicemail \u2014 we'd love to show you our universal AI agent platform. Call us back at +1 415 555 0143 or visit shogo.ai. Have a great day.", time_in_call_secs: 1 }
  ],
  mock_conv_003: [
    { role: "agent", message: "Hi, this is an AI assistant calling on behalf of Russell at Shogo \u2014", time_in_call_secs: 1 },
    { role: "user", message: "Remove me from your list.", time_in_call_secs: 8 },
    { role: "agent", message: "Understood. You're removed and we won't call again. Thank you.", time_in_call_secs: 11 }
  ]
};
var warned = false;
var MockTelephonyClient = class {
  constructor(opts = {}) {
    this.mode = "mock";
    this.callCounter = 0;
    this.projectId = opts.projectId ?? "demo-project";
    this.phoneNumber = opts.phoneNumber ?? DEFAULT_PHONE;
    this.fixtures = opts.callFixtures ?? DEFAULT_CALL_FIXTURES;
    if (!warned) {
      console.warn(
        "[shogo] voice mock mode active \u2014 telephony returns canned data, no real provider calls are made."
      );
      warned = true;
    }
  }
  async provisionNumber(_opts = {}) {
    return {
      phoneNumber: this.phoneNumber,
      twilioPhoneSid: DEFAULT_TWILIO_SID,
      elevenlabsPhoneId: DEFAULT_EL_PHONE_ID,
      setupBilledUsd: 0,
      monthlyBilledUsd: 0,
      usageDebited: { setup: false, monthly: false }
    };
  }
  async outboundCall(_opts) {
    this.callCounter++;
    const seq = String(this.callCounter).padStart(32, "0").slice(-32);
    return {
      callSid: `CAmock${seq}`,
      conversationId: `mock_conv_${String(this.callCounter).padStart(3, "0")}`,
      estimatedBilledUsd: 0,
      billedUsdPerMinute: 0
    };
  }
  async releaseNumber() {
    return { released: true };
  }
  async getUsage(_range) {
    const outboundCalls = this.fixtures.filter((c) => c.direction === "outbound").length;
    const inboundCalls = this.fixtures.length - outboundCalls;
    const totalMinutes = this.fixtures.reduce((sum, c) => sum + c.billedMinutes, 0);
    return {
      projectId: this.projectId,
      range: { from: null, to: null },
      totals: {
        minutesInbound: 0,
        minutesOutbound: totalMinutes,
        billedUsdInbound: 0,
        billedUsdOutbound: 0,
        billedUsdNumbers: 0,
        billedUsd: 0,
        calls: this.fixtures.length,
        inboundCalls,
        outboundCalls
      },
      events: []
    };
  }
  async listCalls(opts = {}) {
    const limit = opts.limit ?? this.fixtures.length;
    const slice = this.fixtures.slice(0, limit);
    if (!opts.includeTranscript) return slice.map((c) => ({ ...c }));
    return slice.map((c) => ({
      ...c,
      transcript: DEFAULT_TRANSCRIPTS[c.conversationId ?? ""] ?? null
    }));
  }
  async getCall(callId) {
    const found = this.fixtures.find(
      (c) => c.id === callId || c.conversationId === callId || c.callSid === callId
    );
    if (!found) {
      return {
        id: callId,
        conversationId: callId,
        callSid: callId,
        direction: "outbound",
        durationSeconds: 0,
        billedMinutes: 0,
        startedAt: (/* @__PURE__ */ new Date()).toISOString(),
        endedAt: (/* @__PURE__ */ new Date()).toISOString(),
        createdAt: (/* @__PURE__ */ new Date()).toISOString(),
        billed: false,
        transcriptSummary: null,
        transcript: null
      };
    }
    const { hasTranscript: _hasTranscript, ...rest } = found;
    return {
      ...rest,
      transcript: DEFAULT_TRANSCRIPTS[found.conversationId ?? ""] ?? null
    };
  }
};
function isVoiceMockEnv() {
  if (typeof process === "undefined") return false;
  const value = process.env?.SHOGO_VOICE_MODE;
  if (!value) return false;
  const v = value.toLowerCase();
  return v === "mock" || v === "demo" || v === "fake";
}
export {
  AUDIO_TAGS,
  AUDIO_TAG_GROUPS,
  CONVAI_SUPPORTED_TTS_MODELS,
  CONVAI_TTS_MODEL_FALLBACK,
  DEFAULT_ALLOWED_TAGS,
  DEFAULT_ELEVENLABS_BASE_URL,
  DEFAULT_MEMORY_BLOCK,
  DEFAULT_MUSIC_MODEL,
  DEFAULT_TWILIO_BASE_URL,
  DEFAULT_VOICE_SETTINGS,
  DirectTelephonyClient,
  EXPRESSIVITY_BLOCK_CLOSE,
  EXPRESSIVITY_BLOCK_OPEN,
  EXPRESSIVITY_OPTIONS,
  ElevenLabsApiError,
  ElevenLabsClient,
  HostedRuntimeTokenClient,
  HostedTelephonyClient,
  MEMORY_CLIENT_TOOLS,
  MockTelephonyClient,
  TelephonyApiError,
  TelephonyConfigError,
  TwilioApiError,
  TwilioClient,
  buildPreviewLine,
  composeAgentPrompt,
  composeExpressivityBlock,
  createTelephonyClient,
  extractBasePrompt,
  isVoiceMockEnv,
  normalizeAudioTags,
  normalizeExpressivity,
  normalizeVoiceSettings,
  readAudioTags,
  readExpressivity,
  readVoiceSettings,
  resolveConvaiTtsModel,
  stripAudioTags,
  stripExpressivityBlock,
  stripMemoryBlock
};
