// @shogo-ai/email/dist/server.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/server.ts
var server_exports = {};
__export(server_exports, {
  createEmail: () => createEmail,
  createEmailOptional: () => createEmailOptional
});
module.exports = __toCommonJS(server_exports);

// src/types.ts
function formatEmailAddress(address) {
  if (typeof address === "string") {
    return address;
  }
  return address.name ? `"${address.name}" <${address.email}>` : address.email;
}
var EmailError = class _EmailError extends Error {
  constructor(code, message, cause) {
    super(message);
    this.name = "EmailError";
    this.code = code;
    this.cause = cause;
  }
  static configMissing(field) {
    return new _EmailError("config_missing", `Missing required config: ${field}`);
  }
  static templateNotFound(name) {
    return new _EmailError("template_not_found", `Template not found: ${name}`);
  }
  static providerNotConfigured() {
    return new _EmailError("provider_not_configured", "Email provider is not configured");
  }
};

// src/providers/smtp.ts
var SmtpProvider = class {
  // nodemailer.Transporter - lazy loaded
  constructor(config) {
    this.validateConfig(config);
    this.config = config;
  }
  /**
   * Validate required config fields
   */
  validateConfig(config) {
    const required = ["host", "port", "user", "password"];
    for (const field of required) {
      if (!config[field]) {
        throw EmailError.configMissing(`smtp.${field}`);
      }
    }
  }
  /**
   * Lazy-load nodemailer and create transporter
   */
  async getTransporter() {
    if (this.transporter) {
      return this.transporter;
    }
    const nodemailer = await import("nodemailer");
    this.transporter = nodemailer.createTransport({
      host: this.config.host,
      port: this.config.port,
      secure: this.config.secure ?? this.config.port === 465,
      auth: {
        user: this.config.user,
        pass: this.config.password
      }
    });
    return this.transporter;
  }
  /**
   * Format recipients for nodemailer
   */
  formatRecipients(addresses) {
    const list = Array.isArray(addresses) ? addresses : [addresses];
    return list.map(formatEmailAddress).join(", ");
  }
  /**
   * Strip HTML tags to create plain text version
   */
  stripHtml(html) {
    return html.replace(/<style[^>]*>.*?<\/style>/gi, "").replace(/<script[^>]*>.*?<\/script>/gi, "").replace(/<[^>]+>/g, "").replace(/\s+/g, " ").trim();
  }
  /**
   * Check if provider is configured
   */
  isConfigured() {
    return !!(this.config.host && this.config.port && this.config.user && this.config.password);
  }
  /**
   * Send an email via SMTP
   */
  async send(params) {
    const { to, subject, html, text, from, replyTo, cc, bcc } = params;
    if (!from) {
      throw EmailError.configMissing("from address");
    }
    try {
      const transporter = await this.getTransporter();
      const mailOptions = {
        from: formatEmailAddress(from),
        to: this.formatRecipients(to),
        subject,
        html,
        text: text || this.stripHtml(html)
      };
      if (replyTo) {
        mailOptions.replyTo = formatEmailAddress(replyTo);
      }
      if (cc) {
        mailOptions.cc = this.formatRecipients(cc);
      }
      if (bcc) {
        mailOptions.bcc = this.formatRecipients(bcc);
      }
      const info = await transporter.sendMail(mailOptions);
      return {
        success: true,
        messageId: info.messageId
      };
    } catch (error) {
      const errorMessage = error.message || "Failed to send email";
      const errorCode = this.mapErrorCode(error);
      return {
        success: false,
        error: `[${errorCode}] ${errorMessage}`
      };
    }
  }
  /**
   * Map nodemailer error to error code
   */
  mapErrorCode(error) {
    const code = error.code || error.errno || "";
    const message = (error.message || "").toLowerCase();
    if (code === "ECONNREFUSED" || code === "ENOTFOUND" || message.includes("connection")) {
      return "connection_error";
    }
    if (code === "EAUTH" || message.includes("auth") || message.includes("credentials")) {
      return "authentication_error";
    }
    if (message.includes("recipient") || message.includes("mailbox") || message.includes("not found")) {
      return "invalid_recipient";
    }
    return "send_failed";
  }
};
function createSmtpProviderFromEnv() {
  const host = process.env.SMTP_HOST;
  const port = process.env.SMTP_PORT;
  const user = process.env.SMTP_USER;
  const password = process.env.SMTP_PASSWORD;
  if (!host || !port || !user || !password) {
    return null;
  }
  return new SmtpProvider({
    host,
    port: parseInt(port, 10),
    user,
    password,
    secure: process.env.SMTP_SECURE === "true" ? true : void 0
  });
}

// src/providers/ses.ts
var SesProvider = class {
  // SESClient - lazy loaded
  constructor(config) {
    this.validateConfig(config);
    this.config = config;
  }
  /**
   * Validate required config fields
   */
  validateConfig(config) {
    if (!config.region) {
      throw EmailError.configMissing("ses.region");
    }
  }
  /**
   * Lazy-load AWS SDK and create client
   */
  async getClient() {
    if (this.client) {
      return this.client;
    }
    const { SESClient } = await import("@aws-sdk/client-ses");
    const clientConfig = {
      region: this.config.region
    };
    if (this.config.accessKeyId && this.config.secretAccessKey) {
      clientConfig.credentials = {
        accessKeyId: this.config.accessKeyId,
        secretAccessKey: this.config.secretAccessKey
      };
    }
    this.client = new SESClient(clientConfig);
    return this.client;
  }
  /**
   * Format recipients as array of strings
   */
  formatRecipients(addresses) {
    const list = Array.isArray(addresses) ? addresses : [addresses];
    return list.map(
      (addr) => typeof addr === "string" ? addr : addr.email
    );
  }
  /**
   * Strip HTML tags to create plain text version
   */
  stripHtml(html) {
    return html.replace(/<style[^>]*>.*?<\/style>/gi, "").replace(/<script[^>]*>.*?<\/script>/gi, "").replace(/<[^>]+>/g, "").replace(/\s+/g, " ").trim();
  }
  /**
   * Check if provider is configured
   */
  isConfigured() {
    return !!this.config.region;
  }
  /**
   * Send an email via SES
   */
  async send(params) {
    const { to, subject, html, text, from, replyTo, cc, bcc } = params;
    if (!from) {
      throw EmailError.configMissing("from address");
    }
    try {
      const client = await this.getClient();
      const { SendEmailCommand } = await import("@aws-sdk/client-ses");
      const destination = {
        ToAddresses: this.formatRecipients(to)
      };
      if (cc) {
        destination.CcAddresses = this.formatRecipients(cc);
      }
      if (bcc) {
        destination.BccAddresses = this.formatRecipients(bcc);
      }
      const command = new SendEmailCommand({
        Source: formatEmailAddress(from),
        Destination: destination,
        Message: {
          Subject: {
            Data: subject,
            Charset: "UTF-8"
          },
          Body: {
            Html: {
              Data: html,
              Charset: "UTF-8"
            },
            Text: {
              Data: text || this.stripHtml(html),
              Charset: "UTF-8"
            }
          }
        },
        ReplyToAddresses: replyTo ? [formatEmailAddress(replyTo)] : void 0
      });
      const response = await client.send(command);
      return {
        success: true,
        messageId: response.MessageId
      };
    } catch (error) {
      const errorMessage = error.message || "Failed to send email";
      return {
        success: false,
        error: `[ses_error] ${errorMessage}`
      };
    }
  }
};
function createSesProviderFromEnv() {
  const region = process.env.SES_REGION || process.env.AWS_REGION;
  if (!region) {
    return null;
  }
  return new SesProvider({
    region,
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
  });
}

// src/providers/oci-email.ts
var OciEmailProvider = class {
  constructor(config) {
    this.fromAddress = config.fromAddress;
    this.smtp = new SmtpProvider({
      host: config.host,
      port: config.port,
      user: config.user,
      password: config.password,
      secure: config.port === 465
    });
  }
  isConfigured() {
    return this.smtp.isConfigured();
  }
  async send(params) {
    const effective = {
      ...params,
      from: params.from ?? this.fromAddress
    };
    return this.smtp.send(effective);
  }
};
function createOciEmailProviderFromEnv() {
  const host = process.env.OCI_EMAIL_SMTP_HOST;
  const user = process.env.OCI_EMAIL_SMTP_USER;
  const password = process.env.OCI_EMAIL_SMTP_PASS;
  if (!host || !user || !password) {
    return null;
  }
  const port = parseInt(process.env.OCI_EMAIL_SMTP_PORT || "587", 10);
  return new OciEmailProvider({
    host,
    port,
    user,
    password,
    fromAddress: process.env.OCI_EMAIL_FROM_ADDRESS
  });
}

// src/providers/index.ts
function createProviderFromEnv() {
  const explicitProvider = process.env.EMAIL_PROVIDER;
  if (explicitProvider === "smtp") {
    return createSmtpProviderFromEnv();
  }
  if (explicitProvider === "ses") {
    return createSesProviderFromEnv();
  }
  if (explicitProvider === "oci-email") {
    return createOciEmailProviderFromEnv();
  }
  if (process.env.OCI_EMAIL_SMTP_HOST) {
    return createOciEmailProviderFromEnv();
  }
  if (process.env.SMTP_HOST) {
    return createSmtpProviderFromEnv();
  }
  if (process.env.SES_REGION || process.env.AWS_REGION) {
    return createSesProviderFromEnv();
  }
  return null;
}

// src/templates.ts
function escapeHtml(value) {
  return value.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
}
function interpolate(template, data, options = {}) {
  return template.replace(/\{\{(\w+(?:\.\w+)*)\}\}/g, (match, path) => {
    const value = getNestedValue(data, path);
    if (value === void 0) return match;
    const str = String(value);
    return options.escapeValues ? escapeHtml(str) : str;
  });
}
function getNestedValue(obj, path) {
  const keys = path.split(".");
  let current = obj;
  for (const key of keys) {
    if (current === null || current === void 0) {
      return void 0;
    }
    current = current[key];
  }
  return current;
}
function htmlToText(html) {
  return html.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, "").replace(/<script[^>]*>[\s\S]*?<\/script>/gi, "").replace(/<br\s*\/?>/gi, "\n").replace(/<\/p>/gi, "\n\n").replace(/<\/div>/gi, "\n").replace(/<\/h[1-6]>/gi, "\n\n").replace(/<\/li>/gi, "\n").replace(/<[^>]+>/g, "").replace(/&nbsp;/gi, " ").replace(/&amp;/gi, "&").replace(/&lt;/gi, "<").replace(/&gt;/gi, ">").replace(/&quot;/gi, '"').replace(/&#39;/gi, "'").replace(/[ \t]+/g, " ").replace(/\n{3,}/g, "\n\n").trim();
}
var EmailTemplateRegistry = class {
  constructor() {
    this.templates = {};
  }
  /**
   * Register a template
   */
  register(template) {
    this.templates[template.name] = template;
  }
  /**
   * Get a template by name
   */
  get(name) {
    return this.templates[name];
  }
  /**
   * Check if a template exists
   */
  has(name) {
    return name in this.templates;
  }
  /**
   * Render a template with data
   */
  render(name, data) {
    const template = this.templates[name];
    if (!template) {
      throw EmailError.templateNotFound(name);
    }
    const mergedData = { ...template.defaults, ...data };
    const allData = {
      ...mergedData,
      currentYear: (/* @__PURE__ */ new Date()).getFullYear()
    };
    const subject = interpolate(template.subject, allData);
    const html = interpolate(template.html, allData, { escapeValues: true });
    const text = template.text ? interpolate(template.text, allData) : htmlToText(interpolate(template.html, allData));
    return { subject, html, text };
  }
  /**
   * List all registered template names
   */
  list() {
    return Object.keys(this.templates);
  }
};

// src/templates/_layout.ts
var EMAIL_CONSTANTS = {
  APP_NAME: "Shogo",
  APP_URL: "https://shogo.ai",
  SUPPORT_EMAIL: "support@shogo.ai",
  LOGO_URL: "https://shogo.ai/assets/images/shogo-logo-email.png",
  LOGO_DARK_URL: "https://shogo.ai/assets/images/shogo-logo-dark-email.png",
  FOOTER_TEXT: "&copy; {{currentYear}} Shogo. All rights reserved"
};
var LIGHT_STYLES = `
  body, html { margin: 0; padding: 0; }
  .email-body {
    font-family: 'Inter', system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif;
    background-color: #fafafa;
    color: #2a2929;
    line-height: 1.6;
  }
  .email-wrapper { padding: 40px 20px; }
  .email-container {
    max-width: 560px;
    margin: 0 auto;
    background: #ffffff;
    border-radius: 16px;
    border: 1px solid #e5e7eb;
    overflow: hidden;
    box-shadow: 0 1px 3px rgba(0,0,0,0.04);
  }
  .email-header {
    background: #ffffff;
    padding: 24px 32px;
    border-bottom: 1px solid #e5e7eb;
  }
  .email-logo {
    font-size: 18px;
    font-weight: 700;
    color: #2a2929;
    text-decoration: none;
    letter-spacing: -0.3px;
  }
  .email-content { padding: 32px; }
  .email-h1 {
    font-size: 22px;
    font-weight: 700;
    color: #2a2929;
    margin: 0 0 16px;
  }
  .email-text {
    font-size: 15px;
    color: #555;
    margin: 0 0 16px;
  }
  .email-muted {
    font-size: 13px;
    color: #888;
    margin: 0 0 12px;
  }
  .email-btn {
    display: inline-block;
    padding: 12px 32px;
    background: #e8853d;
    color: #ffffff;
    text-decoration: none;
    font-weight: 600;
    font-size: 14px;
    border-radius: 9999px;
    margin: 8px 0 16px;
  }
  .email-btn-outline {
    display: inline-block;
    padding: 12px 32px;
    background: #ffffff;
    color: #2a2929;
    text-decoration: none;
    font-weight: 600;
    font-size: 14px;
    border-radius: 9999px;
    border: 1px solid #d4d4d4;
    margin: 8px 0 16px;
  }
  .email-btn-danger {
    display: inline-block;
    padding: 12px 32px;
    background: #ef4444;
    color: #ffffff;
    text-decoration: none;
    font-weight: 600;
    font-size: 14px;
    border-radius: 9999px;
    margin: 8px 0 16px;
  }
  .email-divider {
    border: none;
    border-top: 1px solid #f1f5f9;
    margin: 24px 0;
  }
  .email-footer {
    padding: 20px 32px;
    background: #ffffff;
    border-top: 1px solid #e5e7eb;
    text-align: center;
  }
  .email-footer-text {
    font-size: 12px;
    color: #888;
    margin: 0;
  }
  .email-badge {
    display: inline-block;
    padding: 4px 10px;
    background: #fff7ed;
    color: #e8853d;
    font-size: 12px;
    font-weight: 600;
    border-radius: 9999px;
    border: 1px solid rgba(232,133,61,0.15);
  }
  .email-detail-row {
    display: flex;
    justify-content: space-between;
    padding: 8px 0;
    border-bottom: 1px solid #f4f4f5;
  }
  .email-detail-label {
    font-size: 13px;
    color: #888;
  }
  .email-detail-value {
    font-size: 13px;
    font-weight: 600;
    color: #2a2929;
  }
  .email-logo-dark { display: none !important; }`;
var DARK_STYLES = `
  @media (prefers-color-scheme: dark) {
    .email-body { background-color: #1a1a1e; color: #e4e4e7; }
    .email-container { background: #27272a; border-color: #3f3f46; box-shadow: 0 1px 4px rgba(0,0,0,0.3); }
    .email-header { background: #27272a; border-bottom-color: #3f3f46; }
    .email-logo { color: #fafafa; }
    .email-h1 { color: #fafafa; }
    .email-text { color: #d4d4d8; }
    .email-muted { color: #a1a1aa; }
    .email-btn-outline { background: #3f3f46; color: #fafafa; border-color: #52525b; }
    .email-divider { border-top-color: #3f3f46; }
    .email-footer { background: #27272a; border-top-color: #3f3f46; }
    .email-footer-text { color: #71717a; }
    .email-badge { background: rgba(232,133,61,0.15); color: #f0a970; border-color: rgba(232,133,61,0.25); }
    .email-detail-row { border-bottom-color: #3f3f46; }
    .email-detail-label { color: #a1a1aa; }
    .email-detail-value { color: #fafafa; }
    .email-logo-light { display: none !important; }
    .email-logo-dark { display: block !important; }
  }`;
var STYLES = `
<style>${LIGHT_STYLES}
${DARK_STYLES}
</style>`;
var DARK_STYLE_OVERRIDES = DARK_STYLES.replace(
  "@media (prefers-color-scheme: dark) {",
  ""
).replace(/\}\s*$/, "");
function wrapInLayout(content) {
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  ${STYLES}
</head>
<body class="email-body">
  <div class="email-wrapper">
    <div class="email-container">
      <div class="email-header">
        <a href="${EMAIL_CONSTANTS.APP_URL}" style="display:inline-block;text-decoration:none;">
          <img class="email-logo-light" src="${EMAIL_CONSTANTS.LOGO_URL}" alt="${EMAIL_CONSTANTS.APP_NAME}" width="169" height="58" style="display:block;height:40px;width:auto;" />
          <img class="email-logo-dark" src="${EMAIL_CONSTANTS.LOGO_DARK_URL}" alt="${EMAIL_CONSTANTS.APP_NAME}" width="169" height="58" style="display:none;height:40px;width:auto;" />
        </a>
      </div>
      <div class="email-content">
        ${content}
      </div>
      <div class="email-footer">
        <p class="email-footer-text">&copy; {{currentYear}} Shogo. All rights reserved</p>
      </div>
    </div>
  </div>
</body>
</html>`;
}

// src/templates/auth/welcome.ts
var welcomeTemplate = {
  name: "welcome",
  subject: "Welcome to {{appName}}!",
  html: wrapInLayout(`
    <h1 class="email-h1">Welcome to {{appName}}!</h1>
    <p class="email-text">Hi {{name}},</p>
    <p class="email-text">
      Thanks for signing up. You now have access to build, deploy, and
      collaborate on AI-powered apps \u2014 all from one place.
    </p>
    <a href="{{loginUrl}}" class="email-btn" style="color:#ffffff;text-decoration:none;">Get Started</a>
    <hr class="email-divider">
    <p class="email-muted">
      If you have any questions, just reply to this email \u2014 we're happy to help.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    loginUrl: EMAIL_CONSTANTS.APP_URL
  }
};

// src/templates/auth/password-reset.ts
var passwordResetTemplate = {
  name: "password-reset",
  subject: "Reset your {{appName}} password",
  html: wrapInLayout(`
    <h1 class="email-h1">Reset Your Password</h1>
    <p class="email-text">
      We received a request to reset the password for your account.
      Click the button below to choose a new one:
    </p>
    <a href="{{resetUrl}}" class="email-btn" style="color:#ffffff;text-decoration:none;">Reset Password</a>
    <hr class="email-divider">
    <p class="email-muted">This link expires in {{expiresIn}}.</p>
    <p class="email-muted">
      If you didn't request this, you can safely ignore this email.
      Your password will remain unchanged.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    expiresIn: "1 hour"
  }
};

// src/templates/auth/email-verification.ts
var emailVerificationTemplate = {
  name: "email-verification",
  subject: "Verify your email for {{appName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Verify Your Email</h1>
    <p class="email-text">
      Please confirm your email address by clicking the button below.
      This helps us keep your account secure.
    </p>
    <a href="{{verifyUrl}}" class="email-btn" style="color:#ffffff;text-decoration:none;">Verify Email</a>
    <hr class="email-divider">
    <p class="email-muted">This link expires in {{expiresIn}}.</p>
    <p class="email-muted">
      If you didn't create an account, you can safely ignore this email.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    expiresIn: "24 hours"
  }
};

// src/templates/invitation/workspace-invite.ts
var workspaceInviteTemplate = {
  name: "workspace-invite",
  subject: "{{inviterName}} invited you to {{workspaceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">You're Invited!</h1>
    <p class="email-text">
      <strong>{{inviterName}}</strong> has invited you to join the
      <strong>{{workspaceName}}</strong> workspace as
      <span class="email-badge">{{role}}</span>.
    </p>
    <p class="email-text">
      Accept the invitation to start collaborating on projects, share
      resources, and build together.
    </p>
    <a href="{{acceptUrl}}" class="email-btn" style="color:#ffffff;text-decoration:none;">Accept Invitation</a>
    <hr class="email-divider">
    <p class="email-muted">
      If you weren't expecting this, you can safely ignore this email.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    role: "Editor"
  }
};

// src/templates/invitation/project-invite.ts
var projectInviteTemplate = {
  name: "project-invite",
  subject: "{{inviterName}} invited you to {{projectName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Project Invitation</h1>
    <p class="email-text">
      <strong>{{inviterName}}</strong> has invited you to the
      <strong>{{projectName}}</strong> project as
      <span class="email-badge">{{role}}</span>.
    </p>
    <a href="{{acceptUrl}}" class="email-btn" style="color:#ffffff;text-decoration:none;">Accept Invitation</a>
    <hr class="email-divider">
    <p class="email-muted">
      If you weren't expecting this, you can safely ignore this email.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    role: "Editor"
  }
};

// src/templates/invitation/invite-accepted.ts
var inviteAcceptedTemplate = {
  name: "invite-accepted",
  subject: "{{inviteeName}} accepted your invitation to {{resourceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Invitation Accepted</h1>
    <p class="email-text">
      <strong>{{inviteeName}}</strong> ({{inviteeEmail}}) has accepted your
      invitation to join <strong>{{resourceName}}</strong>.
    </p>
    <p class="email-text">
      They now have access to the {{resourceType}} and can start collaborating.
    </p>
    <a href="{{dashboardUrl}}" class="email-btn-outline">View Team</a>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    resourceType: "workspace"
  }
};

// src/templates/billing/plan-upgraded.ts
var planUpgradedTemplate = {
  name: "plan-upgraded",
  subject: "{{workspaceName}} has been upgraded to {{planName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Plan Upgraded!</h1>
    <p class="email-text">
      Your workspace <strong>{{workspaceName}}</strong> is now on the
      <span class="email-badge">{{planName}}</span> plan.
    </p>
    <table width="100%" cellpadding="0" cellspacing="0">
      <tr>
        <td class="email-detail-label">Plan</td>
        <td class="email-detail-value" align="right">{{planName}}</td>
      </tr>
      <tr>
        <td class="email-detail-label">Billing</td>
        <td class="email-detail-value" align="right">{{billingInterval}}</td>
      </tr>
      <tr>
        <td class="email-detail-label">Seats</td>
        <td class="email-detail-value" align="right">{{seats}}</td>
      </tr>
      <tr>
        <td class="email-detail-label">Monthly included usage</td>
        <td class="email-detail-value" align="right">{{includedUsdTotal}}</td>
      </tr>
    </table>
    <p class="email-text" style="margin-top:18px;">
      Every AI request is billed at the provider's raw cost plus a flat 20% markup \u2014
      no credits, no unit conversions. You can add or remove seats anytime from
      Settings &rarr; Billing; we prorate the change automatically.
    </p>
    <br>
    <a href="{{dashboardUrl}}" class="email-btn" style="color:#ffffff;text-decoration:none;">Go to Dashboard</a>
    <hr class="email-divider">
    <p class="email-muted">
      Manage your subscription anytime from Settings &rarr; Billing.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    billingInterval: "Monthly",
    seats: "1"
  }
};

// src/templates/billing/payment-receipt.ts
var paymentReceiptTemplate = {
  name: "payment-receipt",
  subject: "Payment receipt for {{workspaceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Payment Received</h1>
    <p class="email-text">
      Thank you! Here's your receipt for <strong>{{workspaceName}}</strong>.
    </p>
    <table width="100%" cellpadding="0" cellspacing="0">
      <tr>
        <td class="email-detail-label">Plan</td>
        <td class="email-detail-value" align="right">{{planName}}</td>
      </tr>
      <tr>
        <td class="email-detail-label">Amount</td>
        <td class="email-detail-value" align="right">{{currency}}{{amount}}</td>
      </tr>
      <tr>
        <td class="email-detail-label">Date</td>
        <td class="email-detail-value" align="right">{{invoiceDate}}</td>
      </tr>
    </table>
    <br>
    <a href="{{invoiceUrl}}" class="email-btn-outline">View Invoice</a>
    <hr class="email-divider">
    <p class="email-muted">
      If you have billing questions, reply to this email or contact support.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    currency: "$"
  }
};

// src/templates/billing/payment-failed.ts
var paymentFailedTemplate = {
  name: "payment-failed",
  subject: "Payment failed for {{workspaceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Payment Failed</h1>
    <p class="email-text">
      We were unable to process the payment of <strong>{{currency}}{{amount}}</strong>
      for <strong>{{workspaceName}}</strong> ({{planName}} plan).
    </p>
    <p class="email-text">
      Please update your payment method to keep your plan active and avoid
      any interruption to your workspace.
    </p>
    <a href="{{retryUrl}}" class="email-btn-danger" style="color:#ffffff;text-decoration:none;">Update Payment Method</a>
    <hr class="email-divider">
    <p class="email-muted">
      If you believe this is a mistake, please reply to this email or
      contact support.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    currency: "$"
  }
};

// src/templates/billing/usage-overage-charged.ts
var usageOverageChargedTemplate = {
  name: "usage-overage-charged",
  subject: "Usage charge of {{currency}}{{amount}} for {{workspaceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">On-demand usage charged</h1>
    <p class="email-text">
      <strong>{{workspaceName}}</strong> has used more than its included monthly
      usage, so we charged an on-demand usage block. We bill these in blocks as
      you go rather than one large amount at the end of the period.
    </p>
    <table width="100%" cellpadding="0" cellspacing="0">
      <tr>
        <td class="email-detail-label">Charged now</td>
        <td class="email-detail-value" align="right">{{currency}}{{amount}}</td>
      </tr>
      <tr>
        <td class="email-detail-label">Overage this period</td>
        <td class="email-detail-value" align="right">{{currency}}{{periodOverageUsd}}</td>
      </tr>
    </table>
    <br>
    <a href="{{invoiceUrl}}" class="email-btn-outline">View invoice</a>
    <hr class="email-divider">
    <p class="email-text">
      Want to cap on-demand spend? You can set a monthly spending limit any time.
    </p>
    <a href="{{manageUrl}}" class="email-btn-outline">Set a spending limit</a>
    <p class="email-muted">
      If you have billing questions, reply to this email or contact support.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    currency: "$"
  }
};

// src/templates/billing/usage-approaching-limit.ts
var usageApproachingLimitTemplate = {
  name: "usage-approaching-limit",
  subject: "Usage alert: {{workspaceName}} is at {{percentUsed}}% of its {{limitLabel}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Approaching your {{limitLabel}}</h1>
    <p class="email-text">
      <strong>{{workspaceName}}</strong> has used
      <strong>{{currency}}{{usedUsd}}</strong> of its
      <strong>{{currency}}{{limitUsd}}</strong> {{limitLabel}} this period
      (<strong>{{percentUsed}}%</strong>).
    </p>
    <p class="email-text">
      Once the included amount is used up, additional usage is billed on demand.
      You can review usage or set a monthly spending limit below.
    </p>
    <a href="{{manageUrl}}" class="email-btn">Review usage &amp; limits</a>
    <hr class="email-divider">
    <p class="email-muted">
      You're receiving this because you're a billing admin for this workspace.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    currency: "$"
  }
};

// src/templates/workspace/member-joined.ts
var memberJoinedTemplate = {
  name: "member-joined",
  subject: "{{memberName}} joined {{workspaceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">New Team Member</h1>
    <p class="email-text">
      <strong>{{memberName}}</strong> ({{memberEmail}}) has joined
      <strong>{{workspaceName}}</strong> as
      <span class="email-badge">{{role}}</span>.
    </p>
    <a href="{{dashboardUrl}}" class="email-btn-outline">View Team</a>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME,
    role: "Editor"
  }
};

// src/templates/workspace/member-removed.ts
var memberRemovedTemplate = {
  name: "member-removed",
  subject: "You have been removed from {{workspaceName}}",
  html: wrapInLayout(`
    <h1 class="email-h1">Workspace Access Removed</h1>
    <p class="email-text">
      You no longer have access to the <strong>{{workspaceName}}</strong>
      workspace. All associated projects and data in that workspace are no
      longer accessible from your account.
    </p>
    <hr class="email-divider">
    <p class="email-muted">
      If you believe this was a mistake, please contact the workspace owner
      or reply to this email.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME
  }
};

// src/templates/workspace/account-deleted.ts
var accountDeletedTemplate = {
  name: "account-deleted",
  subject: "Your {{appName}} account has been deleted",
  html: wrapInLayout(`
    <h1 class="email-h1">Account Deleted</h1>
    <p class="email-text">
      Your {{appName}} account associated with <strong>{{email}}</strong>
      has been permanently deleted. All your data, workspaces, and projects
      have been removed.
    </p>
    <p class="email-text">
      If you ever want to come back, you can create a new account at any
      time.
    </p>
    <hr class="email-divider">
    <p class="email-muted">
      If you didn't request this deletion, please contact us immediately
      by replying to this email.
    </p>
  `),
  defaults: {
    appName: EMAIL_CONSTANTS.APP_NAME
  }
};

// src/templates/index.ts
var allTemplates = [
  welcomeTemplate,
  passwordResetTemplate,
  emailVerificationTemplate,
  workspaceInviteTemplate,
  projectInviteTemplate,
  inviteAcceptedTemplate,
  planUpgradedTemplate,
  paymentReceiptTemplate,
  paymentFailedTemplate,
  usageOverageChargedTemplate,
  usageApproachingLimitTemplate,
  memberJoinedTemplate,
  memberRemovedTemplate,
  accountDeletedTemplate
];
function createShogoTemplateRegistry() {
  const registry = new EmailTemplateRegistry();
  for (const template of allTemplates) {
    registry.register(template);
  }
  return registry;
}

// src/server.ts
var EmailService = class {
  constructor(provider, templates, defaultFrom) {
    this.provider = provider;
    this.templates = templates;
    this.defaultFrom = defaultFrom;
  }
  /**
   * Check if the service is configured and ready
   */
  isConfigured() {
    return this.provider.isConfigured();
  }
  /**
   * Register a custom email template
   */
  registerTemplate(template) {
    this.templates.register(template);
  }
  /**
   * Send a raw email
   */
  async send(params) {
    const emailParams = {
      ...params,
      from: params.from ?? this.defaultFrom
    };
    return this.provider.send(emailParams);
  }
  /**
   * Send a templated email
   */
  async sendTemplate(params) {
    const { template: templateName, data, ...rest } = params;
    const { subject, html, text } = this.templates.render(templateName, data);
    return this.send({
      ...rest,
      subject,
      html,
      text
    });
  }
};
function createEmail(options = {}) {
  const { config, defaultFrom, includeBuiltins = true } = options;
  let provider = null;
  if (config) {
    if (config.provider === "smtp" && config.smtp) {
      provider = new SmtpProvider(config.smtp);
    } else if (config.provider === "ses" && config.ses) {
      provider = new SesProvider(config.ses);
    } else if (config.provider === "oci-email" && config.ociEmail) {
      provider = new OciEmailProvider(config.ociEmail);
    }
  } else {
    provider = createProviderFromEnv();
  }
  if (!provider) {
    throw EmailError.providerNotConfigured();
  }
  const from = defaultFrom ?? config?.defaultFrom ?? process.env.EMAIL_FROM ?? process.env.SMTP_FROM_EMAIL ?? process.env.SES_FROM_EMAIL ?? process.env.OCI_EMAIL_FROM_ADDRESS;
  if (!from) {
    throw EmailError.configMissing("defaultFrom (or EMAIL_FROM env var)");
  }
  const templates = includeBuiltins ? createShogoTemplateRegistry() : new EmailTemplateRegistry();
  return new EmailService(provider, templates, from);
}
function createEmailOptional(options = {}) {
  try {
    return createEmail(options);
  } catch {
    return null;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createEmail,
  createEmailOptional
});
